---
title: "ALE indexing"
author: "Yeonui+Kwak"
date: "8/7/2020"
output: html_document


ALE  was indexed based on GENCODE stop codon containing exon start site.
---
```{R}
cps.ref<-read.table("final.30980cps.in.3UTR.1kb.ext.txt",header = T,stringsAsFactors = F)
ref<-read.table("./Reference/transcripts.bed13",header = F,stringsAsFactors = F)
```

1. Define ALE position by the start site of Exon containing stopcodon.
```{R}
Exmp=ref%>%filter(V6=="+")%>%
  #filter(V7!=V8)%>%
  mutate(ALEpos=NA)
Exmp
for (i in 1:nrow(Exmp)){
  blockstarts=as.numeric(unlist(strsplit(str_trim(Exmp[i,12],side="both"),",")))
  blocksizes=as.numeric(unlist(strsplit(str_trim(Exmp[i,11],side="both"),",")))
  stopcodon=as.numeric(Exmp[i,8])
  Txnstart=as.numeric(Exmp[i,2])
  Exoncounts=as.numeric(Exmp[i,10])
  count=1
  while (stopcodon >= (Txnstart+blockstarts[count])&count<=Exoncounts){
    count=count+1
    #blocksizes[length(blockstarts)]=Newend-(Txnstart+(blockstarts[length(blockstarts)]))
    #Exmp[i,11]=str_replace_all(paste(blocksizes, ",",collapse =""), fixed(" "), "")
  }
  count=count-1
  Exmp[i,14]=Txnstart+blockstarts[count]
  }

pl=Exmp
Exmp=ref%>%filter(V6=="-")%>%
  #filter(V7!=V8)%>%
  mutate(ALEpos=NA)
Exmp
for (i in 1:nrow(Exmp)){
  blockstarts=as.numeric(unlist(strsplit(str_trim(Exmp[i,12],side="both"),",")))
  blocksizes=as.numeric(unlist(strsplit(str_trim(Exmp[i,11],side="both"),",")))
  stopcodon=as.numeric(Exmp[i,7])
  Txnstart=as.numeric(Exmp[i,2])
  Exoncounts=as.numeric(Exmp[i,10])
  count=1
  while (stopcodon >= (Txnstart+blockstarts[count])&count<=Exoncounts){
    count=count+1
  }
  count=count-1
  Exmp[i,14]=Txnstart+blockstarts[count]+blocksizes[count]
  }
head(Exmp)
mn=Exmp
ALEpos=bind_rows(pl,mn)%>%arrange(V1,V2,V3)%>%mutate(ALEpos=as.numeric(ALEpos))
ALEpos%>%filter(V13=="UBE2J2")
write.table(ALEpos%>%select(V4,ALEpos)%>%rename(id=V4),"ALEposannotation_withNCRNA.txt",col.names = T,row.names = F,quote=F,sep="\t")
```
Fig1G
```{R}
alepos<-read.table("ALEposannotation_withNCRNA.txt",header=T,stringsAsFactors = F)
head(cps.ref)
cps.ref2=cps.ref%>%inner_join(alepos,by="id")
ALEindex=cps.ref2%>%group_by(hgnc)%>% arrange(hgnc,ALEpos)%>%distinct_at(vars(ALEpos),.keep_all =T)%>%
  mutate(ALE=ifelse(strand=="+",rank(ALEpos),rank(-ALEpos)))%>%select(hgnc,ALEpos,ALE)%>%unite("newid",c(hgnc,ALEpos),sep=":")
cps.ref2=cps.ref2%>%unite("newid",c(hgnc,ALEpos),sep=":")%>%inner_join(ALEindex,by="newid")%>%separate(newid,c("hgnc","ALEpos"),sep=":")

cps.collapsed=cps.ref2%>%ungroup()%>%arrange(hgnc,CPSid,relpos,ALE,id)%>%distinct_at(vars(hgnc,CPSid,relpos),.keep_all = T)
head(cps.collapsed)
cps.collapsed%>%ungroup()%>%summarise(n_distinct(CPSid)) #23036 (CODING RNA) #25997 (Coding +Noncoding RNA)
cps.collapsed%>%ungroup()%>%arrange(hgnc,id,refpos)%>%
  distinct_at(vars(hgnc,refpos),.keep_all = T)%>%nrow()

count1=cps.collapsed%>%filter(abs(relpos)<10)%>%summarise(n_distinct(CPSid)) #10157/30980
id1=cps.collapsed%>%filter(abs(relpos)<10)%>%select(CPSid)%>%arrange(CPSid)%>%distinct()%>%unlist()
count2=cps.collapsed%>%filter(!(CPSid %in% id1))%>%filter(abs(relpos)<300&abs(relpos)>=10)%>%summarise(n_distinct(CPSid)) #10111/30980
id2=cps.collapsed%>%filter(!(CPSid %in% id))%>%filter(abs(relpos)<300&abs(relpos)>=10)%>%select(CPSid)%>%arrange(CPSid)%>%distinct()%>%unlist()
count3=cps.collapsed%>%filter(!(CPSid %in% c(id1,id2)))%>%filter(abs(relpos)>=300)%>%summarise(n_distinct(CPSid)) #10822
cps.collapsed%>%ungroup()%>%summarise(n_distinct(CPSid))
stat=data.frame("Location"=c("within 10 nt","b/w 10 nt and 300 nt", "beyond 300 nt"),"PAS sites"=c(unlist(count1),unlist(count2),unlist(count3)))%>%mutate(`Location`=factor(Location,levels =c("within 10 nt","b/w 10 nt and 300 nt", "beyond 300 nt")))
pdf("Stat_PAS_VERSION2.pdf",width = 4,height = 3)
ggplot(stat,aes(x=Location,y=PAS.sites,fill=Location))+geom_bar(stat="identity")+scale_fill_manual(name="Poly(A) tail length",values=yk.col(5)[c(2,3,4)])+theme_bw()+ylab("Number of PAS sites")+xlab("Position relative to \n the GENCODE PAS annotation")+theme(axis.text.x = element_blank())+ylim(c(0,11000))
dev.off()

```

```{R}

tmp=cps%>%select(-CPSpos,-strand,-cps.ratio.perhgnc,-refpos)%>%group_by(CPSid,id,hgnc,cpsNo,time,relpos)%>%
  summarise(mean.rc=mean(readcount))%>%ungroup()%>%
  distinct_at(vars(hgnc,time,CPSid),.keep_all = T)%>%
  filter(mean.rc>=0)%>%
  ungroup()%>%
  group_by(hgnc, time)%>%
   mutate(ratio=(mean.rc+1)/sum(mean.rc+1))%>%
  #filter(ratio>=0.70)%>%
  ungroup()%>%select(hgnc,CPSid,time,relpos)%>%
  spread(time,relpos)%>% 
  filter_all(all_vars(!is.na(.)))%>%ungroup()%>%group_by(hgnc)%>%summarise(t=n_distinct(CPSid))%>%group_by(t)%>%summarise(n())
tmp

```



Number of genes with different numbers of tandem poly(A) sites in the 3'UTR+1kb.
```{R}
tmp=cps.ref%>%group_by(hgnc)%>%summarise(t=n_distinct(CPSid))%>%group_by(t)%>%summarise(n())
tmp=tmp%>%mutate(count=ifelse(t<=15,t,">15"))%>%group_by(count)%>%summarise(sum=sum(`n()`))%>%mutate(count=factor(count,levels=c(1:15,">15")))
write.table(tmp,"summaryoftallyof genes withCPSNo_version2.txt",sep="\t",quote=F,row.names = F,col.names = T)
pdf("Number of genes with different numbers of poly(A) sites_Version2.pdf",width=4,height=3)
ggplot(tmp,aes(x=count,y=sum))+geom_bar(stat = "identity")+ylab("Number of genes")+xlab("3Â´CPS counts")+
  theme_bw()
dev.off()

```