---
title: "3UTR 1kb Extension"
author: "Yeonui+Kwak"
date: "8/7/2020"
output: html_document
---
Steps:
1)Extend last exon by 1 kb
2)Extract 3'UTR region
3)Separate rows.





1.Set a working directory
```{R}
setwd("~/Desktop/Publication_Mar/Final_Fig1/Rcode")

```

2. Setup color and load packages.
```{r}
#install.packages("raster")
library(RColorBrewer)
library(raster)
library(tidyverse)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggbeeswarm)
library(gridExtra)
source("~/Desktop/Publication_Mar/Final_Supplement/Rsourcecode/scatterPlot.R")
```

3. Load V26 Gencode transcriptome annotation in bed12 format.

```{R}
ref<-read.table("./Reference/transcripts.bed13",header=F,stringsAsFactors = F)
```

4. For + strand transcripts

```{R}
Exmp=ref%>%filter(V6=="+")%>%filter(V7!=V8)
Exmp
for (i in 1:nrow(Exmp)){
  blockstarts=as.numeric(unlist(strsplit(str_trim(Exmp[i,12],side="both"),",")))
  blocksizes=as.numeric(unlist(strsplit(str_trim(Exmp[i,11],side="both"),",")))
  stopcodon=as.numeric(Exmp[i,8])
  Txnstart=as.numeric(Exmp[i,2])
  Txnend=as.numeric(Exmp[i,3])
  Exoncounts=as.numeric(Exmp[i,10])
  #1KB extension
  Txnend=Txnend+1000
  blocksizes[Exoncounts]=blocksizes[Exoncounts]+1000
  #Extract 3UTR region
  count=1
  while (stopcodon >= (Txnstart+blockstarts[count])&count<=Exoncounts){
    count=count+1
    #blocksizes[length(blockstarts)]=Newend-(Txnstart+(blockstarts[length(blockstarts)]))
    #Exmp[i,11]=str_replace_all(paste(blocksizes, ",",collapse =""), fixed(" "), "")
  }
  count=count-1 #stopcodon containing exon.
  exonend=Txnstart+blockstarts[count]+blocksizes[count]
  blockstarts[count]=stopcodon-Txnstart
  blocksizes[count]=exonend-stopcodon
  #leave 3'UTR regions only
  blockstarts=blockstarts[count:Exoncounts]
  blocksizes=blocksizes[count:Exoncounts]
  Exoncounts=Exoncounts-(count-1)
  Txnstart=blockstarts[1]+Txnstart
  blockstarts=blockstarts-blockstarts[1]
  #Exmp[i,11]=str_replace_all(paste(blocksizes, ",",collapse =""), fixed(" "), "")
  #Exmp[i,12]=str_replace_all(paste(blockstarts, ",",collapse =""), fixed(" "), "")
  Exmp[i,14]=str_replace_all(paste(paste(blocksizes,blockstarts,sep=":"),",",collapse =""), fixed(" "), "")
  Exmp[i,10]=Exoncounts
  Exmp[i,2]=Txnstart
  Exmp[i,8]=Exmp[i,2]
  Exmp[i,7]=Exmp[i,2]
  Exmp[i,3]=Txnstart+blockstarts[Exoncounts]+blocksizes[Exoncounts]
}
Exmp%>%filter(V4=="ENST00000487053.5")%>%
  select(V1,V2,V4,V5,V6,V10,V13,V14)%>%
  separate_rows(V14,sep=",")%>%
  filter(V14!="")%>%separate(V14,c("V11","V12"),sep=":")
pl=Exmp%>%
  select(V1,V2,V4,V5,V6,V10,V13,V14)%>%
  separate_rows(V14,sep=",")%>%
  filter(V14!="")%>%separate(V14,c("V11","V12"),sep=":")%>%mutate_at(vars(V2,V11,V12,V10),.funs=as.numeric)%>%mutate(V2=V2+V12)%>%mutate(V3=V2+V11)%>%select(-V11,-V12)%>%select(V1,V2,V3,V4,V5,V6,V10,V13)%>% drop_na()
```
For - strand transcripts
```{R}
#select mRNA
Exmp=ref%>%filter(V6=="-")%>%filter(V7!=V8)
#before=Exmp%>%filter(V4=="ENST00000417652.5")
#Exmp%>%filter(V4=="ENST00000417652.5")
for (i in 1:nrow(Exmp)){
  blockstarts=as.numeric(unlist(strsplit(str_trim(Exmp[i,12],side="both"),",")))
  blocksizes=as.numeric(unlist(strsplit(str_trim(Exmp[i,11],side="both"),",")))
  stopcodon=as.numeric(Exmp[i,7])
  Txnstart=as.numeric(Exmp[i,2])
  Exoncounts=as.numeric(Exmp[i,10])
  #1KB extension
  Txnstart=Txnstart-1000
  blockstarts=blockstarts+1000
  blockstarts[1]=0
  blocksizes[1]=blocksizes[1]+1000
  #Extract 3UTR region
  count=1
  while (stopcodon >= (Txnstart+blockstarts[count])&count<=Exoncounts){
    count=count+1
  }
  count=count-1
  exonstart=Txnstart+blockstarts[count]
  blocksizes[count]=stopcodon-exonstart
  #leave 3'UTR regions only
  blockstarts=blockstarts[1:count]
  blocksizes=blocksizes[1:count]
  Exoncounts=count
  #Exmp[i,11]=str_replace_all(paste(blocksizes, ",",collapse =""), fixed(" "), "")
  #Exmp[i,12]=str_replace_all(paste(blockstarts, ",",collapse =""), fixed(" "), "")
  Exmp[i,14]=str_replace_all(paste(paste(blocksizes,blockstarts,sep=":"),",",collapse =""), fixed(" "), "")
  Exmp[i,10]=Exoncounts
  Exmp[i,2]=Txnstart
  Exmp[i,3]=Txnstart+blockstarts[Exoncounts]+blocksizes[Exoncounts]
  Exmp[i,8]=Exmp[i,3]
  Exmp[i,7]=Exmp[i,3]
}

#3. Separate rows
mn=Exmp%>%select(V1,V2,V4,V5,V6,V10,V13,V14)%>%
  separate_rows(V14,sep=",")%>%
  filter(V14!="")%>%separate(V14,c("V11","V12"),sep=":")%>%mutate_at(vars(V2,V11,V12,V10),.funs=as.numeric)%>%mutate(V2=V2+V12)%>%mutate(V3=V2+V11)%>%select(-V11,-V12)%>%select(V1,V2,V3,V4,V5,V6,V10,V13)%>% drop_na()

UTR.1kb.bed=bind_rows(pl,mn)%>%arrange(V1,V2,V3)%>%drop_na()

write.table(UTR.1kb.bed,"3UTR.1kb.extension.bed",quote=F,col.names = F,row.names = F,sep="\t")
```


