---
title: "Step1_calculating meanPAL"
author: "Yeonui+Kwak"
date: "8/13/2020"
output: html_document
---

1.Set a working directory
```{R}
setwd("~/Desktop/Publication_Mar/Final_fig4/Rcode/step1")

```
2.Load the required packages,and set up colorcode.
```{r}
library(RColorBrewer)
#nstall.packages("raster")
library(raster)
library(tidyverse)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggbeeswarm)
yk.col = function(n = 5, sat = 0.7, lum = 1) {
	col = c("royalblue4", "deepskyblue3", "turquoise3",
	   "grey62", "goldenrod2", "orange2", "orangered2") %>%
	col2rgb %>%
	rgb2hsv
	return(colorRampPalette(hsv(col[1,], col[2,] * sat,
				    col[3,] * lum))(n)) }
```
3. Load the files for ks.test
Spike-in Reference.
#normalize by spike-in
ted$rep1_0=ted$rep1_0+13
ted$rep1_1=ted$rep1_1+15
ted$rep1_2=ted$rep1_2+13
ted$rep1_4=ted$rep1_4+19
ted$rep2_0=ted$rep2_0+16
ted$rep2_1=ted$rep2_1+13
ted$rep2_2=ted$rep2_2+12
ted$rep2_4=ted$rep2_4+8
ted=as_tibble(ted)


3.Statistical Test to identify genes with Significant PAL changes.
3.1 Load the TED-seq position matrix 3.2 Put four matrix to each variable.
Rep1.
```{r}
t0<-read.csv("tedseq.custom.rep1.0hr.txt",sep="\t",header=F, stringsAsFactors = F)
t1<-read.csv("tedseq.custom.rep1.1hr.txt",sep="\t",header=F, stringsAsFactors = F)
t2<-read.csv("tedseq.custom.rep1.2hr.txt",sep="\t",header=F, stringsAsFactors = F)
t4<-read.csv("tedseq.custom.rep1.4hr.txt",sep="\t",header=F, stringsAsFactors = F)

First=t0
Second=t1
Third=t2
Four=t4

#rep1
spike_first=13
spike_second=15
spike_third=13
spike_fourth=19
```


3.3 filter transcripts with more than 50 read counts.
```{R}
start=Sys.time()
dat2=as_tibble(First);dat2$sum=rowSums(dat2[203:502])
d0h=dat2%>%filter(sum>50)%>%
  select(-sum)%>%
  select(1,203:502)

dat2=as_tibble(Second);dat2$sum=rowSums(dat2[203:502])
d1h=dat2%>%filter(sum>50)%>%
  select(-sum)%>%select(1,203:502)

dat2=as_tibble(Third);dat2$sum=rowSums(dat2[203:502])
d2h=dat2%>%filter(sum>50)%>%
  select(-sum)%>%select(1,203:502)
dat2=as_tibble(Four);dat2$sum=rowSums(dat2[203:502])
d4h=dat2%>%filter(sum>50)%>%
  select(-sum)%>%select(1,203:502)

nrow(d4h)
```


innerjoin all samples
```{R}
m=list(d0h,d1h,d2h,d4h)%>%reduce(inner_join,by="V1")
m%>%drop_na%>%nrow() #11868 isoforms. rep2:12258
m=m%>%separate(V1,c("id","refpos","chr","cpspos"),sep=":",remove=F)%>%unite("newid",c(id,cpspos),sep=":")%>%select(-chr,-refpos)
head(m)
```
1. Calculate mean PAL, and ks.test pvalue after normalizing spike-in values.

rep1
```{R}
start=Sys.time()
mean1=c();mean2=c();mean3=c();mean4=c()

for (i in 1:nrow(m)){
  vec_1=c()
  for (j in 1:300){
    vec_1=c(vec_1,rep(j+spike_first,m[i,j+2]))
  }
  vec_2=c()
  for (k in 1:300){
    vec_2=c(vec_2,rep(k+spike_second, m[i,k+302]))
  }
  vec_3=c()
  for (k in 1:300){
    vec_3=c(vec_3,rep(k+spike_third, m[i,k+602]))
  }
  vec_4=c()
  for (k in 1:300){
    vec_4=c(vec_4,rep(k+spike_fourth, m[i,k+902]))
  }
   mean1=c(mean1,mean(vec_1))
   mean2=c(mean2,mean(vec_2))
   mean3=c(mean3,mean(vec_3))
   mean4=c(mean4,mean(vec_4))

}
m$`meanpal_0`=mean1
m$`meanpal_1`=mean2
m$`meanpal_2`=mean3
m$`meanpal_4`=mean4

end=Sys.time()
cat("Running time for loading the files was",end-start,"mins")
head(m)
```

```{R}
stat.final=m%>%select(newid,V1,contains("meanpal"))#%>%inner_join(final.pas,by="newid")
nrow(stat.final) #11868
write.table(stat.final,"11868.PASisoforms_meanPAL_rep1.txt",col.names = T,row.names = F,quote = F,sep="\t")

```

Replicate2


3.1 Load the TED-seq position matrix
#Rep2.
```{r}
start=Sys.time()
t2.0<-read.csv("tedseq.custom.rep2.0hr.txt",sep="\t",header=F, stringsAsFactors = F)
t2.1<-read.csv("tedseq.custom.rep2.1hr.txt",sep="\t",header=F, stringsAsFactors = F)
t2.2<-read.csv("tedseq.custom.rep2.2hr.txt",sep="\t",header=F, stringsAsFactors = F)
t2.4<-read.csv("tedseq.custom.rep2.4hr.txt",sep="\t",header=F, stringsAsFactors = F)
end=Sys.time()
cat("Running time for loading the files was",end-start,"mins") #2mins
#rep2
First=t2.0
Second=t2.1
Third=t2.2
Four=t2.4

#rep2
spike_first=16
spike_second=13
spike_third=12
spike_fourth=8
```


3.3 filter transcripts with more than 50 read counts.
```{R}
start=Sys.time()
dat2=as_tibble(First);dat2$sum=rowSums(dat2[203:502])
d0h=dat2%>%filter(sum>50)%>%
  select(-sum)%>%
  select(1,203:502)

dat2=as_tibble(Second);dat2$sum=rowSums(dat2[203:502])
d1h=dat2%>%filter(sum>50)%>%
  select(-sum)%>%select(1,203:502)

dat2=as_tibble(Third);dat2$sum=rowSums(dat2[203:502])
d2h=dat2%>%filter(sum>50)%>%
  select(-sum)%>%select(1,203:502)
dat2=as_tibble(Four);dat2$sum=rowSums(dat2[203:502])
d4h=dat2%>%filter(sum>50)%>%
  select(-sum)%>%select(1,203:502)

nrow(d4h)
```


innerjoin all samples
```{R}
m=list(d0h,d1h,d2h,d4h)%>%reduce(inner_join,by="V1")
m%>%drop_na%>%nrow() #11868 isoforms. rep2:12258
m=m%>%separate(V1,c("id","refpos","chr","cpspos"),sep=":",remove=F)%>%unite("newid",c(id,cpspos),sep=":")%>%select(-chr,-refpos)
head(m)
```
1. Calculate mean PAL, and ks.test pvalue after normalizing spike-in values.

rep1
```{R}
start=Sys.time()
mean1=c();mean2=c();mean3=c();mean4=c()

for (i in 1:nrow(m)){
  vec_1=c()
  for (j in 1:300){
    vec_1=c(vec_1,rep(j+spike_first,m[i,j+2]))
  }
  vec_2=c()
  for (k in 1:300){
    vec_2=c(vec_2,rep(k+spike_second, m[i,k+302]))
  }
  vec_3=c()
  for (k in 1:300){
    vec_3=c(vec_3,rep(k+spike_third, m[i,k+602]))
  }
  vec_4=c()
  for (k in 1:300){
    vec_4=c(vec_4,rep(k+spike_fourth, m[i,k+902]))
  }
   mean1=c(mean1,mean(vec_1))
   mean2=c(mean2,mean(vec_2))
   mean3=c(mean3,mean(vec_3))
   mean4=c(mean4,mean(vec_4))

}
m$`meanpal_0`=mean1
m$`meanpal_1`=mean2
m$`meanpal_2`=mean3
m$`meanpal_4`=mean4

end=Sys.time()
cat("Running time for loading the files was",end-start,"mins")
head(m)
```

```{R}
stat.final=m%>%select(newid,V1,contains("meanpal"))#%>%inner_join(final.pas,by="newid")
head(stat.final)
nrow(stat.final) #12258
write.table(stat.final,"12258.PASisoforms_meanPAL_rep2.txt",col.names = T,row.names = F,quote = F,sep="\t")


```
