---
title: "Step6:ALE statistics!"
author: "Yeonui+Kwak"
date: "8/8/2020"
output: html_document
---
1.Set a working directory
```{R}
setwd("~/Desktop/Publication_Mar/Final_Fig2/Rcode/")

```

#set up color
```{r}
library(RColorBrewer)
#nstall.packages("raster")
library(raster)
library(tidyverse)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggbeeswarm)
yk.col = function(n = 5, sat = 0.7, lum = 1) {
	col = c("royalblue4", "deepskyblue3", "turquoise3",
	   "grey62", "goldenrod2", "orange2", "orangered2") %>%
	col2rgb %>%
	rgb2hsv
	return(colorRampPalette(hsv(col[1,], col[2,] * sat,
				    col[3,] * lum))(n)) }
```

1.Select the genes with multiple tandem UTR( 2 and 3 only)

```{R}
final=read.table("custom.pal.withoutnearbyPAS.txt",header=T,stringsAsFactors = F)
colnames(final)
```

2. Statistics of ALE

```{R}

t=final%>%
  filter(meanpal<300)%>%
  group_by(hgnc)%>%distinct_at(vars(ALE),.keep_all = T)%>%summarise(ALE_count=n())%>%
  #filter(tandem_count>=1)%>%
  #unite(newid,c(hgnc,ALE),sep=":")%>%
  group_by(ALE_count)%>%summarise(gene.count=n())%>%
mutate(group=ifelse(ALE_count==1,"single last exon","multiple ALEs"))%>%
  mutate(count=factor(ALE_count,levels = c(1:7)))
write.table(t,"Fig2G_table.txt",col.names = T,row.names = F,quote = F,sep="\t")
#t=bind_rows(t,c(6,0,"multiple tandem 3´UTRs",6))
pdf("Fig2G_ALE_statistic.pdf",width=4,height=3)
ggplot(t,aes(x=as.factor(ALE_count),y=gene.count,fill=group))+geom_bar(stat="identity")+theme_bw()+ylab("Number of genes")+xlab("Number of ALEs per gene")+theme_minimal()+scale_fill_manual(values =yk.col(2)[c(1,2)])+ scale_x_discrete(limits=c("1", "2","3","4","5"))
dev.off()

```
Select genes with multiple ALEs (2,3)
#new ratio: readcountofa peak/sum(peaks)
```{R}
t=final%>%
  filter(meanpal<300)%>%
  group_by(hgnc)%>%distinct_at(vars(ALE),.keep_all = T)%>%summarise(ALE_count=n())%>%filter(ALE_count%in%c(2,3))
t%>%group_by(ALE_count)%>%summarise(n())
ALE=final%>%
  filter(meanpal<300)%>%
  inner_join(t,by="hgnc")%>%arrange(hgnc)


custom4=ALE%>%ungroup()%>%group_by(hgnc)%>%mutate(newratio=readcount/sum(readcount))

```

2. Caclualte three parameters

RNA per ALE
weighted mean PAL
weighted mean UTR length

```{r}
ALE.pal=custom4%>%
  #inner_join(newCPSid,by="hgnc:ALE:tandem")%>%
  ungroup()%>%
  group_by(hgnc,ALE)%>%
  summarise(weightpallen=weighted.mean(meanpal,newratio))%>%ungroup()
ALE.sum=custom4%>%group_by(hgnc,ALE)%>%summarise(sum=sum(newratio))%>%ungroup()
ALE.utrlen=custom4%>%group_by(hgnc,ALE)%>%
  summarise(weightutrlen=weighted.mean(utrlen,newratio))
ALE.pal=ALE.pal%>%unite("newid",c(hgnc,ALE),sep=":")%>%inner_join(ALE.sum%>%unite("newid",c(hgnc,ALE),sep=":"),by="newid")%>%
  inner_join(ALE.utrlen%>%unite("newid",c(hgnc,ALE),sep=":"),by="newid")%>%
  separate(newid,c("hgnc","ALE"),sep=":")
```


```{R}
pdf("ALE_stat.pdf",width=4,height=3)
ALE.stat=ALE.pal%>%group_by(hgnc)%>%summarise(ALE_count=n())%>%ungroup()
ALE.stat
ggplot(ALE.stat,aes(ALE_count))+geom_bar(stat="count")+theme_bw()+xlab("Number of ALE isoforms")+ylab("Number of genes")
dev.off()
#ALE.tmp=ALE.pal%>%inner_join(t,by="hgnc")%>%mutate(ALE=as.numeric(ALE))%>%group_by(hgnc)%>%arrange(hgnc,ALE)%>%mutate(newALE=row_number())
```





4. Gene ontologies for genes with multiple ALEs
Gene ontology for genes with multiple ALE isoform expression.
```{R}
ALE.stat=ALE.pal%>%group_by(hgnc)%>%summarise(ALE_count=n())#%>%
t=ALE.stat%>%filter(ALE_count>1)
write.table(t%>%select(hgnc),"List_625geneswithmultipleALEs.txt",col.names = F,row.names = F,quote = F)
t=final%>%
  filter(meanpal<300)%>%
  group_by(hgnc)%>%distinct_at(vars(ALE),.keep_all = T)%>%summarise(ALE_count=n())
nrow(t) 
write.table(t%>%select(hgnc),"List_background6174genes.txt",col.names = F,row.names = F,quote = F)
```

5. Pairwise comparison (n=2)
```{R}
t=ALE.stat%>%filter(ALE_count==2)

ale.pal.len=ALE.pal%>%ungroup()%>%inner_join(t,by="hgnc")%>%
  mutate(ALE=as.numeric(ALE))%>%group_by(hgnc)%>%arrange(hgnc,ALE)%>%mutate(newALE=row_number())%>%ungroup()%>%select(hgnc,newALE,weightpallen)%>%filter(weightpallen<200)%>%
  #select(-ALE,-ALEpos,-ALE_count.x,-ALE_count.y,-sum,-weightutrlen)%>%
   #mutate_at(scale, .vars = vars(weightpallen))%>%
  spread(newALE,weightpallen)%>%mutate(diff=`2`-`1`)%>%
  mutate(type="Poly(A) tail length")
t.test(ale.pal.len$diff) #p-value = 8.451e-06   #less than 200 nt :p-value = 0.02971


ale.utr.len=ALE.pal%>%ungroup()%>%inner_join(t,by="hgnc")%>%filter(weightpallen<200)%>%
  mutate(ALE=as.numeric(ALE))%>%group_by(hgnc)%>%arrange(hgnc,ALE)%>%mutate(newALE=row_number())%>%ungroup()%>%select(hgnc,newALE,weightutrlen)%>%
  spread(newALE,weightutrlen)%>%mutate(diff=`2`-`1`)%>%mutate(type="3´UTR length")%>%ungroup()

t.test(ale.utr.len$diff,alternative = "two.sided") #p-value = 0.04973;less than 200 nt :p-value = 0.6151

ale.rna=ALE.pal%>%ungroup()%>%inner_join(t,by="hgnc")%>%filter(weightpallen<200)%>%
  mutate(ALE=as.numeric(ALE))%>%group_by(hgnc)%>%arrange(hgnc,ALE)%>%mutate(newALE=row_number())%>%ungroup()%>%select(hgnc,newALE,sum)%>%
  spread(newALE,sum)%>%mutate(diff=`2`-`1`)%>%mutate(type="RNA abundance")%>%ungroup()

t.test(ale.rna$diff) #p-value = 2.57e-10  #lessthan200 :p-value = 4.515e-07


p1<-ggplot(ale.utr.len,aes(x=type,y=diff,fill=type))+
  geom_boxplot(notch = T)+
  #geom_dotplot(alpha=1,binaxis = "y", stackgroups = TRUE, binwidth =1,dotsize =2, stackdir="center")
  theme_bw()+theme(legend.position = "none")+xlab("")+
scale_fill_manual(values=yk.col(9)[c(9)])+theme(axis.title.x = element_text(size = 10, # also adjust text size if needed
                                    margin = margin(t = 10, r = 0, b = 0, l = 0,
                                                    unit = "mm")))+ylim(c(-3000,3000))+
 ylab(expression(Delta*"3´UTR length (distal- proximal)"))+xlab("")


p2<-ggplot(ale.pal.len,aes(x=type,y=diff,fill=type))+
  geom_boxplot(notch = T)+
  #geom_dotplot(alpha=1,binaxis = "y", stackgroups = TRUE, binwidth =1,dotsize =2, stackdir="center")
  theme_bw()+theme(legend.position = "none")+xlab("")+
scale_fill_manual(values=yk.col(9)[c(7)])+theme(axis.title.x = element_text(size = 10, # also adjust text size if needed
                                    margin = margin(t = 10, r = 0, b = 0, l = 0,
                                                    unit = "mm")))+ylim(c(-50,50))+
 ylab(expression(Delta*"PAL (distal- proximal)"))


p3<-ggplot(ale.rna,aes(x=type,y=diff,fill=type))+
  geom_boxplot(notch = T)+

  #geom_dotplot(alpha=1,binaxis = "y", stackgroups = TRUE, binwidth =1,dotsize =2, stackdir="center")
  theme_bw()+theme(legend.position = "none")+xlab("")+
scale_fill_manual(values=yk.col(9)[c(3)])+theme(axis.title.x = element_text(size = 10, # also adjust text size if needed
                                    margin = margin(t = 10, r = 0, b = 0, l = 0,
                                                    unit = "mm")))+#ylim(c(-2,2))+
 ylab(expression(Delta*"RNA (distal- proximal)"))
pdf("Fig2I_ALE.summary.pdf",width=6,height=4)
grid.arrange(
  p1,
  p2,
  p3,
  nrow = 1,
  top= "Pairwise comparison of ALE 3´UTR pairs"
  )

dev.off()





```

n=3 ALE
```{R}
t=ALE.stat%>%filter(ALE_count==3)
nrow(t)#59
ale.pal.len=ALE.pal%>%ungroup()%>%inner_join(t,by="hgnc")%>%filter(weightpallen<200)%>%
  mutate(ALE=as.numeric(ALE))%>%group_by(hgnc)%>%arrange(hgnc,ALE)%>%mutate(newALE=row_number())%>%ungroup()%>%select(hgnc,newALE,weightpallen)%>%
  #select(-ALE,-ALEpos,-ALE_count.x,-ALE_count.y,-sum,-weightutrlen)%>%
   #mutate_at(scale, .vars = vars(weightpallen))%>%
  spread(newALE,weightpallen)%>%mutate(diff=`3`-`1`)%>%
  mutate(type="Poly(A) tail length")
t.test(ale.pal.len$diff, alternative = "two.sided") #p-value = 0.002816 #lessthan200:p-value = 0.6327

t.test
ale.utr.len=ALE.pal%>%ungroup()%>%inner_join(t,by="hgnc")%>%filter(weightpallen<200)%>%
  mutate(ALE=as.numeric(ALE))%>%group_by(hgnc)%>%arrange(hgnc,ALE)%>%mutate(newALE=row_number())%>%ungroup()%>%select(hgnc,newALE,weightutrlen)%>%
  spread(newALE,weightutrlen)%>%mutate(diff=`3`-`1`)%>%mutate(type="3´UTR length")%>%ungroup()

t.test(ale.utr.len$diff) #p-value = 0.5379 #lessthan200 p-value = 0.7276

ale.rna=ALE.pal%>%ungroup()%>%inner_join(t,by="hgnc")%>%filter(weightpallen<200)%>%
  mutate(ALE=as.numeric(ALE))%>%group_by(hgnc)%>%arrange(hgnc,ALE)%>%mutate(newALE=row_number())%>%ungroup()%>%select(hgnc,newALE,sum)%>%
  spread(newALE,sum)%>%mutate(diff=`3`-`1`)%>%mutate(type="RNA abundance")%>%ungroup()

t.test(ale.rna$diff) #p-value = 0.002758 #less than 200 p-value = 0.1619

p4<-ggplot(ale.utr.len,aes(x=type,y=diff,fill=type))+
  geom_boxplot(notch = T)+
  #geom_dotplot(alpha=1,binaxis = "y", stackgroups = TRUE, binwidth =1,dotsize =2, stackdir="center")
  theme_bw()+theme(legend.position = "none")+xlab("")+
scale_fill_manual(values=yk.col(9)[c(9)])+theme(axis.title.x = element_text(size = 10, # also adjust text size if needed
                                    margin = margin(t = 10, r = 0, b = 0, l = 0,
                                                    unit = "mm")))+#ylim(c(-3000,3000))+
 ylab(expression(Delta*"3´UTR length (distal- proximal)"))+xlab("")


p5<-ggplot(ale.pal.len,aes(x=type,y=diff,fill=type))+
  geom_boxplot(notch = T)+
  #geom_dotplot(alpha=1,binaxis = "y", stackgroups = TRUE, binwidth =1,dotsize =2, stackdir="center")
  theme_bw()+theme(legend.position = "none")+xlab("")+
scale_fill_manual(values=yk.col(9)[c(7)])+theme(axis.title.x = element_text(size = 10, # also adjust text size if needed
                                    margin = margin(t = 10, r = 0, b = 0, l = 0,
                                                    unit = "mm")))+#ylim(c(-2,2))+
 ylab(expression(Delta*"PAL (distal- proximal)"))


p6<-ggplot(ale.rna,aes(x=type,y=diff,fill=type))+
  geom_boxplot(notch = T)+

  #geom_dotplot(alpha=1,binaxis = "y", stackgroups = TRUE, binwidth =1,dotsize =2, stackdir="center")
  theme_bw()+theme(legend.position = "none")+xlab("")+
scale_fill_manual(values=yk.col(9)[c(3)])+theme(axis.title.x = element_text(size = 10, # also adjust text size if needed
                                    margin = margin(t = 10, r = 0, b = 0, l = 0,
                                                    unit = "mm")))+#ylim(c(-2,2))+
 ylab(expression(Delta*"RNA (distal- proximal)"))

pdf("ALE.summary2.pdf",width=5,height=6)
grid.arrange(
  p1,
  p2,
  p3,
  p4,
  p5,
  p6,
  nrow = 2,
  top= "Pairwise comparison of ALE pairs"
  )

dev.off()
```
#Dotplot
  
```{R}
t=ALE.stat%>%filter(ALE_count %in% c(2,3))
ALE.pal2=ALE.pal%>%ungroup()%>%inner_join(t,by="hgnc")%>%
  mutate(ALE=as.numeric(ALE))%>%group_by(hgnc)%>%arrange(hgnc,ALE)%>%mutate(newALE=row_number())%>%ungroup()%>%select(hgnc,newALE,sum,weightpallen,ALE_count)
pdf("Fig2H_CustomTEDseq_ALE_PAL_dotplot.pdf",width=6,height=3)
ggplot(ALE.pal2,aes(x=as.factor(newALE),y=weightpallen))+geom_quasirandom(method = "smiley",aes(colour=(sum)),size=0.4) + 
  #geom_beeswarm()+
  scale_colour_gradientn(colours = yk.col(1000),name="RNA expression level")+
  #geom_dotplot(aes(fill=as.factor(ratio),colour=as.factor(ratio)),alpha=1,binaxis = "y", stackdir="center",stackgroups = TRUE, binwidth =1,dotsize =2)+
  facet_wrap(.~ALE_count,ncol=5,scale="free_x")+
  theme_bw()+xlab("")+
  
theme(axis.title.x = element_text(size = 10, # also adjust text size if needed
                                    margin = margin(t = 10, r = 0, b = 0, l = 0,
                                                    unit = "mm")))+
xlab((expression("ALE index (proximal"*symbol('\256')*"distal)")))+ylab("Mean poly(A) tail length (nt)")
dev.off()

#statistics
t=ALE.stat%>%filter(ALE_count %in% c(2))
ALE.pal3=ALE.pal%>%ungroup()%>%inner_join(t,by="hgnc")%>%
  mutate(ALE=as.numeric(ALE))%>%group_by(hgnc)%>%arrange(hgnc,ALE)%>%mutate(newALE=row_number())%>%ungroup()%>%select(hgnc,newALE,sum,weightpallen,ALE_count)%>%filter(weightpallen<=180)

t.test(ALE.pal3$weightpallen[ALE.pal3$newALE==1],ALE.pal3$weightpallen[ALE.pal3$newALE==2], alternative = "two.sided")

t=ALE.stat%>%filter(ALE_count %in% c(3))
ALE.pal3=ALE.pal%>%ungroup()%>%inner_join(t,by="hgnc")%>%
  mutate(ALE=as.numeric(ALE))%>%group_by(hgnc)%>%arrange(hgnc,ALE)%>%mutate(newALE=row_number())%>%ungroup()%>%select(hgnc,newALE,sum,weightpallen,ALE_count)%>%filter(weightpallen<=200)


t.test(ALE.pal3$weightpallen[ALE.pal3$newALE==1],ALE.pal3$weightpallen[ALE.pal3$newALE==3], alternative = "two.sided")
```
Extremely long poly(A)
```{R}
ALE.pal2
over200=ALE.pal2%>%mutate(group=ifelse(weightpallen>=200,"extreme long","≤200 nt"))%>%
  group_by(ALE_count,newALE,group)%>%
  summarise(RNA=mean(sum))


df.summary2 <- ALE.pal2%>%mutate(PALgroup=ifelse(weightpallen>=200,"extreme long","≤200 nt"))%>%
  group_by(ALE_count,newALE,PALgroup)%>%
  summarise(RNA=mean(sum), 
            sd = sd(sum),
            n=n())

pdf("FigS2_ALE_PALgt200_RNAabundance.pdf",width=6,height=3 )
ggplot(df.summary2%>%ungroup()%>%mutate(Tandem=as.factor(newALE)), aes(as.factor(newALE), RNA)) +
  geom_col(aes(fill = PALgroup), position = position_dodge(0.8), width = 0.7)+
  geom_errorbar(
    aes(ymin = RNA, ymax = RNA+sd, group = PALgroup),
    width = 0.2, position = position_dodge(0.8)
    )+facet_grid(.~ALE_count,scale="free",space="free_x")+theme_bw()+scale_fill_manual(name="Poly(A) tail length",values=yk.col(9)[c(7,3)])+ylim(c(0,1.1))+xlab((expression("ALE index (proximal"*symbol('\256')*"distal)")))+ylab("Mean of normalized RNA abundance")
dev.off()





pdf("ALE_mean RNA abundance of extremely long poly(A).pdf", width=5,height=3)
ggplot(over200%>%filter(ALE_count %in% c(2,3)),aes(y=RNA,x=group,fill=group))+facet_wrap(.~ALE_count,scale="free_x")+ geom_bar(stat="identity",position="dodge")+scale_fill_manual(name="Poly(A) tail length",values=yk.col(9)[c(1,2)])+theme_bw()+
  #xlab((expression("tandem 3´UTRs (proximal"*symbol('\256')*"distal)")))+
  ylab("Mean of normalized RNA abundance")+xlab("")
dev.off()  
write.table(df.summary2,"Summary_GT200count.ALEgroup.txt",col.names = T,row.names = F,sep="\t",quote=F)

over200=ALE.pal2%>%group_by(ALE_count,newALE)%>%filter(weightpallen>=200)%>%summarise(number=n())

over200 



pdf("FicS2_Number of ALE with PAL greater than 200.pdf",width=4,height=3)
ggplot(over200%>%filter(ALE_count %in% c(2,3)),aes(x=as.factor(newALE),y=number,fill=as.factor(newALE)))+facet_wrap(.~ALE_count,ncol=4,scale="free_x")+ geom_bar(stat="identity")+scale_fill_manual(values=yk.col(9)[c(9,7,5,4)])+theme_bw()+xlab((expression("ALE index")))+ylab("Number of ALEs \n with PAL ≥200 nt")+theme(legend.position = "none")
#+scale_x_discrete(limits=c("1", "2","3","4"))
dev.off()

```
  
 