{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ef9ea9a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#packageurl=\"https://cran.r-project.org/src/contrib/Archive/rphast/rphast_1.6.9.tar.gz\"\n",
    "#install.packages(packageurl, repos=NULL, type=\"source\",INSTALL_opts = '--no-lock')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "60e91e39",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(rphast)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f9fc4e2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "also installing the dependency 'snow'\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#install.packages(\"snowfall\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "407b848c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#packageurl <- \"https://cran.r-project.org/src/contrib/Archive/rtfbs/rtfbs_0.3.15.tar.gz\"\n",
    "#install.packages(packageurl, repos=NULL, type=\"source\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1db8a7a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: snow\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(rtfbs)\n",
    "library(snowfall)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ff1d86b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "also installing the dependencies 'credentials', 'zip', 'gitcreds', 'ini', 'gert', 'gh', 'whisker', 'xopen', 'brew', 'usethis', 'memoise', 'pkgbuild', 'rcmdcheck', 'remotes', 'roxygen2', 'rversions', 'sessioninfo'\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#install.packages(\"devtools\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6f6997bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: usethis\n",
      "\n"
     ]
    }
   ],
   "source": [
    "require(snowfall)\n",
    "require(rtfbs)\n",
    "\n",
    "#devtools::install_github(\"hadley/devtools\")\n",
    "#assuming it is not already installed\n",
    "library(grid)\n",
    "library(devtools)\n",
    "\n",
    "#install_github('andreacirilloac/updateR')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a8ecc55b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading GitHub repo Danko-Lab/rtfbs_db@HEAD\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sm           (NA -> 2.2-5.7) [CRAN]\n",
      "jpeg         (NA -> 0.1-9  ) [CRAN]\n",
      "png          (NA -> 0.1-7  ) [CRAN]\n",
      "vioplot      (NA -> 0.3.7  ) [CRAN]\n",
      "apcluster    (NA -> 1.4.8  ) [CRAN]\n",
      "latticeExtra (NA -> 0.6-29 ) [CRAN]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Skipping 2 packages not available: rtfbs, rphast\n",
      "\n",
      "Installing 6 packages: sm, jpeg, png, vioplot, apcluster, latticeExtra\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mv\u001b[39m  \u001b[90mchecking for file '/private/var/folders/tt/c6wcdvv17lg0nn6yt9fj7dm80000gn/T/RtmpCALie4/remotes417370225113/Danko-Lab-rtfbs_db-a29f73b/rtfbsdb/DESCRIPTION'\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m-\u001b[39m\u001b[90m  \u001b[39m\u001b[90mpreparing 'rtfbsdb':\u001b[39m\u001b[36m\u001b[36m (363ms)\u001b[36m\u001b[39m\n",
      "\u001b[32mv\u001b[39m  \u001b[90mchecking DESCRIPTION meta-information\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m-\u001b[39m\u001b[90m  \u001b[39m\u001b[90mexcluding invalid files\u001b[39m\u001b[36m\u001b[39m\n",
      "   Subdirectory 'man' contains invalid file names:\n",
      "     'tfbs.getDistanceMatrix.Rdxx'\n",
      "\u001b[90m-\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for LF line-endings in source and make files and shell scripts\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m-\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for empty or unneeded directories\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m-\u001b[39m\u001b[90m  \u001b[39m\u001b[90mbuilding 'rtfbsdb_0.4.5.tar.gz'\u001b[39m\u001b[36m\u001b[39m\n",
      "   \n",
      "\r"
     ]
    }
   ],
   "source": [
    "#install.packages(\"ggrepel\")\n",
    "#install_github(\"Danko-Lab/rtfbs_db/rtfbsdb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "36bf893a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: ggplot2\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>bedtools</dt><dd>'/usr/local/bin/bedtools'</dd><dt>bedops</dt><dd>'/usr/local/bin/bedops'</dd><dt>twobitToFa</dt><dd>'/usr/local/bin/twobitToFa'</dd><dt>awk</dt><dd>'/usr/local/bin/awk'</dd><dt>sort-bed</dt><dd>'/usr/local/bin/sort-bed'</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[bedtools] '/usr/local/bin/bedtools'\n",
       "\\item[bedops] '/usr/local/bin/bedops'\n",
       "\\item[twobitToFa] '/usr/local/bin/twobitToFa'\n",
       "\\item[awk] '/usr/local/bin/awk'\n",
       "\\item[sort-bed] '/usr/local/bin/sort-bed'\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "bedtools\n",
       ":   '/usr/local/bin/bedtools'bedops\n",
       ":   '/usr/local/bin/bedops'twobitToFa\n",
       ":   '/usr/local/bin/twobitToFa'awk\n",
       ":   '/usr/local/bin/awk'sort-bed\n",
       ":   '/usr/local/bin/sort-bed'\n",
       "\n"
      ],
      "text/plain": [
       "                   bedtools                      bedops \n",
       "  \"/usr/local/bin/bedtools\"     \"/usr/local/bin/bedops\" \n",
       "                 twobitToFa                         awk \n",
       "\"/usr/local/bin/twobitToFa\"        \"/usr/local/bin/awk\" \n",
       "                   sort-bed \n",
       "  \"/usr/local/bin/sort-bed\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#\n",
    "library(cluster)\n",
    "library(updateR)\n",
    "library(ggrepel)\n",
    "#updateR(admin_password = 'Admin user password')\n",
    "#Install rtfbsdb\n",
    "#install.packages(\"rtfbsdb\")\n",
    "library(rtfbsdb)\n",
    "#Sys.setenv( PATH = paste(Sys.getenv('PATH'), '/Users/yeonuikwak', sep=':') )\n",
    "Sys.which(c('bedtools','bedops','twobitToFa','awk','sort-bed'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e547fd43",
   "metadata": {},
   "outputs": [],
   "source": [
    "setClass(\"rbpbs.db\",\n",
    "\trepresentation(\n",
    "\t\tspecies = \"character\"\n",
    "\t)\n",
    ")\n",
    "\n",
    "setClass(\"CisRBP.db\",\n",
    "\trepresentation(\n",
    "\t\tzip.url     = \"character\",  ## http url for data download\n",
    "\t\tzip.file    = \"character\",  ## zip file including pwm files and TF_information.txt\n",
    "\t\tzip.date    = \"character\",  ## download date\n",
    "\t\tfile.rbpinfo = \"character\"   ## \"TF_Information.txt\"\n",
    "\t), contains = \"rbpbs.db\"\n",
    ")\n",
    "\n",
    "setMethod(\"show\", \"CisRBP.db\", function(object){\n",
    "\tcat(\"Species: \", object@species, \"\\n\");\n",
    "\tcat(\"Data Source: \", object@zip.url, \"\\n\");\n",
    "\tcat(\"Zip file: \", object@zip.file, \"\\n\");\n",
    "\tcat(\"Download date: \", object@zip.date, \"\\n\");\n",
    "\tcat(\"RBP information: \", object@file.rbpinfo, \"\\n\");\n",
    "\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a8b4b878",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Subfunctions.\n",
    "#' Select the motif table for CisBP.create and CisBP.group function.\n",
    "#'\n",
    "#' Two RBP information files in zip file\n",
    "#'\n",
    "#' 1: RBP_Information.txt : (direct motifs) or (no direct but inferred motifs with 90%)\n",
    "#' 2: RBP_Information_all_motifs.txt: (direct motifs) and (inferred motifs above the threshold)\n",
    "#'\n",
    "#' @param cisrbp.db: CisRBP.db object\n",
    "#' @param rbp.information.type: 1,or 2 indicate the index of above files.\n",
    "#'\n",
    "#' @return: temporary file;\n",
    "CisRBP.active_motif_info<-function( cisrbp.db, rbp.information.type=1 )\n",
    "{\n",
    "\tmotif_infos <- c( \"RBP_Information.txt\", \"RBP_Information_all_motifs.txt\");\n",
    "\n",
    "\tif(rbp.information.type<1 || rbp.information.type>2)\n",
    "\t{\n",
    "\t\tcat(\"! The meta data can be stored in TF_Information.txt(1), TF_Information_all_motifs.txt(2) or TF_Information_all_motifs_plus.txt(3).\");\n",
    "\t\treturn(cisrbp.db@file.rbpinfo);\n",
    "\t}\n",
    "\n",
    "\tfile.rbpinfo <- motif_infos[ rbp.information.type ];\n",
    "\n",
    "\ttmp.dir <- tempdir();\n",
    "\tr.file <- unzip(cisrbp.db@zip.file, c(file.rbpinfo), exdir=tmp.dir );\n",
    "\tif(length(r.file)<1)\n",
    "\t{\n",
    "\t\tcat(\"! The meta file (\", file.rbpinfo ,\") can not be found in the zip file.\");\n",
    "\t\treturn( cisrbp.db@file.rbpinfo );\n",
    "\t}\n",
    "\n",
    "\treturn( paste(tmp.dir, file.rbpinfo, sep=\"/\") );\n",
    "}\n",
    "\n",
    "#function rbpbs and rbp_creatFromCisRBP\n",
    "rbpbs <- function(filenames=NULL, names=NULL, species=\"Homo_sapiens\", rbp_info=NULL, rbp_missing=NULL, ...)\n",
    "{\n",
    "\tstopifnot( length(names) == length(filenames) );\n",
    "\n",
    "\tpwms <- list();\n",
    "\tidx.pwm <- c();\n",
    "\tmgisymbols <- c();\n",
    "\tif( !is.null(filenames) )\n",
    "\t{\n",
    "\t\tfor(i in 1:length(filenames)) {\n",
    "\t\t\tpwms[[i]] <- read.motif(filenames[i], ...)\n",
    "\t\t}\n",
    "\n",
    "\t\tidx.pwm <- which( unlist( lapply( pwms, function(x) !is.null(x) ) ) );\n",
    "\t\tif( length(idx.pwm)==0 )\n",
    "\t\t\treturn( NULL );\n",
    "\n",
    "\t\tpwm        = pwms[ idx.pwm ];\n",
    "\t\tfilename   = filenames[ idx.pwm ];\n",
    "\t\tmgisymbols = names[ idx.pwm ];\n",
    "\t}\n",
    "\telse\n",
    "\t\tfilenames <- c();\n",
    "\n",
    "\tif( is.null(rbp_info) )\n",
    "\t\trbp_info  <- data.frame(Motif_ID = mgisymbols, RBP_Name = paste(\"RBP_\", mgisymbols, sep=\"\"))\n",
    "\telse\n",
    "\t\trbp_info  <- rbp_info[idx.pwm,,drop=F];\n",
    "\n",
    "\tif( is.null(rbp_missing) )\n",
    "\t\trbp_missing  <- as.data.frame(NULL)\n",
    "\n",
    "\tnew(\"rbpbs\",\n",
    "\t\tspecies        = species,\n",
    "\t\tnrbps           = as.integer(length(filenames)),\n",
    "\t\tpwm            = pwms,\n",
    "\t\tfilename       = as.character(filenames),\n",
    "\t\tmgisymbols     = as.character(mgisymbols),\n",
    "\t\trbp_info        = rbp_info,\n",
    "\t\trbp_missing     = rbp_missing,\n",
    "\t\tdistancematrix = matrix(, nrow=0, ncol=0),\n",
    "\t\tcluster        = matrix(, nrow=0, ncol=0),\n",
    "\t\texpressionlevel= as.data.frame(NULL) )\n",
    "}\n",
    "#set class rbpbs\n",
    "setClass(\"rbpbs\",\n",
    "\trepresentation(\n",
    "\t\tspecies        = \"character\",    ## such as, Homo_sapiens or Mus_musculus.\n",
    "\t\tnrbps           = \"integer\",      ## Number of motifs in matrix.\n",
    "\t\tmgisymbols     = \"character\",    ## Unique gene symbols for TF i.\n",
    "\t\tfilename       = \"character\",    ## The filename of the PWM.\n",
    "\t\tpwm            = \"list\",         ## PWM for TF i.\n",
    "\t\trbp_info        = \"data.frame\",   ## TF information for PWMs, it maybe different with motif database\n",
    "\t\trbp_missing     = \"data.frame\",   ## Missing TF information for PWMs, it maybe different with motif database\n",
    "\t\tdistancematrix = \"matrix\",       ## Distance matrix between motifs\n",
    "\t\tcluster        = \"matrix\",       ## The number of the cluster that this TF is included in.\n",
    "\t\texpressionlevel= \"data.frame\"    ## Expression level.\n",
    "\t\t#TFID           = \"character\",   ## A non-unique ID for TF i.\n",
    "\t\t#usemotifs     = \"integer\",      ## The indices of TFs to be used for analyses, such as scanning DNA sequences.\n",
    "  ),\n",
    ")\n",
    "\n",
    "setMethod(\"show\", \"rbpbs\", function(object)\n",
    "{\n",
    "\tcat(\"Species: \", object@species, \"\\n\");\n",
    "\tcat(\"RBP number: \", object@nrbps, \"\\n\");\n",
    "\n",
    "\tif( NROW(object@distancematrix)==0 )\n",
    "\t\tcat( \"Distance Matrix:  NULL\\n\" )\n",
    "\telse\n",
    "\t\tcat( \"Distance Matrix:  [\", NROW(object@distancematrix), \",\", NCOL(object@distancematrix), \"]\\n\" );\n",
    "\n",
    "\tif( NROW(object@cluster)==0 )\n",
    "\t\tcat( \"Cluster Matrix:  NULL\\n\" )\n",
    "\telse\n",
    "\t\tcat( \"Cluster Matrix:  [\", NROW(object@cluster), \",\", NCOL(object@cluster), \"]\\n\" );\n",
    "\n",
    "\tif(NROW(object@expressionlevel)==0)\n",
    "\t\tcat( \"Expression:  NULL\\n\" )\n",
    "\telse\n",
    "\t\tcat( \"Expression:  [\", NROW(object@expressionlevel), \",\", NCOL(object@expressionlevel), \"]\\n\" );\n",
    "\n",
    "\tdf <- NULL;\n",
    "\tif( NROW(object@rbp_info) >0 )\n",
    "\t{\n",
    "\t\tdf.list <- lapply ( c(\"Motif_ID\", \"RBP_Name\", \"Family_Name\", \"DBID\", \"Motif_Type\"), function(x){\n",
    "\t\t\t\t\tif(x %in% colnames(object@rbp_info) ) return(as.character(object@rbp_info[,x])) else NA; } );\n",
    "\n",
    "\t\tdf1 <- do.call( cbind, df.list );\n",
    "\t\tcolnames(df1) <- c(\"Motif_ID\", \"RBP_Name\", \"Family_Name\", \"DBID\", \"Motif_Type\");\n",
    "\n",
    "\t\tdf <- data.frame(df1, filename=basename(object@filename), stringsAsFactors=FALSE);\n",
    "\t}\n",
    "\telse\n",
    "\t\tdf <- data.frame(Motif_ID=object@mgisymbols, filename=basename(object@filename), stringsAsFactors=FALSE);\n",
    "\n",
    "\tif( NROW(object@expressionlevel) > 0 )\n",
    "\t\tdf <- data.frame(df, p.pois = object@expressionlevel[,c(\"p.pois\")] );\n",
    "\n",
    "\tcat(\"\\nPartial list of TFs\\n\");\n",
    "\tshow(head(df, 20));\n",
    "})\n",
    "                                         #' Find the subset by querying the motif table\n",
    "#'\n",
    "#' @param cisbp.db: cisbp.db object\n",
    "#' @param tf_name: string, the query value for tf_name\n",
    "#' @param tf_status: string, the query value for tf_status\n",
    "#' @param family_name: string, the query value for family_name\n",
    "#' @param motif_type: string, the query value for motif_type\n",
    "#' @param msource_id: string, the query value for msource_id\n",
    "#' @param tf.information.type: 1,2 or 3 indicate which motif file will be used.\n",
    "#'\n",
    "#' @return: NULL or tfbs object;\n",
    "#' cisrbp.db=db\n",
    "rbpbs_createFromCisRBP <- function ( cisrbp.db,\n",
    "\t\t\t\t\tmotif_id    = NULL,\n",
    "\t\t\t\t\trbp_name     = NULL,\n",
    "\t\t\t\t\trbp_status   = NULL,\n",
    "\t\t\t\t\tfamily_name = NULL,\n",
    "\t\t\t\t\tmotif_type  = NULL,\n",
    "\t\t\t\t\tmsource_id  = NULL,\n",
    "\t\t\t\t\trbp.information.type = 1 )\n",
    "{\n",
    "\tcisrbp.db@file.rbpinfo <- CisRBP.active_motif_info( cisrbp.db, rbp.information.type );\n",
    "\n",
    "\ttbm <- try( read.csv(cisrbp.db@file.rbpinfo, header=T, sep=\"\\t\") );\n",
    "\tif( class(tbm) == \"try-error\" )\n",
    "\t\tstop(\"Can not open RBP information file:\", cisrbp.db@file.rbpinfo, \".\\n\" );\n",
    "\n",
    "\ttbm_f <- c();\n",
    "\n",
    "\tif(!is.null(motif_id))   tbm_f <- c( tbm_f, paste( \"tbm$Motif_ID=='\", motif_id, \"'\", sep=\"\") );\n",
    "\tif(!is.null(rbp_name))    tbm_f <- c( tbm_f, paste( \"tbm$RBP_Name=='\", rbp_name, \"'\", sep=\"\") );\n",
    "\tif(!is.null(rbp_status))  tbm_f <- c( tbm_f, paste( \"tbm$RBP_Status=='\", rbp_status, \"'\", sep=\"\") );\n",
    "\tif(!is.null(family_name))tbm_f <- c( tbm_f, paste( \"tbm$Family_Name=='\", family_name, \"'\", sep=\"\") );\n",
    "\tif(!is.null(motif_type)) tbm_f <- c( tbm_f, paste( \"tbm$Motif_Type=='\", motif_type, \"'\", sep=\"\") );\n",
    "\tif(!is.null(msource_id)) tbm_f <- c( tbm_f, paste( \"tbm$MSource_Identifier=='\", msource_id, \"'\", sep=\"\") );\n",
    "\n",
    "\tif(length(tbm_f)>0)\n",
    "\t{\n",
    "\t\ttbm_f_all <- paste(tbm_f, collapse=\" & \" );\n",
    "\t\tnidx <-  eval(parse(text=paste(\"which(\", tbm_f_all, \")\")));\n",
    "\t\tif(length(nidx)<1) return( NULL );\n",
    "\t}\n",
    "\telse\n",
    "\t{\n",
    "\t\tnidx <- c(1:NROW(tbm));\n",
    "\t\ttbm_f_all <- \"All\";\n",
    "\t}\n",
    "\n",
    "\ttbm <- tbm[nidx,, drop=F];\n",
    "\tcat( \" \", NROW(tbm), \"PWM(s) are defined in the CisRBP dataset.\\n\");\n",
    "\n",
    "\ttmp.dir <- tempdir();\n",
    "\tpwm.files <- c();\n",
    "\tnidx.motif <- c();\n",
    "\tnames <- c();\n",
    "\n",
    "\terr.missing <- 0;\n",
    "\terr.empty <- 0;\n",
    "\n",
    "\tfor (i in 1:NROW(tbm) )\n",
    "\t{\n",
    "\t\tmotif_id <- as.character(tbm$Motif_ID[i]);\n",
    "\n",
    "\t\tif( as.character(motif_id)==\".\")\n",
    "\t\t{\n",
    "\t\t\t# cat(\"! No ID for this motif(.).\\n\");\n",
    "\t\t\terr.missing <- err.missing + 1;\n",
    "\t\t\tnext;\n",
    "\t\t}\n",
    "\n",
    "\t\tpwm.file <- paste( \"pwms_all_motifs/\", motif_id, \".txt\", sep=\"\");\n",
    "\t\tr.file <- unzip( cisrbp.db@zip.file, c(pwm.file), exdir=tmp.dir  );\n",
    "\t\tif(length(r.file)<1)\n",
    "\t\t{\n",
    "\t\t\t# cat(\"! Can not find PWM file for motif ID=\", motif_id, \".\\n\" );\n",
    "\t\t\terr.missing <- err.missing + 1;\n",
    "\t\t\tnext;\n",
    "\t\t}\n",
    "\n",
    "\t\ttb <- try( read.table(paste(tmp.dir, pwm.file, sep=\"/\"), header=T, sep=\"\\t\", row.names=1), TRUE );\n",
    "\t\tif(class(tb)==\"try-error\")\n",
    "\t\t{\n",
    "\t\t\t# cat(\"! Can not find PWM file for motif ID=\", motif_id, \".\\n\" );\n",
    "\t\t\terr.missing <- err.missing + 1;\n",
    "\t\t\tnext;\n",
    "\t\t}\n",
    "\n",
    "\t\tif(NROW(tb)==0)\n",
    "\t\t{\n",
    "\t\t\t# cat(\"! No A C G T values in the PWM file for motif ID=\", motif_id, \".\\n\" );\n",
    "\t\t\terr.empty <- err.empty + 1;\n",
    "\t\t\tnext;\n",
    "\t\t}\n",
    "\n",
    "\t\tpwm.files <- c( pwm.files, paste(tmp.dir, pwm.file, sep=\"/\") );\n",
    "\t\tnidx.motif<- c( nidx.motif, i );\n",
    "\t\tnames     <- c( names, motif_id);\n",
    "\t}\n",
    "\n",
    "\tRBP_info <- NULL;\n",
    "\tRBP_missing <- NULL;\n",
    "\tif(length(nidx.motif)>0)\n",
    "\t{\n",
    "\t\tRBP_info <- tbm[ nidx.motif, , drop=F];\n",
    "\t\tif(NROW(RBP_info)>0) rownames(RBP_info)<- c(1:NROW(RBP_info));\n",
    "\t\tRBP_missing <- tbm[ -nidx.motif, , drop=F];\n",
    "\t\tif(NROW(RBP_missing)>0) rownames(RBP_missing)<- c(1:NROW(RBP_missing));\n",
    "\t}\n",
    "\n",
    "\tif( err.empty + err.missing > 0 )\n",
    "\t\tcat(\"!\", err.empty + err.missing, \"PWM file(s) are failed to be loaded ( Missing PWMs :\", err.missing, \", Empty PWMs :\", err.empty, \").\\n\");\n",
    "\n",
    "\tif(length(pwm.files)==0)\n",
    "\t{\n",
    "\t\tcat(\"! No PWM file to create a tfbs object.\\n\");\n",
    "\t\treturn(NULL);\n",
    "\t}\n",
    "\telse\n",
    "\t\tcat(\"*\", length(pwm.files), \"PWM(s) in the tfbs object.\\n\");\n",
    "\n",
    "\trbps <- rbpbs(\n",
    "\t\tspecies    = cisrbp.db@species,\n",
    "\t\tfilenames  = pwm.files,\n",
    "\t\tnames      = names,\n",
    "\t\trbp_info    = RBP_info,\n",
    "\t\trbp_missing = RBP_missing,\n",
    "\t\theader=T, sep=\"\\t\" , row.names=1 );\n",
    "\n",
    "\n",
    "\treturn(rbps);\n",
    "}\n",
    "                                         \n",
    "                                        \n",
    "                                         "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3e5bae97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'CisRBP.getRBPinformation'"
      ],
      "text/latex": [
       "'CisRBP.getRBPinformation'"
      ],
      "text/markdown": [
       "'CisRBP.getRBPinformation'"
      ],
      "text/plain": [
       "[1] \"CisRBP.getRBPinformation\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "'rbpbs.createFromCisRBP'"
      ],
      "text/latex": [
       "'rbpbs.createFromCisRBP'"
      ],
      "text/markdown": [
       "'rbpbs.createFromCisRBP'"
      ],
      "text/plain": [
       "[1] \"rbpbs.createFromCisRBP\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "#'\n",
    "#' @param cisrbp.db: CisRBP.db object\n",
    "#' @param rbp.information.type: 1,2 or 3 indicate which motif file will be used.\n",
    "#'\n",
    "#' @return: data.frame;\n",
    "\n",
    "setGeneric(\"CisRBP.getRBPinformation\",\n",
    "\tdef=function(cisrbp.db, rbp.information.type = NA ) {\n",
    "\t\tstandardGeneric(\"CisRBP.getRBPinformation\")\n",
    "\t})\n",
    "\n",
    "setMethod(\"CisRBP.getRBPinformation\", signature(cisrbp.db=\"CisRBP.db\"),\n",
    "\tfunction(cisrbp.db, rbp.information.type = NA )\n",
    "\t{\n",
    "\t\tfile.rbpinfo <- cisrbp.db@file.rbpinfo;\n",
    "\n",
    "\t\tif(!missing(rbp.information.type))\n",
    "\t\t\tfile.rbpinfo <- CisRBP.active_motif_info( cisrbp.db, rbp.information.type );\n",
    "\n",
    "\t\ttbm <- try( read.csv(file.rbpinfo , header=T, sep=\"\\t\") );\n",
    "\t\tif( class(tbm) == \"try-error\" )\n",
    "\t\t\tstop(\"Can not open RBP information file:\", file.rbpinfo );\n",
    "\n",
    "\t\ttmp.dir <- tempdir();\n",
    "\t\tmotif.existing <- c();\n",
    "\n",
    "\t\tfor (i in 1:NROW(tbm))\n",
    "\t\t{\n",
    "\t\t\tmotif_id <- as.character(tbm$Motif_ID[i]);\n",
    "\n",
    "\t\t\tif( as.character(motif_id)==\".\")\n",
    "\t\t\t{\n",
    "\t\t\t\tmotif.existing <- c(motif.existing, FALSE);\n",
    "\t\t\t\tnext;\n",
    "\t\t\t}\n",
    "\n",
    "\t\t\tpwm.file <- paste( \"pwms_all_motifs/\", motif_id, \".txt\", sep=\"\");\n",
    "\t\t\tr.file <- unzip( cisrbp.db@zip.file, c(pwm.file), exdir=tmp.dir  );\n",
    "\t\t\tif(length(r.file)<1)\n",
    "\t\t\t{\n",
    "\t\t\t\tmotif.existing <- c(motif.existing, FALSE);\n",
    "\t\t\t\tnext;\n",
    "\t\t\t}\n",
    "\n",
    "\t\t\trbp.motif <- try( read.table(paste(tmp.dir, pwm.file, sep=\"/\"), header=T, sep=\"\\t\", row.names=1), TRUE );\n",
    "\t\t\tif(class(rbp.motif)==\"try-error\")\n",
    "\t\t\t{\n",
    "\t\t\t\tmotif.existing <- c(motif.existing, FALSE);\n",
    "\t\t\t\tnext;\n",
    "\t\t\t}\n",
    "\n",
    "\t\t\tif(NROW(rbp.motif)==0)\n",
    "\t\t\t{\n",
    "\t\t\t\tmotif.existing <- c(motif.existing, FALSE);\n",
    "\t\t\t\tnext;\n",
    "\t\t\t}\n",
    "\n",
    "\t\t\tmotif.existing <- c(motif.existing, TRUE);\n",
    "\t\t}\n",
    "\n",
    "\n",
    "\t\tcat(\"  Total motif:\", NROW(tbm), \", Missing or empty PWMs:\", length(which(!motif.existing)), \".\\n\");\n",
    "\n",
    "\t\treturn( cbind(tbm, Motif_Existing=motif.existing) );\n",
    "\t})\n",
    "\n",
    "\n",
    "setGeneric(\"rbpbs.createFromCisRBP\",\n",
    "\t\tdef=function(cisrbp.db, motif_id = NULL, rbp_name = NULL, rbp_status = NULL, family_name = NULL,\n",
    "\t\t\t\t\tmotif_type = NULL, msource_id =NULL, rbp.information.type = 1){\n",
    "\t\t\tstopifnot(class(cisrbp.db) == \"CisRBP.db\")\n",
    "\t\t\tstandardGeneric(\"rbpbs.createFromCisRBP\")\n",
    "\t})\n",
    "\n",
    "setMethod(\"rbpbs.createFromCisRBP\", c( cisrbp.db = \"CisRBP.db\"),rbpbs_createFromCisRBP )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "82053f80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# build function RBP.zipload which take a zip file and species as an input\n",
    "RBP.zipload<-function(zip.file,species){\n",
    "  tmp.dir<-tempdir() #tempdir() makes temporary directory\n",
    "  r.file<-unzip(zip.file,\"RBP_Information.txt\",exdir=tmp.dir) #extract the file specifed in c() from zip.file and save it to exdir.\n",
    "  if (length(r.file)<1){\n",
    "    cat(\"!RBP_Information.txt can not be found in the zip file(\",zip.file,\")\\n\")\n",
    "  return(NULL)\n",
    "  }\n",
    "\n",
    "  new(\"CisRBP.db\",species=species, file.rbpinfo=paste(tmp.dir, \"RBP_Information.txt\",sep=\"/\"),zip.file=zip.file, zip.date=format(file.info(zip.file)$mtime,\"%m%d%Y\"),zip.url=\"\") \n",
    "  #new is the funciton that make object of class defined in the first arg.\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "aba49bba",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "read.motif <- function(motif_path, pseudocount= -7, force_even= FALSE, ...)\n",
    "{\n",
    "\t## Read the pwm and sequence file.\n",
    "\tmotif <- tryCatch({\n",
    "\t\t\tas.matrix(read.table(motif_path, ...))\n",
    "\t\t},\n",
    "\t\terror= function(e)\n",
    "\t\t{\n",
    "\t\t\tshow(e);\n",
    "\t\t\t#read.pwm(motif_path);\n",
    "\t\t\treturn( NULL ) ;\n",
    "\t\t})\n",
    "\n",
    "\tif( is.null(motif) || NROW(motif) == 0 )\n",
    "\t{\n",
    "\t\twarning(paste( \"No matrix data in the file (\", motif_path, \".\", sep=\"\"))\n",
    "\t\treturn( NULL );\n",
    "\t}\n",
    "\n",
    "\tif( NCOL(motif)>=6 || !is.numeric(motif) )\n",
    "\t{\n",
    "\t\twarning(paste( \"File (\", motif_path, \") maybe not a correct PWM file.\", sep=\"\"))\n",
    "\t\treturn(NULL);\n",
    "\t}\n",
    "\n",
    "\t## remove left position column if 5 columns\n",
    "\tif( NCOL(motif)==5) motif <- motif[,-1]\n",
    "\n",
    "\tif(sum(motif[1,])>0) {\n",
    "\t\tmotif <- log(motif/rowSums(motif)) # Divide by counts.\n",
    "\t}\n",
    "\n",
    "\t## Set a pseudocount of ~0.1%.\n",
    "\tmotif[motif==-Inf] <- pseudocount;\n",
    "\n",
    "\t## Off-by-one bug for odd motifs.  Not sure why?!  For now hack it.\n",
    "\tif((NROW(motif) %% 2) == 1 & force_even)\n",
    "\t\tmotif <- rbind(motif, log(c(0.25, 0.25, 0.25, 0.25)));\n",
    "\n",
    "\treturn(motif);\n",
    "}\n",
    "\n",
    "## Reverse complement.\n",
    "## Assume motif columns in order: A, C, G, T\n",
    "reverse.complement <- function(motif)\n",
    "{\n",
    "\tcolnames(motif) <- colnames(motif)[c(4:1)]\n",
    "\n",
    "\treturn( motif[c(NROW(motif):1),c(4:1)] ) ## Reverse complement.\n",
    "}\n",
    "\n",
    "## Correlation between motifs w/ equal size.\n",
    "cor.motif.eq.size <- function(motif1, motif2)\n",
    "{\n",
    "\tstopifnot(NROW(motif1) == NROW(motif2))\n",
    "\n",
    "\treturn(cor(as.vector(motif1), as.vector(motif2)))\n",
    "}\n",
    "\n",
    "## Extends a shorter motif w/ a pre-defined background (BG)= {A, C, G, T}.\n",
    "extend.motif <- function(motif2, left, right, BG)\n",
    "{\n",
    "\tleftBG  <- t(matrix(rep(BG, left), nrow=4))\n",
    "\trightBG <- t(matrix(rep(BG, right), nrow=4))\n",
    "\n",
    "\treturn(rbind(leftBG, motif2, rightBG))\n",
    "}\n",
    "\n",
    "## Compares two motifs.  Returns Pearson's R for the log-values.\n",
    "compare.motifs.inner <- function(motif1, motif2, BG=log(c(0.25, 0.25, 0.25, 0.25)))\n",
    "{\n",
    "\t## REQUIRE: NROW(motif1) >= NROW(motif2)\n",
    "\tif(NROW(motif1) < NROW(motif2)) {\n",
    "\t\ttmp <- motif1\n",
    "\t\tmotif1 <- motif2\n",
    "\t\tmotif2 <- tmp\n",
    "\t\tremove(tmp)\n",
    "\t}\n",
    "\n",
    "\t## Create two vectors\n",
    "\tld <- (NROW(motif1)-NROW(motif2)+1)\n",
    "\tsize_small <- NROW(motif2)\n",
    "\n",
    "\t## Compare natural.\n",
    "\tmax_plus <- sapply(c(1:ld), function(x){ cor.motif.eq.size(motif1[c(x:(x+size_small-1)),], motif2) })\n",
    "\n",
    "\t## Also reverse complement.\n",
    "\tmotif1_rc <- reverse.complement(motif1)\n",
    "\tmax_minus <- sapply(c(1:ld), function(x){ cor.motif.eq.size(motif1_rc[c(x:(x+size_small-1)),], motif2) })\n",
    "\n",
    "\t## Get the maximum value.\n",
    "\tif(max(max_plus) >= max(max_minus)) {\n",
    "\t\tmax_pos <- which.max(max_plus)\n",
    "\t\tmaxStr_mot1 <- motif1\n",
    "\t}\n",
    "\telse\n",
    "\t{\n",
    "\t\tmax_pos <- which.max(max_minus)\n",
    "\t\tmaxStr_mot1 <- motif1_rc\n",
    "\t}\n",
    "\n",
    "\t## One we have the optimal placement, extend the motifs with background.\n",
    "\tmotif_score <- cor.motif.eq.size(maxStr_mot1, extend.motif(motif2, (max_pos-1), (ld-max_pos), BG))\n",
    "\n",
    "\treturn(motif_score)\n",
    "}\n",
    "\n",
    "\n",
    "compare.motifs <- function(motif1, motif2, BG=log(c(0.25, 0.25, 0.25, 0.25)))\n",
    "{\n",
    "\talign_strong_part <- function(motif.x, motif.y)\n",
    "\t{\n",
    "\t\tn.max1 <- 0;\n",
    "\t\tscore <- NA;\n",
    "\n",
    "\t\tratio_bp1 <- apply(exp(motif.x), 1, max);\n",
    "\t\tif( length(which(ratio_bp1>=0.6)) > 0 ) n.max1 <- min( min( which(ratio_bp1>=0.6) ), floor( length(ratio_bp1) * 0.25) );\n",
    "\t\tif( n.max1>0 ) score <- unlist(lapply(1:n.max1, function(k) { compare.motifs.inner( motif.x[-c(1:k),], motif.y, BG ); } ) ) ;\n",
    "\n",
    "\t\treturn(score);\n",
    "\t}\n",
    "\n",
    "\tscore1 <- c( align_strong_part( motif1, motif2 ), align_strong_part( motif2, motif1) );\n",
    "\tmotif1_rc <- reverse.complement(motif1);\n",
    "\tmotif2_rc <- reverse.complement(motif2);\n",
    "\tscore2 <- c( align_strong_part( motif1_rc, motif2 ), align_strong_part( motif2_rc, motif1) );\n",
    "\n",
    "\tscore <- c(score1, score2, compare.motifs.inner( motif1, motif2, BG ) );\n",
    "\n",
    "\treturn(max(score, na.rm=T));\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "88ca8d89",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: 'parallel'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:snow':\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, clusterSplit, makeCluster, parApply,\n",
      "    parCapply, parLapply, parRapply, parSapply, splitIndices,\n",
      "    stopCluster\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "'rbpbs.clusterMotifs'"
      ],
      "text/latex": [
       "'rbpbs.clusterMotifs'"
      ],
      "text/markdown": [
       "'rbpbs.clusterMotifs'"
      ],
      "text/plain": [
       "[1] \"rbpbs.clusterMotifs\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "\n",
    "## Clusters TFs based on DNA sequence preferences.\n",
    "library(parallel);library(cluster)\n",
    "check_folder_writable<-function(file.prefix)\n",
    "{\n",
    "\tif( !file.exists(dirname(file.prefix) ) )\n",
    "\t\tdir.create( dirname(file.prefix), showWarnings = TRUE, recursive = TRUE );\n",
    "\t\n",
    "\tfile.temp <- tempfile( tmpdir = dirname(file.prefix), fileext = \"\")\n",
    "\tr.try <- try( file.create(file.temp) )\n",
    "\tif( class(r.try)==\"try-error\" || r.try==FALSE )\n",
    "\t\treturn(FALSE);\n",
    "\n",
    "\tunlink(file.temp);\n",
    "\treturn(TRUE);\n",
    "}\n",
    "\n",
    "setGeneric(\"rbpbs.clusterMotifs\",\n",
    "\tdef = function(rbpbs,\n",
    "\t\t\tmethod = c(\"agnes\", \"apcluster\"),\n",
    "\t\t\tpdf.heatmap = NA,\n",
    "\t\t\tgroup.k = NA,\n",
    "\t\t\tapcluster.q = 0.99,\n",
    "\t\t\tncores = 1,\n",
    "\t\t\tplot.style=c(\"rtfbsdb\", \"apcluster\"),\n",
    "\t\t\tBG=log( c(0.25, 0.25, 0.25, 0.25)),\n",
    "\t\t\t... )\n",
    "\t{\n",
    "\t\tstopifnot(class(rbpbs) == \"rbpbs\")\n",
    "\t\tstandardGeneric(\"rbpbs.clusterMotifs\")\n",
    "\t})\n",
    "\n",
    "rbpbs_clusterMotifs <- function(rbpbs, method=c(\"agnes\", \"apcluster\"), pdf.heatmap=NA, group.k=NA, apcluster.q=0.99, ncores=1, plot.style=c(\"rtfbsdb\", \"apcluster\"), BG=log(c(0.25, 0.25, 0.25, 0.25) ), ... )\n",
    "{\n",
    "\tif( missing(method) )\n",
    "\t\tmethod <- \"apcluster\";\n",
    "\n",
    "\tif( missing(plot.style) )\n",
    "\t\tplot.style <- \"rtfbsdb\";\n",
    "\n",
    "\tif( !is.na( pdf.heatmap ) )\n",
    "\t\tif( !check_folder_writable( pdf.heatmap ) )\n",
    "\t\t\tcat(\"! Can not create pdf file: \", pdf.heatmap, \"\\n\");\n",
    "\n",
    "\tmat <- rbpbs@distancematrix;\n",
    "\tif ( is.null(mat) || NROW(mat)==0 ){\n",
    "\t\tdist.list <- mclapply(1:rbpbs@nrbps, function(i) {\n",
    "\t\t\t\tunlist( lapply(1:rbpbs@nrbps, function(j) {compare.motifs(rbpbs@pwm[[i]], rbpbs@pwm[[j]], BG=BG)} )) }, mc.cores=ncores);\n",
    "\t\trbpbs@distancematrix <- as.matrix( do.call( rbind, dist.list ) );\n",
    "\t\tmat <- rbpbs@distancematrix;\n",
    "\t}\n",
    "\n",
    "\tsubset <- 1:NROW(mat);\n",
    "\n",
    "#\tif(method==\"cors\")\n",
    "#\t{\n",
    "#\t\topt_list <- tfbs_corclustering_bic_optim(mat);\n",
    "#\n",
    "#\t\tbics <- lapply(opt_list, function(l)return(l$BIC) );\n",
    "#\t\tmin.grps <- which.min( bics);\n",
    "#\t\tclusters <- opt_list[[min.grps]]$clusters;\n",
    "#\n",
    "#\t\tcat(\"MIN CLUSRERS=\", min.grps, max( opt_list[[min.grps]]$clusters))\n",
    "#\n",
    "#\t\tif(!is.na( pdf.heatmap ))\n",
    "#\t\t{\n",
    "#\t\t\tr.try <- try ( pdf( pdf.heatmap ) );\n",
    "#\t\t\tif( class(r.try) == \"try-error\")\n",
    "#\t\t\t\tcat(\"! Failed to write PDF file:\", pdf.heatmap, \"\\n\")\n",
    "#\t\t\telse\n",
    "#\t\t\t{\n",
    "#\t\t\t\ttfbs_drawheatmapForClusters(tfbs, mat, clusters);\n",
    "#\t\t\t\tdev.off();\n",
    "#\t\t\t}\n",
    "#\t\t}\n",
    "#\t}\n",
    "\t\n",
    "\tif(method==\"apcluster\")\n",
    "\t{\n",
    "\t\tr.ap <- apcluster(mat, q = apcluster.q, ... );\n",
    "\n",
    "\t\tclusters <- rep(NA, length(subset));\n",
    "\t\tfor(i in 1:length(r.ap@clusters))\n",
    "\t\t\tclusters[ r.ap@clusters[[i]] ] <- i;\n",
    "\n",
    "\t\tif(!is.na( pdf.heatmap ))\n",
    "\t\t{\n",
    "\t\t\tr.try <- try ( pdf( pdf.heatmap ) );\n",
    "\t\t\tif( class(r.try) == \"try-error\")\n",
    "\t\t\t\tcat(\"! Failed to write PDF file:\", pdf.heatmap, \"\\n\")\n",
    "\t\t\telse\n",
    "\t\t\t{\n",
    "\t\t\t\tif(plot.style==\"apcluster\")\n",
    "\t\t\t\t\ttry( apcluster::heatmap( r.ap, mat ) )\n",
    "\t\t\t\telse\n",
    "\t\t\t\t\ttryCatch({\n",
    "\t\t\t\t\t\taggres <- aggExCluster(mat, r.ap);\n",
    "\t\t\t\t\t\tdend <- as.dendrogram(aggres, base=0.05, useNames=FALSE)\n",
    "\t\t\t\t\t\ttfbs_draw_heatmap_style(dend, mat, clusters);\n",
    "\t\t\t\t\t\t},\n",
    "\t\t\t\t\t\terror= function(e)\n",
    "\t\t\t\t\t\t{\n",
    "\t\t\t\t\t\t\tshow(e);\n",
    "\t\t\t\t\t\t\treturn( NULL ) ;\n",
    "\t\t\t\t\t\t});\n",
    "\n",
    "\t\t\t\tdev.off();\n",
    "\t\t\t}\n",
    "\t\t}\n",
    "\t}\n",
    "\telse\n",
    "\t{\n",
    "\t\tif(missing(group.k))\n",
    "\t\t\tgroup.k <- round( NROW(mat)/3 );\n",
    "\n",
    "\t\thc1 <- agnes(as.dist((1-mat)^5), diss=TRUE)\n",
    "\t\tclusters <- cutree(hc1, k=group.k)\n",
    "\n",
    "\t\tif(!is.na( pdf.heatmap ))\n",
    "\t\t{\n",
    "\t\t\tr.try <- try ( pdf( pdf.heatmap ) );\n",
    "\t\t\tif( class(r.try) == \"try-error\")\n",
    "\t\t\t\tcat(\"! Failed to write PDF file:\", pdf.heatmap, \"\\n\")\n",
    "\t\t\telse\n",
    "\t\t\t{\n",
    "\t\t\t\thc1 <- as.dendrogram(hc1)\n",
    "\t\t\t\ttfbs_draw_heatmap_style(hc1, mat, clusters);\n",
    "\t\t\t\tdev.off();\n",
    "\t\t\t}\n",
    "\t\t}\n",
    "\t}\n",
    "\n",
    "\trbpbs@cluster <- cbind(subset, clusters);\n",
    "\n",
    "\n",
    "\treturn( rbpbs );\n",
    "}\n",
    "\n",
    "setMethod(\"rbpbs.clusterMotifs\", c(rbpbs=\"rbpbs\"), rbpbs_clusterMotifs);\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b1085f96",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: 'latticeExtra'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:ggplot2':\n",
      "\n",
      "    layer\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#install.packages(\"latticeExtra\")\n",
    "library(lattice)\n",
    "library(stats)\n",
    "library(latticeExtra)\n",
    "tfbs_draw_heatmap_style <- function(hc1, mat, clusters)\n",
    "{\n",
    "\tord.hc1 <- as.numeric(order.dendrogram(hc1))\n",
    "\thc2 <- reorder(hc1, mat[ord.hc1])\n",
    "\tord.hc2 <- as.numeric(order.dendrogram(hc2))\n",
    "\n",
    "\tpal100 <- c(\"#7E291B\",\"#66E52C\",\"#8F66F0\",\"#58DBE8\",\"#396526\",\"#EAABC1\",\"#E1C33C\",\"#3E3668\",\"#EB3F90\",\"#C3E6A8\",\"#E74618\",\"#66A2E9\",\"#3E7774\",\"#DF9056\",\"#3C2C21\",\"#DF40D7\",\"#6CEF92\",\"#8C5A6B\",\"#BC8AE2\",\"#A03B99\",\"#56AC2D\",\"#389C6C\",\"#E26B7E\",\"#706B4B\",\"#D2E374\",\"#A0A560\",\"#7B1C3E\",\"#49F7DB\",\"#C8C6E9\",\"#414FA2\",\"#95A590\",\"#8A669A\",\"#98A62F\",\"#9E792C\",\"#D69489\",\"#547FE5\",\"#DF6340\",\"#849BAB\",\"#E63A61\",\"#8A386D\",\"#DAE338\",\"#263715\",\"#BBDEE3\",\"#3F1324\",\"#A1E03B\",\"#383544\",\"#76C2E8\",\"#794D38\",\"#DD74E2\",\"#D7BF8E\",\"#E366B9\",\"#894D19\",\"#D57221\",\"#D9B764\",\"#B0303D\",\"#D6BFBC\",\"#757A28\",\"#D991CC\",\"#356344\",\"#E3A22D\",\"#223C36\",\"#83B664\",\"#D8E4C5\",\"#AE9ED9\",\"#37576D\",\"#CF7398\",\"#6BEAB2\",\"#6C9266\",\"#B33662\",\"#4B340E\",\"#57E65B\",\"#E23635\",\"#9B464B\",\"#757089\",\"#578BBA\",\"#A6311B\",\"#B2714E\",\"#457F20\",\"#4EA3B1\",\"#B93286\",\"#73D463\",\"#531914\",\"#6B78C2\",\"#E07467\",\"#8D786E\",\"#515018\",\"#361E40\",\"#AA4FCC\",\"#90D2C4\",\"#71469B\",\"#419C4C\",\"#37558A\",\"#B393B0\",\"#9BE090\",\"#856BDA\",\"#66B593\",\"#47CA84\",\"#5D205B\",\"#672F3F\",\"#59D8C2\")\n",
    "\n",
    "\tpal500 <- rep(pal100, ceiling(length(clusters)/100) )\n",
    "\n",
    "\tfill.col<- rep(\"white\", length(ord.hc2));\n",
    "\tfill.col[ord.hc2] <- pal500[ clusters ];\n",
    "\n",
    "\t#pl <- levelplot((mat)[ord.hc2, ord.hc2], col.regions= yb.sig.pal(100, scale=3), xlab=\"\", ylab=\"\",\n",
    "\tprint( levelplot((mat)[ord.hc2, ord.hc2], col.regions= yb.sig.pal(100, scale=3), xlab=\"\", ylab=\"\",\n",
    "\t\tcolorkey = list(space=\"left\", labels=list(cex=1.5)),\n",
    "\t\tlegend = list(\n",
    "\t\t\ttop = list(fun = dendrogramGrob,\n",
    "\t\t\targs = list(x = hc2, ord = ord.hc2, side = \"top\", #lwd=2,\n",
    "\t\t\tsize = 7, size.add = 0.5,\n",
    "\t\t\tadd = list(rect = list(col = \"transparent\", fill = fill.col )),\n",
    "\t\t\ttype = \"rectangle\")))) );\n",
    "}\n",
    "#drawing level plot\n",
    "## For drawing a levelplot.\n",
    "yb.sig.pal <- function(n, scale=10) {\n",
    "\tints<- c(0:(n-1))/(n-1)   ## Linear scale from 0:1 x N values.\n",
    "\tints<- 1/(1+exp(scale*(0.5-ints)))## Transfer to sigmoidal scale.\n",
    "\tb<- min(ints)\n",
    "\tm<- 2*b/(n-1)\n",
    "\tints<- ints+(m*c(0:(n-1)) -b)## Transform by linear function to fill colors out to maxes.\n",
    "\n",
    "\t## Transfer to colorspace.\n",
    "\t# Yellow: 255, 255, 0\n",
    "\t# White:  255, 255, 255\n",
    "\t# Blue:   0, 0, 255\n",
    "\tYW <- ints[ints < 0.5] *2\n",
    "\tWB <- (ints[ints >= 0.5]-0.5) *2\n",
    "\tYW[YW<0] <- 0; WB[WB>1] <- 1\n",
    "\tc(rgb(1, 1, YW), rgb(1-WB, 1-WB, 1))\n",
    "}\n",
    "\n",
    "\n",
    "write.starchbed <- function(bed, file.starch) {\n",
    "\t# pipe bed into starch file\n",
    "\tr <- try( write.table(bed, file = pipe(paste(\"sort-bed - | starch - >\", file.starch)), quote=FALSE, row.names=FALSE, col.names=FALSE, sep=\"\\t\"), silent=F );\n",
    "\tif(class(r)==\"try-error\")\n",
    "\t{\n",
    "\t\tcat(\"Failed to write starch file using 'sort-bed' and 'starch' command.\\n\")\n",
    "\t\treturn(\"ERROR\");\n",
    "\t}\n",
    "\n",
    "\tif( !file.exists(file.starch) )\n",
    "\t{\n",
    "\t\tcat(\"! Failed to write starch file (\", file.starch, \") using the sort-bed and starch commands.\\n\");\n",
    "\t\treturn(\"ERROR\");\n",
    "\t}\n",
    "\telse\n",
    "\t\treturn(file.starch);\n",
    "}\n",
    "\n",
    "# get chromosome size based on genemo data of 2 bit file.\n",
    "#\n",
    "# @file.twoBit, e.g. hh19.2bit, mm10.2bit\n",
    "#\n",
    "get_chromosome_size <- function(file.twoBit)\n",
    "{\n",
    "\tfile.tmp <- tempfile();\n",
    "\n",
    "\terr_code <- system(paste(\"twoBitInfo\", file.twoBit, file.tmp, sep=\" \"));\n",
    "\tchromInfo <- read.table( file.tmp );\n",
    "\tunlink( file.tmp );\n",
    "\n",
    "\tif( err_code!=0 ) cat(\"! Failed to call the twoBitInfo command\\n\");\n",
    "\n",
    "\treturn(chromInfo);\n",
    "}\n",
    "\n",
    "# save the chromosme size into one ftable file\n",
    "#\n",
    "# save_chromosome_size(\"mm10.2bit\", \"chrom_info_mm19.tab\")\n",
    "#\n",
    "save_chromosome_size <- function( file.twoBit, file.tab)\n",
    "{\n",
    "\twrite.table(get_chromosome_size(file.twoBit), file=file.tab, row.names=F, col.names=F, quote=F);\n",
    "}\n",
    "\n",
    "check_folder_writable<-function(file.prefix)\n",
    "{\n",
    "\tif( !file.exists(dirname(file.prefix) ) )\n",
    "\t\tdir.create( dirname(file.prefix), showWarnings = TRUE, recursive = TRUE );\n",
    "\t\n",
    "\tfile.temp <- tempfile( tmpdir = dirname(file.prefix), fileext = \"\")\n",
    "\tr.try <- try( file.create(file.temp) )\n",
    "\tif( class(r.try)==\"try-error\" || r.try==FALSE )\n",
    "\t\treturn(FALSE);\n",
    "\n",
    "\tunlink(file.temp);\n",
    "\treturn(TRUE);\n",
    "}\n",
    "\n",
    "check_bed<-function( df.bed )\n",
    "{\n",
    "\tif (NCOL(df.bed)<3 )\n",
    "\t{\n",
    "\t\twarning(\"At least 3 columns in BED file.\");\n",
    "\t\treturn (FALSE);\n",
    "\t}\n",
    "\n",
    "\tif (any(is.na(df.bed[,c(1:3)])))\n",
    "\t{\n",
    "\t\twarning(\"NA values in first three columns in BED file.\");\n",
    "\t\treturn (FALSE);\n",
    "\t}\n",
    "\n",
    "\tif (  !(class(df.bed[,3]) == \"numeric\" || class(df.bed[,3]) == \"integer\")\n",
    "\t   || !(class(df.bed[,2]) == \"numeric\" || class(df.bed[,2]) == \"integer\") )\n",
    "\t{\n",
    "\t\twarning(\"The 2nd column or 3rd column are not numeric in BED file.\");\n",
    "\t\treturn (FALSE);\n",
    "\t}\n",
    "\n",
    "\tif ( any(df.bed[,2] > df.bed[,3]))\n",
    "\t{\n",
    "\t\tidx.equal <- which (df.bed[,2] > df.bed[,3]);\n",
    "\t\twarning(paste(\"The 2nd column is greater than the 3rd column in BED data. (e.g. Index =\", idx.equal[1], \"\\n\") );\n",
    "\t\treturn (FALSE);\n",
    "\t}\n",
    "\n",
    "\tif ( NCOL(df.bed)==6 )\n",
    "\t{\n",
    "\t\tif (!(unique(df.bed[,6]) %in% c(\"+\", \"-\", \".\")))\n",
    "\t\t{\n",
    "\t\t\twarning(\"Three values are avalaible for the 6th column in BED file, '+', '-' or '.'.\");\n",
    "\t\t\treturn (FALSE);\n",
    "\t\t}\n",
    "\t}\n",
    "\n",
    "\treturn(TRUE);\n",
    "}\n",
    "\n",
    "get_os <- function()\n",
    "{\n",
    "  \tsysinf <- Sys.info()\n",
    "  \tif (!is.null(sysinf)){\n",
    "    \tos <- sysinf['sysname']\n",
    "    \tif (os == 'Darwin')\n",
    "      \t\tos <- \"osx\"\n",
    "  \t}\n",
    "  \telse\n",
    "  \t{ ## mystery machine\n",
    "  \t  \tos <- .Platform$OS.type\n",
    "    \t\tif (grepl(\"^darwin\", R.version$os))\n",
    "      \t\tos <- \"osx\"\n",
    "    \tif (grepl(\"linux-gnu\", R.version$os))\n",
    "      \t\tos <- \"linux\"\n",
    "  \t}\n",
    "  \ttolower(os);\n",
    "}\n",
    "\n",
    "check_command_error<-function( r.tryerror, commands)\n",
    "{\n",
    "\tif(class(r.tryerror)==\"try-error\")\n",
    "\t{\n",
    "\t\tif( any( Sys.which(commands)==\"\") )\n",
    "\t\t{\n",
    "\t\t\tccommand.str <- paste(commands, collapse=\"','\")\n",
    "\t\t\tstop(paste(\"Failed to call '\", ccommand.str , \"' command, use Sys.which() to check its path or system() to verify its function.\", sep=\"\"));\n",
    "\t\t}\n",
    "\t\telse\n",
    "\t\t\tstop(as.character(r.tryerror));\n",
    "\t}\n",
    "}\n",
    "#tfbs.drawLogosForClusters\n",
    "rbpbs_drawLogosForClusters <- function(tfbs,  file.pdf=NULL, nrow.per.page=6, vertical=TRUE )\n",
    "{\n",
    "\tif( NROW( tfbs@cluster )  == 0)\n",
    "\t\tstop(\"No cluster information in the tfbs object.\")\n",
    "\n",
    "\tif( !is.null(file.pdf) && !is.na(file.pdf) )\n",
    "\t\tif( !check_folder_writable( file.pdf ) )\n",
    "\t\t\tstop( paste(\"Can not create pdf file: \",file.pdf ) );\n",
    "\n",
    "\tif( !is.null(file.pdf) && !is.na(file.pdf) )\n",
    "\t{\n",
    "\t\tr.try <- try ( pdf( file.pdf ) );\n",
    "\t\tif(class(r.try)==\"try-error\")\n",
    "\t\t\tstop(\"Failed to create PDF file for motif logos.\\n\");\n",
    "\t}\n",
    "\n",
    "\tdraw_rbp_name <- function(tfbs, idx, motif.perpage)\n",
    "\t{\n",
    "\t\trbp_name <- tfbs@mgisymbols[idx];\n",
    "\t\tmotif_id <- tfbs@mgisymbols[idx];\n",
    "\t\tif ( NROW(tfbs@rbp_info) > 0 )\n",
    "\t\t{\n",
    "\t\t\trbp_name <- tfbs@rbp_info[ idx, \"RBP_Name\"];\n",
    "\t\t\tmotif_id <- tfbs@rbp_info[ idx, \"Motif_ID\"];\n",
    "\t\t}\n",
    "\n",
    "\t\tn.block  <- round(motif.perpage/2);\n",
    "\n",
    "\t\t# draw TF_Name in biggger size\n",
    "\t\tpushViewport( viewport(x=0.01, y=0, width=0.05, height=1, just=c(\"left\",\"bottom\")) );\n",
    "\n",
    "\t\tstr.inches <- strwidth(rbp_name, units = 'in');\n",
    "\t\ty.scale <- convertHeight(unit(str.inches, \"inches\"), \"native\", valueOnly=T );\n",
    "\n",
    "\t\tcex <- 1;\n",
    "\t\tif( y.scale> 0.8 ) cex <- 0.8/y.scale;\n",
    "\t\tif( cex > 1 ) cex <- 1;\n",
    "\n",
    "\t\tgrid.text( rbp_name, rot=90, gp=gpar(cex=cex), check.overlap=T);\n",
    "\n",
    "\t\tpopViewport();\n",
    "\n",
    "\t\t# draw Motif_ID in smaller size\n",
    "\t\tpushViewport( viewport(x=0.06, y=0, width=0.02, height=1, just=c(\"left\",\"bottom\")) );\n",
    "\n",
    "\t\tstr.inches <- strwidth(motif_id, units = 'in');\n",
    "\t\ty.scale <- convertHeight(unit(str.inches, \"inches\"), \"native\", valueOnly=T );\n",
    "\n",
    "\t\tcex <- 1/2;\n",
    "\t\tif( y.scale> 0.5 ) cex <- 0.5/y.scale;\n",
    "\t\tif( cex > 1/2 ) cex <- 1/2;\n",
    "\n",
    "\t\tgrid.text( motif_id, rot=90, gp=gpar(cex=cex), check.overlap=T);\n",
    "\t\tpopViewport();\n",
    "\n",
    "\t}\n",
    "\n",
    "\tcluster.mat <- tfbs@cluster;\n",
    "\n",
    "\tfor(cls in 1:max(cluster.mat[,2]))\n",
    "\t{\n",
    "\t\tmotifs_in_cluster <- cluster.mat[which(cluster.mat[,2] == cls), 1];\n",
    "\t\tnmotifs <- length(motifs_in_cluster);\n",
    "\n",
    "\t\t# at least4 motif space will be designed in each page.\n",
    "\t\tif( nmotifs <= 4 ) motif.perpage <- 4 else motif.perpage <- nmotifs;\n",
    "\t\tif(!is.null(nrow.per.page) && !is.na(nrow.per.page))\n",
    "\t\t\tmotif.perpage <- nrow.per.page * 2;\n",
    "\n",
    "\t\tnpage <- ceiling( nmotifs / motif.perpage );\n",
    "\t\tfor(p in 1:npage)\n",
    "\t\t{\n",
    "\t\t\tgrid.newpage();\n",
    "\n",
    "\t\t\tpushViewport( viewport(x=0.88, y=0.98, width=0.1, height=0.02, just=c(\"left\",\"bottom\")) );\n",
    "\t\t\tstr.pageno <- paste(\"Cluster:\", cls, sep=\"\");\n",
    "\t\t\tif(p>1) str.pageno <- paste(str.pageno, \"/p:\", p, sep=\"\");\n",
    "\t\t\tgrid.text( str.pageno, gp=gpar(cex=0.6), check.overlap=F);\n",
    "\t\t\tpopViewport();\n",
    "\n",
    "\t\t\tnrow.page <- ceiling(motif.perpage/2);\n",
    "\t\t\tfor(r in 1:nrow.page)\n",
    "\t\t\t{\n",
    "\t\t\t\tvp1 <- viewport(x=0, y = 1 - r/nrow.page, width=0.5, height=1/nrow.page, just=c(\"left\",\"bottom\"));\n",
    "\t\t\t\tpushViewport(vp1);\n",
    "\n",
    "\t\t\t\t#seqLogo(makePWM(exp(t(tfbs@pwm[[motifs_in_cluster[j]]]))), xaxis = FALSE, yaxis = FALSE);\n",
    "\t\t\t\tif(!vertical)\n",
    "\t\t\t\t\tidx.motif <- 2*r-1 + (p-1)* motif.perpage\n",
    "\t\t\t\telse\n",
    "\t\t\t\t\tidx.motif <- r + (p-1)* motif.perpage;\n",
    "\n",
    "\t\t\t\tif ( idx.motif <= nmotifs )\n",
    "\t\t\t\t{\n",
    "\t\t\t\t\tdraw_rbp_name(tfbs, motifs_in_cluster[ idx.motif ], motif.perpage);\n",
    "\t\t\t\t\tpushViewport( viewport(x=0.07, y=0, width=0.93, height=1, just=c(\"left\",\"bottom\")) );\n",
    "\t\t\t\t\tseqLogo(exp(t(tfbs@pwm[[motifs_in_cluster[ idx.motif ]]])), xaxis = FALSE, yaxis = FALSE);\n",
    "\t\t\t\t\tpopViewport();\n",
    "\t\t\t\t}\n",
    "\t\t\t\tpopViewport();\n",
    "\n",
    "\t\t\t\tvp1 <- viewport(x=0.5, y = 1 - r/nrow.page, width=0.5, height=1/nrow.page, just=c(\"left\",\"bottom\"));\n",
    "\t\t\t\tpushViewport(vp1);\n",
    "\t\t\t\t#seqLogo(makePWM(exp(t(tfbs@pwm[[motifs_in_cluster[j]]]))), xaxis = FALSE, yaxis = FALSE);\n",
    "\t\t\t\tif(!vertical)\n",
    "\t\t\t\t\tidx.motif <- 2*r + (p-1)* motif.perpage\n",
    "\t\t\t\telse\n",
    "\t\t\t\t\tidx.motif <- r + nrow.per.page + (p-1)* motif.perpage;\n",
    "\t\t\t\tif ( idx.motif <= nmotifs )\n",
    "\t\t\t\t{\n",
    "\t\t\t\t\tdraw_rbp_name(tfbs, motifs_in_cluster[ idx.motif ], motif.perpage);\n",
    "\t\t\t\t\tpushViewport( viewport(x=0.07, y=0, width=0.93, height=1, just=c(\"left\",\"bottom\")) );\n",
    "\t\t\t\t\tseqLogo( exp(t(tfbs@pwm[[motifs_in_cluster[ idx.motif ]]])), xaxis = FALSE, yaxis = FALSE);\n",
    "\t\t\t\t\tpopViewport();\n",
    "\t\t\t\t}\n",
    "\n",
    "\t\t\t\tpopViewport();\n",
    "\t\t\t}\n",
    "\t\t}\n",
    "\t}\n",
    "\n",
    "\tif( !is.null(file.pdf) && !is.na(file.pdf) )\n",
    "\t\tdev.off();\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b203a8aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Species:  Homo_sapiens \n",
      "Data Source:   \n",
      "Zip file:  ./Homo_sapiens_2020_06_22_10_25_pm.zip \n",
      "Download date:  06222020 \n",
      "RBP information:  /var/folders/tt/c6wcdvv17lg0nn6yt9fj7dm80000gn/T//RtmpvQaTEG/RBP_Information.txt \n",
      "  472 PWM(s) are defined in the CisRBP dataset.\n",
      "! 270 PWM file(s) are failed to be loaded ( Missing PWMs : 270 , Empty PWMs : 0 ).\n",
      "* 202 PWM(s) in the tfbs object.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 202 × 26</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>RBP_ID</th><th scope=col>Family_ID</th><th scope=col>RSource_ID</th><th scope=col>Motif_ID</th><th scope=col>MSource_ID</th><th scope=col>DBID</th><th scope=col>RBP_Name</th><th scope=col>RBP_Species</th><th scope=col>RBP_Status</th><th scope=col>Family_Name</th><th scope=col>⋯</th><th scope=col>MSource_Type</th><th scope=col>MSource_Author</th><th scope=col>MSource_Year</th><th scope=col>PMID</th><th scope=col>MSource_Version</th><th scope=col>RBPSource_Name</th><th scope=col>RBPSource_URL</th><th scope=col>RBPSource_Year</th><th scope=col>RBPSource_Month</th><th scope=col>RBPSource_Day</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>T02631_0.6</td><td>F17_0.6</td><td>TS19_0.6</td><td>M037_0.6</td><td>MS2_0.6</td><td>ENSG00000076770</td><td>MBNL3  </td><td>Homo_sapiens</td><td>I</td><td>CCCH ZF    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>T02638_0.6</td><td>F17_0.6</td><td>TS19_0.6</td><td>M269_0.6</td><td>MS1_0.6</td><td>ENSG00000128016</td><td>ZFP36  </td><td>Homo_sapiens</td><td>D</td><td>CCCH ZF    </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>T02638_0.6</td><td>F17_0.6</td><td>TS19_0.6</td><td>M350_0.6</td><td>MS1_0.6</td><td>ENSG00000128016</td><td>ZFP36  </td><td>Homo_sapiens</td><td>D</td><td>CCCH ZF    </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>T02643_0.6</td><td>F17_0.6</td><td>TS19_0.6</td><td>M083_0.6</td><td>MS2_0.6</td><td>ENSG00000135482</td><td>ZC3H10 </td><td>Homo_sapiens</td><td>D</td><td>CCCH ZF    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>T02646_0.6</td><td>F17_0.6</td><td>TS19_0.6</td><td>M037_0.6</td><td>MS2_0.6</td><td>ENSG00000139793</td><td>MBNL2  </td><td>Homo_sapiens</td><td>I</td><td>CCCH ZF    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>T02652_0.6</td><td>F17_0.6</td><td>TS19_0.6</td><td>M037_0.6</td><td>MS2_0.6</td><td>ENSG00000152601</td><td>MBNL1  </td><td>Homo_sapiens</td><td>D</td><td>CCCH ZF    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>T02652_0.6</td><td>F17_0.6</td><td>TS19_0.6</td><td>M320_0.6</td><td>MS1_0.6</td><td>ENSG00000152601</td><td>MBNL1  </td><td>Homo_sapiens</td><td>D</td><td>CCCH ZF    </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>T06788_0.6</td><td>F20_0.6</td><td>TS19_0.6</td><td>M035_0.6</td><td>MS2_0.6</td><td>ENSG00000131914</td><td>LIN28A </td><td>Homo_sapiens</td><td>D</td><td>CCHC ZF,CSD</td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>T06788_0.6</td><td>F20_0.6</td><td>TS19_0.6</td><td>M153_0.6</td><td>MS2_0.6</td><td>ENSG00000131914</td><td>LIN28A </td><td>Homo_sapiens</td><td>D</td><td>CCHC ZF,CSD</td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>T06789_0.6</td><td>F20_0.6</td><td>TS19_0.6</td><td>M035_0.6</td><td>MS2_0.6</td><td>ENSG00000187772</td><td>LIN28B </td><td>Homo_sapiens</td><td>I</td><td>CCHC ZF,CSD</td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>T07243_0.6</td><td>F25_0.6</td><td>TS19_0.6</td><td>M071_0.6</td><td>MS2_0.6</td><td>ENSG00000115875</td><td>SRSF7  </td><td>Homo_sapiens</td><td>D</td><td>CCHC ZF,RRM</td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>T07243_0.6</td><td>F25_0.6</td><td>TS19_0.6</td><td>M331_0.6</td><td>MS1_0.6</td><td>ENSG00000115875</td><td>SRSF7  </td><td>Homo_sapiens</td><td>D</td><td>CCHC ZF,RRM</td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>T07243_0.6</td><td>F25_0.6</td><td>TS19_0.6</td><td>M353_0.6</td><td>MS1_0.6</td><td>ENSG00000115875</td><td>SRSF7  </td><td>Homo_sapiens</td><td>D</td><td>CCHC ZF,RRM</td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>14</th><td>T07244_0.6</td><td>F25_0.6</td><td>TS19_0.6</td><td>M085_0.6</td><td>MS2_0.6</td><td>ENSG00000139168</td><td>ZCRB1  </td><td>Homo_sapiens</td><td>D</td><td>CCHC ZF,RRM</td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>15</th><td>T07245_0.6</td><td>F25_0.6</td><td>TS19_0.6</td><td>M050_0.6</td><td>MS2_0.6</td><td>ENSG00000173914</td><td>RBM4B  </td><td>Homo_sapiens</td><td>I</td><td>CCHC ZF,RRM</td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>16</th><td>T07246_0.6</td><td>F25_0.6</td><td>TS19_0.6</td><td>M050_0.6</td><td>MS2_0.6</td><td>ENSG00000173933</td><td>RBM4   </td><td>Homo_sapiens</td><td>I</td><td>CCHC ZF,RRM</td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>17</th><td>T16852_0.6</td><td>F30_0.6</td><td>TS19_0.6</td><td>M082_0.6</td><td>MS2_0.6</td><td>ENSG00000006047</td><td>YBX2   </td><td>Homo_sapiens</td><td>D</td><td>CSD        </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>18</th><td>T16854_0.6</td><td>F30_0.6</td><td>TS19_0.6</td><td>M081_0.6</td><td>MS2_0.6</td><td>ENSG00000060138</td><td>CSDA   </td><td>Homo_sapiens</td><td>I</td><td>CSD        </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>19</th><td>T16855_0.6</td><td>F30_0.6</td><td>TS19_0.6</td><td>M081_0.6</td><td>MS2_0.6</td><td>ENSG00000065978</td><td>YB-1   </td><td>Homo_sapiens</td><td>D</td><td>CSD        </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>20</th><td>T16855_0.6</td><td>F30_0.6</td><td>TS19_0.6</td><td>M111_0.6</td><td>MS2_0.6</td><td>ENSG00000065978</td><td>YB-1   </td><td>Homo_sapiens</td><td>D</td><td>CSD        </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>21</th><td>T18786_0.6</td><td>F35_0.6</td><td>TS19_0.6</td><td>M188_0.6</td><td>MS2_0.6</td><td>ENSG00000090097</td><td>PCBP4  </td><td>Homo_sapiens</td><td>I</td><td>KH         </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>22</th><td>T18787_0.6</td><td>F35_0.6</td><td>TS19_0.6</td><td>M016_0.6</td><td>MS2_0.6</td><td>ENSG00000102081</td><td>FMR1   </td><td>Homo_sapiens</td><td>D</td><td>KH         </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>23</th><td>T18788_0.6</td><td>F35_0.6</td><td>TS19_0.6</td><td>M323_0.6</td><td>MS1_0.6</td><td>ENSG00000104967</td><td>NOVA2  </td><td>Homo_sapiens</td><td>D</td><td>KH         </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>24</th><td>T18788_0.6</td><td>F35_0.6</td><td>TS19_0.6</td><td>M325_0.6</td><td>MS1_0.6</td><td>ENSG00000104967</td><td>NOVA2  </td><td>Homo_sapiens</td><td>D</td><td>KH         </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>25</th><td>T18791_0.6</td><td>F35_0.6</td><td>TS19_0.6</td><td>M176_0.6</td><td>MS2_0.6</td><td>ENSG00000112232</td><td>KHDRBS2</td><td>Homo_sapiens</td><td>D</td><td>KH         </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>26</th><td>T18792_0.6</td><td>F35_0.6</td><td>TS19_0.6</td><td>M046_0.6</td><td>MS2_0.6</td><td>ENSG00000112531</td><td>QKI    </td><td>Homo_sapiens</td><td>D</td><td>KH         </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>27</th><td>T18792_0.6</td><td>F35_0.6</td><td>TS19_0.6</td><td>M262_0.6</td><td>MS1_0.6</td><td>ENSG00000112531</td><td>QKI    </td><td>Homo_sapiens</td><td>D</td><td>KH         </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>28</th><td>T18793_0.6</td><td>F35_0.6</td><td>TS19_0.6</td><td>M152_0.6</td><td>MS2_0.6</td><td>ENSG00000114416</td><td>FXR1   </td><td>Homo_sapiens</td><td>D</td><td>KH         </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>29</th><td>T18797_0.6</td><td>F35_0.6</td><td>TS19_0.6</td><td>M160_0.6</td><td>MS2_0.6</td><td>ENSG00000121774</td><td>KHDRBS1</td><td>Homo_sapiens</td><td>D</td><td>KH         </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>30</th><td>T18798_0.6</td><td>F35_0.6</td><td>TS19_0.6</td><td>M020_0.6</td><td>MS2_0.6</td><td>ENSG00000129245</td><td>FXR2   </td><td>Homo_sapiens</td><td>D</td><td>KH         </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>173</th><td>T37316_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M344_0.6</td><td>MS1_0.6</td><td>ENSG00000170748</td><td>RBMXL2         </td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>174</th><td>T37319_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M162_0.6</td><td>MS2_0.6</td><td>ENSG00000174740</td><td>PABPC5         </td><td>Homo_sapiens</td><td>D</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>175</th><td>T37320_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M344_0.6</td><td>MS1_0.6</td><td>ENSG00000175718</td><td>RBMXL3         </td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>176</th><td>T37324_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M158_0.6</td><td>MS2_0.6</td><td>ENSG00000179172</td><td>HNRNPCL1       </td><td>Homo_sapiens</td><td>D</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>177</th><td>T37328_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M070_0.6</td><td>MS2_0.6</td><td>ENSG00000180771</td><td>ENSG00000180771</td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>178</th><td>T37341_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M019_0.6</td><td>MS2_0.6</td><td>ENSG00000188529</td><td>Fusip1         </td><td>Homo_sapiens</td><td>D</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>179</th><td>T37341_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M086_0.6</td><td>MS2_0.6</td><td>ENSG00000188529</td><td>Fusip1         </td><td>Homo_sapiens</td><td>D</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>180</th><td>T37341_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M087_0.6</td><td>MS2_0.6</td><td>ENSG00000188529</td><td>Fusip1         </td><td>Homo_sapiens</td><td>D</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>181</th><td>T37341_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M088_0.6</td><td>MS2_0.6</td><td>ENSG00000188529</td><td>Fusip1         </td><td>Homo_sapiens</td><td>D</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>182</th><td>T37350_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M148_0.6</td><td>MS2_0.6</td><td>ENSG00000205022</td><td>PABPN1L        </td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>183</th><td>T37357_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M143_0.6</td><td>MS2_0.6</td><td>ENSG00000213250</td><td>ENSG00000213250</td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>184</th><td>T37358_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M344_0.6</td><td>MS1_0.6</td><td>ENSG00000213516</td><td>RBMXL1         </td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>185</th><td>T37360_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M089_0.6</td><td>MS2_0.6</td><td>ENSG00000215042</td><td>ENSG00000215042</td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>186</th><td>T37361_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M022_0.6</td><td>MS2_0.6</td><td>ENSG00000215492</td><td>ENSG00000215492</td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>187</th><td>T37362_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M019_0.6</td><td>MS2_0.6</td><td>ENSG00000215699</td><td>BX511012.1     </td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>188</th><td>T37363_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M246_0.6</td><td>MS1_0.6</td><td>ENSG00000226941</td><td>RBMY1J         </td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>189</th><td>T37366_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M022_0.6</td><td>MS2_0.6</td><td>ENSG00000231942</td><td>ENSG00000231942</td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>190</th><td>T37368_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M246_0.6</td><td>MS1_0.6</td><td>ENSG00000234414</td><td>RBMY1A1        </td><td>Homo_sapiens</td><td>D</td><td>RRM    </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>191</th><td>T37368_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M247_0.6</td><td>MS1_0.6</td><td>ENSG00000234414</td><td>RBMY1A1        </td><td>Homo_sapiens</td><td>D</td><td>RRM    </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>192</th><td>T37369_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M050_0.6</td><td>MS2_0.6</td><td>ENSG00000239306</td><td>RBM4           </td><td>Homo_sapiens</td><td>D</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>193</th><td>T37369_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M109_0.6</td><td>MS2_0.6</td><td>ENSG00000239306</td><td>RBM4           </td><td>Homo_sapiens</td><td>D</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>194</th><td>T37370_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M246_0.6</td><td>MS1_0.6</td><td>ENSG00000242389</td><td>RBMY1E         </td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>195</th><td>T37371_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M246_0.6</td><td>MS1_0.6</td><td>ENSG00000242875</td><td>RBMY1B         </td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>196</th><td>T37372_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M246_0.6</td><td>MS1_0.6</td><td>ENSG00000244395</td><td>RBMY1D         </td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>197</th><td>T37374_0.6</td><td>F49_0.6</td><td>TS19_0.6</td><td>M042_0.6</td><td>MS2_0.6</td><td>ENSG00000250177</td><td>ENSG00000250177</td><td>Homo_sapiens</td><td>I</td><td>RRM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>198</th><td>T60363_0.6</td><td>F52_0.6</td><td>TS19_0.6</td><td>M231_0.6</td><td>MS2_0.6</td><td>ENSG00000134001</td><td>EIF2S1         </td><td>Homo_sapiens</td><td>I</td><td>S1     </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>199</th><td>T61784_0.6</td><td>F53_0.6</td><td>TS19_0.6</td><td>M061_0.6</td><td>MS2_0.6</td><td>ENSG00000020577</td><td>SAMD4A         </td><td>Homo_sapiens</td><td>D</td><td>SAM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>200</th><td>T61785_0.6</td><td>F53_0.6</td><td>TS19_0.6</td><td>M061_0.6</td><td>MS2_0.6</td><td>ENSG00000179134</td><td>SAMD4B         </td><td>Homo_sapiens</td><td>I</td><td>SAM    </td><td>⋯</td><td>RNAcompete</td><td>Ray </td><td>2013</td><td>NULL    </td><td>NULL  </td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>201</th><td>T61860_0.6</td><td>F54_0.6</td><td>TS19_0.6</td><td>M256_0.6</td><td>MS1_0.6</td><td>ENSG00000122729</td><td>ACO1           </td><td>Homo_sapiens</td><td>D</td><td>Unknown</td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>202</th><td>T62023_0.6</td><td>F55_0.6</td><td>TS19_0.6</td><td>M354_0.6</td><td>MS1_0.6</td><td>ENSG00000083896</td><td>YTHDC1         </td><td>Homo_sapiens</td><td>D</td><td>YTH    </td><td>⋯</td><td>RBPDB     </td><td>Cook</td><td>2011</td><td>21036867</td><td>v1.2.1</td><td>Ensembl</td><td>http://www.ensembl.org/</td><td>2011</td><td>Oct</td><td>26</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 202 × 26\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & RBP\\_ID & Family\\_ID & RSource\\_ID & Motif\\_ID & MSource\\_ID & DBID & RBP\\_Name & RBP\\_Species & RBP\\_Status & Family\\_Name & ⋯ & MSource\\_Type & MSource\\_Author & MSource\\_Year & PMID & MSource\\_Version & RBPSource\\_Name & RBPSource\\_URL & RBPSource\\_Year & RBPSource\\_Month & RBPSource\\_Day\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & T02631\\_0.6 & F17\\_0.6 & TS19\\_0.6 & M037\\_0.6 & MS2\\_0.6 & ENSG00000076770 & MBNL3   & Homo\\_sapiens & I & CCCH ZF     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t2 & T02638\\_0.6 & F17\\_0.6 & TS19\\_0.6 & M269\\_0.6 & MS1\\_0.6 & ENSG00000128016 & ZFP36   & Homo\\_sapiens & D & CCCH ZF     & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t3 & T02638\\_0.6 & F17\\_0.6 & TS19\\_0.6 & M350\\_0.6 & MS1\\_0.6 & ENSG00000128016 & ZFP36   & Homo\\_sapiens & D & CCCH ZF     & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t4 & T02643\\_0.6 & F17\\_0.6 & TS19\\_0.6 & M083\\_0.6 & MS2\\_0.6 & ENSG00000135482 & ZC3H10  & Homo\\_sapiens & D & CCCH ZF     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t5 & T02646\\_0.6 & F17\\_0.6 & TS19\\_0.6 & M037\\_0.6 & MS2\\_0.6 & ENSG00000139793 & MBNL2   & Homo\\_sapiens & I & CCCH ZF     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t6 & T02652\\_0.6 & F17\\_0.6 & TS19\\_0.6 & M037\\_0.6 & MS2\\_0.6 & ENSG00000152601 & MBNL1   & Homo\\_sapiens & D & CCCH ZF     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t7 & T02652\\_0.6 & F17\\_0.6 & TS19\\_0.6 & M320\\_0.6 & MS1\\_0.6 & ENSG00000152601 & MBNL1   & Homo\\_sapiens & D & CCCH ZF     & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t8 & T06788\\_0.6 & F20\\_0.6 & TS19\\_0.6 & M035\\_0.6 & MS2\\_0.6 & ENSG00000131914 & LIN28A  & Homo\\_sapiens & D & CCHC ZF,CSD & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t9 & T06788\\_0.6 & F20\\_0.6 & TS19\\_0.6 & M153\\_0.6 & MS2\\_0.6 & ENSG00000131914 & LIN28A  & Homo\\_sapiens & D & CCHC ZF,CSD & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t10 & T06789\\_0.6 & F20\\_0.6 & TS19\\_0.6 & M035\\_0.6 & MS2\\_0.6 & ENSG00000187772 & LIN28B  & Homo\\_sapiens & I & CCHC ZF,CSD & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t11 & T07243\\_0.6 & F25\\_0.6 & TS19\\_0.6 & M071\\_0.6 & MS2\\_0.6 & ENSG00000115875 & SRSF7   & Homo\\_sapiens & D & CCHC ZF,RRM & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t12 & T07243\\_0.6 & F25\\_0.6 & TS19\\_0.6 & M331\\_0.6 & MS1\\_0.6 & ENSG00000115875 & SRSF7   & Homo\\_sapiens & D & CCHC ZF,RRM & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t13 & T07243\\_0.6 & F25\\_0.6 & TS19\\_0.6 & M353\\_0.6 & MS1\\_0.6 & ENSG00000115875 & SRSF7   & Homo\\_sapiens & D & CCHC ZF,RRM & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t14 & T07244\\_0.6 & F25\\_0.6 & TS19\\_0.6 & M085\\_0.6 & MS2\\_0.6 & ENSG00000139168 & ZCRB1   & Homo\\_sapiens & D & CCHC ZF,RRM & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t15 & T07245\\_0.6 & F25\\_0.6 & TS19\\_0.6 & M050\\_0.6 & MS2\\_0.6 & ENSG00000173914 & RBM4B   & Homo\\_sapiens & I & CCHC ZF,RRM & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t16 & T07246\\_0.6 & F25\\_0.6 & TS19\\_0.6 & M050\\_0.6 & MS2\\_0.6 & ENSG00000173933 & RBM4    & Homo\\_sapiens & I & CCHC ZF,RRM & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t17 & T16852\\_0.6 & F30\\_0.6 & TS19\\_0.6 & M082\\_0.6 & MS2\\_0.6 & ENSG00000006047 & YBX2    & Homo\\_sapiens & D & CSD         & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t18 & T16854\\_0.6 & F30\\_0.6 & TS19\\_0.6 & M081\\_0.6 & MS2\\_0.6 & ENSG00000060138 & CSDA    & Homo\\_sapiens & I & CSD         & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t19 & T16855\\_0.6 & F30\\_0.6 & TS19\\_0.6 & M081\\_0.6 & MS2\\_0.6 & ENSG00000065978 & YB-1    & Homo\\_sapiens & D & CSD         & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t20 & T16855\\_0.6 & F30\\_0.6 & TS19\\_0.6 & M111\\_0.6 & MS2\\_0.6 & ENSG00000065978 & YB-1    & Homo\\_sapiens & D & CSD         & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t21 & T18786\\_0.6 & F35\\_0.6 & TS19\\_0.6 & M188\\_0.6 & MS2\\_0.6 & ENSG00000090097 & PCBP4   & Homo\\_sapiens & I & KH          & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t22 & T18787\\_0.6 & F35\\_0.6 & TS19\\_0.6 & M016\\_0.6 & MS2\\_0.6 & ENSG00000102081 & FMR1    & Homo\\_sapiens & D & KH          & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t23 & T18788\\_0.6 & F35\\_0.6 & TS19\\_0.6 & M323\\_0.6 & MS1\\_0.6 & ENSG00000104967 & NOVA2   & Homo\\_sapiens & D & KH          & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t24 & T18788\\_0.6 & F35\\_0.6 & TS19\\_0.6 & M325\\_0.6 & MS1\\_0.6 & ENSG00000104967 & NOVA2   & Homo\\_sapiens & D & KH          & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t25 & T18791\\_0.6 & F35\\_0.6 & TS19\\_0.6 & M176\\_0.6 & MS2\\_0.6 & ENSG00000112232 & KHDRBS2 & Homo\\_sapiens & D & KH          & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t26 & T18792\\_0.6 & F35\\_0.6 & TS19\\_0.6 & M046\\_0.6 & MS2\\_0.6 & ENSG00000112531 & QKI     & Homo\\_sapiens & D & KH          & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t27 & T18792\\_0.6 & F35\\_0.6 & TS19\\_0.6 & M262\\_0.6 & MS1\\_0.6 & ENSG00000112531 & QKI     & Homo\\_sapiens & D & KH          & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t28 & T18793\\_0.6 & F35\\_0.6 & TS19\\_0.6 & M152\\_0.6 & MS2\\_0.6 & ENSG00000114416 & FXR1    & Homo\\_sapiens & D & KH          & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t29 & T18797\\_0.6 & F35\\_0.6 & TS19\\_0.6 & M160\\_0.6 & MS2\\_0.6 & ENSG00000121774 & KHDRBS1 & Homo\\_sapiens & D & KH          & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t30 & T18798\\_0.6 & F35\\_0.6 & TS19\\_0.6 & M020\\_0.6 & MS2\\_0.6 & ENSG00000129245 & FXR2    & Homo\\_sapiens & D & KH          & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t173 & T37316\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M344\\_0.6 & MS1\\_0.6 & ENSG00000170748 & RBMXL2          & Homo\\_sapiens & I & RRM     & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t174 & T37319\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M162\\_0.6 & MS2\\_0.6 & ENSG00000174740 & PABPC5          & Homo\\_sapiens & D & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t175 & T37320\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M344\\_0.6 & MS1\\_0.6 & ENSG00000175718 & RBMXL3          & Homo\\_sapiens & I & RRM     & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t176 & T37324\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M158\\_0.6 & MS2\\_0.6 & ENSG00000179172 & HNRNPCL1        & Homo\\_sapiens & D & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t177 & T37328\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M070\\_0.6 & MS2\\_0.6 & ENSG00000180771 & ENSG00000180771 & Homo\\_sapiens & I & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t178 & T37341\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M019\\_0.6 & MS2\\_0.6 & ENSG00000188529 & Fusip1          & Homo\\_sapiens & D & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t179 & T37341\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M086\\_0.6 & MS2\\_0.6 & ENSG00000188529 & Fusip1          & Homo\\_sapiens & D & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t180 & T37341\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M087\\_0.6 & MS2\\_0.6 & ENSG00000188529 & Fusip1          & Homo\\_sapiens & D & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t181 & T37341\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M088\\_0.6 & MS2\\_0.6 & ENSG00000188529 & Fusip1          & Homo\\_sapiens & D & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t182 & T37350\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M148\\_0.6 & MS2\\_0.6 & ENSG00000205022 & PABPN1L         & Homo\\_sapiens & I & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t183 & T37357\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M143\\_0.6 & MS2\\_0.6 & ENSG00000213250 & ENSG00000213250 & Homo\\_sapiens & I & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t184 & T37358\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M344\\_0.6 & MS1\\_0.6 & ENSG00000213516 & RBMXL1          & Homo\\_sapiens & I & RRM     & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t185 & T37360\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M089\\_0.6 & MS2\\_0.6 & ENSG00000215042 & ENSG00000215042 & Homo\\_sapiens & I & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t186 & T37361\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M022\\_0.6 & MS2\\_0.6 & ENSG00000215492 & ENSG00000215492 & Homo\\_sapiens & I & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t187 & T37362\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M019\\_0.6 & MS2\\_0.6 & ENSG00000215699 & BX511012.1      & Homo\\_sapiens & I & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t188 & T37363\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M246\\_0.6 & MS1\\_0.6 & ENSG00000226941 & RBMY1J          & Homo\\_sapiens & I & RRM     & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t189 & T37366\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M022\\_0.6 & MS2\\_0.6 & ENSG00000231942 & ENSG00000231942 & Homo\\_sapiens & I & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t190 & T37368\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M246\\_0.6 & MS1\\_0.6 & ENSG00000234414 & RBMY1A1         & Homo\\_sapiens & D & RRM     & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t191 & T37368\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M247\\_0.6 & MS1\\_0.6 & ENSG00000234414 & RBMY1A1         & Homo\\_sapiens & D & RRM     & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t192 & T37369\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M050\\_0.6 & MS2\\_0.6 & ENSG00000239306 & RBM4            & Homo\\_sapiens & D & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t193 & T37369\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M109\\_0.6 & MS2\\_0.6 & ENSG00000239306 & RBM4            & Homo\\_sapiens & D & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t194 & T37370\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M246\\_0.6 & MS1\\_0.6 & ENSG00000242389 & RBMY1E          & Homo\\_sapiens & I & RRM     & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t195 & T37371\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M246\\_0.6 & MS1\\_0.6 & ENSG00000242875 & RBMY1B          & Homo\\_sapiens & I & RRM     & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t196 & T37372\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M246\\_0.6 & MS1\\_0.6 & ENSG00000244395 & RBMY1D          & Homo\\_sapiens & I & RRM     & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t197 & T37374\\_0.6 & F49\\_0.6 & TS19\\_0.6 & M042\\_0.6 & MS2\\_0.6 & ENSG00000250177 & ENSG00000250177 & Homo\\_sapiens & I & RRM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t198 & T60363\\_0.6 & F52\\_0.6 & TS19\\_0.6 & M231\\_0.6 & MS2\\_0.6 & ENSG00000134001 & EIF2S1          & Homo\\_sapiens & I & S1      & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t199 & T61784\\_0.6 & F53\\_0.6 & TS19\\_0.6 & M061\\_0.6 & MS2\\_0.6 & ENSG00000020577 & SAMD4A          & Homo\\_sapiens & D & SAM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t200 & T61785\\_0.6 & F53\\_0.6 & TS19\\_0.6 & M061\\_0.6 & MS2\\_0.6 & ENSG00000179134 & SAMD4B          & Homo\\_sapiens & I & SAM     & ⋯ & RNAcompete & Ray  & 2013 & NULL     & NULL   & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t201 & T61860\\_0.6 & F54\\_0.6 & TS19\\_0.6 & M256\\_0.6 & MS1\\_0.6 & ENSG00000122729 & ACO1            & Homo\\_sapiens & D & Unknown & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\t202 & T62023\\_0.6 & F55\\_0.6 & TS19\\_0.6 & M354\\_0.6 & MS1\\_0.6 & ENSG00000083896 & YTHDC1          & Homo\\_sapiens & D & YTH     & ⋯ & RBPDB      & Cook & 2011 & 21036867 & v1.2.1 & Ensembl & http://www.ensembl.org/ & 2011 & Oct & 26\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 202 × 26\n",
       "\n",
       "| <!--/--> | RBP_ID &lt;chr&gt; | Family_ID &lt;chr&gt; | RSource_ID &lt;chr&gt; | Motif_ID &lt;chr&gt; | MSource_ID &lt;chr&gt; | DBID &lt;chr&gt; | RBP_Name &lt;chr&gt; | RBP_Species &lt;chr&gt; | RBP_Status &lt;chr&gt; | Family_Name &lt;chr&gt; | ⋯ ⋯ | MSource_Type &lt;chr&gt; | MSource_Author &lt;chr&gt; | MSource_Year &lt;chr&gt; | PMID &lt;chr&gt; | MSource_Version &lt;chr&gt; | RBPSource_Name &lt;chr&gt; | RBPSource_URL &lt;chr&gt; | RBPSource_Year &lt;int&gt; | RBPSource_Month &lt;chr&gt; | RBPSource_Day &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | T02631_0.6 | F17_0.6 | TS19_0.6 | M037_0.6 | MS2_0.6 | ENSG00000076770 | MBNL3   | Homo_sapiens | I | CCCH ZF     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 2 | T02638_0.6 | F17_0.6 | TS19_0.6 | M269_0.6 | MS1_0.6 | ENSG00000128016 | ZFP36   | Homo_sapiens | D | CCCH ZF     | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 3 | T02638_0.6 | F17_0.6 | TS19_0.6 | M350_0.6 | MS1_0.6 | ENSG00000128016 | ZFP36   | Homo_sapiens | D | CCCH ZF     | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 4 | T02643_0.6 | F17_0.6 | TS19_0.6 | M083_0.6 | MS2_0.6 | ENSG00000135482 | ZC3H10  | Homo_sapiens | D | CCCH ZF     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 5 | T02646_0.6 | F17_0.6 | TS19_0.6 | M037_0.6 | MS2_0.6 | ENSG00000139793 | MBNL2   | Homo_sapiens | I | CCCH ZF     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 6 | T02652_0.6 | F17_0.6 | TS19_0.6 | M037_0.6 | MS2_0.6 | ENSG00000152601 | MBNL1   | Homo_sapiens | D | CCCH ZF     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 7 | T02652_0.6 | F17_0.6 | TS19_0.6 | M320_0.6 | MS1_0.6 | ENSG00000152601 | MBNL1   | Homo_sapiens | D | CCCH ZF     | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 8 | T06788_0.6 | F20_0.6 | TS19_0.6 | M035_0.6 | MS2_0.6 | ENSG00000131914 | LIN28A  | Homo_sapiens | D | CCHC ZF,CSD | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 9 | T06788_0.6 | F20_0.6 | TS19_0.6 | M153_0.6 | MS2_0.6 | ENSG00000131914 | LIN28A  | Homo_sapiens | D | CCHC ZF,CSD | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 10 | T06789_0.6 | F20_0.6 | TS19_0.6 | M035_0.6 | MS2_0.6 | ENSG00000187772 | LIN28B  | Homo_sapiens | I | CCHC ZF,CSD | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 11 | T07243_0.6 | F25_0.6 | TS19_0.6 | M071_0.6 | MS2_0.6 | ENSG00000115875 | SRSF7   | Homo_sapiens | D | CCHC ZF,RRM | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 12 | T07243_0.6 | F25_0.6 | TS19_0.6 | M331_0.6 | MS1_0.6 | ENSG00000115875 | SRSF7   | Homo_sapiens | D | CCHC ZF,RRM | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 13 | T07243_0.6 | F25_0.6 | TS19_0.6 | M353_0.6 | MS1_0.6 | ENSG00000115875 | SRSF7   | Homo_sapiens | D | CCHC ZF,RRM | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 14 | T07244_0.6 | F25_0.6 | TS19_0.6 | M085_0.6 | MS2_0.6 | ENSG00000139168 | ZCRB1   | Homo_sapiens | D | CCHC ZF,RRM | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 15 | T07245_0.6 | F25_0.6 | TS19_0.6 | M050_0.6 | MS2_0.6 | ENSG00000173914 | RBM4B   | Homo_sapiens | I | CCHC ZF,RRM | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 16 | T07246_0.6 | F25_0.6 | TS19_0.6 | M050_0.6 | MS2_0.6 | ENSG00000173933 | RBM4    | Homo_sapiens | I | CCHC ZF,RRM | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 17 | T16852_0.6 | F30_0.6 | TS19_0.6 | M082_0.6 | MS2_0.6 | ENSG00000006047 | YBX2    | Homo_sapiens | D | CSD         | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 18 | T16854_0.6 | F30_0.6 | TS19_0.6 | M081_0.6 | MS2_0.6 | ENSG00000060138 | CSDA    | Homo_sapiens | I | CSD         | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 19 | T16855_0.6 | F30_0.6 | TS19_0.6 | M081_0.6 | MS2_0.6 | ENSG00000065978 | YB-1    | Homo_sapiens | D | CSD         | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 20 | T16855_0.6 | F30_0.6 | TS19_0.6 | M111_0.6 | MS2_0.6 | ENSG00000065978 | YB-1    | Homo_sapiens | D | CSD         | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 21 | T18786_0.6 | F35_0.6 | TS19_0.6 | M188_0.6 | MS2_0.6 | ENSG00000090097 | PCBP4   | Homo_sapiens | I | KH          | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 22 | T18787_0.6 | F35_0.6 | TS19_0.6 | M016_0.6 | MS2_0.6 | ENSG00000102081 | FMR1    | Homo_sapiens | D | KH          | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 23 | T18788_0.6 | F35_0.6 | TS19_0.6 | M323_0.6 | MS1_0.6 | ENSG00000104967 | NOVA2   | Homo_sapiens | D | KH          | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 24 | T18788_0.6 | F35_0.6 | TS19_0.6 | M325_0.6 | MS1_0.6 | ENSG00000104967 | NOVA2   | Homo_sapiens | D | KH          | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 25 | T18791_0.6 | F35_0.6 | TS19_0.6 | M176_0.6 | MS2_0.6 | ENSG00000112232 | KHDRBS2 | Homo_sapiens | D | KH          | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 26 | T18792_0.6 | F35_0.6 | TS19_0.6 | M046_0.6 | MS2_0.6 | ENSG00000112531 | QKI     | Homo_sapiens | D | KH          | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 27 | T18792_0.6 | F35_0.6 | TS19_0.6 | M262_0.6 | MS1_0.6 | ENSG00000112531 | QKI     | Homo_sapiens | D | KH          | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 28 | T18793_0.6 | F35_0.6 | TS19_0.6 | M152_0.6 | MS2_0.6 | ENSG00000114416 | FXR1    | Homo_sapiens | D | KH          | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 29 | T18797_0.6 | F35_0.6 | TS19_0.6 | M160_0.6 | MS2_0.6 | ENSG00000121774 | KHDRBS1 | Homo_sapiens | D | KH          | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 30 | T18798_0.6 | F35_0.6 | TS19_0.6 | M020_0.6 | MS2_0.6 | ENSG00000129245 | FXR2    | Homo_sapiens | D | KH          | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 173 | T37316_0.6 | F49_0.6 | TS19_0.6 | M344_0.6 | MS1_0.6 | ENSG00000170748 | RBMXL2          | Homo_sapiens | I | RRM     | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 174 | T37319_0.6 | F49_0.6 | TS19_0.6 | M162_0.6 | MS2_0.6 | ENSG00000174740 | PABPC5          | Homo_sapiens | D | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 175 | T37320_0.6 | F49_0.6 | TS19_0.6 | M344_0.6 | MS1_0.6 | ENSG00000175718 | RBMXL3          | Homo_sapiens | I | RRM     | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 176 | T37324_0.6 | F49_0.6 | TS19_0.6 | M158_0.6 | MS2_0.6 | ENSG00000179172 | HNRNPCL1        | Homo_sapiens | D | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 177 | T37328_0.6 | F49_0.6 | TS19_0.6 | M070_0.6 | MS2_0.6 | ENSG00000180771 | ENSG00000180771 | Homo_sapiens | I | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 178 | T37341_0.6 | F49_0.6 | TS19_0.6 | M019_0.6 | MS2_0.6 | ENSG00000188529 | Fusip1          | Homo_sapiens | D | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 179 | T37341_0.6 | F49_0.6 | TS19_0.6 | M086_0.6 | MS2_0.6 | ENSG00000188529 | Fusip1          | Homo_sapiens | D | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 180 | T37341_0.6 | F49_0.6 | TS19_0.6 | M087_0.6 | MS2_0.6 | ENSG00000188529 | Fusip1          | Homo_sapiens | D | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 181 | T37341_0.6 | F49_0.6 | TS19_0.6 | M088_0.6 | MS2_0.6 | ENSG00000188529 | Fusip1          | Homo_sapiens | D | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 182 | T37350_0.6 | F49_0.6 | TS19_0.6 | M148_0.6 | MS2_0.6 | ENSG00000205022 | PABPN1L         | Homo_sapiens | I | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 183 | T37357_0.6 | F49_0.6 | TS19_0.6 | M143_0.6 | MS2_0.6 | ENSG00000213250 | ENSG00000213250 | Homo_sapiens | I | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 184 | T37358_0.6 | F49_0.6 | TS19_0.6 | M344_0.6 | MS1_0.6 | ENSG00000213516 | RBMXL1          | Homo_sapiens | I | RRM     | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 185 | T37360_0.6 | F49_0.6 | TS19_0.6 | M089_0.6 | MS2_0.6 | ENSG00000215042 | ENSG00000215042 | Homo_sapiens | I | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 186 | T37361_0.6 | F49_0.6 | TS19_0.6 | M022_0.6 | MS2_0.6 | ENSG00000215492 | ENSG00000215492 | Homo_sapiens | I | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 187 | T37362_0.6 | F49_0.6 | TS19_0.6 | M019_0.6 | MS2_0.6 | ENSG00000215699 | BX511012.1      | Homo_sapiens | I | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 188 | T37363_0.6 | F49_0.6 | TS19_0.6 | M246_0.6 | MS1_0.6 | ENSG00000226941 | RBMY1J          | Homo_sapiens | I | RRM     | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 189 | T37366_0.6 | F49_0.6 | TS19_0.6 | M022_0.6 | MS2_0.6 | ENSG00000231942 | ENSG00000231942 | Homo_sapiens | I | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 190 | T37368_0.6 | F49_0.6 | TS19_0.6 | M246_0.6 | MS1_0.6 | ENSG00000234414 | RBMY1A1         | Homo_sapiens | D | RRM     | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 191 | T37368_0.6 | F49_0.6 | TS19_0.6 | M247_0.6 | MS1_0.6 | ENSG00000234414 | RBMY1A1         | Homo_sapiens | D | RRM     | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 192 | T37369_0.6 | F49_0.6 | TS19_0.6 | M050_0.6 | MS2_0.6 | ENSG00000239306 | RBM4            | Homo_sapiens | D | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 193 | T37369_0.6 | F49_0.6 | TS19_0.6 | M109_0.6 | MS2_0.6 | ENSG00000239306 | RBM4            | Homo_sapiens | D | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 194 | T37370_0.6 | F49_0.6 | TS19_0.6 | M246_0.6 | MS1_0.6 | ENSG00000242389 | RBMY1E          | Homo_sapiens | I | RRM     | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 195 | T37371_0.6 | F49_0.6 | TS19_0.6 | M246_0.6 | MS1_0.6 | ENSG00000242875 | RBMY1B          | Homo_sapiens | I | RRM     | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 196 | T37372_0.6 | F49_0.6 | TS19_0.6 | M246_0.6 | MS1_0.6 | ENSG00000244395 | RBMY1D          | Homo_sapiens | I | RRM     | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 197 | T37374_0.6 | F49_0.6 | TS19_0.6 | M042_0.6 | MS2_0.6 | ENSG00000250177 | ENSG00000250177 | Homo_sapiens | I | RRM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 198 | T60363_0.6 | F52_0.6 | TS19_0.6 | M231_0.6 | MS2_0.6 | ENSG00000134001 | EIF2S1          | Homo_sapiens | I | S1      | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 199 | T61784_0.6 | F53_0.6 | TS19_0.6 | M061_0.6 | MS2_0.6 | ENSG00000020577 | SAMD4A          | Homo_sapiens | D | SAM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 200 | T61785_0.6 | F53_0.6 | TS19_0.6 | M061_0.6 | MS2_0.6 | ENSG00000179134 | SAMD4B          | Homo_sapiens | I | SAM     | ⋯ | RNAcompete | Ray  | 2013 | NULL     | NULL   | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 201 | T61860_0.6 | F54_0.6 | TS19_0.6 | M256_0.6 | MS1_0.6 | ENSG00000122729 | ACO1            | Homo_sapiens | D | Unknown | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "| 202 | T62023_0.6 | F55_0.6 | TS19_0.6 | M354_0.6 | MS1_0.6 | ENSG00000083896 | YTHDC1          | Homo_sapiens | D | YTH     | ⋯ | RBPDB      | Cook | 2011 | 21036867 | v1.2.1 | Ensembl | http://www.ensembl.org/ | 2011 | Oct | 26 |\n",
       "\n"
      ],
      "text/plain": [
       "    RBP_ID     Family_ID RSource_ID Motif_ID MSource_ID DBID           \n",
       "1   T02631_0.6 F17_0.6   TS19_0.6   M037_0.6 MS2_0.6    ENSG00000076770\n",
       "2   T02638_0.6 F17_0.6   TS19_0.6   M269_0.6 MS1_0.6    ENSG00000128016\n",
       "3   T02638_0.6 F17_0.6   TS19_0.6   M350_0.6 MS1_0.6    ENSG00000128016\n",
       "4   T02643_0.6 F17_0.6   TS19_0.6   M083_0.6 MS2_0.6    ENSG00000135482\n",
       "5   T02646_0.6 F17_0.6   TS19_0.6   M037_0.6 MS2_0.6    ENSG00000139793\n",
       "6   T02652_0.6 F17_0.6   TS19_0.6   M037_0.6 MS2_0.6    ENSG00000152601\n",
       "7   T02652_0.6 F17_0.6   TS19_0.6   M320_0.6 MS1_0.6    ENSG00000152601\n",
       "8   T06788_0.6 F20_0.6   TS19_0.6   M035_0.6 MS2_0.6    ENSG00000131914\n",
       "9   T06788_0.6 F20_0.6   TS19_0.6   M153_0.6 MS2_0.6    ENSG00000131914\n",
       "10  T06789_0.6 F20_0.6   TS19_0.6   M035_0.6 MS2_0.6    ENSG00000187772\n",
       "11  T07243_0.6 F25_0.6   TS19_0.6   M071_0.6 MS2_0.6    ENSG00000115875\n",
       "12  T07243_0.6 F25_0.6   TS19_0.6   M331_0.6 MS1_0.6    ENSG00000115875\n",
       "13  T07243_0.6 F25_0.6   TS19_0.6   M353_0.6 MS1_0.6    ENSG00000115875\n",
       "14  T07244_0.6 F25_0.6   TS19_0.6   M085_0.6 MS2_0.6    ENSG00000139168\n",
       "15  T07245_0.6 F25_0.6   TS19_0.6   M050_0.6 MS2_0.6    ENSG00000173914\n",
       "16  T07246_0.6 F25_0.6   TS19_0.6   M050_0.6 MS2_0.6    ENSG00000173933\n",
       "17  T16852_0.6 F30_0.6   TS19_0.6   M082_0.6 MS2_0.6    ENSG00000006047\n",
       "18  T16854_0.6 F30_0.6   TS19_0.6   M081_0.6 MS2_0.6    ENSG00000060138\n",
       "19  T16855_0.6 F30_0.6   TS19_0.6   M081_0.6 MS2_0.6    ENSG00000065978\n",
       "20  T16855_0.6 F30_0.6   TS19_0.6   M111_0.6 MS2_0.6    ENSG00000065978\n",
       "21  T18786_0.6 F35_0.6   TS19_0.6   M188_0.6 MS2_0.6    ENSG00000090097\n",
       "22  T18787_0.6 F35_0.6   TS19_0.6   M016_0.6 MS2_0.6    ENSG00000102081\n",
       "23  T18788_0.6 F35_0.6   TS19_0.6   M323_0.6 MS1_0.6    ENSG00000104967\n",
       "24  T18788_0.6 F35_0.6   TS19_0.6   M325_0.6 MS1_0.6    ENSG00000104967\n",
       "25  T18791_0.6 F35_0.6   TS19_0.6   M176_0.6 MS2_0.6    ENSG00000112232\n",
       "26  T18792_0.6 F35_0.6   TS19_0.6   M046_0.6 MS2_0.6    ENSG00000112531\n",
       "27  T18792_0.6 F35_0.6   TS19_0.6   M262_0.6 MS1_0.6    ENSG00000112531\n",
       "28  T18793_0.6 F35_0.6   TS19_0.6   M152_0.6 MS2_0.6    ENSG00000114416\n",
       "29  T18797_0.6 F35_0.6   TS19_0.6   M160_0.6 MS2_0.6    ENSG00000121774\n",
       "30  T18798_0.6 F35_0.6   TS19_0.6   M020_0.6 MS2_0.6    ENSG00000129245\n",
       "<U+22EE>   <U+22EE>   <U+22EE>  <U+22EE>   <U+22EE> <U+22EE>   <U+22EE>       \n",
       "173 T37316_0.6 F49_0.6   TS19_0.6   M344_0.6 MS1_0.6    ENSG00000170748\n",
       "174 T37319_0.6 F49_0.6   TS19_0.6   M162_0.6 MS2_0.6    ENSG00000174740\n",
       "175 T37320_0.6 F49_0.6   TS19_0.6   M344_0.6 MS1_0.6    ENSG00000175718\n",
       "176 T37324_0.6 F49_0.6   TS19_0.6   M158_0.6 MS2_0.6    ENSG00000179172\n",
       "177 T37328_0.6 F49_0.6   TS19_0.6   M070_0.6 MS2_0.6    ENSG00000180771\n",
       "178 T37341_0.6 F49_0.6   TS19_0.6   M019_0.6 MS2_0.6    ENSG00000188529\n",
       "179 T37341_0.6 F49_0.6   TS19_0.6   M086_0.6 MS2_0.6    ENSG00000188529\n",
       "180 T37341_0.6 F49_0.6   TS19_0.6   M087_0.6 MS2_0.6    ENSG00000188529\n",
       "181 T37341_0.6 F49_0.6   TS19_0.6   M088_0.6 MS2_0.6    ENSG00000188529\n",
       "182 T37350_0.6 F49_0.6   TS19_0.6   M148_0.6 MS2_0.6    ENSG00000205022\n",
       "183 T37357_0.6 F49_0.6   TS19_0.6   M143_0.6 MS2_0.6    ENSG00000213250\n",
       "184 T37358_0.6 F49_0.6   TS19_0.6   M344_0.6 MS1_0.6    ENSG00000213516\n",
       "185 T37360_0.6 F49_0.6   TS19_0.6   M089_0.6 MS2_0.6    ENSG00000215042\n",
       "186 T37361_0.6 F49_0.6   TS19_0.6   M022_0.6 MS2_0.6    ENSG00000215492\n",
       "187 T37362_0.6 F49_0.6   TS19_0.6   M019_0.6 MS2_0.6    ENSG00000215699\n",
       "188 T37363_0.6 F49_0.6   TS19_0.6   M246_0.6 MS1_0.6    ENSG00000226941\n",
       "189 T37366_0.6 F49_0.6   TS19_0.6   M022_0.6 MS2_0.6    ENSG00000231942\n",
       "190 T37368_0.6 F49_0.6   TS19_0.6   M246_0.6 MS1_0.6    ENSG00000234414\n",
       "191 T37368_0.6 F49_0.6   TS19_0.6   M247_0.6 MS1_0.6    ENSG00000234414\n",
       "192 T37369_0.6 F49_0.6   TS19_0.6   M050_0.6 MS2_0.6    ENSG00000239306\n",
       "193 T37369_0.6 F49_0.6   TS19_0.6   M109_0.6 MS2_0.6    ENSG00000239306\n",
       "194 T37370_0.6 F49_0.6   TS19_0.6   M246_0.6 MS1_0.6    ENSG00000242389\n",
       "195 T37371_0.6 F49_0.6   TS19_0.6   M246_0.6 MS1_0.6    ENSG00000242875\n",
       "196 T37372_0.6 F49_0.6   TS19_0.6   M246_0.6 MS1_0.6    ENSG00000244395\n",
       "197 T37374_0.6 F49_0.6   TS19_0.6   M042_0.6 MS2_0.6    ENSG00000250177\n",
       "198 T60363_0.6 F52_0.6   TS19_0.6   M231_0.6 MS2_0.6    ENSG00000134001\n",
       "199 T61784_0.6 F53_0.6   TS19_0.6   M061_0.6 MS2_0.6    ENSG00000020577\n",
       "200 T61785_0.6 F53_0.6   TS19_0.6   M061_0.6 MS2_0.6    ENSG00000179134\n",
       "201 T61860_0.6 F54_0.6   TS19_0.6   M256_0.6 MS1_0.6    ENSG00000122729\n",
       "202 T62023_0.6 F55_0.6   TS19_0.6   M354_0.6 MS1_0.6    ENSG00000083896\n",
       "    RBP_Name        RBP_Species  RBP_Status Family_Name <U+22EF>        MSource_Type\n",
       "1   MBNL3           Homo_sapiens I          CCCH ZF     <U+22EF> RNAcompete  \n",
       "2   ZFP36           Homo_sapiens D          CCCH ZF     <U+22EF> RBPDB       \n",
       "3   ZFP36           Homo_sapiens D          CCCH ZF     <U+22EF> RBPDB       \n",
       "4   ZC3H10          Homo_sapiens D          CCCH ZF     <U+22EF> RNAcompete  \n",
       "5   MBNL2           Homo_sapiens I          CCCH ZF     <U+22EF> RNAcompete  \n",
       "6   MBNL1           Homo_sapiens D          CCCH ZF     <U+22EF> RNAcompete  \n",
       "7   MBNL1           Homo_sapiens D          CCCH ZF     <U+22EF> RBPDB       \n",
       "8   LIN28A          Homo_sapiens D          CCHC ZF,CSD <U+22EF> RNAcompete  \n",
       "9   LIN28A          Homo_sapiens D          CCHC ZF,CSD <U+22EF> RNAcompete  \n",
       "10  LIN28B          Homo_sapiens I          CCHC ZF,CSD <U+22EF> RNAcompete  \n",
       "11  SRSF7           Homo_sapiens D          CCHC ZF,RRM <U+22EF> RNAcompete  \n",
       "12  SRSF7           Homo_sapiens D          CCHC ZF,RRM <U+22EF> RBPDB       \n",
       "13  SRSF7           Homo_sapiens D          CCHC ZF,RRM <U+22EF> RBPDB       \n",
       "14  ZCRB1           Homo_sapiens D          CCHC ZF,RRM <U+22EF> RNAcompete  \n",
       "15  RBM4B           Homo_sapiens I          CCHC ZF,RRM <U+22EF> RNAcompete  \n",
       "16  RBM4            Homo_sapiens I          CCHC ZF,RRM <U+22EF> RNAcompete  \n",
       "17  YBX2            Homo_sapiens D          CSD         <U+22EF> RNAcompete  \n",
       "18  CSDA            Homo_sapiens I          CSD         <U+22EF> RNAcompete  \n",
       "19  YB-1            Homo_sapiens D          CSD         <U+22EF> RNAcompete  \n",
       "20  YB-1            Homo_sapiens D          CSD         <U+22EF> RNAcompete  \n",
       "21  PCBP4           Homo_sapiens I          KH          <U+22EF> RNAcompete  \n",
       "22  FMR1            Homo_sapiens D          KH          <U+22EF> RNAcompete  \n",
       "23  NOVA2           Homo_sapiens D          KH          <U+22EF> RBPDB       \n",
       "24  NOVA2           Homo_sapiens D          KH          <U+22EF> RBPDB       \n",
       "25  KHDRBS2         Homo_sapiens D          KH          <U+22EF> RNAcompete  \n",
       "26  QKI             Homo_sapiens D          KH          <U+22EF> RNAcompete  \n",
       "27  QKI             Homo_sapiens D          KH          <U+22EF> RBPDB       \n",
       "28  FXR1            Homo_sapiens D          KH          <U+22EF> RNAcompete  \n",
       "29  KHDRBS1         Homo_sapiens D          KH          <U+22EF> RNAcompete  \n",
       "30  FXR2            Homo_sapiens D          KH          <U+22EF> RNAcompete  \n",
       "<U+22EE>   <U+22EE>        <U+22EE>     <U+22EE>   <U+22EE>    <U+22F1> <U+22EE>    \n",
       "173 RBMXL2          Homo_sapiens I          RRM         <U+22EF> RBPDB       \n",
       "174 PABPC5          Homo_sapiens D          RRM         <U+22EF> RNAcompete  \n",
       "175 RBMXL3          Homo_sapiens I          RRM         <U+22EF> RBPDB       \n",
       "176 HNRNPCL1        Homo_sapiens D          RRM         <U+22EF> RNAcompete  \n",
       "177 ENSG00000180771 Homo_sapiens I          RRM         <U+22EF> RNAcompete  \n",
       "178 Fusip1          Homo_sapiens D          RRM         <U+22EF> RNAcompete  \n",
       "179 Fusip1          Homo_sapiens D          RRM         <U+22EF> RNAcompete  \n",
       "180 Fusip1          Homo_sapiens D          RRM         <U+22EF> RNAcompete  \n",
       "181 Fusip1          Homo_sapiens D          RRM         <U+22EF> RNAcompete  \n",
       "182 PABPN1L         Homo_sapiens I          RRM         <U+22EF> RNAcompete  \n",
       "183 ENSG00000213250 Homo_sapiens I          RRM         <U+22EF> RNAcompete  \n",
       "184 RBMXL1          Homo_sapiens I          RRM         <U+22EF> RBPDB       \n",
       "185 ENSG00000215042 Homo_sapiens I          RRM         <U+22EF> RNAcompete  \n",
       "186 ENSG00000215492 Homo_sapiens I          RRM         <U+22EF> RNAcompete  \n",
       "187 BX511012.1      Homo_sapiens I          RRM         <U+22EF> RNAcompete  \n",
       "188 RBMY1J          Homo_sapiens I          RRM         <U+22EF> RBPDB       \n",
       "189 ENSG00000231942 Homo_sapiens I          RRM         <U+22EF> RNAcompete  \n",
       "190 RBMY1A1         Homo_sapiens D          RRM         <U+22EF> RBPDB       \n",
       "191 RBMY1A1         Homo_sapiens D          RRM         <U+22EF> RBPDB       \n",
       "192 RBM4            Homo_sapiens D          RRM         <U+22EF> RNAcompete  \n",
       "193 RBM4            Homo_sapiens D          RRM         <U+22EF> RNAcompete  \n",
       "194 RBMY1E          Homo_sapiens I          RRM         <U+22EF> RBPDB       \n",
       "195 RBMY1B          Homo_sapiens I          RRM         <U+22EF> RBPDB       \n",
       "196 RBMY1D          Homo_sapiens I          RRM         <U+22EF> RBPDB       \n",
       "197 ENSG00000250177 Homo_sapiens I          RRM         <U+22EF> RNAcompete  \n",
       "198 EIF2S1          Homo_sapiens I          S1          <U+22EF> RNAcompete  \n",
       "199 SAMD4A          Homo_sapiens D          SAM         <U+22EF> RNAcompete  \n",
       "200 SAMD4B          Homo_sapiens I          SAM         <U+22EF> RNAcompete  \n",
       "201 ACO1            Homo_sapiens D          Unknown     <U+22EF> RBPDB       \n",
       "202 YTHDC1          Homo_sapiens D          YTH         <U+22EF> RBPDB       \n",
       "    MSource_Author MSource_Year PMID     MSource_Version RBPSource_Name\n",
       "1   Ray            2013         NULL     NULL            Ensembl       \n",
       "2   Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "3   Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "4   Ray            2013         NULL     NULL            Ensembl       \n",
       "5   Ray            2013         NULL     NULL            Ensembl       \n",
       "6   Ray            2013         NULL     NULL            Ensembl       \n",
       "7   Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "8   Ray            2013         NULL     NULL            Ensembl       \n",
       "9   Ray            2013         NULL     NULL            Ensembl       \n",
       "10  Ray            2013         NULL     NULL            Ensembl       \n",
       "11  Ray            2013         NULL     NULL            Ensembl       \n",
       "12  Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "13  Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "14  Ray            2013         NULL     NULL            Ensembl       \n",
       "15  Ray            2013         NULL     NULL            Ensembl       \n",
       "16  Ray            2013         NULL     NULL            Ensembl       \n",
       "17  Ray            2013         NULL     NULL            Ensembl       \n",
       "18  Ray            2013         NULL     NULL            Ensembl       \n",
       "19  Ray            2013         NULL     NULL            Ensembl       \n",
       "20  Ray            2013         NULL     NULL            Ensembl       \n",
       "21  Ray            2013         NULL     NULL            Ensembl       \n",
       "22  Ray            2013         NULL     NULL            Ensembl       \n",
       "23  Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "24  Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "25  Ray            2013         NULL     NULL            Ensembl       \n",
       "26  Ray            2013         NULL     NULL            Ensembl       \n",
       "27  Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "28  Ray            2013         NULL     NULL            Ensembl       \n",
       "29  Ray            2013         NULL     NULL            Ensembl       \n",
       "30  Ray            2013         NULL     NULL            Ensembl       \n",
       "<U+22EE>   <U+22EE>       <U+22EE>     <U+22EE> <U+22EE>        <U+22EE>      \n",
       "173 Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "174 Ray            2013         NULL     NULL            Ensembl       \n",
       "175 Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "176 Ray            2013         NULL     NULL            Ensembl       \n",
       "177 Ray            2013         NULL     NULL            Ensembl       \n",
       "178 Ray            2013         NULL     NULL            Ensembl       \n",
       "179 Ray            2013         NULL     NULL            Ensembl       \n",
       "180 Ray            2013         NULL     NULL            Ensembl       \n",
       "181 Ray            2013         NULL     NULL            Ensembl       \n",
       "182 Ray            2013         NULL     NULL            Ensembl       \n",
       "183 Ray            2013         NULL     NULL            Ensembl       \n",
       "184 Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "185 Ray            2013         NULL     NULL            Ensembl       \n",
       "186 Ray            2013         NULL     NULL            Ensembl       \n",
       "187 Ray            2013         NULL     NULL            Ensembl       \n",
       "188 Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "189 Ray            2013         NULL     NULL            Ensembl       \n",
       "190 Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "191 Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "192 Ray            2013         NULL     NULL            Ensembl       \n",
       "193 Ray            2013         NULL     NULL            Ensembl       \n",
       "194 Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "195 Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "196 Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "197 Ray            2013         NULL     NULL            Ensembl       \n",
       "198 Ray            2013         NULL     NULL            Ensembl       \n",
       "199 Ray            2013         NULL     NULL            Ensembl       \n",
       "200 Ray            2013         NULL     NULL            Ensembl       \n",
       "201 Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "202 Cook           2011         21036867 v1.2.1          Ensembl       \n",
       "    RBPSource_URL           RBPSource_Year RBPSource_Month RBPSource_Day\n",
       "1   http://www.ensembl.org/ 2011           Oct             26           \n",
       "2   http://www.ensembl.org/ 2011           Oct             26           \n",
       "3   http://www.ensembl.org/ 2011           Oct             26           \n",
       "4   http://www.ensembl.org/ 2011           Oct             26           \n",
       "5   http://www.ensembl.org/ 2011           Oct             26           \n",
       "6   http://www.ensembl.org/ 2011           Oct             26           \n",
       "7   http://www.ensembl.org/ 2011           Oct             26           \n",
       "8   http://www.ensembl.org/ 2011           Oct             26           \n",
       "9   http://www.ensembl.org/ 2011           Oct             26           \n",
       "10  http://www.ensembl.org/ 2011           Oct             26           \n",
       "11  http://www.ensembl.org/ 2011           Oct             26           \n",
       "12  http://www.ensembl.org/ 2011           Oct             26           \n",
       "13  http://www.ensembl.org/ 2011           Oct             26           \n",
       "14  http://www.ensembl.org/ 2011           Oct             26           \n",
       "15  http://www.ensembl.org/ 2011           Oct             26           \n",
       "16  http://www.ensembl.org/ 2011           Oct             26           \n",
       "17  http://www.ensembl.org/ 2011           Oct             26           \n",
       "18  http://www.ensembl.org/ 2011           Oct             26           \n",
       "19  http://www.ensembl.org/ 2011           Oct             26           \n",
       "20  http://www.ensembl.org/ 2011           Oct             26           \n",
       "21  http://www.ensembl.org/ 2011           Oct             26           \n",
       "22  http://www.ensembl.org/ 2011           Oct             26           \n",
       "23  http://www.ensembl.org/ 2011           Oct             26           \n",
       "24  http://www.ensembl.org/ 2011           Oct             26           \n",
       "25  http://www.ensembl.org/ 2011           Oct             26           \n",
       "26  http://www.ensembl.org/ 2011           Oct             26           \n",
       "27  http://www.ensembl.org/ 2011           Oct             26           \n",
       "28  http://www.ensembl.org/ 2011           Oct             26           \n",
       "29  http://www.ensembl.org/ 2011           Oct             26           \n",
       "30  http://www.ensembl.org/ 2011           Oct             26           \n",
       "<U+22EE>   <U+22EE>                <U+22EE>       <U+22EE>        <U+22EE>     \n",
       "173 http://www.ensembl.org/ 2011           Oct             26           \n",
       "174 http://www.ensembl.org/ 2011           Oct             26           \n",
       "175 http://www.ensembl.org/ 2011           Oct             26           \n",
       "176 http://www.ensembl.org/ 2011           Oct             26           \n",
       "177 http://www.ensembl.org/ 2011           Oct             26           \n",
       "178 http://www.ensembl.org/ 2011           Oct             26           \n",
       "179 http://www.ensembl.org/ 2011           Oct             26           \n",
       "180 http://www.ensembl.org/ 2011           Oct             26           \n",
       "181 http://www.ensembl.org/ 2011           Oct             26           \n",
       "182 http://www.ensembl.org/ 2011           Oct             26           \n",
       "183 http://www.ensembl.org/ 2011           Oct             26           \n",
       "184 http://www.ensembl.org/ 2011           Oct             26           \n",
       "185 http://www.ensembl.org/ 2011           Oct             26           \n",
       "186 http://www.ensembl.org/ 2011           Oct             26           \n",
       "187 http://www.ensembl.org/ 2011           Oct             26           \n",
       "188 http://www.ensembl.org/ 2011           Oct             26           \n",
       "189 http://www.ensembl.org/ 2011           Oct             26           \n",
       "190 http://www.ensembl.org/ 2011           Oct             26           \n",
       "191 http://www.ensembl.org/ 2011           Oct             26           \n",
       "192 http://www.ensembl.org/ 2011           Oct             26           \n",
       "193 http://www.ensembl.org/ 2011           Oct             26           \n",
       "194 http://www.ensembl.org/ 2011           Oct             26           \n",
       "195 http://www.ensembl.org/ 2011           Oct             26           \n",
       "196 http://www.ensembl.org/ 2011           Oct             26           \n",
       "197 http://www.ensembl.org/ 2011           Oct             26           \n",
       "198 http://www.ensembl.org/ 2011           Oct             26           \n",
       "199 http://www.ensembl.org/ 2011           Oct             26           \n",
       "200 http://www.ensembl.org/ 2011           Oct             26           \n",
       "201 http://www.ensembl.org/ 2011           Oct             26           \n",
       "202 http://www.ensembl.org/ 2011           Oct             26           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "db<-RBP.zipload(\"./Homo_sapiens_2020_06_22_10_25_pm.zip\",species=\"Homo_sapiens\")\n",
    "show(db)\n",
    "\n",
    "rbps<-rbpbs_createFromCisRBP(db)\n",
    "rbps@rbp_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "40fe17ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'rbpbs.selectByRandom'"
      ],
      "text/latex": [
       "'rbpbs.selectByRandom'"
      ],
      "text/markdown": [
       "'rbpbs.selectByRandom'"
      ],
      "text/plain": [
       "[1] \"rbpbs.selectByRandom\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Select a representative motif randomly from each cluster\n",
    "setGeneric(\"rbpbs.selectByRandom\",\n",
    "\tdef=function( rbpbs ) {\n",
    "\t\tstopifnot(class(rbpbs) == \"rbpbs\")\n",
    "\t\tstandardGeneric(\"rbpbs.selectByRandom\")\n",
    "\t})\n",
    "\n",
    "\n",
    "setMethod(\"rbpbs.selectByRandom\", c(rbpbs=\"rbpbs\"),\n",
    "\tfunction( rbpbs ) {\n",
    "\n",
    "\t\tstopifnot( NROW(rbpbs@cluster)>0 );\n",
    "\n",
    "\t\tcluster <- rbpbs@cluster[,2]\n",
    "\n",
    "\t\tusemotifs <- sapply(1:max(cluster), function(x) {\n",
    "\t\t\ta <- which(cluster == x)\n",
    "\t\t\tif(length(a) > 1) {\n",
    "\t\t\t\treturn(sample(a, 1)) ## DANGEROUS!! If length(a) == 1, samples from 1:a[1].\n",
    "\t\t\t} else {\n",
    "\t\t\t\treturn(a)\n",
    "\t\t}})\n",
    "\n",
    "\t\tselect.col <- rep( 0, NROW(rbpbs@cluster) );\n",
    "\t\tselect.col[ usemotifs ]<-1;\n",
    "\n",
    "\t\tif(NCOL(rbpbs@cluster)==2)\n",
    "\t\t\trbpbs@cluster <- cbind( rbpbs@cluster, selected=select.col )\n",
    "\t\telse\n",
    "\t\t\trbpbs@cluster[,3] <- select.col;\n",
    "\n",
    "\t\treturn( rbpbs );\n",
    "\t})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1574015d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: 'apcluster'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    heatmap\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Species:  Homo_sapiens \n",
       "RBP number:  202 \n",
       "Distance Matrix:  NULL\n",
       "Cluster Matrix:  NULL\n",
       "Expression:  NULL\n",
       "\n",
       "Partial list of TFs\n",
       "   Motif_ID RBP_Name Family_Name            DBID Motif_Type     filename\n",
       "1  M037_0.6    MBNL3     CCCH ZF ENSG00000076770 RNAcompete M037_0.6.txt\n",
       "2  M269_0.6    ZFP36     CCCH ZF ENSG00000128016      SELEX M269_0.6.txt\n",
       "3  M350_0.6    ZFP36     CCCH ZF ENSG00000128016      SELEX M350_0.6.txt\n",
       "4  M083_0.6   ZC3H10     CCCH ZF ENSG00000135482 RNAcompete M083_0.6.txt\n",
       "5  M037_0.6    MBNL2     CCCH ZF ENSG00000139793 RNAcompete M037_0.6.txt\n",
       "6  M037_0.6    MBNL1     CCCH ZF ENSG00000152601 RNAcompete M037_0.6.txt\n",
       "7  M320_0.6    MBNL1     CCCH ZF ENSG00000152601      SELEX M320_0.6.txt\n",
       "8  M035_0.6   LIN28A CCHC ZF,CSD ENSG00000131914 RNAcompete M035_0.6.txt\n",
       "9  M153_0.6   LIN28A CCHC ZF,CSD ENSG00000131914 RNAcompete M153_0.6.txt\n",
       "10 M035_0.6   LIN28B CCHC ZF,CSD ENSG00000187772 RNAcompete M035_0.6.txt\n",
       "11 M071_0.6    SRSF7 CCHC ZF,RRM ENSG00000115875 RNAcompete M071_0.6.txt\n",
       "12 M331_0.6    SRSF7 CCHC ZF,RRM ENSG00000115875      SELEX M331_0.6.txt\n",
       "13 M353_0.6    SRSF7 CCHC ZF,RRM ENSG00000115875      SELEX M353_0.6.txt\n",
       "14 M085_0.6    ZCRB1 CCHC ZF,RRM ENSG00000139168 RNAcompete M085_0.6.txt\n",
       "15 M050_0.6    RBM4B CCHC ZF,RRM ENSG00000173914 RNAcompete M050_0.6.txt\n",
       "16 M050_0.6     RBM4 CCHC ZF,RRM ENSG00000173933 RNAcompete M050_0.6.txt\n",
       "17 M082_0.6     YBX2         CSD ENSG00000006047 RNAcompete M082_0.6.txt\n",
       "18 M081_0.6     CSDA         CSD ENSG00000060138 RNAcompete M081_0.6.txt\n",
       "19 M081_0.6     YB-1         CSD ENSG00000065978 RNAcompete M081_0.6.txt\n",
       "20 M111_0.6     YB-1         CSD ENSG00000065978 RNAcompete M111_0.6.txt"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n",
      "Warning message in setmid(x, type = type):\n",
      "\"midcache() of non-binary dendrograms only partly implemented\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Species:  Homo_sapiens \n",
       "RBP number:  202 \n",
       "Distance Matrix:  [ 202 , 202 ]\n",
       "Cluster Matrix:  [ 202 , 3 ]\n",
       "Expression:  NULL\n",
       "\n",
       "Partial list of TFs\n",
       "   Motif_ID RBP_Name Family_Name            DBID Motif_Type     filename\n",
       "1  M037_0.6    MBNL3     CCCH ZF ENSG00000076770 RNAcompete M037_0.6.txt\n",
       "2  M269_0.6    ZFP36     CCCH ZF ENSG00000128016      SELEX M269_0.6.txt\n",
       "3  M350_0.6    ZFP36     CCCH ZF ENSG00000128016      SELEX M350_0.6.txt\n",
       "4  M083_0.6   ZC3H10     CCCH ZF ENSG00000135482 RNAcompete M083_0.6.txt\n",
       "5  M037_0.6    MBNL2     CCCH ZF ENSG00000139793 RNAcompete M037_0.6.txt\n",
       "6  M037_0.6    MBNL1     CCCH ZF ENSG00000152601 RNAcompete M037_0.6.txt\n",
       "7  M320_0.6    MBNL1     CCCH ZF ENSG00000152601      SELEX M320_0.6.txt\n",
       "8  M035_0.6   LIN28A CCHC ZF,CSD ENSG00000131914 RNAcompete M035_0.6.txt\n",
       "9  M153_0.6   LIN28A CCHC ZF,CSD ENSG00000131914 RNAcompete M153_0.6.txt\n",
       "10 M035_0.6   LIN28B CCHC ZF,CSD ENSG00000187772 RNAcompete M035_0.6.txt\n",
       "11 M071_0.6    SRSF7 CCHC ZF,RRM ENSG00000115875 RNAcompete M071_0.6.txt\n",
       "12 M331_0.6    SRSF7 CCHC ZF,RRM ENSG00000115875      SELEX M331_0.6.txt\n",
       "13 M353_0.6    SRSF7 CCHC ZF,RRM ENSG00000115875      SELEX M353_0.6.txt\n",
       "14 M085_0.6    ZCRB1 CCHC ZF,RRM ENSG00000139168 RNAcompete M085_0.6.txt\n",
       "15 M050_0.6    RBM4B CCHC ZF,RRM ENSG00000173914 RNAcompete M050_0.6.txt\n",
       "16 M050_0.6     RBM4 CCHC ZF,RRM ENSG00000173933 RNAcompete M050_0.6.txt\n",
       "17 M082_0.6     YBX2         CSD ENSG00000006047 RNAcompete M082_0.6.txt\n",
       "18 M081_0.6     CSDA         CSD ENSG00000060138 RNAcompete M081_0.6.txt\n",
       "19 M081_0.6     YB-1         CSD ENSG00000065978 RNAcompete M081_0.6.txt\n",
       "20 M111_0.6     YB-1         CSD ENSG00000065978 RNAcompete M111_0.6.txt"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(parallel)\n",
    "#install.packages(\"apcluster\")\n",
    "library(apcluster)\n",
    "library(cluster)\n",
    "#Draw heatmap\n",
    "rbps\n",
    "tfs1<- rbpbs.clusterMotifs(rbps, method=\"apcluster\", group.k=30, \tpdf.heatmap=\"agnes.hm.pdf\" )\n",
    "#tfs2<- rbpbs.clusterMotifs(rbps, method=\"aplcuster\", group.k=50, \tpdf.heatmap=\"apcluster.hm.pdf\" )\n",
    "\n",
    "\n",
    "tfs1.randem<-rbpbs.selectByRandom(tfs1)\n",
    "tfs1.randem\n",
    "#tfs2 <- rbpbs.selectByRandom(tfs2)\n",
    "#show(tfs2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "36d74a61",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "'rbpbs.scanRBPsite'"
      ],
      "text/latex": [
       "'rbpbs.scanRBPsite'"
      ],
      "text/markdown": [
       "'rbpbs.scanRBPsite'"
      ],
      "text/plain": [
       "[1] \"rbpbs.scanRBPsite\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "letterA <- function(x.pos,y.pos,ht,wt,id=NULL){\n",
    "  \n",
    "  x <- c(0,4,6,10,8,6.8,3.2,2,0,3.6,5,6.4,3.6)\n",
    "  y <- c(0,10,10,0,0,3,3,0,0,4,7.5,4,4)\n",
    "  x <- 0.1*x\n",
    "  y <- 0.1*y\n",
    "\n",
    "  x <- x.pos + wt*x\n",
    "  y <- y.pos + ht*y\n",
    "\n",
    "  if (is.null(id)){\n",
    "    id <- c(rep(1,9),rep(2,4))\n",
    "  }else{\n",
    "    id <- c(rep(id,9),rep(id+1,4))\n",
    "  }\n",
    "    \n",
    "  fill <- c(\"green\",\"white\")\n",
    "    \n",
    "  list(x=x,y=y,id=id,fill=fill)\n",
    "}\n",
    "\n",
    "## U\n",
    "letterU <- function(x.pos,y.pos,ht,wt,id=NULL){\n",
    "  \n",
    "  x <- c(0,10,10,6,6,4,4,0)\n",
    "  y <- c(10,10,9,9,0,0,9,9)\n",
    "  x <- 0.1*x\n",
    "  y <- 0.1*y\n",
    "\n",
    "  x <- x.pos + wt*x\n",
    "  y <- y.pos + ht*y\n",
    "\n",
    "  if (is.null(id)){\n",
    "    id <- rep(1,8)\n",
    "  }else{\n",
    "    id <- rep(id,8)\n",
    "  }\n",
    "  \n",
    "  fill <- \"red\"\n",
    "    \n",
    "  list(x=x,y=y,id=id,fill=fill)\n",
    "}\n",
    "\n",
    "## C\n",
    "letterC <- function(x.pos,y.pos,ht,wt,id=NULL){\n",
    "  angle1 <- seq(0.3+pi/2,pi,length=100)\n",
    "  angle2 <- seq(pi,1.5*pi,length=100)\n",
    "  x.l1 <- 0.5 + 0.5*sin(angle1)\n",
    "  y.l1 <- 0.5 + 0.5*cos(angle1)\n",
    "  x.l2 <- 0.5 + 0.5*sin(angle2)\n",
    "  y.l2 <- 0.5 + 0.5*cos(angle2)\n",
    "  \n",
    "  x.l <- c(x.l1,x.l2)\n",
    "  y.l <- c(y.l1,y.l2)\n",
    "  \n",
    "  x <- c(x.l,rev(x.l))\n",
    "  y <- c(y.l,1-rev(y.l))\n",
    "  \n",
    "  x.i1 <- 0.5 +0.35*sin(angle1)\n",
    "  y.i1 <- 0.5 +0.35*cos(angle1)\n",
    "  x.i1 <- x.i1[y.i1<=max(y.l1)]\n",
    "  y.i1 <- y.i1[y.i1<=max(y.l1)]\n",
    "  y.i1[1] <- max(y.l1)\n",
    "  \n",
    "  x.i2 <- 0.5 +0.35*sin(angle2)\n",
    "  y.i2 <- 0.5 +0.35*cos(angle2)\n",
    "  \n",
    "  x.i <- c(x.i1,x.i2)\n",
    "  y.i <- c(y.i1,y.i2)\n",
    "  \n",
    "  x1 <- c(x.i,rev(x.i))\n",
    "  y1 <- c(y.i,1-rev(y.i))\n",
    "  \n",
    "  x <- c(x,rev(x1))\n",
    "  y <- c(y,rev(y1))\n",
    "  \n",
    "  x <- x.pos + wt*x\n",
    "  y <- y.pos + ht*y\n",
    "\n",
    "  if (is.null(id)){\n",
    "    id <- rep(1,length(x))\n",
    "  }else{\n",
    "    id <- rep(id,length(x))\n",
    "  }\n",
    "  \n",
    "  fill <- \"blue\"\n",
    "    \n",
    "  list(x=x,y=y,id=id,fill=fill)\n",
    "}\n",
    "\n",
    "\n",
    "## G\n",
    "letterG <- function(x.pos,y.pos,ht,wt,id=NULL){\n",
    "  angle1 <- seq(0.3+pi/2,pi,length=100)\n",
    "  angle2 <- seq(pi,1.5*pi,length=100)\n",
    "  x.l1 <- 0.5 + 0.5*sin(angle1)\n",
    "  y.l1 <- 0.5 + 0.5*cos(angle1)\n",
    "  x.l2 <- 0.5 + 0.5*sin(angle2)\n",
    "  y.l2 <- 0.5 + 0.5*cos(angle2)\n",
    "  \n",
    "  x.l <- c(x.l1,x.l2)\n",
    "  y.l <- c(y.l1,y.l2)\n",
    "  \n",
    "  x <- c(x.l,rev(x.l))\n",
    "  y <- c(y.l,1-rev(y.l))\n",
    "  \n",
    "  x.i1 <- 0.5 +0.35*sin(angle1)\n",
    "  y.i1 <- 0.5 +0.35*cos(angle1)\n",
    "  x.i1 <- x.i1[y.i1<=max(y.l1)]\n",
    "  y.i1 <- y.i1[y.i1<=max(y.l1)]\n",
    "  y.i1[1] <- max(y.l1)\n",
    "  \n",
    "  x.i2 <- 0.5 +0.35*sin(angle2)\n",
    "  y.i2 <- 0.5 +0.35*cos(angle2)\n",
    "  \n",
    "  x.i <- c(x.i1,x.i2)\n",
    "  y.i <- c(y.i1,y.i2)\n",
    "  \n",
    "  x1 <- c(x.i,rev(x.i))\n",
    "  y1 <- c(y.i,1-rev(y.i))\n",
    "  \n",
    "  x <- c(x,rev(x1))\n",
    "  y <- c(y,rev(y1))\n",
    "  \n",
    "  h1 <- max(y.l1)\n",
    "  r1 <- max(x.l1)\n",
    "  \n",
    "  h1 <- 0.4\n",
    "  x.add <- c(r1,0.5,0.5,r1-0.2,r1-0.2,r1,r1)\n",
    "  y.add <- c(h1,h1,h1-0.1,h1-0.1,0,0,h1)\n",
    "\n",
    "  \n",
    "\n",
    "  if (is.null(id)){\n",
    "    id <- c(rep(1,length(x)),rep(2,length(x.add)))\n",
    "  }else{\n",
    "    id <- c(rep(id,length(x)),rep(id+1,length(x.add)))\n",
    "  }\n",
    "\n",
    "  x <- c(rev(x),x.add)\n",
    "  y <- c(rev(y),y.add)\n",
    "    \n",
    "  x <- x.pos + wt*x\n",
    "  y <- y.pos + ht*y\n",
    "  \n",
    "  \n",
    "  fill <- c(\"orange\",\"orange\")\n",
    "    \n",
    "  list(x=x,y=y,id=id,fill=fill)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "addLetter <- function(letters,which,x.pos,y.pos,ht,wt){\n",
    "  \n",
    "  if (which == \"A\"){\n",
    "    letter <- letterA(x.pos,y.pos,ht,wt)\n",
    "  }else if (which == \"C\"){\n",
    "    letter <- letterC(x.pos,y.pos,ht,wt)    \n",
    "  }else if (which == \"G\"){\n",
    "    letter <- letterG(x.pos,y.pos,ht,wt)    \n",
    "  }else if (which == \"U\"){\n",
    "    letter <- letterU(x.pos,y.pos,ht,wt)    \n",
    "  }else{\n",
    "    stop(\"which must be one of A,C,G,U\")\n",
    "  }\n",
    "\n",
    "  letters$x <- c(letters$x,letter$x)\n",
    "  letters$y <- c(letters$y,letter$y)\n",
    "\n",
    "  lastID <- ifelse(is.null(letters$id),0,max(letters$id))\n",
    "  letters$id <- c(letters$id,lastID+letter$id)\n",
    "  letters$fill <- c(letters$fill,letter$fill)\n",
    "  letters\n",
    "}\n",
    "\n",
    "\n",
    "## plot a sequence logo\n",
    "seqLogo <- function(pwm, ic.scale=TRUE, xaxis=TRUE, yaxis=TRUE, xfontsize=15, yfontsize=15, newpage=FALSE){\n",
    "\n",
    "#  if (class(pwm) == \"pwm\"){\n",
    "#    pwm <- pwm@pwm    \n",
    "#  }else \n",
    "  if (class(pwm) == \"data.frame\"){\n",
    "    pwm <- as.matrix(pwm)\n",
    "  }else if (class(pwm) != \"matrix\"){\n",
    "    stop(\"pwm must be of class matrix or data.frame\")\n",
    "  }\n",
    "\n",
    "  if (any(abs(1 - apply(pwm,2,sum)) > 0.01))\n",
    "    stop(\"Columns of PWM must add up to 1.0\")\n",
    "\n",
    "  \n",
    "  chars <- c(\"A\",\"C\",\"G\",\"U\")\n",
    "  letters <- list(x=NULL,y=NULL,id=NULL,fill=NULL)\n",
    "  npos <- ncol(pwm)\n",
    "\n",
    "  \n",
    "  if (ic.scale){\n",
    "    ylim <- 2\n",
    "    ylab <- \"Information content\"\n",
    "    facs <- pwm2ic(pwm)\n",
    "  }else{\n",
    "    ylim <- 1\n",
    "    ylab <- \"Probability\"\n",
    "    facs <- rep(1, npos)\n",
    "  }\n",
    "  \n",
    "  wt <- 1  \n",
    "  x.pos <- 0  \n",
    "  for (j in 1:npos){\n",
    "    \n",
    "    column <- pwm[,j]\n",
    "    hts <- column*facs[j]*0.95\n",
    "    letterOrder <- order(hts)\n",
    "        \n",
    "    y.pos <- 0    \n",
    "    for (i in 1:4){\n",
    "      letter <- chars[letterOrder[i]]\n",
    "      ht <- hts[letterOrder[i]]\n",
    "      if (ht>0) letters <- addLetter(letters,letter,x.pos,y.pos,ht,wt)\n",
    "      y.pos <- y.pos + ht + 0.01\n",
    "    }\n",
    "    x.pos <- x.pos + wt\n",
    "  }\n",
    "\n",
    "  if(newpage) grid.newpage()\n",
    "  bottomMargin = ifelse(xaxis, 0 + xfontsize/3.5, 0)\n",
    "  leftMargin = ifelse(yaxis, 0 + yfontsize/3.5, 0)\n",
    "  pushViewport(plotViewport(c(bottomMargin,leftMargin,0,0)))\n",
    "  pushViewport(dataViewport(0:ncol(pwm),0:ylim,name=\"vp1\"))\n",
    "  grid.polygon(x=unit(letters$x,\"native\"), y=unit(letters$y,\"native\"),\n",
    "               id=letters$id,\n",
    "               gp=gpar(fill=letters$fill,col=\"transparent\"))\n",
    "  if (xaxis){\n",
    "    grid.xaxis(at=seq(0.5,ncol(pwm)-0.5),label=1:ncol(pwm), gp=gpar(fontsize=xfontsize))\n",
    "    grid.text(\"Position\",y=unit(-3,\"lines\"), gp=gpar(fontsize=xfontsize))\n",
    "  }\n",
    "  if (yaxis){\n",
    "    grid.yaxis(gp=gpar(fontsize=yfontsize))\n",
    "    grid.text(ylab,x=unit(-3,\"lines\"),rot=90, gp=gpar(fontsize=yfontsize))\n",
    "  }\n",
    "  popViewport()\n",
    "  popViewport()\n",
    "  par(ask=FALSE)\n",
    "}\n",
    "\n",
    "## get information content profile from PWM\n",
    "pwm2ic<-function(pwm) {\n",
    "    npos<-ncol(pwm)\n",
    "    ic<-numeric(length=npos)\n",
    "    for (i in 1:npos) {\n",
    "        ic[i]<-2 + sum(sapply(pwm[, i], function(x) {\n",
    "            if (x > 0) { x*log2(x) } else { 0 }\n",
    "        }))\n",
    "    }\n",
    "    ic\n",
    "}\n",
    "\n",
    "## get consensus sequence from PWM\n",
    "pwm2cons<-function(pwm) {\n",
    "    if (class(pwm)!=\"matrix\") {warning(\"pwm argument must be of class matrix\")}\n",
    "    letters <- c(\"A\", \"C\", \"G\", \"U\")\n",
    "    paste(apply(pwm, 2, function(x){letters[rev(order(x))[1]]}), collapse=\"\")\n",
    "}\n",
    "\n",
    "#scanDB\n",
    "install.packages(\"snowfall\")\n",
    "#install.packages(\"rtfbs\")\n",
    "require(snowfall)\n",
    "require(\"rtfbs\")\n",
    "scanDb_rtfbs <- function(rbpbs,\n",
    "\t\t\t\t\t\tfasta=NULL,\n",
    "\t\t\t\t\t\treturn.type = \"matches\",\n",
    "\t\t\t\t\t\tfile.prefix = NA,\n",
    "\t\t\t\t\t\tusemotifs = NA,\n",
    "\t\t\t\t\t\tncores = 1,\n",
    "\t\t\t\t\t\tfdr.threshold = NA,\n",
    "\t\t\t\t\t\tscore.threshold = 6,\n",
    "\t\t\t\t\t\tgc.groups = NA,\n",
    "\t\t\t\t\t\tbackground.order = 2,\n",
    "\t\t\t\t\t\tbackground.length = 100000,...)\n",
    "{\n",
    "\tstopifnot(class(rbpbs) == \"rbpbs\")\n",
    "  require(\"rtfbs\")\n",
    "  seq.ms<- read.ms(fasta)\n",
    "\t#seq.ms  <- read.seqfile.from.bed( extBed, file.twoBit);\n",
    "\tbgModel <- build.mm( seq.ms, 3);\n",
    "\n",
    "\t## Swiched rtfbs to returning posteriors.\n",
    "\treturn.posteriors <- ( return.type %in% c(\"posteriors\", \"maxposterior\", \"maxscore\" ) );\n",
    "\t#\n",
    "get_binding_site <- function( bgModel1,\n",
    "\t\t\t\t\t\t\tseq.ms,\n",
    "\t\t\t\t\t\t\tPWM,\n",
    "\t\t\t\t\t\t\treturn.posteriors,\n",
    "\t\t\t\t\t\t\tscore.threshold = 6,\n",
    "\t\t\t\t\t\t\tfdr.threshold = NA,\n",
    "\t\t\t\t\t\t\tgc.groups = NA,\n",
    "\t\t\t\t\t\t\tbackground.order = 2,\n",
    "\t\t\t\t\t\t\tbackground.length = 100000)\n",
    "{\n",
    "\tif ( is.na(score.threshold) &&  is.na(fdr.threshold)) score.threshold <- 6;\n",
    "\tif (!is.na(score.threshold) && !is.na(fdr.threshold)) score.threshold <- NA;\n",
    "\n",
    "\tif( is.na( gc.groups) )\n",
    "\t{\n",
    "\t\tif(is.na(fdr.threshold))\n",
    "\t\t\tbinding <- score.ms( seq.ms, PWM, bgModel1, strand=\"+\",return_posteriors=return.posteriors, threshold=score.threshold )\n",
    "\t\telse\n",
    "\t\t{\n",
    "\t\t\tseq.score <- score.ms( seq.ms, PWM, bgModel1, return_posteriors=return.posteriors,strand=\"+\", threshold=0 );\n",
    "\t\t\tsimu.ms   <- simulate.ms( bgModel1, background.length );\n",
    "\t\t\tsimu.score<- score.ms( simu.ms, PWM, bgModel1, threshold=0 ,strand=\"+\");\n",
    "\t\t\tfdrMap    <- calc.fdr( seq.ms, seq.score, simu.ms, simu.score );\n",
    "\t\t\tbinding   <- output.sites( seq.score, fdrScoreMap  = fdrMap, fdrThreshold = fdr.threshold);\n",
    "\t\t}\n",
    "\t}\n",
    "\telse\n",
    "\t{\n",
    "\t\tmsGroups <- groupByGC.ms( seq.ms, gc.groups);\n",
    "\n",
    "\t\tbgModels <- lapply(1:length(msGroups),\n",
    "\t\t\t\t\t\tfunction(i) { build.mm(msGroups[[i]], background.order) } );\n",
    "\n",
    "\t\tseq.score <- lapply(1:length(msGroups),\n",
    "\t\t\t\t\t\tfunction(i) { score.ms(msGroups[[i]],\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tPWM,\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tbgModels[[i]],\n",
    "\t\t\t\t\t\t\t\t\t\t\t\treturn_posteriors=return.posteriors,strand=\"+\",\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tthreshold=ifelse(!is.na(fdr.threshold), 0, score.threshold));});\n",
    "\n",
    "\t\tif(is.na(fdr.threshold))\n",
    "\t\t{\n",
    "\t\t\tbinding <- do.call(\"rbind\", seq.score);\n",
    "\t\t}\n",
    "\t\telse\n",
    "\t\t{\n",
    "\t\t\tsimu.ms <- lapply( 1:length(msGroups),\n",
    "\t\t\t\t\t\t\tfunction(i){ simulate.ms(bgModels[[i]], background.length)});\n",
    "\t\t\tsimu.score <- lapply( 1:length(msGroups),\n",
    "\t\t\t\t\t\t\tfunction(i){ score.ms(simu.ms[[i]], PWM, bgModels[[i]],strand=\"+\")});\n",
    "\t\t\tfdrMap  <- lapply( 1:length(msGroups),\n",
    "\t\t\t\t\t\t\tfunction(i) { calc.fdr(msGroups[[i]], seq.score[[i]], simu.ms[[i]], simu.score[[i]])});\n",
    "\t\t\tbinding <- lapply( 1:length(msGroups),\n",
    "\t\t\t\t\t\t\tfunction(i) { output.sites(seq.score[[i]], fdrScoreMap  = fdrMap[[i]], fdrThreshold = fdr.threshold);} );\n",
    "\n",
    "\t\t\tbinding <- do.call(\"rbind\", binding);\n",
    "\t\t}\n",
    "\n",
    "\t}\n",
    "\n",
    "\treturn( binding );\n",
    "}\n",
    "#\n",
    "\tscan_each_motif <- function (i )\n",
    "\t{\n",
    "\t\trequire(rtfbsdb);\n",
    "\t\trequire(rtfbs);\n",
    "\n",
    "\t\toptions(\"scipen\"=100, \"digits\"=4)\n",
    "\t\tbinding<- NULL;\n",
    "\n",
    "\t\t## Read the pwm and sequence file.\n",
    "\t\tPWM <- rbpbs@pwm[[i]];\n",
    "\n",
    "\t\t## Parse binding site into genomic coordinates.\n",
    "\t\tif( return.posteriors )\n",
    "\t\t{\n",
    "\t\t\tbinding <- score.ms( seq.ms, PWM, bgModel, return_posteriors=TRUE, threshold = score.threshold,strand=\"+\" );\n",
    "\n",
    "\t\t\tif( is.null(binding) || NROW(binding) == 0 )\n",
    "\t\t\t\treturn(NA);\n",
    "\n",
    "\t\t\tif( return.type == \"maxscore\" )\n",
    "\t\t\t\treturn(sapply(1:NROW(binding), function(x) {\n",
    "\t\t\t\t\tmax(c(binding[[x]]$MotifModel$Forward - binding[[x]]$Background, binding[[x]]$MotifModel$Reverse - binding[[x]]$Background)) }));\n",
    "\n",
    "\t\t\tif( return.type == \"maxposterior\" )\n",
    "\t\t\t\treturn(sapply(1:NROW(binding), function(x) {\n",
    "\t\t\t\t\tmax(c(binding[[x]]$MotifModel$Forward, binding[[x]]$MotifModel$Reverse )) }));\n",
    "\t\t}\n",
    "\n",
    "\t\t## Parse binding site into genomic coordinates.\n",
    "\t\tif(return.posteriors == FALSE )\n",
    "\t\t{\n",
    "\t\t\tbinding <- get_binding_site( bgModel,\n",
    "\t\t\t\t\t\t\tseq.ms,\n",
    "\t\t\t\t\t\t\tPWM,\n",
    "\t\t\t\t\t\t\tFALSE,\n",
    "\t\t\t\t\t\t\tscore.threshold = score.threshold,\n",
    "\t\t\t\t\t\t\tfdr.threshold = fdr.threshold,\n",
    "\t\t\t\t\t\t\tgc.groups = gc.groups,\n",
    "\t\t\t\t\t\t\tbackground.order = 2,\n",
    "\t\t\t\t\t\t\tbackground.length = 100000 );\n",
    "\n",
    "\t\t\tif( NROW(binding) > 0 )\n",
    "\t\t\t{\n",
    "\t\t\t\t\n",
    "\n",
    "\t\t\t\t#binding <- data.frame(  transcript      =binding$seqname,\n",
    "\t\t\t\t\t\t\t\t\t#\tname       = rbpbs@mgisymbols[i], # binding$motif_id\n",
    "\t\t\t\t\t\t\t\t\t#\tscore      = binding$score,\n",
    "\t\t\t\t\t\t\t\t\t#\tstrand     = binding$strand,\n",
    "\t\t\t\t\t\t\t\t\t#\t)\n",
    "\n",
    "\t\t\t\tif(return.type == \"writedb\") {\n",
    "\t\t\t\t\tfile.starch <- paste(file.prefix,i,\".bed.tmp.starch\", sep=\"\");\n",
    "\t\t\t\t\twrite.table(binding, file.starch);\n",
    "\t\t\t\t\treturn(file.starch);\n",
    "\t\t\t\t}\n",
    "\t\t\t}\n",
    "\t\t\telse\n",
    "\t\t\t\tbinding <- NULL;\n",
    "\t\t}\n",
    "\n",
    "\t\treturn(binding);\n",
    "\t}\n",
    "\n",
    "    if(ncores>1)\n",
    "    {\n",
    "   \t    require(snowfall);\n",
    "        sfInit(parallel = TRUE, cpus = ncores, type = \"SOCK\" )\n",
    "        sfExport(\"rbpbs\", \"seq.ms\", \"bgModel\", \"return.posteriors\",\n",
    "\t\t\t\"gfasta\",\n",
    "\t\t\t\"return.type\",\n",
    "\t\t\t\"file.prefix\",\n",
    "\t\t\t\"fdr.threshold\",\n",
    "\t\t\t\"score.threshold\",\n",
    "\t\t\t\"gc.groups\",\n",
    "\t\t\t\"background.order\",\n",
    "\t\t\t\"background.length\",\n",
    "\t\t\t\"get_binding_site\");\n",
    "\n",
    "        fun <- as.function(scan_each_motif);\n",
    "        environment(fun)<-globalenv();\n",
    "\n",
    "        binding_all <- sfLapply( usemotifs, fun);sfStop();\n",
    "    }\n",
    "    else\n",
    "        binding_all <- lapply( usemotifs, scan_each_motif);\n",
    "    \n",
    "\tif(return.type == \"writedb\")\n",
    "\t{\n",
    "\t\tcat_files <- paste(unlist(binding_all), collapse=\" \");\n",
    "\t\tfinal_starch <- paste(file.prefix,\".db.starch\", sep=\"\");\n",
    "\n",
    "\t\t#err_code <- system(paste(\"starchcat \", cat_files, \" > \", final_starch, sep=\"\"));\n",
    "\t\t#system(paste(\"rm \",file.prefix,\"*.bed.tmp.starch\",sep=\"\"))\n",
    "\t\t#if( err_code != 0 )\n",
    "\t\t#\twarning(\"Failed to call the starchcat command to generate starch file.\\n\");\n",
    "\n",
    "\t\t#err_code <- smallres_starchcat( unlist(binding_all), final_starch, ncores);\n",
    "\t\tbinding_all <- final_starch;\n",
    "\t}\n",
    "\n",
    "\tif(return.type %in% c( \"maxposterior\", \"maxscore\") )\n",
    "\t{\n",
    "\t\tseqList <- unlist(lapply(seq.ms, function(ms) {ms[1]}));\n",
    "\t\tseqList.org <- paste(extBed[,1],\":\", as.integer(extBed[,2]), \"-\", as.integer(extBed[,3]), sep=\"\");\n",
    "\t\tidx.bed <- match(seqList, seqList.org);\n",
    "\n",
    "\t\tbinding_mat <- matrix(NA, nrow= NROW(extBed), ncol= NROW(usemotifs))\n",
    "\t\tbinding_mat[idx.bed, ] <- matrix(unlist(binding_all), ncol= NROW(usemotifs));\n",
    "\t\tbinding_all <- binding_mat;\n",
    "\t}\n",
    "\n",
    "\treturn(binding_all)\n",
    "}\n",
    "\n",
    "setGeneric(\"rbpbs.scanRBPsite\",\n",
    "\tdef = function( rbpbs,\n",
    "\t\t\t\tfasta=NULL,\n",
    "\t\t\t\t\treturn.type = c(\"matches\", \"maxscore\", \"posteriors\", \"maxposterior\", \"writedb\"),\n",
    "\t\t\t\t\tfile.prefix = NA,\n",
    "\t\t\t\t\tusemotifs = NA,\n",
    "\t\t\t\t\tncores = 1,\n",
    "\t\t\t\t\tthreshold = 6,\n",
    "\t\t\t\t\tthreshold.type = c(\"score\", \"fdr\"),\n",
    "\t\t\t\t\tgc.groups = NA,\n",
    "\t\t\t\t\tbackground.order = 2,\n",
    "\t\t\t\t\tbackground.length = 100000,\n",
    "\t\t\t\t\texclude_offset = 250,\n",
    "\t\t\t\t\texclude_chromosome=\"_|chrM|chrY|chrX\" )\n",
    "\t{\n",
    "\t\tstopifnot(class(rbpbs) == \"rbpbs\")\n",
    "\t\tstandardGeneric(\"rbpbs.scanRBPsite\")\n",
    "\t})\n",
    "\n",
    "rbpbs_scanRBPsite <- function( rbpbs, fasta=NULL,\n",
    "\t\t\t\t\t\t\treturn.type=\"matches\",\n",
    "\t\t\t\t\t\t\tfile.prefix = NA,\n",
    "\t\t\t\t\t\t\tusemotifs = NA,\n",
    "\t\t\t\t\t\t\tncores = 1,\n",
    "\t\t\t\t\t\t\tthreshold = 6,\n",
    "\t\t\t\t\t\t\tthreshold.type = c(\"score\", \"fdr\"),\n",
    "\t\t\t\t\t\t\tgc.groups = NA,\n",
    "\t\t\t\t\t\t\tbackground.order = 2,\n",
    "\t\t\t\t\t\t\tbackground.length = 100000,\n",
    "\t\t\t\t\t\t\texclude_offset = 250,\n",
    "\t\t\t\t\t\t\texclude_chromosome=\"_|chrM|chrY|chrX\" )\n",
    "{\n",
    "\tstopifnot(class(rbpbs) == \"rbpbs\")\n",
    "\tstopifnot(return.type %in% c(\"matches\", \"maxscore\", \"posteriors\", \"maxposterior\", \"writedb\") );\n",
    "\n",
    "\tif( !file.exists (fasta)  )\n",
    "\t\tstop(paste(\"fasta file is not accessible, File=\", file.genome));\n",
    "\n",
    "\tgfasta =fasta;\n",
    "\n",
    "\tif( missing(usemotifs)) usemotifs =c(1:rbpbs@nrbps);\n",
    "\tif( missing(file.prefix) ) file.prefix=\"scan.db\";\n",
    "\tif( missing(return.type) ) return.type=\"matches\";\n",
    "\tif( missing(ncores) ) ncores= 1;\n",
    "\n",
    "\tscore.threshold <- NA;\n",
    "\tfdr.threshold   <- NA;\n",
    "\tif( missing( threshold.type ) ) threshold.type <- \"score\";\n",
    "\tif( threshold.type == \"score\" )\n",
    "\t\tif( missing( threshold ) )\n",
    "\t\t\tscore.threshold <- 6\n",
    "\t\telse\n",
    "\t\t\tscore.threshold <- threshold ;\n",
    "\n",
    "\tif( threshold.type == \"fdr\" )\n",
    "\t\tif( missing( threshold ) )\n",
    "\t\t\tfdr.threshold <- 0.1\n",
    "\t\telse\n",
    "\t\t\tfdr.threshold <- threshold ;\n",
    "\n",
    "\tr.ret <- scanDb_rtfbs( rbpbs, gfasta,\n",
    "\t\t\t\t\tfile.prefix      = file.prefix,\n",
    "\t\t\t\t\treturn.type      = return.type,\n",
    "\t\t\t\t\tusemotifs        = usemotifs,\n",
    "\t\t\t\t\tncores           = ncores,\n",
    "\t\t\t\t\tfdr.threshold    = fdr.threshold,\n",
    "\t\t\t\t\tscore.threshold  = score.threshold,\n",
    "\t\t\t\t\tgc.groups        = gc.groups,\n",
    "\t\t\t\t\tbackground.order = background.order,\n",
    "\t\t\t\t\tbackground.length = background.length );\n",
    "\n",
    "\tr.parm <- list(file.genome       = gfasta,\n",
    "\t\t\t\t\tfile.prefix      = file.prefix,\n",
    "\t\t\t\t\treturn.type      = return.type,\n",
    "\t\t\t\t\tusemotifs        = usemotifs,\n",
    "\t\t\t\t\tncores           = ncores,\n",
    "\t\t\t\t\tgc.groups        = gc.groups,\n",
    "\t\t\t\t\tthreshold        = ifelse(threshold.type==\"score\", score.threshold, fdr.threshold ),\n",
    "\t\t\t\t\tthreshold.type   = threshold.type,\n",
    "\t\t\t\t\tbackground.order = background.order,\n",
    "\t\t\t\t\tbackground.length = background.length);\n",
    "\n",
    "\tsum.match <- data.frame(\n",
    "\t\t\t\t\tMotif_ID = rbpbs@rbp_info$Motif_ID[ usemotifs ],\n",
    "\t\t\t\t\tRBP_Name  = rbpbs@rbp_info$RBP_Name[  usemotifs ] );\n",
    "\n",
    "\tif( return.type == \"matches\" )\n",
    "\t{\n",
    "\t\tsum.match <- do.call(\"rbind\", lapply( 1:length(usemotifs), function(i){\n",
    "\n",
    "\t\t\t\tx <- r.ret[[i]];\n",
    "\t\t\t\tmotif.id <- as.character( rbpbs@rbp_info$Motif_ID[ usemotifs[i] ] );\n",
    "\t\t\t\trbp.name  <- as.character( rbpbs@rbp_info$RBP_Name[  usemotifs[i] ] );\n",
    "\n",
    "\t\t\t\tif (NROW(x)==0) return( data.frame( motif.id, rbp.name, count=0 ) );\n",
    "\n",
    "\t\t\t\tstopifnot( as.character( x$name[1] ) == motif.id );\n",
    "\n",
    "\t\t\t\treturn( data.frame(motif.id, rbp.name, count=NROW(x) ) );\n",
    "\t\t} ) );\n",
    "\n",
    "\t\tcolnames(sum.match) <- c(\"Motif_ID\", \"RBP_Name\", \"Count\");\n",
    "\t}\n",
    "\n",
    "\tif( return.type %in% c(\"maxposterior\", \"maxscore\") )\n",
    "\t{\n",
    "\t\tsum.match <- data.frame(\n",
    "\t\t\t\t\trbpbbs@rbp_info$Motif_ID[ usemotifs ],\n",
    "\t\t\t\t\trbpbs@rbp_info$RBP_Name[ usemotifs ],\n",
    "\t\t\t\t\tcolMeans( r.ret, na.rm=T) );\n",
    "\n",
    "\t\tcolnames(sum.match) <- c(\"Motif_ID\", \"RBP_Name\", \"Mean\" );\n",
    "\t}\n",
    "\n",
    "\tr.scan <- list( parm = r.parm, result = r.ret, summary=sum.match );\n",
    "\tclass( r.scan ) <- c( \"rbpbs.finding\" );\n",
    "\n",
    "\treturn( r.scan );\n",
    "}\n",
    "\n",
    "print.tfbs.finding<-function(x, ...)\n",
    "{\n",
    "\tr.scan <- x;\n",
    "\n",
    "\tcat(\"Return type: \", r.scan$parm$return.type, \"\\n\");\n",
    "\tif(r.scan$parm$return.type %in% c(\"matches\", \"writedb\") )\n",
    "\t{\n",
    "\t\tcat(\"Threshold Type: \", r.scan$parm$threshold.type, \"\\n\");\n",
    "\t\tcat(\"Threshold: \",   r.scan$parm$threshold, \"\\n\");\n",
    "\t}\n",
    "\n",
    "\tif( r.scan$parm$return.type == \"matches\" )\n",
    "\t{\n",
    "\t\tdf.allfinding <- do.call(\"rbind\", r.scan$result );\n",
    "\t\tcat(\"Motifs count: \", length(r.scan$result), \"\\n\");\n",
    "\t\tcat(\"Binding sites: \", NROW(df.allfinding), \"\\n\");\n",
    "\n",
    "\t\tsummary <- r.scan$summary[order(r.scan$summary[,3], decreasing = TRUE ),];\n",
    "\t\tif( NROW( summary )>20)\n",
    "\t\t\tsummary <- summary[c(1:20),];\n",
    "\t\tshow( summary );\n",
    "\t}\n",
    "\n",
    "\tif( r.scan$parm$return.type == \"writedb\" )\n",
    "\t\tcat(\"Binary Bed file: \", r.scan$result, \"\\n\");\n",
    "\n",
    "\tif( r.scan$parm$return.type == \"posteriors\" )\n",
    "\t{\n",
    "\t\tdf.allfinding <- do.call(\"rbind\", r.scan$result );\n",
    "\t\tcat(\"Motifs count: \", length(r.scan$result), \"\\n\");\n",
    "\t\tcat(\"Binding sites: \", NROW(df.allfinding), \"\\n\");\n",
    "\t}\n",
    "\n",
    "\tif(r.scan$parm$return.type %in% c(\"maxposterior\", \"maxscore\") )\n",
    "\t\tcat(\"Matrix posterior: \", NROW(r.scan$result), \"*\", NCOL(r.scan$result), \"\\n\");\n",
    "}\n",
    "\n",
    "summary.tfbs.finding<-function( object, ... )\n",
    "{\n",
    "\tr.scan <- object;\n",
    "\n",
    "\tif( r.scan$parm$return.type == \"matches\" )\n",
    "\t\treturn( r.scan$summary )\n",
    "\telse\n",
    "\t\treturn(NULL);\n",
    "}\n",
    "\n",
    "tfbs.reportFinding<-function( tfbs, r.scan, file.pdf = NA, report.size = \"letter\", report.title = \"\" )\n",
    "{\n",
    "\tstopifnot(class(tfbs) == \"tfbs\" && class(r.scan) == \"tfbs.finding\")\n",
    "\n",
    "\tif( !is.null(file.pdf) && !is.na(file.pdf) )\n",
    "\t\tif( !check_folder_writable( file.pdf ) )\n",
    "\t\t\tstop( paste(\"Can not create pdf file: \",file.pdf ) );\n",
    "\n",
    "\tif( r.scan$parm$return.type != \"matches\" )\n",
    "\t\tcat( \"! No summary information for the report.\\n\" )\n",
    "\telse\n",
    "\t{\n",
    "\t\tsummary <- r.scan$summary[order(r.scan$summary[,3], decreasing = TRUE ),];\n",
    "\n",
    "\t\tr.scan.sum <- data.frame( No=c(1:NROW(summary)), summary, summary[,1] );\n",
    "\n",
    "\t\tdf.style <- data.frame( position = numeric(0),\n",
    "\t\t\t\t\t\t\t\twidth    = numeric(0),\n",
    "\t\t\t\t\t\t\t\theader   = character(0),\n",
    "\t\t\t\t\t\t\t\thjust    = character(0),\n",
    "\t\t\t\t\t\t\t\tstyle    = character(0),\n",
    "\t\t\t\t\t\t\t\textra1   = character(0),\n",
    "\t\t\t\t\t\t\t\textra2   = character(0),\n",
    "\t\t\t\t\t\t\t\textra3   = character(0),\n",
    "\t\t\t\t\t\t\t\textra4   = character(0));\n",
    "\t\tdf.style <- rbind( df.style,\n",
    "\t\t\t\t\tdata.frame( position = 0.00,\n",
    "\t\t\t\t\t\t\t\twidth    = 0.04,\n",
    "\t\t\t\t\t\t\t\theader   = \"No.\",\n",
    "\t\t\t\t\t\t\t\thjust    = \"left\",\n",
    "\t\t\t\t\t\t\t\tstyle    = \"text\",\n",
    "\t\t\t\t\t\t\t\textra1   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra2   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra3   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra4   = \"0\") );\n",
    "\t\tdf.style <- rbind( df.style,\n",
    "\t\t\t\t\tdata.frame( position = 0.04,\n",
    "\t\t\t\t\t\t\t\twidth    = 0.10,\n",
    "\t\t\t\t\t\t\t\theader   = \"Motif ID\",\n",
    "\t\t\t\t\t\t\t\thjust    = \"left\",\n",
    "\t\t\t\t\t\t\t\tstyle    = \"text\",\n",
    "\t\t\t\t\t\t\t\textra1   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra2   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra3   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra4   = \"0\") );\n",
    "\t\tdf.style <- rbind( df.style,\n",
    "\t\t\t\t\tdata.frame( position = 0.14,\n",
    "\t\t\t\t\t\t\t\twidth    = 0.10,\n",
    "\t\t\t\t\t\t\t\theader   = \"TF Name\",\n",
    "\t\t\t\t\t\t\t\thjust    = \"left\",\n",
    "\t\t\t\t\t\t\t\tstyle    = \"text\",\n",
    "\t\t\t\t\t\t\t\textra1   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra2   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra3   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra4   = \"0\") );\n",
    "\t\tdf.style <- rbind( df.style,\n",
    "\t\t\t\t\tdata.frame( position = 0.24,\n",
    "\t\t\t\t\t\t\t\twidth    = 0.10,\n",
    "\t\t\t\t\t\t\t\theader   = \"Count\",\n",
    "\t\t\t\t\t\t\t\thjust    = \"centre\",\n",
    "\t\t\t\t\t\t\t\tstyle    = \"text\",\n",
    "\t\t\t\t\t\t\t\textra1   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra2   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra3   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra4   = \"0\") );\n",
    "\t\tdf.style <- rbind( df.style,\n",
    "\t\t\t\t\tdata.frame( position = 0.34,\n",
    "\t\t\t\t\t\t\t\twidth    = 0.49,\n",
    "\t\t\t\t\t\t\t\theader   = \"Motif Logo\",\n",
    "\t\t\t\t\t\t\t\thjust    = \"centre\",\n",
    "\t\t\t\t\t\t\t\tstyle    = \"logo\",\n",
    "\t\t\t\t\t\t\t\textra1   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra2   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra3   = \"0\",\n",
    "\t\t\t\t\t\t\t\textra4   = \"0\") );\n",
    "\n",
    "\t\toutput_motif_report( tfbs, r.scan.sum, file.pdf, report.size, report.title, df.style );\n",
    "\t}\n",
    "}\n",
    "\n",
    "smallres_starchcat<-function( starch.files, final.starch, ncores)\n",
    "{\n",
    "    reduce4<-function(mfiles4)\n",
    "    {\n",
    "\t\tL <- mclapply( 1:ceiling(NROW(mfiles4)/4), function(i)\n",
    "\t\t\t{\n",
    "\t\t\t    tempf <- tempfile(fileext=\".starch\");\n",
    "\t\t\t    temp0 <- starch.files[ (i-1)*4+c(1:4) ];\n",
    "\t\t\t    temp0 <- temp0[!is.na(temp0)];\n",
    "\t\t\t\terr_code <- system(paste(\"starchcat \", paste(temp0, collapse=\" \"), \" > \", tempf));\t\t\n",
    "\t\t\t\tif( err_code != 0 )\n",
    "\t\t\t\t\twarning(\"Failed to call the starchcat command to generate starch file.\\n\");\n",
    "\t\t         \n",
    "\t\t\t\tunlink(temp0);\n",
    "\t\t        return(tempf);\n",
    "\t\t\t}, mc.cores=ncores);\n",
    "\t\treturn(unlist(L));\t\n",
    "\t}\n",
    "\t\n",
    "\twhile( NROW(starch.files)>1 )\n",
    "\t{\n",
    "\t\tstarch.files <- reduce4(starch.files);\n",
    "\t}\n",
    "\n",
    "\terr_code <- system(paste(\"cp\", starch.files, final.starch));\n",
    "\tunlink(starch.files);\n",
    "\treturn(0);\n",
    "}\n",
    "\n",
    "\n",
    "setMethod(\"rbpbs.scanRBPsite\", c(rbpbs=\"rbpbs\"), rbpbs_scanRBPsite )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "abdc8804",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "output_motif_report<-function( tfbs, df.motif, file.pdf, report.size, report.title, report.style, report.note=\"\")\n",
    "{\n",
    "\tscheme1 <- data.frame( min.col = c( 0,0,1 ), max.col = c( 0.9, 0.9, 0.9 ) );\n",
    "\tscheme2 <- data.frame( min.col = c( 0,1,0 ), max.col = c( 0.6, 0.3, 0.6 ) );\n",
    "\n",
    "\tget_rgbcol<-function(pvalue, p.min=0.001, p.max=1, log10=T, scheme=1)\n",
    "\t{\n",
    "\t\tif(pvalue > p.max) pvalue <- p.max;\n",
    "\t\tif(pvalue < p.min) pvalue <- p.min;\n",
    "\n",
    "\t\tif(log10)\n",
    "\t\t{\n",
    "\t\t\tpvalue <- log10(pvalue)\n",
    "\t\t\tp.min  <- log10(p.min)\n",
    "\t\t\tp.max  <- log10(p.max)\n",
    "\t\t}\n",
    "\n",
    "\t\tif(scheme==1) col.scheme <- scheme1 else col.scheme <- scheme2;\n",
    "\n",
    "\t\tc1 <- (col.scheme$min.col[1] - col.scheme$max.col[1]) / (p.min-p.max)*(pvalue-p.max) + col.scheme$max.col[1];\n",
    "\t\tc2 <- (col.scheme$min.col[2] - col.scheme$max.col[2]) / (p.min-p.max)*(pvalue-p.max) + col.scheme$max.col[2];\n",
    "\t\tc3 <- (col.scheme$min.col[3] - col.scheme$max.col[3]) / (p.min-p.max)*(pvalue-p.max) + col.scheme$max.col[3];\n",
    "\n",
    "\t\tc <- rgb(c1,c2,c3)\n",
    "\n",
    "\t\treturn(c);\n",
    "\t}\n",
    "\n",
    "\tdrawlegend<-function( x0, y0, width, height, title, pv.min=1e-6, pv.max=1, pv.log10=T, scheme=1)\n",
    "\t{\n",
    "\t\tbar.len <- 50;\n",
    "\t\tif(pv.log10)\n",
    "\t\t\tpv <- 10^seq(log10(pv.min), log10(pv.max), length.out=bar.len)\n",
    "\t\telse\n",
    "\t\t\tpv <- seq(pv.min, pv.max, length.out=bar.len);\n",
    "\n",
    "\t\tbar.width <- width*0.6;\n",
    "\t\tfor( i in 1:length(pv) )\n",
    "\t\t{\n",
    "\t\t\tcol <- get_rgbcol( pv[i], pv.min, pv.max, log10=pv.log10, scheme=scheme);\n",
    "\t\t\tgrid.rect(  x0 + 0.3*width + i*bar.width/bar.len,\n",
    "\t\t\t\t\t\ty0,\n",
    "\t\t\t\t\t\twidth = bar.width/bar.len,\n",
    "\t\t\t\t\t\theight = height,\n",
    "\t\t\t\t\t\tgp = gpar(fill=col, col=NA),\n",
    "\t\t\t\t\t\tjust = c(\"right\", \"centre\") );\n",
    "\t\t}\n",
    "\n",
    "\t\tif( pv.log10 )\n",
    "\t\t{\n",
    "\t\t\tp1<- (log10(pv.max) - log10(0.05) )*bar.width/(log10(pv.max) - log10(pv.min) );\n",
    "\n",
    "\t\t\tgrid.text(  \".05\",\n",
    "\t\t\t\t\t\tp1 + x0 + 0.3*width,\n",
    "\t\t\t\t\t\ty0 - height,\n",
    "\t\t\t\t\t\tgp = gpar(cex=0.4),\n",
    "\t\t\t\t\t\tjust = c(\"centre\",  \"centre\"));\n",
    "\t\t}\n",
    "\t\telse\n",
    "\t\t{\n",
    "\t\t\tp1 <- ( 1 - pv.min) * bar.width/(pv.max - pv.min );\n",
    "\n",
    "\t\t\tgrid.text(1, p1+x0+0.3*width,  y0-height, gp=gpar(cex=0.4), just=c(\"centre\",  \"centre\"));\n",
    "\t\t}\n",
    "\n",
    "\t\tgrid.lines( x = c( p1 + x0 + 0.3*width, p1 + x0 + 0.3*width),\n",
    "\t\t\t\t\ty = c( y0 - height/2, y0 + height/2),\n",
    "\t\t\t\t\tgp = gpar(col=\"white\", fill = \"white\", lwd = 0.01) );\n",
    "\n",
    "\t\tgrid.text(  title,\n",
    "\t\t\t\t\tx0 + 0.25*width,\n",
    "\t\t\t\t\ty0,\n",
    "\t\t\t\t\tgp = gpar(cex=0.4),\n",
    "\t\t\t\t\tjust = c(\"right\",  \"centre\"));\n",
    "\t\tgrid.text( sprintf(\"%1g\", pv.min),\n",
    "\t\t\t\t\tx0 + 0.28*width,\n",
    "\t\t\t\t\ty0 - height,\n",
    "\t\t\t\t\tgp = gpar(cex=0.4),\n",
    "\t\t\t\t\tjust = c(\"centre\", \"centre\"));\n",
    "\t\tgrid.text( sprintf(\"%1g\", pv.max),\n",
    "\t\t\t\t\tx0 + 0.9*width,\n",
    "\t\t\t\t\ty0 - height,\n",
    "\t\t\t\t\tgp = gpar(cex=0.4),\n",
    "\t\t\t\t\tjust = c(\"centre\", \"centre\"));\n",
    "\t}\n",
    "\n",
    "\tget_short_value<-function(val, log=T)\n",
    "\t{\n",
    "\t\tif(log)\n",
    "\t\t{\n",
    "\t\t\tn.log <- ceiling(log10(val));\n",
    "\t\t\tif(n.log > -2)\n",
    "\t\t\t\tstr <- round(val, digits=2)\n",
    "\t\t\telse\n",
    "\t\t\t\tstr <- sprintf(\"<%.1g\", val);\n",
    "\t\t\tif(val>=0.05) str<-\"\";\n",
    "\t\t\treturn(str);\n",
    "\t\t}\n",
    "\t\telse\n",
    "\t\t\treturn(round(val, digits=1));\n",
    "\t}\n",
    "\n",
    "\tdraw_top <- function()\n",
    "\t{\n",
    "\t\tgrid.text( report.title, x=0.5, y= 0.6,  gp=gpar(cex=0.8), just=c(\"centre\", \"centre\"));\n",
    "\n",
    "\t\tgrid.lines( x = c(0,1), y=c(0.4,0.4), gp=gpar(col=\"black\", lwd=1) );\n",
    "\n",
    "\t\tfor( k in 1:NROW(report.style) )\n",
    "\t\t{\n",
    "\t\t\tif(as.character(report.style$hjust[k])==\"left\")\n",
    "\t\t\t\tgrid.text(  report.style$header[k],\n",
    "\t\t\t\t\t\t\tx = report.style$position[k]+0,\n",
    "\t\t\t\t\t\t\ty = 0.2,\n",
    "\t\t\t\t\t\t\trot = 0,\n",
    "\t\t\t\t\t\t\tgp = gpar(cex=0.5),\n",
    "\t\t\t\t\t\t\tjust=c(\"left\", \"centre\") );\n",
    "\t\t\tif(as.character(report.style$hjust[k])==\"centre\")\n",
    "\t\t\t\tgrid.text(  report.style$header[k],\n",
    "\t\t\t\t\t\t\tx = report.style$position[k]+report.style$width[k]/2,\n",
    "\t\t\t\t\t\t\ty = 0.2,\n",
    "\t\t\t\t\t\t\trot = 0,\n",
    "\t\t\t\t\t\t\tgp = gpar(cex=0.5),\n",
    "\t\t\t\t\t\t\tjust = c(\"centre\", \"centre\") );\n",
    "\t\t\tif(as.character(report.style$hjust[k])==\"right\")\n",
    "\t\t\t\tgrid.text( report.style$header[k],\n",
    "\t\t\t\t\t\t\tx = report.style$position[k]+report.style$width[k],\n",
    "\t\t\t\t\t\t\ty = 0.2,\n",
    "\t\t\t\t\t\t\trot = 0,\n",
    "\t\t\t\t\t\t\tgp = gpar(cex=0.5),\n",
    "\t\t\t\t\t\t\tjust = c(\"right\", \"centre\") );\n",
    "\t\t}\n",
    "\n",
    "\t\tgrid.lines(x = c(0,1), y=c(0, 0), gp=gpar(col=\"black\", lwd=1) );\n",
    "\t}\n",
    "\n",
    "\tdraw_bottom <- function()\n",
    "\t{\n",
    "\t\tgrid.text(report.note, x=1, y=0.95,  gp=gpar(cex=0.4), just=c(\"right\", \"top\"));\n",
    "\n",
    "\t\tx0 <- 0.8;\n",
    "\t\tfor( k in 1:NROW( report.style) )\n",
    "\t\t{\n",
    "\t\t\tif(report.style$style[k]==\"bar\")\n",
    "\t\t\t{\n",
    "\t\t\t\tp.min  = as.numeric(as.character(report.style$extra1[k]));\n",
    "\t\t\t\tp.max  = as.numeric(as.character(report.style$extra2[k]));\n",
    "\t\t\t\tlog10  = as.numeric(as.character(report.style$extra3[k]));\n",
    "\t\t\t\tscheme = as.numeric(as.character(report.style$extra4[k]));\n",
    "\n",
    "\t\t\t\tdrawlegend( x0, 0.5, width=0.2, height=0.3, title=report.style$header[k], p.min, p.max, log10, scheme );\n",
    "\t\t\t\tx0 <- x0 - 0.2;\n",
    "\t\t\t}\n",
    "\t\t}\n",
    "\t}\n",
    "\n",
    "\tdraw_item<-function(r.comp.sort, idx.start, idx.stop)\n",
    "\t{\n",
    "\t\tM40 <- view.lines;\n",
    "\n",
    "\t\ty0 <- 0;\n",
    "\t\tfor(i in idx.start:idx.stop)\n",
    "\t\t{\n",
    "\t\t\ty0 <- 1 - (i - idx.start +1 )/M40;\n",
    "\t\t\tfor(k in 1:NCOL(df.motif))\n",
    "\t\t\t{\n",
    "\t\t\t\tif(report.style$style[k]==\"text\")\n",
    "\t\t\t\t{\n",
    "\t\t\t\t\tif(as.character(report.style$hjust[k]) == \"left\")\n",
    "\t\t\t\t\t\tgrid.text(  df.motif[i, k],\n",
    "\t\t\t\t\t\t\t\t\tx = report.style$position[k]+0,\n",
    "\t\t\t\t\t\t\t\t\ty = y0+0.005,\n",
    "\t\t\t\t\t\t\t\t\trot = 0,\n",
    "\t\t\t\t\t\t\t\t\tgp = gpar(cex=0.5),\n",
    "\t\t\t\t\t\t\t\t\tcheck.overlap = F,\n",
    "\t\t\t\t\t\t\t\t\tjust = c(\"left\", \"centre\") );\n",
    "\t\t\t\t\tif(as.character(report.style$hjust[k]) == \"centre\")\n",
    "\t\t\t\t\t\tgrid.text( df.motif[i, k],\n",
    "\t\t\t\t\t\t\t\t\tx = report.style$position[k]+report.style$width[k]/2,\n",
    "\t\t\t\t\t\t\t\t\ty = y0+0.005,\n",
    "\t\t\t\t\t\t\t\t\trot = 0,\n",
    "\t\t\t\t\t\t\t\t\tgp = gpar(cex=0.5),\n",
    "\t\t\t\t\t\t\t\t\tcheck.overlap = F,\n",
    "\t\t\t\t\t\t\t\t\tjust=c(\"centre\", \"centre\") );\n",
    "\t\t\t\t\tif(as.character(report.style$hjust[k])==\"right\")\n",
    "\t\t\t\t\t\tgrid.text( df.motif[i, k],\n",
    "\t\t\t\t\t\t\t\t\tx = report.style$position[k]+report.style$width[k],\n",
    "\t\t\t\t\t\t\t\t\ty = y0+0.005,\n",
    "\t\t\t\t\t\t\t\t\trot = 0,\n",
    "\t\t\t\t\t\t\t\t\tgp = gpar(cex=0.5),\n",
    "\t\t\t\t\t\t\t\t\tcheck.overlap = F,\n",
    "\t\t\t\t\t\t\t\t\tjust = c(\"right\", \"centre\") );\n",
    "\t\t\t\t}\n",
    "\n",
    "\t\t\t\tif(report.style$style[k]==\"bar\")\n",
    "\t\t\t\t{\n",
    "\t\t\t\t\tcol.fill <- get_rgbcol( df.motif[i,k],\n",
    "\t\t\t\t\t\t\t\t\t\tp.min = as.numeric(as.character(report.style$extra1[k])),\n",
    "\t\t\t\t\t\t\t\t\t\tp.max = as.numeric(as.character(report.style$extra2[k])),\n",
    "\t\t\t\t\t\t\t\t\t\tlog10 = as.numeric(as.character(report.style$extra3[k])),\n",
    "\t\t\t\t\t\t\t\t\t\tscheme = as.numeric(as.character(report.style$extra4[k]))) ;\n",
    "\t\t\t\t\tgrid.rect(  x = report.style$position[k] + report.style$width[k]/2.0,\n",
    "\t\t\t\t\t\t\t\ty = y0+0.005,\n",
    "\t\t\t\t\t\t\t\twidth = report.style$width[k]-0.02,\n",
    "\t\t\t\t\t\t\t\theight = 1/M40-0.006,\n",
    "\t\t\t\t\t\t\t\tgp = gpar(lwd=0.01,\n",
    "\t\t\t\t\t\t\t\tfill = col.fill) );\n",
    "\n",
    "\t\t\t\t\tstr.val <- get_short_value( df.motif[i,k], as.numeric(as.character(report.style$extra3[k])) );\n",
    "\n",
    "\t\t\t\t\tgrid.text(  str.val,\n",
    "\t\t\t\t\t\t\t\tx = report.style$position[k] + report.style$width[k]/2.0,\n",
    "\t\t\t\t\t\t\t\ty = y0+0.005,\n",
    "\t\t\t\t\t\t\t\trot = 0,\n",
    "\t\t\t\t\t\t\t\tgp = gpar( cex=0.4, col=\"white\"),\n",
    "\t\t\t\t\t\t\t\tcheck.overlap = F );\n",
    "\t\t\t\t}\n",
    "\n",
    "\t\t\t\tif(report.style$style[k]==\"logo\" && !is.null(tfbs@rbp_info))\n",
    "\t\t\t\t{\n",
    "\t\t\t\t\tpushViewport( viewport( y = y0,\n",
    "\t\t\t\t\t\t\t\tx = report.style$position[k],\n",
    "\t\t\t\t\t\t\t\twidth = report.style$width[k],\n",
    "\t\t\t\t\t\t\t\theight = 1/M40,\n",
    "\t\t\t\t\t\t\t\tjust = c(\"left\",\"bottom\")));\n",
    "\n",
    "\t\t\t\t\tidx <-which( as.character(df.motif[i,k]) == as.character(tfbs@rbp_info$Motif_ID) );\n",
    "\t\t\t\t\tif(length(idx)>1)\n",
    "\t\t\t\t\t{\n",
    "\t\t\t\t\t\twarning(paste(\"Multiple matrice in the tfbs object for Motif ID:\", as.character(df.motif[i,k]),\n",
    "\t\t\t\t\t\t\t\", first matrix (index:\", idx[1], \") is used to draw logo.\\n\"));\n",
    "\t\t\t\t\t\tidx <- idx[1];\n",
    "\t\t\t\t\t}\n",
    "\n",
    "\t\t\t\t\tseqLogo( exp(t(tfbs@pwm[[idx]])), xaxis = FALSE, yaxis = FALSE);\n",
    "\t\t\t\t\tpopViewport();\n",
    "\t\t\t\t}\n",
    "\n",
    "\t\t\t}\n",
    "\t\t}\n",
    "\n",
    "\t\tgrid.lines( x = c(0, 1), y=c( y0-1/M40, y0-1/M40 ), gp=gpar(col=\"black\", lwd=1) );\n",
    "\t\treturn(1 - ( y0 - 1/M40) );\n",
    "\t}\n",
    "\n",
    "\tview.height <- 1;\n",
    "\tview.lines  <- 28;\n",
    "\tif(!is.na(file.pdf))\n",
    "\t{\n",
    "\t\tif (report.size==\"letter\")\n",
    "\t\t{\n",
    "\t\t\tview.height <- 10/6;\n",
    "\t\t\tview.lines  <- floor(28*10/6);\n",
    "\t\t\tpdf( file.pdf, width=8, height=10 );\n",
    "\t\t}\n",
    "\t\telse\n",
    "\t\t\tpdf( file.pdf, width=6, height=6 );\n",
    "\n",
    "\t}\n",
    "\n",
    "\tfor(i in 1:ceiling(NROW(df.motif)/view.lines))\n",
    "\t{\n",
    "\t\tgrid.newpage();\n",
    "\t\tpushViewport( viewport(x=0, y=0,\n",
    "\t\t\t\t\t\twidth=1, height=1,\n",
    "\t\t\t\t\t\tjust=c(\"left\",\"bottom\")) );\n",
    "\n",
    "\t\tpushViewport( viewport(x=0.15, y=0.97,\n",
    "\t\t\t\t\t\twidth=0.7, height=0.07,\n",
    "\t\t\t\t\t\tjust=c(\"left\",\"top\"),\n",
    "\t\t\t\t\t\txscale = c(0, 1),\n",
    "\t\t\t\t\t\tyscale = c(0, 1) ) )\n",
    "\t\tdraw_top();\n",
    "\t\tpopViewport();\n",
    "\n",
    "\t\tpushViewport( viewport(x=0.15, y=0.90,\n",
    "\t\t\t\t\t\twidth=0.7, height=0.82,\n",
    "\t\t\t\t\t\tjust=c(\"left\",\"top\"),\n",
    "\t\t\t\t\t\txscale = c(0, 1),\n",
    "\t\t\t\t\t\tyscale = c(0, 1) ) )\n",
    "\t\ty0 <- draw_item( df.motif, (i-1)*view.lines+1, min( i*view.lines, NROW(df.motif) ) );\n",
    "\t\tpopViewport();\n",
    "\n",
    "\t\tpushViewport( viewport(x=0.15, y=0.9-0.82*y0,\n",
    "\t\t\t\t\t\twidth=0.7, height=0.04,\n",
    "\t\t\t\t\t\tjust=c(\"left\",\"top\"),\n",
    "\t\t\t\t\t\txscale = c(0, 1),\n",
    "\t\t\t\t\t\tyscale = c(0, 1) ) )\n",
    "\t\tdraw_bottom();\n",
    "\t\tpopViewport();\n",
    "\n",
    "\t}\n",
    "\n",
    "\tpopViewport();\n",
    "\n",
    "\tif(!is.na(file.pdf)) dev.off();\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "95b0b354",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'rbpbs.enrichmentTest'"
      ],
      "text/latex": [
       "'rbpbs.enrichmentTest'"
      ],
      "text/markdown": [
       "'rbpbs.enrichmentTest'"
      ],
      "text/plain": [
       "[1] \"rbpbs.enrichmentTest\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rbpbs_enrichmentTest<-function( rbpbs, \n",
    "                               fasta.pos,\n",
    "                               fasta.neg,\n",
    "                               file.prefix = NA,\n",
    "                               use.cluster = FALSE,\n",
    "                               ncores = 1,\n",
    "                               gc.correction = TRUE,\n",
    "                               gc.correction.pdf = NA,\n",
    "                               gc.min.sample = 500,\n",
    "                               gc.robust.rep = NA,\n",
    "                               threshold = 6,\n",
    "                               threshold.type = c(\"score\", \"fdr\"),\n",
    "                               gc.groups = 1,\n",
    "                               background.order = 2,\n",
    "                               background.length = 100000,\n",
    "                               pv.adj = p.adjust.methods)\n",
    "{\n",
    "  stopifnot(class(rbpbs) == \"rbpbs\")\n",
    "  \n",
    "  #if( !check_bed( positive.bed ) )\n",
    "   # stop(\"Wrong format in the parameter of 'positive.bed', at least three columns including chromosome, strat, stop.\");\n",
    "  \n",
    "#  if( !missing(negative.bed) && !check_bed( negative.bed ) )\n",
    " #   stop(\"Wrong format in the parameter of 'negative.bed', at least three columns including chromosome, strat, stop.\");\n",
    "  \n",
    "  if( missing( threshold.type ) ) threshold.type <- \"score\";\n",
    "  if( threshold.type == \"score\" && missing( threshold ) ) threshold <- 6;\n",
    "  if( threshold.type == \"fdr\" && missing( threshold ) ) threshold <- 0.1;\n",
    "  \n",
    "  if( !missing(pv.adj)) pv.adj <- match.arg(pv.adj)\n",
    "  if( missing(pv.adj) ) pv.adj <- \"bonferroni\";\n",
    "  if( pv.adj == \"fdr\" ) pv.adj <- \"BH\";\n",
    "  \n",
    "  if( missing(gc.robust.rep) || is.na(gc.robust.rep)) gc.robust.rep <- 1;\n",
    "  if( gc.robust.rep >1 && gc.robust.rep<3 ) gc.robust.rep <- 3;\n",
    "  \n",
    "  if( missing( gc.min.sample ) ) gc.min.sample<- 500;\n",
    "  if( missing( ncores) ) ncores <- 1;\n",
    "  if( missing( gc.groups) ) gc.groups <- 1;\n",
    "  if( missing( gc.correction) ) gc.correction <- TRUE;\n",
    "  if( missing( background.order ) ) background.order <- 2;\n",
    "  if( missing( background.length ) ) background.length <- 1000;\n",
    "  if( missing( use.cluster) ) use.cluster <- FALSE;\n",
    "  if( use.cluster && NROW(tfbs@cluster)==0 )\n",
    "    stop(\"No cluster information in the tfbs object\");\n",
    "  if( use.cluster && NCOL(tfbs@cluster)!=3 )\n",
    "    stop(\"No selected motif for each cluster in the tfbs object\");\n",
    "  \n",
    "  #file.twoBit = file.genome;\n",
    "  #if( tolower( file_ext( file.genome ) ) != \"2bit\" )\n",
    "  #{\n",
    "   # file.twoBit = tempfile(fileext=\".2bit\")\n",
    "    \n",
    "    # generate fasta file\n",
    "   # err_code <- system(paste(\"faToTwoBit \", file.genome, \" \", file.twoBit), wait = TRUE);\n",
    "    #if( err_code != 0 || !file.exists (file.twoBit) )\n",
    "     # stop(\"Failed to call faToTwoBit to convert FASTFA file.\");\n",
    "#  }\n",
    "  \n",
    "  if( use.cluster )\n",
    "  {\n",
    "    r.mat <- range( rbpbs@cluster[,1] );\n",
    "    if( r.mat[1]<1 || r.mat[2]>rbpbs@nrbps )\n",
    "      stop(\"The first column of 'tfbs@cluster' exceeds the range of motif data set.\");\n",
    "    \n",
    "    cluster.mat <- rbpbs@cluster[ rbpbs@cluster[,3]==1, c(1,2), drop=F];\n",
    "  }\n",
    "  else\n",
    "    ## if dont use cluster to correct p-value, here each individual is a group\n",
    "    cluster.mat <- cbind( 1:rbpbs@nrbps, 1:rbpbs@nrbps);\n",
    "  \n",
    " # if( missing(negative.bed) )\n",
    "  #{\n",
    "   # negative.bed <- background.generate( positive.bed );\n",
    "    #cat(\"*\", NROW(negative.bed),  \"GC negative loci are randomly generated.\\n\");\n",
    "#  }\n",
    "  \n",
    "  # read sequences\n",
    "  positive.ms = read.ms(fasta.pos)\n",
    "  negative.ms = read.ms(fasta.neg)\n",
    "  \n",
    "  \n",
    "  \n",
    "  r.comp <- comparative_scanDb_rtfbs( rbpbs,\n",
    "                                      fasta.pos,\n",
    "                                      fasta.neg,\n",
    "                                      positive.ms,\n",
    "                                      negative.ms,\n",
    "                                      file.prefix,\n",
    "                                      cluster.mat = cluster.mat,\n",
    "                                      ncores = ncores,\n",
    "                                      gc.correction = gc.correction,\n",
    "                                      gc.correction.pdf = gc.correction.pdf,\n",
    "                                      gc.min.sample = gc.min.sample,\n",
    "                                      gc.correction.verbose = TRUE,\n",
    "                                      threshold = threshold ,\n",
    "                                      threshold.type = threshold.type,\n",
    "                                      gc.groups = gc.groups,\n",
    "                                      background.order = background.order,\n",
    "                                      background.length = background.length,\n",
    "                                      pv.adj=pv.adj );\n",
    "  \n",
    "  if( is.null(r.comp) ) return(NULL);\n",
    "  ret <- r.comp$ret;\n",
    "  \n",
    "  if( !is.null(r.comp$bg.sample)  && gc.robust.rep > 1 )\n",
    "  {\n",
    "    ret.list <- list();\n",
    "    ret.list[[1]] <- ret;\n",
    "    \n",
    "    for(i in 2:gc.robust.rep)\n",
    "    {\n",
    "      cat(\"* GC robust replication for background resampling, loop=\", i, \"\\n\");\n",
    "      r.comp <- compartive_scanDb_rtfbs( rbpbs,\n",
    "                                          fasta.pos,\n",
    "                                          fasta.neg,\n",
    "                                  \n",
    "                                          positive.ms,\n",
    "                                          negative.ms,\n",
    "                                          file.prefix,\n",
    "                                          cluster.mat = cluster.mat,\n",
    "                                          ncores = ncores,\n",
    "                                          gc.correction = gc.correction,\n",
    "                                          gc.correction.pdf = gc.correction.pdf,\n",
    "                                          gc.min.sample = gc.min.sample,\n",
    "                                          gc.correction.verbose = FALSE,\n",
    "                                          threshold = threshold ,\n",
    "                                          threshold.type = threshold.type,\n",
    "                                          gc.groups = gc.groups,\n",
    "                                          background.order = background.order,\n",
    "                                          background.length = background.length,\n",
    "                                          pv.adj=pv.adj );\n",
    "      ret.list[[i]] <- r.comp$ret;\n",
    "    }\n",
    "    \n",
    "    for(i in 1:NROW(ret))\n",
    "    {\n",
    "      Npos.list <- c();\n",
    "      Nneg.list <- c();\n",
    "      \n",
    "      for(k in 1:gc.robust.rep)\n",
    "      {\n",
    "        Npos.list <- c( Npos.list, ret.list[[k]][i,'Npos'] );\n",
    "        Nneg.list <- c( Nneg.list, ret.list[[k]][i,'Nneg'] );\n",
    "      }\n",
    "      \n",
    "      Npos <- median( Npos.list );\n",
    "      Nneg <- median( Nneg.list );\n",
    "      n.gc.pos <- ret.list[[1]][i,'gc.pos'];\n",
    "      n.gc.neg <- ret.list[[1]][i,'gc.neg'];\n",
    "      Nneg.expected <- Nneg * n.gc.pos/ n.gc.neg ;\n",
    "      \n",
    "      Npos <- ifelse( Npos==0, 1, Npos );\n",
    "      Nneg <- ifelse( Nneg==0, 1, Nneg );\n",
    "      \n",
    "      fe.ratio <- ( Npos / n.gc.pos )/( Nneg / n.gc.neg );\n",
    "      \n",
    "      pos.left <-  n.gc.pos - Npos;\n",
    "      if( pos.left<0 ) pos.left  <- 0;\n",
    "      neg.left <- n.gc.neg - Nneg;\n",
    "      if( neg.left<0 ) neg.left  <- 0;\n",
    "      \n",
    "      tbl  = rbind( c( Npos, Nneg ), c( pos.left, neg.left ) )\n",
    "      pval = fisher.test(tbl)$p.value;\n",
    "      \n",
    "      ret[i, 'Npos']     <- Npos;\n",
    "      ret[i, 'Nneg']     <- Nneg;\n",
    "      ret[i, 'expected'] <- round( Nneg.expected, 1);\n",
    "      ret[i, 'fe.ratio'] <- fe.ratio;\n",
    "      ret[i, 'pvalue']   <- pval;\n",
    "    }\n",
    "    \n",
    "    ret$pv.adj <- adjust.pvale( ret$pvalue, cluster.mat, pv.adj );\n",
    "  }\n",
    "  \n",
    "  ret$Nneg   <- NULL;\n",
    "  ret$gc.pos <- NULL;\n",
    "  ret$gc.neg <- NULL;\n",
    "\n",
    "  if(!is.null(ret))\n",
    "  {\n",
    "    if( NROW(rbpbs@rbp_info) > 0 )\n",
    "    {\n",
    "      rbp.motifid   <- unlist(strsplit(as.character( ret$rbp.name ), \"@\"))[seq(1, length(ret$rbp.name)*2-1, 2)];\n",
    "      rbp.idx       <- match( rbp.motifid, rbpbs@rbp_info$Motif_ID );\n",
    "      ret$rbp.name  <- rbpbs@rbp_info$RBP_Name[ rbp.idx ];\n",
    "      ret$motif.id <- rbpbs@rbp_info$Motif_ID[ rbp.idx ];\n",
    "    }\n",
    "  }\n",
    "\n",
    "  r.parm <- list( file.genome   = NA,\n",
    "                  file.prefix       = file.prefix,\n",
    "                  use.cluster       = use.cluster,\n",
    "                  cluster.mat       = cluster.mat,\n",
    "                  ncores            = ncores,\n",
    "                  gc.robust.rep     = gc.robust.rep,\n",
    "                  threshold.type    = threshold.type,\n",
    "                  threshold         = threshold,\n",
    "                  pv.adj            = pv.adj,\n",
    "                  background.order  = background.order,\n",
    "                  background.length = background.length,\n",
    "                  gc.min.sample     = gc.min.sample,\n",
    "                  gc.correction     = gc.correction );\n",
    "  \n",
    "  ## result structure\n",
    "  ## 'motif.id'   Motif ID.\n",
    "  ## 'tf.name'    TF name.\n",
    "  ## 'Npos'       TF site count found in positive ranges.\n",
    "  ## 'expected'   TF site count found in negative ranges.\n",
    "  ## 'fe.ratio'   Ratio of fold enrichment.\n",
    "  ## 'pvalue'     p-value calculated by fisher test.\n",
    "  ## 'pv.adj'     p-value corrected by the multiple correction.\n",
    "  ## 'starch'     Binary filename of detected TF sites.\n",
    "  \n",
    "  r <- list( result = ret, parm = r.parm);\n",
    " # class(r) <- c(\"tfbs.enrichment\");\n",
    "  \n",
    "  return(r);\n",
    "}\n",
    "\n",
    "\n",
    "setGeneric(\"rbpbs.enrichmentTest\",\n",
    "           def = function( rbpbs,\n",
    "                           fasta.pos,\n",
    "                           fasta.neg,\n",
    "                           file.prefix  = NA,\n",
    "                           use.cluster  = FALSE,\n",
    "                           ncores = 1,\n",
    "                           gc.correction = TRUE,\n",
    "                           gc.correction.pdf = NA,\n",
    "                           gc.min.sample = 500,\n",
    "                           gc.robust.rep = NA,\n",
    "                           threshold = 6,\n",
    "                           threshold.type = c(\"score\", \"fdr\"),\n",
    "                           gc.groups=1,\n",
    "                           background.order = 2,\n",
    "                           background.length = 100000,\n",
    "                           pv.adj=p.adjust.methods)\n",
    "           {\n",
    "             stopifnot(class(rbpbs) == \"rbpbs\");\n",
    "             standardGeneric(\"rbpbs.enrichmentTest\");\n",
    "           })\n",
    "setMethod(\"rbpbs.enrichmentTest\", c(rbpbs=\"rbpbs\"), rbpbs_enrichmentTest)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "efa42a7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adjust.pvale<-function( r.pvalue, cluster.mat, pv.adj.method )\n",
    "{\n",
    "\t# If the cluster is used,...No cluster info ==>  cluster.mat[,2]=1\n",
    "\t# cluster.id <- unique( cluster.mat[,2] );\n",
    "\t# for(i in 1:length(cluster.id))\n",
    "\t# {\n",
    "\t#\tcluster.set <- which( cluster.mat[,2] == cluster.id[i] )\n",
    "\t#\n",
    "\t#\t# if the cluster index is called from the compare function, the p.adjust will be used to the cluster range, not all results.\n",
    "\t#\tr.pvalue[cluster.set] <- p.adjust( r.pvalue[cluster.set], method=pv.adj.method );\n",
    "\t# }\n",
    "\n",
    "\t# If the cluster is used,...No cluster info ==>  cluster.mat[,2]=1\n",
    "\tcluster.id <- unique( cluster.mat[,2] );\n",
    "\n",
    "\tpv.mins <- unlist(lapply( 1:length(cluster.id), function(i) {\n",
    "\t\tmin(r.pvalue[which( cluster.mat[,2] == cluster.id[i])]);\n",
    "\t\t}));\n",
    "\n",
    "\tpv.news <- p.adjust( pv.mins, method=pv.adj.method );\n",
    "\tfor(i in cluster.id)\n",
    "\t{\n",
    "\t\tcluster.set <- which( cluster.mat[,2] == i )\n",
    "\t\tr.pvalue[cluster.set] <- pmin( 1, r.pvalue[cluster.set] * pv.news[i]/pv.mins[i] );\n",
    "\t}\n",
    "\n",
    "\treturn(r.pvalue);\n",
    "}\n",
    "\n",
    "#create GC beans\n",
    "joint.gc.quantile.bins <- function(set1.ms, set2.ms, n = 4)\n",
    "{\n",
    "  gc1 = gcContent.ms(set1.ms)\n",
    "  gc2 = gcContent.ms(set2.ms)\n",
    "  \n",
    "  cutoffs = quantile(c(gc1, gc2), seq(from = 0, to = 1, length.out = n + 1), names = FALSE, na.rm=T)\n",
    "  bin1 <- findInterval(gc1, cutoffs, all.inside = TRUE)\n",
    "  bin2 <- findInterval(gc2, cutoffs, all.inside = TRUE)\n",
    "  \n",
    "  # check each bin has at least one sequence\n",
    "  n.seqs.per.bin = sapply(1:n, function(i) min(sum(bin1 == i, na.rm=T), sum(bin2 == i, na.rm=T)))\n",
    "  \n",
    "  stopifnot(all(n.seqs.per.bin > 0))\n",
    "  \n",
    "  return(list(bins1 = bin1, bins2 = bin2))\n",
    "}\n",
    "                          \n",
    "                          \n",
    "                          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8107f1ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "comparative_scan_rtfbs.ms <- function(pwm,\n",
    "\t\t\t\t\t\tpositive.ms,\n",
    "\t\t\t\t\t\tnegative.ms,\n",
    "\t\t\t\t\t\tbackground.ms,\n",
    "\t\t\t\t\t\tbackground.mm,\n",
    "\t\t\t\t\t\tthreshold = NA,\n",
    "\t\t\t\t\t\tthreshold.type = NA,\n",
    "\t\t\t\t\t\tcalc.empirical.pvalue = FALSE )\n",
    "{\n",
    "\tif (threshold.type == \"fdr\")\n",
    "\t{\n",
    "\t\tpos.sites = score.ms(positive.ms, pwm, background.mm, threshold = 0,strand=\"+\")\n",
    "\t\tneg.sites = score.ms(negative.ms, pwm, background.mm, threshold = 0,strand=\"+\")\n",
    "\t\tbg.sites  = score.ms(background.ms, pwm, background.mm, threshold = 0,strand=\"+\")\n",
    "\t\tfdrtbl    = calc.fdr(concat.ms(positive.ms, negative.ms), rbind(pos.sites, neg.sites), background.ms, bg.sites)\n",
    "\n",
    "\t\t## if something wrong\n",
    "\t\tif (is.null(fdrtbl))\n",
    "\t\t\treturn(NULL)\n",
    "\n",
    "\t\t# find first score threshold below FDR threshold\n",
    "\t\tidxs = which(fdrtbl$fdr <= threshold)\n",
    "\t\tif (length(idxs) == 0)\n",
    "\t\t\treturn(NULL)\n",
    "\n",
    "\t\tthresh = fdrtbl[idxs[length(idxs)], 1];\n",
    "\t\tpos.mask = pos.sites$score >= thresh;\n",
    "\t\tneg.mask = neg.sites$score >= thresh;\n",
    "\t\t#Npos = length(pos.sites$seqname[pos.mask]);\n",
    "\t\t#Nneg = length(neg.sites$seqname[neg.mask]);\n",
    "\t\tNpos = length(unique(pos.sites$seqname[pos.mask]));\n",
    "\t\tNneg = length(unique(neg.sites$seqname[neg.mask]));\n",
    "\n",
    "\t\t# NOTE: empirical p-values are computed under the assumption that the negative set\n",
    "\t\t#       is not bound by the specified factor (but is accessible)\n",
    "\t\tif (calc.empirical.pvalue) {\n",
    "\t\t\tepvals = sapply(pos.sites$score[pos.mask], function(thresh) {\n",
    "\t\t\t\tK = sum(neg.sites$score >= thresh)\n",
    "\t\t\t\tZ = length(neg.sites$score)\n",
    "\n",
    "\t\t\t\t(K + 1)/(Z + 1) # empirical p-value\n",
    "\t\t\t})\n",
    "\n",
    "\t\t\treturn(list(Nneg   = Nneg,\n",
    "\t\t\t\t\t\tNpos   = Npos,\n",
    "\t\t\t\t\t\tthresh = thresh,\n",
    "\t\t\t\t\t\tsites  = pos.sites[pos.mask, ],\n",
    "\t\t\t\t\t\tempirical.pvalues = epvals))\n",
    "\t\t}\n",
    "\n",
    "\t\treturn(list(Nneg = Nneg, Npos = Npos, thresh = thresh, sites = pos.sites[pos.mask, ]))\n",
    "\t}\n",
    "\t# threshold.type == score\n",
    "\telse\n",
    "\t{\n",
    "\t\tpos.sites = score.ms(positive.ms, pwm, background.mm, strand=\"+\", threshold = threshold)\n",
    "\t\tneg.sites = score.ms(negative.ms, pwm, background.mm, strand=\"+\",threshold = threshold)\n",
    "\n",
    "\t\t#Npos = length(pos.sites$seqname[pos.mask]);\n",
    "\t\t#Nneg = length(neg.sites$seqname[neg.mask]);\n",
    "\t\t\n",
    "\t\tNpos = length(unique(pos.sites$seqname))\n",
    "\t\tNneg = length(unique(neg.sites$seqname))\n",
    "\n",
    "\t\tif (calc.empirical.pvalue) {\n",
    "\t\t\tepvals = sapply(pos.sites$score, function(thresh) {\n",
    "\t\t\t\tK = sum(neg.sites$score >= thresh)\n",
    "\t\t\t\tZ = length(neg.sites$score)\n",
    "\n",
    "\t\t\t\t(K + 1)/(Z + 1) # empirical p-value\n",
    "\t\t\t})\n",
    "\n",
    "\t\t\treturn(list(Nneg   = Nneg,\n",
    "\t\t\t\t\t\tNpos   = Npos,\n",
    "\t\t\t\t\t\tthresh = threshold,\n",
    "\t\t\t\t\t\tsites  = pos.sites,\n",
    "\t\t\t\t\t\tempirical.pvalues = epvals))\n",
    "\t\t}\n",
    "\n",
    "\t\treturn(list(Nneg = Nneg, Npos = Npos, thresh = threshold, sites = pos.sites))\n",
    "\t}\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c699430d",
   "metadata": {},
   "outputs": [],
   "source": [
    "comparative_scan_rtfbs <- function(pwm, \n",
    "                                   fasta.pos,\n",
    "                                   fasta.neg,\n",
    "                                   threshold = NA,\n",
    "                                   threshold.type = NA,\n",
    "                                   gc.groups = 1,\n",
    "                                   background.order = 2,\n",
    "                                   background.length = 100000,\n",
    "                                   calc.empirical.pvalue = FALSE)\n",
    "{\n",
    "  # read sequences\n",
    "  positive.ms = read.ms(fasta.pos)\n",
    "  negative.ms = read.ms(fasta.neg)\n",
    "  \n",
    "  # compute GC quantile masks\n",
    "  gcBins = joint.gc.quantile.bins(positive.ms, negative.ms, n=gc.groups)\n",
    "  \n",
    "  # get per GC quantile results\n",
    "  result.sites = NULL\n",
    "  Npos   = 0\n",
    "  Nneg   = 0\n",
    "  thresh = NULL\n",
    "  epvals = NULL\n",
    "  \n",
    "  for (idx in 1:gc.groups)\n",
    "  {\n",
    "    # compute background model\n",
    "    pos.ms.i = positive.ms[gcBins$bins1 == idx]\n",
    "    neg.ms.i = negative.ms[gcBins$bins2 == idx]\n",
    "    both.ms  = concat.ms(pos.ms.i, neg.ms.i)\n",
    "    bg.mm    = build.mm(both.ms, background.order)\n",
    "    \n",
    "    # generate background sequences\n",
    "    bg.ms <- simulate.ms(bg.mm, background.length)\n",
    "    \n",
    "    # collect sequences\n",
    "    res = comparative_scan_rtfbs.ms(pwm,\n",
    "                                    pos.ms.i,\n",
    "                                    neg.ms.i,\n",
    "                                    bg.ms,\n",
    "                                    bg.mm,\n",
    "                                    threshold = threshold,\n",
    "                                    threshold.type = threshold.type,\n",
    "                                    calc.empirical.pvalue = calc.empirical.pvalue)\n",
    "    \n",
    "    if (!is.null(res)) {\n",
    "      result.sites = rbind(result.sites, res$sites)\n",
    "      Npos   = Npos + res$Npos\n",
    "      Nneg   = Nneg + res$Nneg\n",
    "      thresh = c(thresh, res$thresh)\n",
    "      \n",
    "      if (calc.empirical.pvalue)\n",
    "        epvals = c(epvals, res$empirical.pvalues)\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  # fisher test\n",
    "  pos.left <- length(pos.ms) - Npos;\n",
    "  if( pos.left<0 ) pos.left <- 0;\n",
    "  neg.left <- length(neg.ms) - Nneg;\n",
    "  if( neg.left<0 ) neg.left <- 0;\n",
    "  \n",
    "  tbl = rbind(c(Npos, Nneg), c( pos.left, neg.left ));\n",
    "  pval = fisher.test(tbl)$p.value;\n",
    "  \n",
    "  # merge results\n",
    "  bed = tfbs_to_bed(result.sites, \"pwm\")\n",
    "  if (!is.null(bed) && dim(bed)[1] > 0) {\n",
    "    ord = order(bed[,1], bed[,2])\n",
    "    bed = bed[ord,]\n",
    "    \n",
    "    if (calc.empirical.pvalue)\n",
    "      epvals = epvals[ord]\n",
    "  }\n",
    "  \n",
    "  if (calc.empirical.pvalue)\n",
    "    result = list(Npos     = Npos,\n",
    "                  Nneg     = Nneg,\n",
    "                  pvalue   = pval,\n",
    "                  thresh   = thresh,\n",
    "                  sites    = bed,\n",
    "                  empirical.pvalues = epvals)\n",
    "  else\n",
    "    result = list(Npos   = Npos,\n",
    "                  Nneg   = Nneg,\n",
    "                  pvalue = pval,\n",
    "                  thresh = thresh,\n",
    "                  sites  = bed)\n",
    "}\n",
    "\n",
    "tfbs_to_bed <- function(sites, rbp.name)\n",
    "{\n",
    "\t## Alternative formulation fails if sites is empty.\n",
    "\tif (NROW(sites) == 0)\n",
    "\t\treturn(NULL)\n",
    "\n",
    "\tspl <- strsplit(as.character(sites$seqname), \":|-\")\n",
    "\n",
    "\tchroms <- sapply(spl, function(pair) pair[1]);\n",
    "\tstarts <- as.integer(sapply(spl, function(pair) pair[2]));\n",
    "\n",
    "\tbed = data.frame(\n",
    "\t\t\tchrom      = chroms,\n",
    "\t\t\tchromStart = as.integer(starts + sites$start - 1), # add site offset\n",
    "\t\t\tchromEnd   = as.integer(starts + sites$end),\n",
    "\t\t\tname       = rbp.name,\n",
    "\t\t\tscore      = sites$score,\n",
    "\t\t\tstrand     = sites$strand )\n",
    "\n",
    "\treturn(bed)\n",
    "}\n",
    "\n",
    "background.check<-function( gc.pos, gc.neg, gc.correction, file.pdf.vioplot=NA, gc.min.sample = 500, verbose=TRUE )\n",
    "{\n",
    "\tpdf.output <- FALSE;\n",
    "\tgc.test <- wilcox.test(gc.pos, gc.neg, conf.int=TRUE, conf.level=0.9 );\n",
    "\n",
    "\t# Actually 0.01 is not good for wilcox.test,\n",
    "\t# We use the customizable pvalue in this function.\n",
    "\tif( gc.test$p.value < 0.01 && verbose)\n",
    "\t{\n",
    "\t\tcat(\"! Difference between GC content in negative and positive TREs (use 'gc.correction.pdf' to see pdf figure):\\n\");\n",
    "\t\tcat(\"  p-value (Wilcoxon-Mann-Whitney test) =\", gc.test$p.value, \"\\n\");\n",
    "\t\tcat(\"  median/sample size of GC positive =\", median(gc.pos), \"/\", length(gc.pos), \"\\n\");\n",
    "\t\tcat(\"  median/sample size of GC negative =\", median(gc.neg), \"/\", length(gc.neg), \"\\n\");\n",
    "\n",
    "\t\tif( !is.null(file.pdf.vioplot) && !is.na( file.pdf.vioplot)  )\n",
    "\t\t{\n",
    "\t\t\tr.try <- try ( pdf(file.pdf.vioplot) );\n",
    "\t\t\tif( class(r.try) != \"try-error\" )\n",
    "\t\t\t{\n",
    "\t\t\t\tvioplot( gc.pos, gc.neg, names=c(\"Positive\", \"Negative\") );\n",
    "\t\t\t\tabline(h=median(gc.pos), lty=\"dotted\");\n",
    "\t\t\t\tpdf.output <- TRUE;\n",
    "\t\t\t}\n",
    "\t\t\telse\n",
    "\t\t\t\tcat(\"  Failed to output the vioplot figure for the results.\\n\");\n",
    "\t\t}\n",
    "\t}\n",
    "\n",
    "\t# No need to do sampling the background data.\n",
    "\tif( gc.test$p.value > 0.01 || gc.correction==FALSE )\n",
    "\t{\n",
    "\t\t# PDF has not been finished until dev.off();\n",
    "\t\tif( pdf.output )\n",
    "\t\t{\n",
    "\t\t\tdev.off();\n",
    "\t\t\tcat(\"* Please check the vioplot figure to make sure, the vioplot figure: \", file.pdf.vioplot, \"\\n\" );\n",
    "\t\t}\n",
    "\t\treturn(NULL);\n",
    "\t}\n",
    "\n",
    "\tif( length(gc.neg)< 5000 )\n",
    "\t{\n",
    "\t\tif(verbose) cat(\"! Failed to make GC correction due to small bed data(size<5000).\\n\");\n",
    "\t\treturn(NULL);\n",
    "\t}\n",
    "#resample :n=10k default was changed to 4000\n",
    "\tresample <- function(ref, orig, n=10000, nbins=10) {\n",
    "\t\t## (1) Break gcContent down into 10 equally sized bins.\n",
    "\t\tbreaks <- seq(0, 1, length.out=nbins) #seq(min(ref), max(ref), length.out=nbins)\n",
    "\n",
    "\t\t## (2) Get the empirical frequencies of each bin.\n",
    "\t\tempir <- sapply(1:(NROW(breaks)-1), function(x) {sum(breaks[x]<= ref & ref < breaks[x+1])/NROW(ref)})\n",
    "\n",
    "\t\t## (3) Re-sample TREs in the BG set w/ probability proportional to the bin.\n",
    "\t\tresamp_prob <- rep(1, NROW(orig))\n",
    "\t\tfor(i in 1:NROW(empir)) {\n",
    "\t\t\tincl <- breaks[i] <= orig & orig < breaks[i+1]\n",
    "\t\t\tresamp_prob[incl] <- empir[i]/ sum(incl)\n",
    "\t\t}\n",
    "\n",
    "\t\tsample(1:NROW(resamp_prob), n, prob= resamp_prob, replace=FALSE)\n",
    "\t}\n",
    "\n",
    "\ttry.sample<-function(n.sample)\n",
    "\t{\n",
    "\t\tindx.bgnew <- resample( gc.pos, gc.neg, n=n.sample)\n",
    "\t\tgc.testx <- wilcox.test(gc.pos, gc.neg[indx.bgnew], conf.int=TRUE, conf.level=0.9 );\n",
    "\t\treturn(gc.testx$p.value);\n",
    "\t}\n",
    "\n",
    "\t## Make n.sample as large as possible, up to 10k sequences.  More BG gains statistical power.\n",
    "\t#n.sample <- min(c(10000, round(length(gc.neg)/10)))\n",
    "\n",
    "\t# In order to get the high p-value, multiple resample sizes are used to do test, find a best one.\n",
    "\t# Need to consider ?\n",
    "\n",
    "\t#ns.sample <- c ( round(length(gc.neg)/c(5,10,15,20,25)), 1000);\n",
    "\t#ns.sample <- ns.sample[ ns.sample>=1000 ];\n",
    "\n",
    "\tns.sample <- c ( round(length(gc.neg)/c( 2:10,15,20,25)), 1000, 800, 500 );\n",
    "\tns.sample <- ns.sample[ ns.sample >= gc.min.sample  ];\n",
    "\n",
    "\tns.pvalue <- unlist(lapply(ns.sample, try.sample));\n",
    "\tn.sample  <- ns.sample[ which.max(ns.pvalue) ];\n",
    "\n",
    "\tindx.bgnew <- resample( gc.pos, gc.neg, n=n.sample)\n",
    "\tgc.test2 <- wilcox.test(gc.pos, gc.neg[indx.bgnew], conf.int=TRUE, conf.level=0.9 );\n",
    "\n",
    "\tif(verbose)\n",
    "\t{\n",
    "\t\tcat(\"* After the resampling from negative TREs:\\n\");\n",
    "\t\tcat(\"  p-value (Wilcoxon-Mann-Whitney test) =\", gc.test2$p.value, \"\\n\");\n",
    "\t\tcat(\"  median/sample size of GC negative =\", median(gc.neg[indx.bgnew]), \"/\", length(indx.bgnew), \"\\n\");\n",
    "\t}\n",
    "\n",
    "\tif( pdf.output )\n",
    "\t{\n",
    "\t\tvioplot(gc.pos, gc.neg, gc.neg[indx.bgnew], names=c(\"Positive\", \"Negative\", \"Negative.resample\"));\n",
    "\t\tabline(h=median(gc.pos), lty=\"dotted\")\n",
    "\t\tdev.off();\n",
    "\t\tcat(\"* Please check the vioplot figure to make sure, the vioplot figure: \", file.pdf.vioplot, \"\\n\" );\n",
    "\t}\n",
    "\n",
    "\t## return sampling background.\n",
    "\treturn( indx.bgnew );\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f975aa8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "comparative_scanDb_rtfbs <- function( rbpbs,\n",
    "                                      fasta.pos,\n",
    "                                      fasta.neg,\n",
    "                                      positive.ms = NULL,\n",
    "                                      negative.ms = NULL,\n",
    "                                      file.prefix = NA,\n",
    "                                      cluster.mat = NA,\n",
    "                                      ncores = 1,\n",
    "                                      gc.correction = TRUE,\n",
    "                                      gc.correction.pdf = NA,\n",
    "                                      gc.min.sample = 500,\n",
    "                                      gc.correction.verbose = TRUE,\n",
    "                                      threshold = NA,\n",
    "                                      threshold.type = NA,\n",
    "                                      gc.groups=1,\n",
    "                                      background.order = 2,\n",
    "                                      background.length = 100000,\n",
    "                                      pv.adj = NA )\n",
    "{\n",
    "  stopifnot(class(rbpbs) == \"rbpbs\")\n",
    "  \n",
    "  if( missing(pv.adj) ) pv.adj <- \"bonferroni\";\n",
    "  \n",
    "  if( !is.na(file.prefix))\n",
    "    if( !check_folder_writable( file.prefix ) )\n",
    "      stop(paste(\"Can not create files starting with the prefix:\", file.prefix));\n",
    "  \n",
    " # if(NROW(negative.bed) < 50)\n",
    "  #  stop(paste(\"Negative BED file contains only \", NROW(negative.bed), \" entries.  Strongly suggest size is increased.\"))\n",
    "  \n",
    "  # read sequences\n",
    "  if(is.null(positive.ms)) positive.ms = read.ms(fasta.pos)\n",
    "  if(is.null(negative.ms)) negative.ms = read.ms(fasta.neg)\n",
    "  \n",
    "  gc.pos <- gcContent.ms(positive.ms);\n",
    "  gc.neg <- gcContent.ms(negative.ms);\n",
    "  \n",
    "  # for unknown reason, some ms objects get NaN values for gc content.\n",
    "  # here we need to remove these data.\n",
    "  if( length(which(is.na(gc.pos)))>0 )\n",
    "  {\n",
    "    positive.ms <- positive.ms [ -which(is.na(gc.pos)) ];\n",
    "    gc.pos <- gc.pos [ -which(is.na(gc.pos)) ];\n",
    "  }\n",
    "  \n",
    "  if( length(which(is.na(gc.neg)))>0 )\n",
    "  {\n",
    "    negative.ms <- negative.ms [ -which(is.na(gc.neg)) ];\n",
    "    gc.neg <- gc.neg [ -which(is.na(gc.neg)) ];\n",
    "  }\n",
    "  \n",
    "  # detect the difference of gcContent between positive and negative TREs,\n",
    "  # if the difference is significant, make a correction for the negative TREs\n",
    "  # based on the resampling method.\n",
    "  \n",
    "  bg.sample <- background.check( gc.pos,\n",
    "                                 gc.neg,\n",
    "                                 gc.correction,\n",
    "                                 gc.correction.pdf,\n",
    "                                 gc.min.sample ,\n",
    "                                 verbose=gc.correction.verbose )\n",
    "  if(!is.null(bg.sample))\n",
    "  {\n",
    "    gc.neg <- gc.neg[ bg.sample ];\n",
    "    negative.ms  <- negative.ms [ bg.sample ];\n",
    "  }\n",
    "  \n",
    "  # compute CG quantile masks\n",
    "  gcBins = joint.gc.quantile.bins(positive.ms, negative.ms, n=gc.groups)\n",
    "  \n",
    "  # partition sequences\n",
    "  seqs.subsets = lapply(1:gc.groups, function(idx) {\n",
    "    pos.ms.i = positive.ms[gcBins$bins1 == idx]\n",
    "    neg.ms.i = negative.ms[gcBins$bins2 == idx]\n",
    "    \n",
    "    list(pos = pos.ms.i, neg = neg.ms.i)\n",
    "  })\n",
    "  \n",
    "  # compute background models\n",
    "  gcBacks = lapply(1:gc.groups, function(idx) {\n",
    "    both.ms = concat.ms(seqs.subsets[[idx]]$pos, seqs.subsets[[idx]]$neg)\n",
    "    build.mm(both.ms, background.order)\n",
    "  })\n",
    "  \n",
    "  # compute background sequences\n",
    "  gcBacks.ms = lapply(1:gc.groups, function(idx) {\n",
    "    simulate.ms(gcBacks[[idx]], background.length)\n",
    "  })\n",
    "  \n",
    "  if(missing(cluster.mat)) cluster.mat <- cbind(1:rbpbs@nrbps, 1);\n",
    "  \n",
    "  # iterate over TF set\n",
    "  binding_all <- mclapply(cluster.mat[,1], function(i, ...) {\n",
    "    # get PWM information\n",
    "    pwm = rbpbs@pwm[[i]];\n",
    "    \n",
    "    pwm.name = paste(rbpbs@mgisymbols[i], \"@\", i, sep='');\n",
    "    \n",
    "    # scan sequences, per GC bin\n",
    "    result.sites = NULL;\n",
    "    Npos = 0;\n",
    "    Nneg = 0;\n",
    "    \n",
    "    for (idx in 1:gc.groups) {\n",
    "      pos.ms.i = seqs.subsets[[idx]]$pos;\n",
    "      neg.ms.i = seqs.subsets[[idx]]$neg;\n",
    "      bg.i.mm  = gcBacks[[idx]];\n",
    "      bg.i.ms  = gcBacks.ms[[idx]];\n",
    "      \n",
    "      res = comparative_scan_rtfbs.ms( pwm,\n",
    "                                       pos.ms.i,\n",
    "                                       neg.ms.i,\n",
    "                                       bg.i.ms,\n",
    "                                       bg.i.mm,\n",
    "                                       threshold = threshold,\n",
    "                                       threshold.type = threshold.type );\n",
    "      \n",
    "      if (!is.null(res)) {\n",
    "        result.sites = rbind(result.sites, res$sites);\n",
    "        Npos = Npos + res$Npos;\n",
    "        Nneg = Nneg + res$Nneg;\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    # process result\n",
    "    #result.bed = tfbs_to_bed(result.sites, pwm.name)\n",
    "    \n",
    "    pos.left <- length(gc.pos) - Npos;\n",
    "    if( pos.left<0 ) pos.left <- 0;\n",
    "    neg.left <- length(gc.neg) - Nneg;\n",
    "    if( neg.left<0 ) neg.left <- 0;\n",
    "    \n",
    "    tbl  = rbind(c(Npos, Nneg), c(pos.left, neg.left))\n",
    "    pval = fisher.test(tbl)$p.value;\n",
    "    \n",
    "    # save sites\n",
    "    starch.file <- NA;\n",
    "    if( !is.na(file.prefix) )\n",
    "    {\n",
    "      starch.file = paste(file.prefix, \".\", i, \".bed.starch\", sep='')\n",
    "      starch.file <- write.starchbed(result.bed, starch.file);\n",
    "    }\n",
    "    \n",
    "    Npos.tmp <- Npos;\n",
    "    if(Npos.tmp==0) Npos.tmp <- 1;\n",
    "    Nneg.tmp <- Nneg;\n",
    "    if(Nneg.tmp==0) Nneg.tmp <- 1;\n",
    "    \n",
    "    Nneg.expected <- Nneg * length(gc.pos)/length(gc.neg);\n",
    "    fe.ratio <- (Npos.tmp/length(gc.pos) )/(Nneg.tmp/length(gc.neg));\n",
    "    \n",
    "    # return info\n",
    "    return(data.frame('motif.id' = \"\",\n",
    "                      'rbp.name'  = pwm.name,\n",
    "                      'Npos'     = Npos,\n",
    "                      'Nneg'     = Nneg,\n",
    "                      'expected' = round( Nneg.expected, 1),\n",
    "                      'fe.ratio' = fe.ratio,\n",
    "                      'pvalue'   = pval,\n",
    "                      'pv.adj'   = pval,\n",
    "                      'starch'   = starch.file,\n",
    "                      'gc.pos'   = length(gc.pos),\n",
    "                      'gc.neg'   = length(gc.neg) ) );\n",
    "  }, mc.cores = ncores)\n",
    "  \n",
    "  # recombine results\n",
    "  r.df <- do.call(\"rbind\", binding_all);\n",
    "  \n",
    "  #put this dummy variable here to pass R CMD check rtfbsdb --as-cran\n",
    "  starch <- NULL;\n",
    "  \n",
    "  if (NROW(r.df)>0)\n",
    "  {\n",
    "    r.df$pv.adj <- adjust.pvale( r.df$pvalue, cluster.mat, pv.adj );\n",
    "    \n",
    "    # the following 'starch' is a column in r.df, not equal NULL;\n",
    "    if( missing(file.prefix) || is.na(file.prefix) )\n",
    "      r.df <- subset( r.df, select = -starch);\n",
    "  }\n",
    "  \n",
    "  return( list(ret=r.df, bg.sample = bg.sample ) );\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "summary.rbpbs.enrichment<-function( object, pv.threshold=0.05, pv.adj=NA, ...)\n",
    "{\n",
    "\tr.comp <- object;\n",
    "\n",
    "\tr <- r.comp$result;\n",
    "\tif(!is.na(pv.adj))\n",
    "\t\tr$pv.adj <- adjust.pvale( r$pvalue, r.comp$parm$cluster.mat, pv.adj )\n",
    "\n",
    "\tr.comp.sum <- r[ order( r$pv.adj), c(\"motif.id\",\"rbp.name\",\"Npos\",\"expected\",\"pv.adj\",\"fe.ratio\") ];\n",
    "\n",
    "\tcolnames(r.comp.sum) <- c(\"Motif id\",\"TF name\",\"Npos\",\"Expected\",\"pv.adj\",\"Fold enrichment\");\n",
    "\n",
    "\treturn( r.comp.sum );\n",
    "}\n",
    "\n",
    "rbpbs.reportEnrichment<-function( tfbs,\n",
    "\t\t\t\t\t\t\t\t r.comp,\n",
    "\t\t\t\t\t\t\t\t file.pdf = NA,\n",
    "\t\t\t\t\t\t\t\t report.size = \"letter\",\n",
    "\t\t\t\t\t\t\t\t report.title = \"\",\n",
    "\t\t\t\t\t\t\t\t enrichment.type = c (\"both\", \"enriched\", \"depleted\"),\n",
    "\t\t\t\t\t\t\t\t sig.only = TRUE,\n",
    "\t\t\t\t\t\t\t\t pv.threshold = 0.05,\n",
    "\t\t\t\t\t\t\t\t pv.adj = NA,\n",
    "\t\t\t\t\t\t\t\t sorted = c (\"pvalue\", \"enrich.ratio\"))\n",
    "{\n",
    "\t#stopifnot(class(tfbs) == \"rbpbs\" && class(r.comp) == \"tfbs.enrichment\" )\n",
    "\n",
    "\tif( !is.null(file.pdf) && !is.na(file.pdf) )\n",
    "\t\tif( !check_folder_writable( file.pdf ) )\n",
    "\t\t\tstop( paste(\"Can not create pdf file: \",file.pdf ) );\n",
    "\n",
    "\tif(!missing(pv.adj))\n",
    "\t\tr.comp$result$pv.adj <- adjust.pvale( r.comp$result$pvalue, r.comp$parm$cluster.mat, pv.adj )\n",
    "\n",
    "\tif( !missing(sorted))\n",
    "\t\tsorted <- match.arg( sorted )\n",
    "\telse\n",
    "\t\tsorted <- \"pvalue\";\n",
    "\n",
    "\tr.comp.sel <- r.comp$result;\n",
    "\tif( sorted == \"enrich.ratio\")\n",
    "\t\tr.comp.sel <- r.comp.sel[ order(r.comp.sel$fe.ratio, decreasing = TRUE), c(\"motif.id\",\"rbp.name\",\"Npos\",\"expected\",\"pv.adj\",\"fe.ratio\"), drop=F ]\n",
    "\telse\n",
    "\t\tr.comp.sel <- r.comp.sel[ order(r.comp.sel$pvalue), c(\"motif.id\",\"rbp.name\",\"Npos\",\"expected\",\"pv.adj\",\"fe.ratio\"), drop=F ];\n",
    "\n",
    "\tif( !missing(enrichment.type))\n",
    "\t\tenrichment.type <- match.arg(enrichment.type)\n",
    "\telse\n",
    "\t\tenrichment.type <- \"both\";\n",
    "\n",
    "\tidx.sel <- 1:NROW(r.comp.sel);\n",
    "\tif( sig.only )\n",
    "\t{\n",
    "\t\tidx.sel <- which( r.comp.sel$pv.adj <= pv.threshold );\n",
    "\n",
    "\t\tif (enrichment.type==\"enriched\")\n",
    "\t\t\tidx.sel <- which( r.comp.sel$pv.adj <= pv.threshold & r.comp.sel$fe.ratio >= 1);\n",
    "\n",
    "\t\tif (enrichment.type==\"depleted\")\n",
    "\t\t\tidx.sel <- which( r.comp.sel$pv.adj <= pv.threshold & r.comp.sel$fe.ratio < 1);\n",
    "\t}\n",
    "\telse\n",
    "\t{\n",
    "\t\tif (enrichment.type==\"enriched\")\n",
    "\t\t\tidx.sel <- which( r.comp.sel$fe.ratio >= 1);\n",
    "\n",
    "\t\tif (enrichment.type==\"depleted\")\n",
    "\t\t\tidx.sel <- which( r.comp.sel$fe.ratio < 1);\n",
    "\t}\n",
    "\n",
    "\tif( length(idx.sel) > 0 )\n",
    "\t\tr.comp.sel <- r.comp.sel[ idx.sel, ,drop=F ]\n",
    "\telse\n",
    "\t{\n",
    "\t\tcat( \"! No motif information for the report.\\n\" );\n",
    "\t\treturn(invisible(NULL));\n",
    "\t}\n",
    "\n",
    "\tr.comp.new <- data.frame( No=c(1:NROW(r.comp.sel)),\n",
    "\t\t\t\t\t\t\t  r.comp.sel[, c(\"motif.id\",\"rbp.name\",\"Npos\",\"expected\",\"pv.adj\",\"fe.ratio\",\"motif.id\"), drop=F ] );\n",
    "\n",
    "\tdf.style <- data.frame(\n",
    "\t\t\tposition = numeric(0),\n",
    "\t\t\twidth    = numeric(0),\n",
    "\t\t\theader   = character(0),\n",
    "\t\t\thjust    = character(0),\n",
    "\t\t\tstyle    = character(0),\n",
    "\t\t\textra1   = character(0),\n",
    "\t\t\textra2   = character(0),\n",
    "\t\t\textra3   = character(0),\n",
    "\t\t\textra4   = character(0));\n",
    "\n",
    "\tdf.style <- rbind(df.style,\n",
    "\t\t\tdata.frame( position = 0.00,\n",
    "\t\t\t\t\t\twidth    = 0.04,\n",
    "\t\t\t\t\t\theader   = \"No.\",\n",
    "\t\t\t\t\t\thjust    = \"left\",\n",
    "\t\t\t\t\t\tstyle    = \"text\",\n",
    "\t\t\t\t\t\textra1   = \"0\",\n",
    "\t\t\t\t\t\textra2   = \"0\",\n",
    "\t\t\t\t\t\textra3   = \"0\",\n",
    "\t\t\t\t\t\textra4   = \"0\"));\n",
    "\tdf.style <- rbind(df.style,\n",
    "\t\t\tdata.frame( position = 0.04,\n",
    "\t\t\t\t\t\twidth    = 0.10,\n",
    "\t\t\t\t\t\theader   = \"Motif ID\",\n",
    "\t\t\t\t\t\thjust    = \"left\",\n",
    "\t\t\t\t\t\tstyle    = \"text\",\n",
    "\t\t\t\t\t\textra1   = \"0\",\n",
    "\t\t\t\t\t\textra2   = \"0\",\n",
    "\t\t\t\t\t\textra3   = \"0\",\n",
    "\t\t\t\t\t\textra4   = \"0\"));\n",
    "\tdf.style <- rbind(df.style,\n",
    "\t\t\tdata.frame( position = 0.14,\n",
    "\t\t\t\t\t\twidth    = 0.09,\n",
    "\t\t\t\t\t\theader   = \"TF Name\",\n",
    "\t\t\t\t\t\thjust    = \"left\",\n",
    "\t\t\t\t\t\tstyle    = \"text\",\n",
    "\t\t\t\t\t\textra1   = \"0\",\n",
    "\t\t\t\t\t\textra2   = \"0\",\n",
    "\t\t\t\t\t\textra3   = \"0\",\n",
    "\t\t\t\t\t\textra4   = \"0\"));\n",
    "\tdf.style <- rbind(df.style,\n",
    "\t\t\tdata.frame( position = 0.23,\n",
    "\t\t\t\t\t\twidth    = 0.06,\n",
    "\t\t\t\t\t\theader   = \"N. Pos.\",\n",
    "\t\t\t\t\t\thjust    = \"centre\",\n",
    "\t\t\t\t\t\tstyle    = \"text\",\n",
    "\t\t\t\t\t\textra1   = \"0\",\n",
    "\t\t\t\t\t\textra2   = \"0\",\n",
    "\t\t\t\t\t\textra3   = \"0\",\n",
    "\t\t\t\t\t\textra4   = \"0\"));\n",
    "\tdf.style <- rbind(df.style,\n",
    "\t\t\tdata.frame( position = 0.29,\n",
    "\t\t\t\t\t\twidth    = 0.06,\n",
    "\t\t\t\t\t\theader   = \"Expected.\",\n",
    "\t\t\t\t\t\thjust    = \"centre\",\n",
    "\t\t\t\t\t\tstyle    = \"text\",\n",
    "\t\t\t\t\t\textra1   = \"0\",\n",
    "\t\t\t\t\t\textra2   = \"0\",\n",
    "\t\t\t\t\t\textra3   = \"0\",\n",
    "\t\t\t\t\t\textra4   = \"0\"));\n",
    "\tdf.style <- rbind(df.style,\n",
    "\t\t\tdata.frame( position = 0.35,\n",
    "\t\t\t\t\t\twidth    = 0.08,\n",
    "\t\t\t\t\t\theader   = \"p-value\",\n",
    "\t\t\t\t\t\thjust    = \"centre\",\n",
    "\t\t\t\t\t\tstyle    = \"bar\",\n",
    "\t\t\t\t\t\textra1   = \"1e-6\",\n",
    "\t\t\t\t\t\textra2   = \"1\",\n",
    "\t\t\t\t\t\textra3   = \"1\",\n",
    "\t\t\t\t\t\textra4   = \"1\"));\n",
    "\tdf.style <- rbind(df.style,\n",
    "\t\t\tdata.frame( position = 0.43,\n",
    "\t\t\t\t\t\twidth    = 0.08,\n",
    "\t\t\t\t\t\theader   = \"*Enrichment\",\n",
    "\t\t\t\t\t\thjust    = \"centre\",\n",
    "\t\t\t\t\t\tstyle    = \"bar\",\n",
    "\t\t\t\t\t\textra1   = \"-10\",\n",
    "\t\t\t\t\t\textra2   = \"10\",\n",
    "\t\t\t\t\t\textra3   = \"0\",\n",
    "\t\t\t\t\t\textra4   = \"2\"));\n",
    "\tdf.style <- rbind(df.style,\n",
    "\t\t\tdata.frame( position = 0.51,\n",
    "\t\t\t\t\t\twidth    = 0.49,\n",
    "\t\t\t\t\t\theader   = \"Motif Logo\",\n",
    "\t\t\t\t\t\thjust    = \"centre\",\n",
    "\t\t\t\t\t\tstyle    = \"logo\",\n",
    "\t\t\t\t\t\textra1   = \"0\",\n",
    "\t\t\t\t\t\textra2   = \"0\",\n",
    "\t\t\t\t\t\textra3   = \"0\",\n",
    "\t\t\t\t\t\textra4   = \"0\"));\n",
    "\n",
    "\toutput_motif_report( tfbs,\n",
    "\t\t\t\t\t\tr.comp.new,\n",
    "\t\t\t\t\t\tfile.pdf,\n",
    "\t\t\t\t\t\treport.size,\n",
    "\t\t\t\t\t\treport.title,\n",
    "\t\t\t\t\t\tdf.style,\n",
    "\t\t\t\t\t\t\"*Enrichment: Enrichment ratio for positive reads against negative reads.\" );\n",
    "}\n",
    "\n",
    "tfbs.draw.enrichment <- function( tfbs, file.pdf, df.ret, enrichment.type, plot.type, df.top, df.bottom, options )\n",
    "{\n",
    "\tscheme1 <- data.frame( min.col = c( 0.9,0.6,0 ), max.col = c( 0, 0, 1 ) );\n",
    "\tscheme2 <- data.frame( min.col = c( 0,0.99,0 ), max.col = c( 0.9, 0.01, 0.9 ) );\n",
    "\n",
    "\tget_rgbcol<-function(pvalue, p.min=0.001, p.max=1, log10=T, scheme=1)\n",
    "\t{\n",
    "\t\tpvalue[ pvalue > p.max] <- p.max;\n",
    "\t\tpvalue[ pvalue < p.min] <- p.min;\n",
    "\n",
    "\t\tif(log10)\n",
    "\t\t{\n",
    "\t\t\tpvalue <- log10(pvalue)\n",
    "\t\t\tp.min  <- log10(p.min)\n",
    "\t\t\tp.max  <- log10(p.max)\n",
    "\t\t}\n",
    "\n",
    "\t\tif(scheme==1) col.scheme <- scheme1 else col.scheme <- scheme2;\n",
    "\n",
    "\t\tc1 <- (col.scheme$min.col[1] - col.scheme$max.col[1]) / (p.min-p.max)*(pvalue-p.max) + col.scheme$max.col[1];\n",
    "\t\tc2 <- (col.scheme$min.col[2] - col.scheme$max.col[2]) / (p.min-p.max)*(pvalue-p.max) + col.scheme$max.col[2];\n",
    "\t\tc3 <- (col.scheme$min.col[3] - col.scheme$max.col[3]) / (p.min-p.max)*(pvalue-p.max) + col.scheme$max.col[3];\n",
    "\n",
    "\t\tc <- rgb(c1,c2,c3)\n",
    "\n",
    "\t\treturn(c);\n",
    "\t}\n",
    "\n",
    "\tdrawlegend<-function( x0, y0, width, height, title, pv.min=0.001, pv.max=1, pv.log10=T, scheme=1)\n",
    "\t{\n",
    "\t\tbar.len <- 50;\n",
    "\t\tif(pv.log10)\n",
    "\t\t\tpv <- 10^seq(log10(pv.min), log10(pv.max), length.out=bar.len)\n",
    "\t\telse\n",
    "\t\t\tpv <- seq(pv.min, pv.max, length.out=bar.len);\n",
    "\n",
    "\t\tbar.width <- width*0.9;\n",
    "\t\tfor( i in 1:length(pv) )\n",
    "\t\t{\n",
    "\t\t\tcol <- get_rgbcol( pv[i], pv.min, pv.max, log10=pv.log10, scheme=scheme);\n",
    "\t\t\trect(  x0 + 0.1*width + i*bar.width/bar.len, y0, x0 + 0.1*width + i*bar.width/bar.len+bar.width/bar.len, y0+height*0.4, col=col,border=NA );\n",
    "\t\t}\n",
    "\n",
    "\t\ttext( x0 + 0.1*width,  y0 + height*0.55, title, cex=2/3*options$zoom.legend.label, adj= c(0, 0.5));\n",
    "\t\ttext( x0 + 0.1*width,  y0 - height*0.15,  sprintf(\"%1g\", pv.min), cex=2/3*options$zoom.legend.label, adj = c(0, 0.5));\n",
    "\t\ttext( x0 + 1*width,  y0 - height*0.15,  sprintf(\"%1g\", pv.max), cex=2/3*options$zoom.legend.label, adj = c(1, 0.5));\n",
    "\t}\n",
    "\n",
    "\tdraw_top_motif<- function( df.motifs )\n",
    "\t{\n",
    "\t\tlabels <- unlist(lapply( 1:NROW(df.motifs) , function(i) {\n",
    "\t\t\tif( is.na( df.motifs$label[i] ) )\n",
    "\t\t\t\tpaste(df.motifs[i,\"motif.id\"], df.motifs[i,\"rbp.name\"], sep=\"/\" )\n",
    "\t\t\telse\n",
    "\t\t\t\tdf.motifs$label[i];\n",
    "\t\t}));\n",
    "\n",
    "\t\told.y <- max( df.motifs[, \"y\"] )+10;\n",
    "\t\tvps <- baseViewports();\n",
    "\t\tpushViewport(vps$inner, vps$figure, vps$plot);\n",
    "\n",
    "\t\tfor(i in 1:NROW(df.motifs))\n",
    "\t\t{\n",
    "\t\t\tif( labels [i] == \"\" ) next;\n",
    "\n",
    "\t\t\tx.pos <- df.motifs[i, \"x\"];\n",
    "\t\t\ty.pos <- distortY( df.motifs[i, \"y\"] );\n",
    "\n",
    "\t\t\tif(y.pos > old.y) y.pos <- old.y\n",
    "\n",
    "\t\t\tr.grid.left <- x.pos - (options$xlim[2]-options$xlim[1])*0.6\n",
    "\t\t\tr.grid.right <-  x.pos - 0.02*(options$xlim[2]-options$xlim[1]);\n",
    "\t\t\tr.grid <- c( r.grid.left, y.pos - motif.logo.height/2, r.grid.right, y.pos + motif.logo.height/2 );\n",
    "\n",
    "\t\t\tpoints( x.pos, distortY( df.motifs[i, \"y\"] ), pch=19, col=\"black\", cex = y.cex/2 );\n",
    "\n",
    "\t\t\tif(r.grid[1] < options$xlim[1] ) r.grid[1] <- options$xlim[1];\n",
    "\t\t\tif(r.grid[3] > options$xlim[2] ) r.grid[3] <- options$xlim[2];\n",
    "\n",
    "\t\t\ttext( (r.grid[1] + r.grid[3])/2,  y.pos - motif.logo.height/2 - motif.label.height*0.95, labels[i], srt=0, adj=c(1/2,0.5),\n",
    "\t\t\t\t\tcex=2/3* options$zoom.motif.label,\n",
    "\t\t\t\t\tcol= \"black\" ); # get_rgbcol( df.motifs$fe.ratio[i], 0.4, 2.5, F, options$color.scheme )\n",
    "\n",
    "\t\t\tidx <-which( as.character(df.motifs[i,\"motif.id\"]) == as.character(tfbs@rbp_info$Motif_ID) );\n",
    "\t\t\tif(length(idx)>1)\n",
    "\t\t\t{\n",
    "\t\t\t\twarning(paste(\"Multiple matrice in the tfbs object for Motif ID:\", as.character(df.motifs[i,\"motif.id\"]),\n",
    "\t\t\t\t\t\", first matrix (index:\", idx[1], \") is used to draw logo.\\n\"));\n",
    "\t\t\t\tidx <- idx[1];\n",
    "\t\t\t}\n",
    "\n",
    "\t\t\tif(length(idx)>0)\n",
    "\t\t\t{\n",
    "\n",
    "\t\t\t\tpushViewport( viewport(\n",
    "\t\t\t\t\t\t\t\tx  = unit(r.grid[1], \"native\"),\n",
    "\t\t\t\t\t\t\t\ty  = unit(r.grid[2], \"native\"),\n",
    "\t\t\t\t\t\t\t\twidth  = unit(r.grid[3] - r.grid[1], \"native\"),\n",
    "\t\t\t\t\t\t\t\theight = unit(r.grid[4] - r.grid[2], \"native\"),\n",
    "\t\t\t\t\t\t\t\tjust = c(\"left\",\"bottom\")));\n",
    "\n",
    "\t\t\t\tseqLogo( exp(t(tfbs@pwm[[idx]])), xaxis = FALSE, yaxis = FALSE);\n",
    "\t\t\t\tpopViewport();\n",
    "\t\t\t}\n",
    "\t\t\telse\n",
    "\t\t\t\twarning(paste(\"No PWM is found in the rbpbs object for Motif ID:\", as.character(df.motifs[i,\"motif.id\"]), \"\\n\"));\n",
    "\n",
    "\t\t\told.y <- y.pos - motif.height;\n",
    "\t\t}\n",
    "\n",
    "\t\tpopViewport();\n",
    "\t}\n",
    "\n",
    "\tdraw_bottom_motif <- function(df.motifs )\n",
    "\t{\n",
    "\t\tlabels <- unlist(lapply( 1:NROW(df.motifs) , function(i) {\n",
    "\t\t\tif( is.na( df.motifs$label[i] ) )\n",
    "\t\t\t\tpaste(df.motifs[i,\"motif.id\"], df.motifs[i,\"rbp.name\"], sep=\"/\" )\n",
    "\t\t\telse\n",
    "\t\t\t\tdf.motifs$label[i];\n",
    "\t\t}));\n",
    "\n",
    "\t\told.y <- min( df.motifs[, \"y\"] )-10;\n",
    "\t\tvps <- baseViewports();\n",
    "\t\tpushViewport(vps$inner, vps$figure, vps$plot);\n",
    "\n",
    "\t\tfor(i in 1:NROW(df.motifs))\n",
    "\t\t{\n",
    "\t\t\tif( labels [i] == \"\" ) next;\n",
    "\n",
    "\t\t\tx.pos <- df.motifs[i, \"x\"];\n",
    "\t\t\ty.pos <- distortY( df.motifs[i, \"y\"] );\n",
    "\t\t\tif(y.pos < old.y) y.pos <- old.y\n",
    "\n",
    "\t\t\tr.grid.left <- x.pos + 2/(options$xlim[2]-options$xlim[1]);\n",
    "\t\t\tr.grid.right <-  x.pos + (options$xlim[2]-options$xlim[1])*0.6\n",
    "\t\t\tr.grid <- c( r.grid.left , y.pos - motif.logo.height/2, r.grid.right, y.pos + motif.logo.height/2 );\n",
    "\n",
    "\t\t\tpoints( x.pos, distortY( df.motifs[i, \"y\"] ), pch=19, col=\"black\", cex = y.cex/2 );\n",
    "\n",
    "\t\t\tif(r.grid[1] < options$xlim[1] ) r.grid[1] <- options$xlim[1];\n",
    "\t\t\tif(r.grid[3] > options$xlim[2] ) r.grid[3] <- options$xlim[2];\n",
    "\n",
    "\t\t\ttext( (r.grid[1] + r.grid[3])/2, y.pos - motif.logo.height/2*1.1, labels[i], srt=0, adj=c( 0.5, 1),\n",
    "\t\t\t\t\tcex = 2/3 * options$zoom.motif.label,\n",
    "\t\t\t\t\tcol= \"black\" ); # get_rgbcol( df.motifs$fe.ratio[i], 0.4, 2.5, F, options$color.scheme )\n",
    "\n",
    "\t\t\tidx <-which( as.character(df.motifs[i,\"motif.id\"]) == as.character(tfbs@rbp_info$Motif_ID) );\n",
    "\t\t\tif(length(idx)>1)\n",
    "\t\t\t{\n",
    "\t\t\t\twarning(paste(\"Multiple matrice in the tfbs object for Motif ID:\", as.character(df.motifs[i,\"motif.id\"]),\n",
    "\t\t\t\t\t\", first matrix (index:\", idx[1], \") is used to draw logo.\\n\"));\n",
    "\t\t\t\tidx <- idx[1];\n",
    "\t\t\t}\n",
    "\n",
    "\t\t\tif(length(idx)>0)\n",
    "\t\t\t{\n",
    "\t\t\t\tpushViewport( viewport(\n",
    "\t\t\t\t\t\t\t\tx  = unit(r.grid[1], \"native\"),\n",
    "\t\t\t\t\t\t\t\ty  = unit(r.grid[2], \"native\"),\n",
    "\t\t\t\t\t\t\t\twidth  = unit(r.grid[3] - r.grid[1], \"native\"),\n",
    "\t\t\t\t\t\t\t\theight = unit(r.grid[4] - r.grid[2], \"native\"),\n",
    "\t\t\t\t\t\t\t\tjust = c(\"left\",\"bottom\")));\n",
    "\n",
    "\t\t\t\tseqLogo( exp(t(tfbs@pwm[[idx]])), xaxis = FALSE, yaxis = FALSE);\n",
    "\t\t\t\tpopViewport();\n",
    "\t\t\t}\n",
    "\t\t\telse\n",
    "\t\t\t\twarning(paste(\"No PWM is found in the tfbs object for Motif ID:\", as.character(df.motifs[i,\"motif.id\"]), \"\\n\"));\n",
    "\n",
    "\t\t\told.y <- y.pos + motif.height;\n",
    "\t\t}\n",
    "\n",
    "\t\tpopViewport();\n",
    "\t}\n",
    "\n",
    "\tget_distortion<-function( df.ret, options )\n",
    "\t{\n",
    "\t\tn.label <- NROW(df.ret) - length(which(df.ret$label==\"\"));\n",
    "\n",
    "\t\tlogo.area.min.height <- motif.height * n.label;\n",
    "\t\tlogo.area.org.height <- max(df.ret$y)-min(df.ret$y) + 0.5*motif.height;\n",
    "\n",
    "\t\tif( logo.area.min.height < logo.area.org.height )\n",
    "\t\t\treturn(1);\n",
    "\n",
    "\t\t##distored curve: y.orinial^a = y.distorted.\n",
    "\t\t##                a = log(y.distored)/log(y.orignial)\n",
    "\t\ty.original  <- logo.area.min.height/(options$ylim[2]);\n",
    "\t\ty.distorted <- logo.area.org.height/(options$ylim[2]);\n",
    "\t\treturn( log(y.distorted)/log(y.original) );\n",
    "\t}\n",
    "\n",
    "\tdistortY<-function( y )\n",
    "\t{\n",
    "\t\ty.sgn <- sapply(1:NROW(y), function(i){ifelse(y[i]>0, 1, -1)} );\n",
    "\t\ty.original  <- ( abs(y) - 0 )/(options$ylim[2]-0);\n",
    "\t\treturn( (y.original ^ distort.a)*options$ylim[2]*y.sgn )\n",
    "\t}\n",
    "\n",
    "\n",
    "\trequire(gridBase);\n",
    "\tplot(NA, NA, type=\"n\",\n",
    "\t\t\txlab = options$xlab,\n",
    "\t\t\tylab = options$ylab,\n",
    "\t\t\tylim = options$ylim,\n",
    "\t\t\txlim = options$xlim,\n",
    "\t\t\tyaxt = \"n\",\n",
    "\t\t\tcex  = 1.0,\n",
    "\t\t\tcex.axis = options$zoom.tick,\n",
    "\t\t\tcex.lab = options$zoom.label,\n",
    "\t\t\tmain = options$title,\n",
    "\t\t\tpch  = 19 );\n",
    "\n",
    "\n",
    "\tmotif.logo.height  <- strheight(\"A\", cex=1.5 * options$zoom.motif.label) * options$zoom.motif.logo;\n",
    "\tmotif.label.height <- strheight(\"A\", cex=2/3 * options$zoom.motif.label)\n",
    "\tmotif.height <- motif.logo.height + motif.label.height + strheight(\"A\", cex=1/3* options$zoom.motif.label);\n",
    "\n",
    "\tdistort.a <- get_distortion( df.top, options);\n",
    "\tif( NROW( df.bottom) > 0 )\n",
    "\t{\n",
    "\t\tdistort.a2 <- get_distortion( df.bottom, options);\n",
    "\t\tif(is.na(distort.a))\n",
    "\t\t\tdistort.a <-distort.a2\n",
    "\t\telse\n",
    "\t\t{\n",
    "\t\t\tif(distort.a2<distort.a) distort.a <- distort.a2;\n",
    "\t\t}\n",
    "\t}\n",
    "\n",
    "\tif(!is.null(options$abline))\n",
    "\t\tabline( h = distortY( options$abline ), lty=22, lwd=0.5);\n",
    "\n",
    "\tif(plot.type==\"polar\")\n",
    "\t{\n",
    "\t\tif(!is.null(options$abline)) abline( h = distortY( -1*options$abline ), lty=22, lwd=0.5);\n",
    "\t\tabline( h = distortY( 0 ), lty=1, lwd=0.5);\n",
    "\t}\n",
    "\n",
    "\taxis(2, at=distortY( axTicks(2) ), labels = axTicks(2), cex.axis = options$zoom.tick, cex.lab = options$zoom.label );\n",
    "\n",
    "\told.xpd <- par( xpd=NA );\n",
    "\n",
    "\ty     <- df.ret$y.log;\n",
    "\ty.cex <- (log(abs(y))+1)/3;\n",
    "\ty.cex [ y.cex < 0.6 ] <- 0.6;\n",
    "\ty.cex [ y.cex > 3 ]   <- 3;\n",
    "\ty.col <- get_rgbcol( df.ret$fe.ratio, 0.4, 2.5, F, options$color.scheme );\n",
    "\n",
    "\tpoints(1:NROW(y), distortY( y ), pch=19, col = y.col, cex=y.cex);\n",
    "\n",
    "\tif( NROW(df.top)>0) draw_top_motif( df.top);\n",
    "\tif( NROW(df.bottom)>0) draw_bottom_motif( df.bottom);\n",
    "\n",
    "    par( xpd = old.xpd );\n",
    "\ty.size <- max( options$ylim )-min( options$ylim );\n",
    "\tdrawlegend( 0, y.size*0.9+min( options$ylim ), width  = NROW(y)*0.2, height = y.size*0.15,  title=\"Enrichment Ratio\", 0.4, 2.5, F, options$color.scheme );\n",
    "\n",
    "\tpopViewport();\n",
    "\n",
    "}\n",
    "\n",
    "tfbs.plotEnrichment <- function( tfbs, r.comp, file.pdf, enrichment.type = c (\"both\", \"enriched\", \"depleted\"), plot.type=c(\"nonpolar\", \"polar\"), options=list() )\n",
    "{\n",
    "\tget_default_options<-function()\n",
    "\t{\n",
    "\t\toptions=list(\n",
    "\t\t\t\tabline = NULL,\n",
    "\t\t\t\ttitle  = \"\",\n",
    "\t\t\t\twidth  = 7,\n",
    "\t\t\t\theight = 7,\n",
    "\t\t\t\txlab   = \"Rank Order\",\n",
    "\t\t\t\tylab   = \"-log10(p-value)\",\n",
    "\t\t\t\ty.max  = NULL,\n",
    "\t\t\t\ttop.motif.labels    = 5,\n",
    "\t\t\t\tbottom.motif.labels = 5,\n",
    "\t\t\t\tcolor.scheme = 2,\n",
    "\t\t\t\tzoom.tick    = 1,\n",
    "\t\t\t\tzoom.label   = 1,\n",
    "\t\t\t\tzoom.motif.logo  = 1,\n",
    "\t\t\t\tzoom.legend.label= 1,\n",
    "\t\t\t\tzoom.motif.label = 1 );\n",
    "\n",
    "\t\treturn(options);\n",
    "\t}\n",
    "\n",
    "\toptions0 <- get_default_options();\n",
    "\toptions0[names(options)] <- options;\n",
    "\toptions <- options0;\n",
    "\n",
    "\tif( options$width > 7 || options$height > 7\t)\n",
    "\t\twarning(\"PDF size exceeds 7 inches.\")\n",
    "\tif(missing(plot.type)) plot.type <- \"nonpolar\";\n",
    "\tif(missing(enrichment.type)) enrichment.type <- \"both\";\n",
    "\tif(!(enrichment.type %in% c(\"both\", \"enriched\", \"depleted\")))\n",
    "\t\tstop(\"The option 'enrichment.type' has three values: both, enriched, depleted.\");\n",
    "\tif(!(plot.type %in% c(\"nonpolar\", \"polar\")))\n",
    "\t\tstop(\"The option 'plot.type' has two values: nonpolar, polar.\");\n",
    "\tif( (enrichment.type == \"enriched\" && plot.type == \"polar\" ) | (enrichment.type == \"depleted\" && plot.type == \"polar\" ) )\n",
    "\t\twarning(\"'polar' figure only can be applied to the motifs enriched and depleted 'both'.\" );\n",
    "\n",
    "\tdf.ret <- r.comp$result[,c('motif.id','rbp.name', 'fe.ratio', 'pvalue' ) ];\n",
    "\tdf.ret$y.log <- -log10( df.ret$pvalue );\n",
    "\tif( is.null( options$y.max ))\n",
    "\t\toptions$y.max <- max( df.ret$y.log [ !is.infinite(df.ret$y.log) ] );\n",
    "\tdf.ret$y.log [ is.infinite(df.ret$y.log) | (df.ret$y.log >= options$y.max) ] <- options$y.max;\n",
    "\n",
    "\tif(enrichment.type == \"enriched\")\n",
    "\t\tdf.ret  <- df.ret [ df.ret$fe.ratio >= 1, ,drop=F ]\n",
    "\telse if (enrichment.type == \"depleted\")\n",
    "\t\tdf.ret  <- df.ret [ df.ret$fe.ratio <1, ,drop=F];\n",
    "\n",
    "\toptions$ylim <- c(0, 1) * options$y.max ;\n",
    "\tif(plot.type==\"polar\")\n",
    "\t{\n",
    "\t\tidx.enrich  <- which( df.ret$fe.ratio >= 1 );\n",
    "\t\tidx.deple   <- which( df.ret$fe.ratio < 1 );\n",
    "\t\tdf.ret$y.log[idx.deple] <- - df.ret$y.log[idx.deple];\n",
    "\t\toptions$ylim <- c(-1,1) * options$y.max;\n",
    "\t}\n",
    "\n",
    "\tdf.ret <- df.ret[order( df.ret$y.log, decreasing = F ),];\n",
    "\n",
    "\toptions$xlim = c(0, NROW(df.ret) );\n",
    "\toptions$abline.h <- -log10(0.05/NROW(df.ret));\n",
    "\n",
    "\tdf.top <- c();\n",
    "\t## exclude this condition: polar && only 'depleted'\n",
    "\tif( !(plot.type== \"polar\"  && enrichment.type == \"depleted\") )\n",
    "\t{\n",
    "\t\tdf.dummy.bottom <- c();\n",
    "\t\tif( is.numeric(options$top.motif.labels))\n",
    "\t\t{\n",
    "\t\t\tif( plot.type == \"polar\" )\n",
    "\t\t\t{\n",
    "\t\t\t\tdf.top <- df.ret[ df.ret[, \"fe.ratio\"] >= 1,,drop=F]\n",
    "\t\t\t\tdf.dummy.bottom <- df.ret[ df.ret[, \"fe.ratio\"] < 1,,drop=F]\n",
    "\t\t\t}\n",
    "\t\t\telse\n",
    "\t\t\t\t\tdf.top <- df.ret[ df.ret[, \"fe.ratio\"] >= 1,,drop=F];\n",
    "\n",
    "\t\t\tif( options$top.motif.labels> NROW(df.top) )\n",
    "\t\t\t\toptions$top.motif.labels <- NROW(df.top);\n",
    "\n",
    "\t\t\tif( options$top.motif.labels>0 )\n",
    "\t\t\t{\n",
    "\t\t\t\tn.revse <- NROW(df.top) - c(1:as.numeric(options$top.motif.labels)) + 1;\n",
    "\t\t\t\tdf.top <- df.top[n.revse,,drop=F]\n",
    "\t\t\t\t\n",
    "\n",
    "\t\t\t\tif(NROW(df.top)>0)\n",
    "\t\t\t\t\t#df.top <- cbind( x=(NROW(df.top)-which(df.ret[, \"fe.ratio\"] >= 1.5)+NROW(df.dummy.bottom)), y=df.top$y.log, label=NA, df.top);\n",
    "\t\t\t\t  df.top <- cbind( x=n.revse+NROW(df.dummy.bottom), y=df.top$y.log, label=NA, df.top);\n",
    "\t\t\t\t \n",
    "\t\t\t}\n",
    "\t\t}\n",
    "\n",
    "\t\tif( is.character(options$top.motif.labels))\n",
    "\t\t{\n",
    "\t\t\tn.motif.labels <- NROW(df.ret);\n",
    "\t\t\tif( NROW(options$top.motif.labels) < NROW(df.ret) )\n",
    "\t\t\t\tn.motif.labels <- NROW(options$top.motif.labels);\n",
    "\n",
    "\t\t\tn.revse <- NROW(df.ret) - c( 1:n.motif.labels ) + 1;\n",
    "\t\t\tdf.top <- df.ret[n.revse,,drop=F]\n",
    "\n",
    "\t\t\tif(NROW(df.top)>0)\n",
    "\t\t\t\tdf.top <- cbind( x = n.revse + NROW(df.dummy.bottom), y=df.top$y.log, label=options$top.motif.labels, df.top);\n",
    "\t\t}\n",
    "\n",
    "\t}\n",
    "#df.top=c()\n",
    "\tdf.bottom <- c();\n",
    "\t## only include this condition: polar && not 'enriched'\n",
    "\tif( plot.type== \"polar\" && enrichment.type != \"enriched\")\n",
    "\t{\n",
    "\t\tif( is.numeric(options$bottom.motif.labels) )\n",
    "\t\t{\n",
    "\t\t\tdf.bottom <- df.ret[ df.ret[, \"fe.ratio\"] < 1.0,,drop=F]\n",
    "\t\t\tif ( options$bottom.motif.labels> NROW(df.bottom) )\n",
    "\t\t\t\toptions$bottom.motif.labels <- NROW(df.bottom);\n",
    "\n",
    "\t\t\tif(options$bottom.motif.labels>0)\n",
    "\t\t\t{\n",
    "\t\t\t\tn.order <-  c(1:as.numeric( options$bottom.motif.labels )) ;\n",
    "\t\t\t\tdf.bottom <- df.bottom[n.order,,drop=F]\n",
    "\n",
    "\t\t\t\tif(NROW(df.bottom)>0)\n",
    "\t\t\t\t\tdf.bottom <- cbind( x=n.order, y=df.bottom$y.log, label=NA, df.bottom);\n",
    "\t\t\t}\n",
    "\t\t}\n",
    "\n",
    "\t\tif( is.character(options$bottom.motif.labels) )\n",
    "\t\t{\n",
    "\t\t\tdf.bottom <- df.ret[ df.ret[, \"fe.ratio\"] < 1.0,,drop=F]\n",
    "\n",
    "\t\t\tn.motif.labels <- NROW(df.bottom);\n",
    "\t\t\tif ( NROW(options$bottom.motif.labels) > NROW(df.bottom) )\n",
    "\t\t\t\tn.motif.labels <- NROW(df.bottom);\n",
    "\n",
    "\t\t\tif(n.motif.labels>0)\n",
    "\t\t\t{\n",
    "\t\t\t\tdf.bottom <- df.bottom[c(1:n.motif.labels ),,drop=F]\n",
    "\t\t\t\tif(NROW(df.bottom)>0)\n",
    "\t\t\t\t\tdf.bottom <- cbind( x=n.order, y=df.bottom$y.log, label=options$bottom.motif.labels, df.bottom);\n",
    "\t\t\t}\n",
    "\t\t}\n",
    "\t}\n",
    "\n",
    "\tpdf( file.pdf, width = options$width, height = options$height );\n",
    "\n",
    "\ttry( tfbs.draw.enrichment ( tfbs, file.pdf, df.ret, enrichment.type, plot.type, df.top, df.bottom, options ) );\n",
    "\n",
    "\tdev.off();\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a7b41b4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "also installing the dependencies 'beeswarm', 'vipor'\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"ggbeeswarm\")\n",
    "install.packages(\"gridExtra\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1ddfda78",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in library(gridExtra): there is no package called 'gridExtra'\n",
     "output_type": "error",
     "traceback": [
      "Error in library(gridExtra): there is no package called 'gridExtra'\nTraceback:\n",
      "1. library(gridExtra)"
     ]
    }
   ],
   "source": [
    "library(RColorBrewer)\n",
    "#install.packages(\"raster\")\n",
    "library(raster)\n",
    "library(tidyverse)\n",
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(ggplot2)\n",
    "#library(ggbeeswarm)\n",
    "yk.col = function(n = 5, sat = 0.7, lum = 1) {\n",
    "\tcol = c(\"royalblue4\", \"deepskyblue3\", \"turquoise3\",\n",
    "\t   \"grey62\", \"goldenrod2\", \"orange2\", \"orangered2\") %>%\n",
    "\tcol2rgb %>%\n",
    "\trgb2hsv\n",
    "\treturn(colorRampPalette(hsv(col[1,], col[2,] * sat,\n",
    "\t\t\t\t    col[3,] * lum))(n)) }\n",
    "\n",
    "\n",
    "#CDF\n",
    "\n",
    "library(gridExtra)\n",
    "par(mar = c(2,2,1,1), mgp = c(1.5,0.5,0))\n",
    "col = yk.col(10)[c(3,5,7)] \n",
    "#col = yk.col(5)[c(1,2,4,5)] \n",
    "cdf_common = ggplot() +\n",
    "  scale_color_manual(values = col) +\n",
    "  theme_bw() +\n",
    "  xlim(0.3, 0.9) +\n",
    "  ylab(\"Cumulative fraction\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59de5496",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "9d8d5678",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Compare motifs between two conditions\n",
    "utr<-read.table(\"custom3UTR.n.31490.tab.fa\",stringsAsFactors = F)\n",
    "utr=utr%>%dplyr::mutate(len=nchar(V2))\n",
    "utr.id=utr%>%\n",
    "  separate(V1,c(\"id\",\"ref\",\"chr\",\"cps\"),sep=\":\")%>%\n",
    "  unite(\"newid2\",c(id,ref,cps),sep=\":\",remove=F)%>%\n",
    "  unite(\"newid\",c(id,cps),sep=\":\",remove=F)%>%\n",
    "mutate(newid=substr(newid,1,nchar(newid)-3))\n",
    "#last 500 regions\n",
    "utr.id=utr.id%>%filter(nchar(V2)>300)%>%\n",
    "dplyr::mutate(V2=ifelse(len>300,substr(V2,len-300,len),V2))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "8fcb740c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>newid2</th><th scope=col>newid</th><th scope=col>id</th><th scope=col>ref</th><th scope=col>chr</th><th scope=col>cps</th><th scope=col>V2</th><th scope=col>len</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENST00000327044.6:944203:944211(-)   </td><td>ENST00000327044.6:944211  </td><td>ENST00000327044.6 </td><td>944203 </td><td>CHR1</td><td>944211(-) </td><td>GGCTGGTGTGGCTGCTGAAGCCCCACAGCTGTGGGCTGCTGAAGTCAGCTCCGCGGGGGAGCTGACCCTGACGTCAGCAGACCGAGACCAGTCCCAGTTCCAGGGGGAGGCCTGCAGGCCCCTGGCCCCTTCCACCACCTCTGCCCTCCGTCTGCAGACCTCGTCCATCTGCACCAGGCTCTGCCTTCACTCCCCCAAGTCTTTGAAAATTTGTTCCTTTCCTTTGAAGTCACATTTTCTTTTAAAATTTTTTGTTTTGCATCCGAAACCGAAAGAAATAAAGCGGTGGGAGGCAGGGCCA</td><td>301</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENST00000327044.6:944203:944245(-)   </td><td>ENST00000327044.6:944245  </td><td>ENST00000327044.6 </td><td>944203 </td><td>CHR1</td><td>944245(-) </td><td>AAAGACCAAACGTATCGAGAGCTGGGCTGGGCTGGGCTGGTGTGGCTGCTGAAGCCCCACAGCTGTGGGCTGCTGAAGTCAGCTCCGCGGGGGAGCTGACCCTGACGTCAGCAGACCGAGACCAGTCCCAGTTCCAGGGGGAGGCCTGCAGGCCCCTGGCCCCTTCCACCACCTCTGCCCTCCGTCTGCAGACCTCGTCCATCTGCACCAGGCTCTGCCTTCACTCCCCCAAGTCTTTGAAAATTTGTTCCTTTCCTTTGAAGTCACATTTTCTTTTAAAATTTTTTGTTTTGCATCCGAA</td><td>301</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENST00000379410.7:975108:975164(+)   </td><td>ENST00000379410.7:975164  </td><td>ENST00000379410.7 </td><td>975108 </td><td>CHR1</td><td>975164(+) </td><td>AGGAGTCAGGGAGGAGACTCGCTGGGAGTGGGAGGGCAGCACGGGCGTGAAGGTCGGAGGACAGAGAAAGGTCAGCAGGGTCAGAGTATGTGAGGTCAGAGGGCATGAGGGTCACAGGTCAGCAAGGTGTGAGGAGCACAAGCCAGGGTGCCCCGAGGAGGAGGGTGGGTGGGTCCTTGTGTGGCCTGGCGCGCACCACAGGGCAGCACGGGAGACGTTGACACCACCGGACGAGAAAGAAAAAAGAGAGAAGAGAGAGAGGAGAAGAGAGAAAAGAAGAGAAGAGAGAAGAGAAGAGAAG</td><td>301</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENST00000379370.6:1056116:1056103(+) </td><td>ENST00000379370.6:1056103 </td><td>ENST00000379370.6 </td><td>1056116</td><td>CHR1</td><td>1056103(+)</td><td>CCTCACTGCAAATCCTCCCCACAGCGTCAGTGAAAGTCGTCCTTGTCTCAGAATGACCAGGGGCCAGCCAGTGTCTGACCAAGGTCAAGGGGCAGGTGCAGAGGTGGCAGGGATGGCTCCGAAGCCAGAAATGCCTTAAACTGCAACGTCCCGTCCCTTCCCCACCCCCATCCCATCCCCACCCCCAGCCCCAGCCCAGTCCTCCTAGGAGCAGGACCCGATGAAGCGGGCGGCGGTGGGGCTGGGTGCCGTGTTACTAACTCTAGTATGTTTCTGTGTCAATCGCTGTGAAATAAAGTCT</td><td>301</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENST00000379339.5:1081817:1081819(-) </td><td>ENST00000379339.5:1081819 </td><td>ENST00000379339.5 </td><td>1081817</td><td>CHR1</td><td>1081819(-)</td><td>CCTGCCAAGGCTGCCCTGGCGCTGAGTCCTGGGGTCCCTCCCGGAGTTCCTGGGAGAAAGGCGCCGTCGTGGCCGCCTCCCGCACGCCAGGCCCGGGCTCCACCGTGGGTCTCAGACGCCCTGCGGCACCGGCACCGTCTGCTTTAGCATGGGACCCCCCTCTGAGGGGTGGCCTGGCCTTCGGGGTCCCCACGCTCCTTTGCGAAGTCCACTGTGGGTGCCATCATGGTCTCCGGGACCTGGGCCAGCGGGAACGTGGGGGCACTGGGTGTGCTGATATAAAGTCGGCATTACTCAAGCT</td><td>301</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENST00000360001.10:1216907:1216883(-)</td><td>ENST00000360001.10:1216883</td><td>ENST00000360001.10</td><td>1216907</td><td>CHR1</td><td>1216883(-)</td><td>CCGGAACCTCTGGGCCCGGCCCCTCCCAGGGCCGCCGCTCCGTGGGAAAAAACAGCTCCTCCATTTCCTTGAAAACTGAACGATTATTAAAAATAGATTAAACTTCGCTGGAAATGAGTAGCCAGGAAGTTCAGGGGAGGGTGCCGGGTCCTTCCCGGGCCTGGCGTGTCGGAGCCACCCAGGTCCCGCAGCTGCCGCTGAGAAAATGCAAATATTTGTTGTGACAAGAATCACATACATTTACTTTAAATATAGTTGCCTTTTTTGGTCAGCTTCATTCTTTGGGCGTGAAACAAAACAC</td><td>301</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & newid2 & newid & id & ref & chr & cps & V2 & len\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & ENST00000327044.6:944203:944211(-)    & ENST00000327044.6:944211   & ENST00000327044.6  & 944203  & CHR1 & 944211(-)  & GGCTGGTGTGGCTGCTGAAGCCCCACAGCTGTGGGCTGCTGAAGTCAGCTCCGCGGGGGAGCTGACCCTGACGTCAGCAGACCGAGACCAGTCCCAGTTCCAGGGGGAGGCCTGCAGGCCCCTGGCCCCTTCCACCACCTCTGCCCTCCGTCTGCAGACCTCGTCCATCTGCACCAGGCTCTGCCTTCACTCCCCCAAGTCTTTGAAAATTTGTTCCTTTCCTTTGAAGTCACATTTTCTTTTAAAATTTTTTGTTTTGCATCCGAAACCGAAAGAAATAAAGCGGTGGGAGGCAGGGCCA & 301\\\\\n",
       "\t2 & ENST00000327044.6:944203:944245(-)    & ENST00000327044.6:944245   & ENST00000327044.6  & 944203  & CHR1 & 944245(-)  & AAAGACCAAACGTATCGAGAGCTGGGCTGGGCTGGGCTGGTGTGGCTGCTGAAGCCCCACAGCTGTGGGCTGCTGAAGTCAGCTCCGCGGGGGAGCTGACCCTGACGTCAGCAGACCGAGACCAGTCCCAGTTCCAGGGGGAGGCCTGCAGGCCCCTGGCCCCTTCCACCACCTCTGCCCTCCGTCTGCAGACCTCGTCCATCTGCACCAGGCTCTGCCTTCACTCCCCCAAGTCTTTGAAAATTTGTTCCTTTCCTTTGAAGTCACATTTTCTTTTAAAATTTTTTGTTTTGCATCCGAA & 301\\\\\n",
       "\t3 & ENST00000379410.7:975108:975164(+)    & ENST00000379410.7:975164   & ENST00000379410.7  & 975108  & CHR1 & 975164(+)  & AGGAGTCAGGGAGGAGACTCGCTGGGAGTGGGAGGGCAGCACGGGCGTGAAGGTCGGAGGACAGAGAAAGGTCAGCAGGGTCAGAGTATGTGAGGTCAGAGGGCATGAGGGTCACAGGTCAGCAAGGTGTGAGGAGCACAAGCCAGGGTGCCCCGAGGAGGAGGGTGGGTGGGTCCTTGTGTGGCCTGGCGCGCACCACAGGGCAGCACGGGAGACGTTGACACCACCGGACGAGAAAGAAAAAAGAGAGAAGAGAGAGAGGAGAAGAGAGAAAAGAAGAGAAGAGAGAAGAGAAGAGAAG & 301\\\\\n",
       "\t4 & ENST00000379370.6:1056116:1056103(+)  & ENST00000379370.6:1056103  & ENST00000379370.6  & 1056116 & CHR1 & 1056103(+) & CCTCACTGCAAATCCTCCCCACAGCGTCAGTGAAAGTCGTCCTTGTCTCAGAATGACCAGGGGCCAGCCAGTGTCTGACCAAGGTCAAGGGGCAGGTGCAGAGGTGGCAGGGATGGCTCCGAAGCCAGAAATGCCTTAAACTGCAACGTCCCGTCCCTTCCCCACCCCCATCCCATCCCCACCCCCAGCCCCAGCCCAGTCCTCCTAGGAGCAGGACCCGATGAAGCGGGCGGCGGTGGGGCTGGGTGCCGTGTTACTAACTCTAGTATGTTTCTGTGTCAATCGCTGTGAAATAAAGTCT & 301\\\\\n",
       "\t5 & ENST00000379339.5:1081817:1081819(-)  & ENST00000379339.5:1081819  & ENST00000379339.5  & 1081817 & CHR1 & 1081819(-) & CCTGCCAAGGCTGCCCTGGCGCTGAGTCCTGGGGTCCCTCCCGGAGTTCCTGGGAGAAAGGCGCCGTCGTGGCCGCCTCCCGCACGCCAGGCCCGGGCTCCACCGTGGGTCTCAGACGCCCTGCGGCACCGGCACCGTCTGCTTTAGCATGGGACCCCCCTCTGAGGGGTGGCCTGGCCTTCGGGGTCCCCACGCTCCTTTGCGAAGTCCACTGTGGGTGCCATCATGGTCTCCGGGACCTGGGCCAGCGGGAACGTGGGGGCACTGGGTGTGCTGATATAAAGTCGGCATTACTCAAGCT & 301\\\\\n",
       "\t6 & ENST00000360001.10:1216907:1216883(-) & ENST00000360001.10:1216883 & ENST00000360001.10 & 1216907 & CHR1 & 1216883(-) & CCGGAACCTCTGGGCCCGGCCCCTCCCAGGGCCGCCGCTCCGTGGGAAAAAACAGCTCCTCCATTTCCTTGAAAACTGAACGATTATTAAAAATAGATTAAACTTCGCTGGAAATGAGTAGCCAGGAAGTTCAGGGGAGGGTGCCGGGTCCTTCCCGGGCCTGGCGTGTCGGAGCCACCCAGGTCCCGCAGCTGCCGCTGAGAAAATGCAAATATTTGTTGTGACAAGAATCACATACATTTACTTTAAATATAGTTGCCTTTTTTGGTCAGCTTCATTCTTTGGGCGTGAAACAAAACAC & 301\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 8\n",
       "\n",
       "| <!--/--> | newid2 &lt;chr&gt; | newid &lt;chr&gt; | id &lt;chr&gt; | ref &lt;chr&gt; | chr &lt;chr&gt; | cps &lt;chr&gt; | V2 &lt;chr&gt; | len &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | ENST00000327044.6:944203:944211(-)    | ENST00000327044.6:944211   | ENST00000327044.6  | 944203  | CHR1 | 944211(-)  | GGCTGGTGTGGCTGCTGAAGCCCCACAGCTGTGGGCTGCTGAAGTCAGCTCCGCGGGGGAGCTGACCCTGACGTCAGCAGACCGAGACCAGTCCCAGTTCCAGGGGGAGGCCTGCAGGCCCCTGGCCCCTTCCACCACCTCTGCCCTCCGTCTGCAGACCTCGTCCATCTGCACCAGGCTCTGCCTTCACTCCCCCAAGTCTTTGAAAATTTGTTCCTTTCCTTTGAAGTCACATTTTCTTTTAAAATTTTTTGTTTTGCATCCGAAACCGAAAGAAATAAAGCGGTGGGAGGCAGGGCCA | 301 |\n",
       "| 2 | ENST00000327044.6:944203:944245(-)    | ENST00000327044.6:944245   | ENST00000327044.6  | 944203  | CHR1 | 944245(-)  | AAAGACCAAACGTATCGAGAGCTGGGCTGGGCTGGGCTGGTGTGGCTGCTGAAGCCCCACAGCTGTGGGCTGCTGAAGTCAGCTCCGCGGGGGAGCTGACCCTGACGTCAGCAGACCGAGACCAGTCCCAGTTCCAGGGGGAGGCCTGCAGGCCCCTGGCCCCTTCCACCACCTCTGCCCTCCGTCTGCAGACCTCGTCCATCTGCACCAGGCTCTGCCTTCACTCCCCCAAGTCTTTGAAAATTTGTTCCTTTCCTTTGAAGTCACATTTTCTTTTAAAATTTTTTGTTTTGCATCCGAA | 301 |\n",
       "| 3 | ENST00000379410.7:975108:975164(+)    | ENST00000379410.7:975164   | ENST00000379410.7  | 975108  | CHR1 | 975164(+)  | AGGAGTCAGGGAGGAGACTCGCTGGGAGTGGGAGGGCAGCACGGGCGTGAAGGTCGGAGGACAGAGAAAGGTCAGCAGGGTCAGAGTATGTGAGGTCAGAGGGCATGAGGGTCACAGGTCAGCAAGGTGTGAGGAGCACAAGCCAGGGTGCCCCGAGGAGGAGGGTGGGTGGGTCCTTGTGTGGCCTGGCGCGCACCACAGGGCAGCACGGGAGACGTTGACACCACCGGACGAGAAAGAAAAAAGAGAGAAGAGAGAGAGGAGAAGAGAGAAAAGAAGAGAAGAGAGAAGAGAAGAGAAG | 301 |\n",
       "| 4 | ENST00000379370.6:1056116:1056103(+)  | ENST00000379370.6:1056103  | ENST00000379370.6  | 1056116 | CHR1 | 1056103(+) | CCTCACTGCAAATCCTCCCCACAGCGTCAGTGAAAGTCGTCCTTGTCTCAGAATGACCAGGGGCCAGCCAGTGTCTGACCAAGGTCAAGGGGCAGGTGCAGAGGTGGCAGGGATGGCTCCGAAGCCAGAAATGCCTTAAACTGCAACGTCCCGTCCCTTCCCCACCCCCATCCCATCCCCACCCCCAGCCCCAGCCCAGTCCTCCTAGGAGCAGGACCCGATGAAGCGGGCGGCGGTGGGGCTGGGTGCCGTGTTACTAACTCTAGTATGTTTCTGTGTCAATCGCTGTGAAATAAAGTCT | 301 |\n",
       "| 5 | ENST00000379339.5:1081817:1081819(-)  | ENST00000379339.5:1081819  | ENST00000379339.5  | 1081817 | CHR1 | 1081819(-) | CCTGCCAAGGCTGCCCTGGCGCTGAGTCCTGGGGTCCCTCCCGGAGTTCCTGGGAGAAAGGCGCCGTCGTGGCCGCCTCCCGCACGCCAGGCCCGGGCTCCACCGTGGGTCTCAGACGCCCTGCGGCACCGGCACCGTCTGCTTTAGCATGGGACCCCCCTCTGAGGGGTGGCCTGGCCTTCGGGGTCCCCACGCTCCTTTGCGAAGTCCACTGTGGGTGCCATCATGGTCTCCGGGACCTGGGCCAGCGGGAACGTGGGGGCACTGGGTGTGCTGATATAAAGTCGGCATTACTCAAGCT | 301 |\n",
       "| 6 | ENST00000360001.10:1216907:1216883(-) | ENST00000360001.10:1216883 | ENST00000360001.10 | 1216907 | CHR1 | 1216883(-) | CCGGAACCTCTGGGCCCGGCCCCTCCCAGGGCCGCCGCTCCGTGGGAAAAAACAGCTCCTCCATTTCCTTGAAAACTGAACGATTATTAAAAATAGATTAAACTTCGCTGGAAATGAGTAGCCAGGAAGTTCAGGGGAGGGTGCCGGGTCCTTCCCGGGCCTGGCGTGTCGGAGCCACCCAGGTCCCGCAGCTGCCGCTGAGAAAATGCAAATATTTGTTGTGACAAGAATCACATACATTTACTTTAAATATAGTTGCCTTTTTTGGTCAGCTTCATTCTTTGGGCGTGAAACAAAACAC | 301 |\n",
       "\n"
      ],
      "text/plain": [
       "  newid2                                newid                     \n",
       "1 ENST00000327044.6:944203:944211(-)    ENST00000327044.6:944211  \n",
       "2 ENST00000327044.6:944203:944245(-)    ENST00000327044.6:944245  \n",
       "3 ENST00000379410.7:975108:975164(+)    ENST00000379410.7:975164  \n",
       "4 ENST00000379370.6:1056116:1056103(+)  ENST00000379370.6:1056103 \n",
       "5 ENST00000379339.5:1081817:1081819(-)  ENST00000379339.5:1081819 \n",
       "6 ENST00000360001.10:1216907:1216883(-) ENST00000360001.10:1216883\n",
       "  id                 ref     chr  cps       \n",
       "1 ENST00000327044.6  944203  CHR1 944211(-) \n",
       "2 ENST00000327044.6  944203  CHR1 944245(-) \n",
       "3 ENST00000379410.7  975108  CHR1 975164(+) \n",
       "4 ENST00000379370.6  1056116 CHR1 1056103(+)\n",
       "5 ENST00000379339.5  1081817 CHR1 1081819(-)\n",
       "6 ENST00000360001.10 1216907 CHR1 1216883(-)\n",
       "  V2                                                                                                                                                                                                                                                                                                           \n",
       "1 GGCTGGTGTGGCTGCTGAAGCCCCACAGCTGTGGGCTGCTGAAGTCAGCTCCGCGGGGGAGCTGACCCTGACGTCAGCAGACCGAGACCAGTCCCAGTTCCAGGGGGAGGCCTGCAGGCCCCTGGCCCCTTCCACCACCTCTGCCCTCCGTCTGCAGACCTCGTCCATCTGCACCAGGCTCTGCCTTCACTCCCCCAAGTCTTTGAAAATTTGTTCCTTTCCTTTGAAGTCACATTTTCTTTTAAAATTTTTTGTTTTGCATCCGAAACCGAAAGAAATAAAGCGGTGGGAGGCAGGGCCA\n",
       "2 AAAGACCAAACGTATCGAGAGCTGGGCTGGGCTGGGCTGGTGTGGCTGCTGAAGCCCCACAGCTGTGGGCTGCTGAAGTCAGCTCCGCGGGGGAGCTGACCCTGACGTCAGCAGACCGAGACCAGTCCCAGTTCCAGGGGGAGGCCTGCAGGCCCCTGGCCCCTTCCACCACCTCTGCCCTCCGTCTGCAGACCTCGTCCATCTGCACCAGGCTCTGCCTTCACTCCCCCAAGTCTTTGAAAATTTGTTCCTTTCCTTTGAAGTCACATTTTCTTTTAAAATTTTTTGTTTTGCATCCGAA\n",
       "3 AGGAGTCAGGGAGGAGACTCGCTGGGAGTGGGAGGGCAGCACGGGCGTGAAGGTCGGAGGACAGAGAAAGGTCAGCAGGGTCAGAGTATGTGAGGTCAGAGGGCATGAGGGTCACAGGTCAGCAAGGTGTGAGGAGCACAAGCCAGGGTGCCCCGAGGAGGAGGGTGGGTGGGTCCTTGTGTGGCCTGGCGCGCACCACAGGGCAGCACGGGAGACGTTGACACCACCGGACGAGAAAGAAAAAAGAGAGAAGAGAGAGAGGAGAAGAGAGAAAAGAAGAGAAGAGAGAAGAGAAGAGAAG\n",
       "4 CCTCACTGCAAATCCTCCCCACAGCGTCAGTGAAAGTCGTCCTTGTCTCAGAATGACCAGGGGCCAGCCAGTGTCTGACCAAGGTCAAGGGGCAGGTGCAGAGGTGGCAGGGATGGCTCCGAAGCCAGAAATGCCTTAAACTGCAACGTCCCGTCCCTTCCCCACCCCCATCCCATCCCCACCCCCAGCCCCAGCCCAGTCCTCCTAGGAGCAGGACCCGATGAAGCGGGCGGCGGTGGGGCTGGGTGCCGTGTTACTAACTCTAGTATGTTTCTGTGTCAATCGCTGTGAAATAAAGTCT\n",
       "5 CCTGCCAAGGCTGCCCTGGCGCTGAGTCCTGGGGTCCCTCCCGGAGTTCCTGGGAGAAAGGCGCCGTCGTGGCCGCCTCCCGCACGCCAGGCCCGGGCTCCACCGTGGGTCTCAGACGCCCTGCGGCACCGGCACCGTCTGCTTTAGCATGGGACCCCCCTCTGAGGGGTGGCCTGGCCTTCGGGGTCCCCACGCTCCTTTGCGAAGTCCACTGTGGGTGCCATCATGGTCTCCGGGACCTGGGCCAGCGGGAACGTGGGGGCACTGGGTGTGCTGATATAAAGTCGGCATTACTCAAGCT\n",
       "6 CCGGAACCTCTGGGCCCGGCCCCTCCCAGGGCCGCCGCTCCGTGGGAAAAAACAGCTCCTCCATTTCCTTGAAAACTGAACGATTATTAAAAATAGATTAAACTTCGCTGGAAATGAGTAGCCAGGAAGTTCAGGGGAGGGTGCCGGGTCCTTCCCGGGCCTGGCGTGTCGGAGCCACCCAGGTCCCGCAGCTGCCGCTGAGAAAATGCAAATATTTGTTGTGACAAGAATCACATACATTTACTTTAAATATAGTTGCCTTTTTTGGTCAGCTTCATTCTTTGGGCGTGAAACAAAACAC\n",
       "  len\n",
       "1 301\n",
       "2 301\n",
       "3 301\n",
       "4 301\n",
       "5 301\n",
       "6 301"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "22064"
      ],
      "text/latex": [
       "22064"
      ],
      "text/markdown": [
       "22064"
      ],
      "text/plain": [
       "[1] 22064"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(utr.id)%>%mutate(len=nchar(V2))\n",
    "nrow(utr.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "4e883ef6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "6876"
      ],
      "text/latex": [
       "6876"
      ],
      "text/markdown": [
       "6876"
      ],
      "text/plain": [
       "[1] 6876"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "77"
      ],
      "text/latex": [
       "77"
      ],
      "text/markdown": [
       "77"
      ],
      "text/plain": [
       "[1] 77"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "752"
      ],
      "text/latex": [
       "752"
      ],
      "text/markdown": [
       "752"
      ],
      "text/plain": [
       "[1] 752"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "3962"
      ],
      "text/latex": [
       "3962"
      ],
      "text/markdown": [
       "3962"
      ],
      "text/plain": [
       "[1] 3962"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 24</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>newid</th><th scope=col>dpal.0.1.x</th><th scope=col>dpal.0.2.x</th><th scope=col>dpal.1.2.x</th><th scope=col>kspval.0.1.x</th><th scope=col>kspval.0.2.x</th><th scope=col>kspval.1.2.x</th><th scope=col>dpal.0.1.y</th><th scope=col>dpal.0.2.y</th><th scope=col>dpal.1.2.y</th><th scope=col>⋯</th><th scope=col>V4</th><th scope=col>ALE</th><th scope=col>rep1_0</th><th scope=col>rep1_1</th><th scope=col>rep1_2</th><th scope=col>rep2_0</th><th scope=col>rep2_1</th><th scope=col>rep2_2</th><th scope=col>tandem.new</th><th scope=col>cluster</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENST00000328089.10:9369527 </td><td> 5.180414</td><td>-5.737671</td><td>-10.918085</td><td>1.071275e-01</td><td>3.714737e-01</td><td>3.388636e-04</td><td> 4.600682</td><td> -1.3205336</td><td> -5.921216</td><td>⋯</td><td>ENST00000328089.10:9369532:chr1:9369527   </td><td>2</td><td> 44.28767</td><td> 49.46809</td><td> 38.55000</td><td> 47.57341</td><td> 52.17409</td><td> 46.25287</td><td>2</td><td>SPSB1:2:2  </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENST00000529905.5:40073187 </td><td> 6.099345</td><td>11.971800</td><td>  5.872455</td><td>9.861732e-03</td><td>1.611132e-05</td><td>8.354466e-02</td><td>25.915931</td><td> 22.0499838</td><td> -3.865947</td><td>⋯</td><td>ENST00000529905.5:40072937:chr1:40073187  </td><td>4</td><td>115.99531</td><td>122.09465</td><td>127.96711</td><td>119.05128</td><td>144.96721</td><td>141.10127</td><td>4</td><td>PPT1:4:1   </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENST00000372653.5:40771114 </td><td>12.777349</td><td> 1.303309</td><td>-11.474040</td><td>2.449807e-12</td><td>8.186515e-04</td><td>8.434666e-10</td><td> 3.733495</td><td>-22.1128374</td><td>-25.846332</td><td>⋯</td><td>ENST00000372653.5:40771106:chr1:40771114  </td><td>8</td><td> 65.64068</td><td> 78.41803</td><td> 66.94399</td><td>103.42475</td><td>107.15825</td><td> 81.31191</td><td>2</td><td>NFYC:8:1   </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENST00000304979.7:42676754 </td><td>22.169459</td><td> 1.353835</td><td>-20.815624</td><td>9.255115e-14</td><td>3.251077e-08</td><td>3.018415e-14</td><td> 7.154819</td><td>  6.0500800</td><td> -1.104739</td><td>⋯</td><td>ENST00000304979.7:42676758:chr1:42676754  </td><td>3</td><td> 41.92429</td><td> 64.09375</td><td> 43.27813</td><td> 45.31574</td><td> 52.47056</td><td> 51.36582</td><td>1</td><td>PPIH:3:1   </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENST00000256646.6:119912062</td><td> 3.505238</td><td> 2.592593</td><td> -0.912645</td><td>8.525272e-02</td><td>1.186867e-06</td><td>8.660299e-05</td><td> 6.500152</td><td>  3.7493314</td><td> -2.750821</td><td>⋯</td><td>ENST00000256646.6:119911552:chr1:119912062</td><td>4</td><td> 56.04527</td><td> 59.55051</td><td> 58.63786</td><td> 57.91472</td><td> 64.41487</td><td> 61.66405</td><td>4</td><td>NOTCH2:4:1 </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENST00000368655.4:153806580</td><td> 7.829808</td><td> 2.833333</td><td> -4.996475</td><td>9.785228e-03</td><td>3.464253e-01</td><td>2.023910e-02</td><td> 6.470399</td><td>  0.3962766</td><td> -6.074123</td><td>⋯</td><td>ENST00000368655.4:153804724:chr1:153806580</td><td>5</td><td> 55.91176</td><td> 63.74157</td><td> 58.74510</td><td> 63.64539</td><td> 70.11579</td><td> 64.04167</td><td>1</td><td>GATAD2B:5:1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 24\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & newid & dpal.0.1.x & dpal.0.2.x & dpal.1.2.x & kspval.0.1.x & kspval.0.2.x & kspval.1.2.x & dpal.0.1.y & dpal.0.2.y & dpal.1.2.y & ⋯ & V4 & ALE & rep1\\_0 & rep1\\_1 & rep1\\_2 & rep2\\_0 & rep2\\_1 & rep2\\_2 & tandem.new & cluster\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <chr> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ENST00000328089.10:9369527  &  5.180414 & -5.737671 & -10.918085 & 1.071275e-01 & 3.714737e-01 & 3.388636e-04 &  4.600682 &  -1.3205336 &  -5.921216 & ⋯ & ENST00000328089.10:9369532:chr1:9369527    & 2 &  44.28767 &  49.46809 &  38.55000 &  47.57341 &  52.17409 &  46.25287 & 2 & SPSB1:2:2  \\\\\n",
       "\t2 & ENST00000529905.5:40073187  &  6.099345 & 11.971800 &   5.872455 & 9.861732e-03 & 1.611132e-05 & 8.354466e-02 & 25.915931 &  22.0499838 &  -3.865947 & ⋯ & ENST00000529905.5:40072937:chr1:40073187   & 4 & 115.99531 & 122.09465 & 127.96711 & 119.05128 & 144.96721 & 141.10127 & 4 & PPT1:4:1   \\\\\n",
       "\t3 & ENST00000372653.5:40771114  & 12.777349 &  1.303309 & -11.474040 & 2.449807e-12 & 8.186515e-04 & 8.434666e-10 &  3.733495 & -22.1128374 & -25.846332 & ⋯ & ENST00000372653.5:40771106:chr1:40771114   & 8 &  65.64068 &  78.41803 &  66.94399 & 103.42475 & 107.15825 &  81.31191 & 2 & NFYC:8:1   \\\\\n",
       "\t4 & ENST00000304979.7:42676754  & 22.169459 &  1.353835 & -20.815624 & 9.255115e-14 & 3.251077e-08 & 3.018415e-14 &  7.154819 &   6.0500800 &  -1.104739 & ⋯ & ENST00000304979.7:42676758:chr1:42676754   & 3 &  41.92429 &  64.09375 &  43.27813 &  45.31574 &  52.47056 &  51.36582 & 1 & PPIH:3:1   \\\\\n",
       "\t5 & ENST00000256646.6:119912062 &  3.505238 &  2.592593 &  -0.912645 & 8.525272e-02 & 1.186867e-06 & 8.660299e-05 &  6.500152 &   3.7493314 &  -2.750821 & ⋯ & ENST00000256646.6:119911552:chr1:119912062 & 4 &  56.04527 &  59.55051 &  58.63786 &  57.91472 &  64.41487 &  61.66405 & 4 & NOTCH2:4:1 \\\\\n",
       "\t6 & ENST00000368655.4:153806580 &  7.829808 &  2.833333 &  -4.996475 & 9.785228e-03 & 3.464253e-01 & 2.023910e-02 &  6.470399 &   0.3962766 &  -6.074123 & ⋯ & ENST00000368655.4:153804724:chr1:153806580 & 5 &  55.91176 &  63.74157 &  58.74510 &  63.64539 &  70.11579 &  64.04167 & 1 & GATAD2B:5:1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 24\n",
       "\n",
       "| <!--/--> | newid &lt;chr&gt; | dpal.0.1.x &lt;dbl&gt; | dpal.0.2.x &lt;dbl&gt; | dpal.1.2.x &lt;dbl&gt; | kspval.0.1.x &lt;dbl&gt; | kspval.0.2.x &lt;dbl&gt; | kspval.1.2.x &lt;dbl&gt; | dpal.0.1.y &lt;dbl&gt; | dpal.0.2.y &lt;dbl&gt; | dpal.1.2.y &lt;dbl&gt; | ⋯ ⋯ | V4 &lt;chr&gt; | ALE &lt;int&gt; | rep1_0 &lt;dbl&gt; | rep1_1 &lt;dbl&gt; | rep1_2 &lt;dbl&gt; | rep2_0 &lt;dbl&gt; | rep2_1 &lt;dbl&gt; | rep2_2 &lt;dbl&gt; | tandem.new &lt;int&gt; | cluster &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | ENST00000328089.10:9369527  |  5.180414 | -5.737671 | -10.918085 | 1.071275e-01 | 3.714737e-01 | 3.388636e-04 |  4.600682 |  -1.3205336 |  -5.921216 | ⋯ | ENST00000328089.10:9369532:chr1:9369527    | 2 |  44.28767 |  49.46809 |  38.55000 |  47.57341 |  52.17409 |  46.25287 | 2 | SPSB1:2:2   |\n",
       "| 2 | ENST00000529905.5:40073187  |  6.099345 | 11.971800 |   5.872455 | 9.861732e-03 | 1.611132e-05 | 8.354466e-02 | 25.915931 |  22.0499838 |  -3.865947 | ⋯ | ENST00000529905.5:40072937:chr1:40073187   | 4 | 115.99531 | 122.09465 | 127.96711 | 119.05128 | 144.96721 | 141.10127 | 4 | PPT1:4:1    |\n",
       "| 3 | ENST00000372653.5:40771114  | 12.777349 |  1.303309 | -11.474040 | 2.449807e-12 | 8.186515e-04 | 8.434666e-10 |  3.733495 | -22.1128374 | -25.846332 | ⋯ | ENST00000372653.5:40771106:chr1:40771114   | 8 |  65.64068 |  78.41803 |  66.94399 | 103.42475 | 107.15825 |  81.31191 | 2 | NFYC:8:1    |\n",
       "| 4 | ENST00000304979.7:42676754  | 22.169459 |  1.353835 | -20.815624 | 9.255115e-14 | 3.251077e-08 | 3.018415e-14 |  7.154819 |   6.0500800 |  -1.104739 | ⋯ | ENST00000304979.7:42676758:chr1:42676754   | 3 |  41.92429 |  64.09375 |  43.27813 |  45.31574 |  52.47056 |  51.36582 | 1 | PPIH:3:1    |\n",
       "| 5 | ENST00000256646.6:119912062 |  3.505238 |  2.592593 |  -0.912645 | 8.525272e-02 | 1.186867e-06 | 8.660299e-05 |  6.500152 |   3.7493314 |  -2.750821 | ⋯ | ENST00000256646.6:119911552:chr1:119912062 | 4 |  56.04527 |  59.55051 |  58.63786 |  57.91472 |  64.41487 |  61.66405 | 4 | NOTCH2:4:1  |\n",
       "| 6 | ENST00000368655.4:153806580 |  7.829808 |  2.833333 |  -4.996475 | 9.785228e-03 | 3.464253e-01 | 2.023910e-02 |  6.470399 |   0.3962766 |  -6.074123 | ⋯ | ENST00000368655.4:153804724:chr1:153806580 | 5 |  55.91176 |  63.74157 |  58.74510 |  63.64539 |  70.11579 |  64.04167 | 1 | GATAD2B:5:1 |\n",
       "\n"
      ],
      "text/plain": [
       "  newid                       dpal.0.1.x dpal.0.2.x dpal.1.2.x kspval.0.1.x\n",
       "1 ENST00000328089.10:9369527   5.180414  -5.737671  -10.918085 1.071275e-01\n",
       "2 ENST00000529905.5:40073187   6.099345  11.971800    5.872455 9.861732e-03\n",
       "3 ENST00000372653.5:40771114  12.777349   1.303309  -11.474040 2.449807e-12\n",
       "4 ENST00000304979.7:42676754  22.169459   1.353835  -20.815624 9.255115e-14\n",
       "5 ENST00000256646.6:119912062  3.505238   2.592593   -0.912645 8.525272e-02\n",
       "6 ENST00000368655.4:153806580  7.829808   2.833333   -4.996475 9.785228e-03\n",
       "  kspval.0.2.x kspval.1.2.x dpal.0.1.y dpal.0.2.y  dpal.1.2.y <U+22EF>       \n",
       "1 3.714737e-01 3.388636e-04  4.600682   -1.3205336  -5.921216 <U+22EF>\n",
       "2 1.611132e-05 8.354466e-02 25.915931   22.0499838  -3.865947 <U+22EF>\n",
       "3 8.186515e-04 8.434666e-10  3.733495  -22.1128374 -25.846332 <U+22EF>\n",
       "4 3.251077e-08 3.018415e-14  7.154819    6.0500800  -1.104739 <U+22EF>\n",
       "5 1.186867e-06 8.660299e-05  6.500152    3.7493314  -2.750821 <U+22EF>\n",
       "6 3.464253e-01 2.023910e-02  6.470399    0.3962766  -6.074123 <U+22EF>\n",
       "  V4                                         ALE rep1_0    rep1_1    rep1_2   \n",
       "1 ENST00000328089.10:9369532:chr1:9369527    2    44.28767  49.46809  38.55000\n",
       "2 ENST00000529905.5:40072937:chr1:40073187   4   115.99531 122.09465 127.96711\n",
       "3 ENST00000372653.5:40771106:chr1:40771114   8    65.64068  78.41803  66.94399\n",
       "4 ENST00000304979.7:42676758:chr1:42676754   3    41.92429  64.09375  43.27813\n",
       "5 ENST00000256646.6:119911552:chr1:119912062 4    56.04527  59.55051  58.63786\n",
       "6 ENST00000368655.4:153804724:chr1:153806580 5    55.91176  63.74157  58.74510\n",
       "  rep2_0    rep2_1    rep2_2    tandem.new cluster    \n",
       "1  47.57341  52.17409  46.25287 2          SPSB1:2:2  \n",
       "2 119.05128 144.96721 141.10127 4          PPT1:4:1   \n",
       "3 103.42475 107.15825  81.31191 2          NFYC:8:1   \n",
       "4  45.31574  52.47056  51.36582 1          PPIH:3:1   \n",
       "5  57.91472  64.41487  61.66405 4          NOTCH2:4:1 \n",
       "6  63.64539  70.11579  64.04167 1          GATAD2B:5:1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#x<-read.table(\"Actd.custom.meanPAL.acrossalltimepoints_insertsize250.txt\",header=T,stringsAsFactors = F)\n",
    "#head(x)\n",
    "#y=read.table(\"ActD_PALchange.transcriptid.list.insert250.txt\",header=T,stringsAsFactors = F,sep=\"\\t\")\n",
    "#head(y)\n",
    "pal.all<-read.table(\"ActD_insert250.txt\",header=T,stringsAsFactors = F)\n",
    "nrow(pal.all)\n",
    "id_=pal.all%>%\n",
    "    filter(`kspval.0.1.x`<0.2&`kspval.0.1.y`<0.2)%>%\n",
    "  filter(`dpal.0.1.x`>3&`dpal.0.1.y`>3)%>%\n",
    "filter((dpal.0.1.x+dpal.0.1.y)/2>=3)\n",
    "nrow(id_)\n",
    "ctrl=pal.all%>%\n",
    "  filter((`dpal.0.1.x`<1&`dpal.0.1.y`<1))\n",
    "nrow(ctrl)\n",
    "bg=pal.all%>%\n",
    "  #separate(newid,c(\"enst\",\"ref\",\"cps\"),sep=\":\")%>%unite(\"newid\",c(enst,cps),sep=\":\")%>%\n",
    "  inner_join(utr.id,by=\"newid\")%>%dplyr::rename(seq=V2)\n",
    "#bg=unlist(bg$newid)\n",
    "nrow(bg) #7100 #7100\n",
    "head(id_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "19f5e651",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "6876"
      ],
      "text/latex": [
       "6876"
      ],
      "text/markdown": [
       "6876"
      ],
      "text/plain": [
       "[1] 6876"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "2763"
      ],
      "text/latex": [
       "2763"
      ],
      "text/markdown": [
       "2763"
      ],
      "text/plain": [
       "[1] 2763"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "6876"
      ],
      "text/latex": [
       "6876"
      ],
      "text/markdown": [
       "6876"
      ],
      "text/plain": [
       "[1] 6876"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(x);nrow(y);nrow(pal.all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "2fcf7538",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>hgnc</th><th scope=col>newid</th><th scope=col>cluster</th><th scope=col>ALE</th><th scope=col>tandem.new</th><th scope=col>X0</th><th scope=col>X1</th><th scope=col>X2</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ISG15  </td><td>ENST00000379389.4:1014536 </td><td>ISG15:1:1  </td><td>1</td><td>1</td><td>39.45536</td><td>39.29251</td><td>40.23275</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>SDF4   </td><td>ENST00000360001.10:1216883</td><td>SDF4:3:1   </td><td>3</td><td>1</td><td>13.56960</td><td>15.12861</td><td>13.95809</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>SDF4   </td><td>ENST00000263741.11:1216924</td><td>SDF4:2:1   </td><td>2</td><td>1</td><td>51.41975</td><td>52.41744</td><td>52.17947</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>B3GALT6</td><td>ENST00000379198.3:1234172 </td><td>B3GALT6:1:2</td><td>1</td><td>2</td><td>52.22210</td><td>53.27926</td><td>55.32442</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>UBE2J2 </td><td>ENST00000347370.6:1253920 </td><td>UBE2J2:5:2 </td><td>5</td><td>2</td><td>89.12181</td><td>91.69267</td><td>92.49213</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>UBE2J2 </td><td>ENST00000400929.6:1254899 </td><td>UBE2J2:5:1 </td><td>5</td><td>1</td><td>43.30606</td><td>47.45721</td><td>46.35648</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & hgnc & newid & cluster & ALE & tandem.new & X0 & X1 & X2\\\\\n",
       "  & <chr> & <chr> & <chr> & <int> & <int> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & ISG15   & ENST00000379389.4:1014536  & ISG15:1:1   & 1 & 1 & 39.45536 & 39.29251 & 40.23275\\\\\n",
       "\t2 & SDF4    & ENST00000360001.10:1216883 & SDF4:3:1    & 3 & 1 & 13.56960 & 15.12861 & 13.95809\\\\\n",
       "\t3 & SDF4    & ENST00000263741.11:1216924 & SDF4:2:1    & 2 & 1 & 51.41975 & 52.41744 & 52.17947\\\\\n",
       "\t4 & B3GALT6 & ENST00000379198.3:1234172  & B3GALT6:1:2 & 1 & 2 & 52.22210 & 53.27926 & 55.32442\\\\\n",
       "\t5 & UBE2J2  & ENST00000347370.6:1253920  & UBE2J2:5:2  & 5 & 2 & 89.12181 & 91.69267 & 92.49213\\\\\n",
       "\t6 & UBE2J2  & ENST00000400929.6:1254899  & UBE2J2:5:1  & 5 & 1 & 43.30606 & 47.45721 & 46.35648\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 8\n",
       "\n",
       "| <!--/--> | hgnc &lt;chr&gt; | newid &lt;chr&gt; | cluster &lt;chr&gt; | ALE &lt;int&gt; | tandem.new &lt;int&gt; | X0 &lt;dbl&gt; | X1 &lt;dbl&gt; | X2 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | ISG15   | ENST00000379389.4:1014536  | ISG15:1:1   | 1 | 1 | 39.45536 | 39.29251 | 40.23275 |\n",
       "| 2 | SDF4    | ENST00000360001.10:1216883 | SDF4:3:1    | 3 | 1 | 13.56960 | 15.12861 | 13.95809 |\n",
       "| 3 | SDF4    | ENST00000263741.11:1216924 | SDF4:2:1    | 2 | 1 | 51.41975 | 52.41744 | 52.17947 |\n",
       "| 4 | B3GALT6 | ENST00000379198.3:1234172  | B3GALT6:1:2 | 1 | 2 | 52.22210 | 53.27926 | 55.32442 |\n",
       "| 5 | UBE2J2  | ENST00000347370.6:1253920  | UBE2J2:5:2  | 5 | 2 | 89.12181 | 91.69267 | 92.49213 |\n",
       "| 6 | UBE2J2  | ENST00000400929.6:1254899  | UBE2J2:5:1  | 5 | 1 | 43.30606 | 47.45721 | 46.35648 |\n",
       "\n"
      ],
      "text/plain": [
       "  hgnc    newid                      cluster     ALE tandem.new X0      \n",
       "1 ISG15   ENST00000379389.4:1014536  ISG15:1:1   1   1          39.45536\n",
       "2 SDF4    ENST00000360001.10:1216883 SDF4:3:1    3   1          13.56960\n",
       "3 SDF4    ENST00000263741.11:1216924 SDF4:2:1    2   1          51.41975\n",
       "4 B3GALT6 ENST00000379198.3:1234172  B3GALT6:1:2 1   2          52.22210\n",
       "5 UBE2J2  ENST00000347370.6:1253920  UBE2J2:5:2  5   2          89.12181\n",
       "6 UBE2J2  ENST00000400929.6:1254899  UBE2J2:5:1  5   1          43.30606\n",
       "  X1       X2      \n",
       "1 39.29251 40.23275\n",
       "2 15.12861 13.95809\n",
       "3 52.41744 52.17947\n",
       "4 53.27926 55.32442\n",
       "5 91.69267 92.49213\n",
       "6 47.45721 46.35648"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "459"
      ],
      "text/latex": [
       "459"
      ],
      "text/markdown": [
       "459"
      ],
      "text/plain": [
       "[1] 459"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "3068"
      ],
      "text/latex": [
       "3068"
      ],
      "text/markdown": [
       "3068"
      ],
      "text/plain": [
       "[1] 3068"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "5047"
      ],
      "text/latex": [
       "5047"
      ],
      "text/markdown": [
       "5047"
      ],
      "text/plain": [
       "[1] 5047"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>hgnc</th><th scope=col>newid</th><th scope=col>cluster</th><th scope=col>ALE</th><th scope=col>tandem.new</th><th scope=col>X0</th><th scope=col>X1</th><th scope=col>X2</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>TNFRSF14</td><td>ENST00000355716.4:2563741 </td><td>TNFRSF14:5:1</td><td>5</td><td>1</td><td>82.40454</td><td>89.17630</td><td>87.48372</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>RERE    </td><td>ENST00000377464.5:8352404 </td><td>RERE:6:1    </td><td>6</td><td>1</td><td>63.57829</td><td>69.72249</td><td>66.60631</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>NECAP2  </td><td>ENST00000443980.6:16460070</td><td>NECAP2:3:1  </td><td>3</td><td>1</td><td>53.70508</td><td>59.10934</td><td>56.41762</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>EMC1    </td><td>ENST00000486405.1:19218951</td><td>EMC1:3:1    </td><td>3</td><td>1</td><td>65.24573</td><td>70.29326</td><td>67.45808</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>DDOST   </td><td>ENST00000375048.7:20652186</td><td>DDOST:2:1   </td><td>2</td><td>3</td><td>56.81589</td><td>63.02540</td><td>62.13230</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>EIF4G3  </td><td>ENST00000602326.5:20806497</td><td>EIF4G3:4:1  </td><td>4</td><td>1</td><td>61.77399</td><td>67.99354</td><td>63.80132</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & hgnc & newid & cluster & ALE & tandem.new & X0 & X1 & X2\\\\\n",
       "  & <chr> & <chr> & <chr> & <int> & <int> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & TNFRSF14 & ENST00000355716.4:2563741  & TNFRSF14:5:1 & 5 & 1 & 82.40454 & 89.17630 & 87.48372\\\\\n",
       "\t2 & RERE     & ENST00000377464.5:8352404  & RERE:6:1     & 6 & 1 & 63.57829 & 69.72249 & 66.60631\\\\\n",
       "\t3 & NECAP2   & ENST00000443980.6:16460070 & NECAP2:3:1   & 3 & 1 & 53.70508 & 59.10934 & 56.41762\\\\\n",
       "\t4 & EMC1     & ENST00000486405.1:19218951 & EMC1:3:1     & 3 & 1 & 65.24573 & 70.29326 & 67.45808\\\\\n",
       "\t5 & DDOST    & ENST00000375048.7:20652186 & DDOST:2:1    & 2 & 3 & 56.81589 & 63.02540 & 62.13230\\\\\n",
       "\t6 & EIF4G3   & ENST00000602326.5:20806497 & EIF4G3:4:1   & 4 & 1 & 61.77399 & 67.99354 & 63.80132\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 8\n",
       "\n",
       "| <!--/--> | hgnc &lt;chr&gt; | newid &lt;chr&gt; | cluster &lt;chr&gt; | ALE &lt;int&gt; | tandem.new &lt;int&gt; | X0 &lt;dbl&gt; | X1 &lt;dbl&gt; | X2 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | TNFRSF14 | ENST00000355716.4:2563741  | TNFRSF14:5:1 | 5 | 1 | 82.40454 | 89.17630 | 87.48372 |\n",
       "| 2 | RERE     | ENST00000377464.5:8352404  | RERE:6:1     | 6 | 1 | 63.57829 | 69.72249 | 66.60631 |\n",
       "| 3 | NECAP2   | ENST00000443980.6:16460070 | NECAP2:3:1   | 3 | 1 | 53.70508 | 59.10934 | 56.41762 |\n",
       "| 4 | EMC1     | ENST00000486405.1:19218951 | EMC1:3:1     | 3 | 1 | 65.24573 | 70.29326 | 67.45808 |\n",
       "| 5 | DDOST    | ENST00000375048.7:20652186 | DDOST:2:1    | 2 | 3 | 56.81589 | 63.02540 | 62.13230 |\n",
       "| 6 | EIF4G3   | ENST00000602326.5:20806497 | EIF4G3:4:1   | 4 | 1 | 61.77399 | 67.99354 | 63.80132 |\n",
       "\n"
      ],
      "text/plain": [
       "  hgnc     newid                      cluster      ALE tandem.new X0      \n",
       "1 TNFRSF14 ENST00000355716.4:2563741  TNFRSF14:5:1 5   1          82.40454\n",
       "2 RERE     ENST00000377464.5:8352404  RERE:6:1     6   1          63.57829\n",
       "3 NECAP2   ENST00000443980.6:16460070 NECAP2:3:1   3   1          53.70508\n",
       "4 EMC1     ENST00000486405.1:19218951 EMC1:3:1     3   1          65.24573\n",
       "5 DDOST    ENST00000375048.7:20652186 DDOST:2:1    2   3          56.81589\n",
       "6 EIF4G3   ENST00000602326.5:20806497 EIF4G3:4:1   4   1          61.77399\n",
       "  X1       X2      \n",
       "1 89.17630 87.48372\n",
       "2 69.72249 66.60631\n",
       "3 59.10934 56.41762\n",
       "4 70.29326 67.45808\n",
       "5 63.02540 62.13230\n",
       "6 67.99354 63.80132"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "x<-read.table(\"Actd.custom.meanPAL.acrossalltimepoints_insertsize250.txt\",header=T,stringsAsFactors = F)\n",
    "head(x)\n",
    "id_=x%>%filter(X1-X0>5)\n",
    "ctrl=x%>%\n",
    "  filter(X1-X0<1)\n",
    "nrow(id_)\n",
    "nrow(ctrl)\n",
    "bg=x%>%\n",
    "  #separate(newid,c(\"enst\",\"ref\",\"cps\"),sep=\":\")%>%unite(\"newid\",c(enst,cps),sep=\":\")%>%\n",
    "  inner_join(utr.id,by=\"newid\")%>%dplyr::rename(seq=V2)\n",
    "#bg=unlist(bg$newid)\n",
    "nrow(bg) #7100\n",
    "head(id_)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "dac820c0",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in parse(text = x, srcfile = src): <text>:2:13: unexpected symbol\n1: #Code4: Build fasta file for control vs condition\n2: All:Without sampling\n               ^\n",
     "output_type": "error",
     "traceback": [
      "Error in parse(text = x, srcfile = src): <text>:2:13: unexpected symbol\n1: #Code4: Build fasta file for control vs condition\n2: All:Without sampling\n               ^\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "#Code4: Build fasta file for control vs condition\n",
    "All:Without sampling\n",
    "PAL: up VS all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "45886a26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "459"
      ],
      "text/latex": [
       "459"
      ],
      "text/markdown": [
       "459"
      ],
      "text/plain": [
       "[1] 459"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "374"
      ],
      "text/latex": [
       "374"
      ],
      "text/markdown": [
       "374"
      ],
      "text/plain": [
       "[1] 374"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "1993"
      ],
      "text/latex": [
       "1993"
      ],
      "text/markdown": [
       "1993"
      ],
      "text/plain": [
       "[1] 1993"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "upgene=id_\n",
    "nrow(upgene)\n",
    "UP=bg%>%filter(newid%in%unlist(upgene$newid))%>%dplyr::select(newid2,seq)%>%mutate(newid2=paste0(\">\",newid2))\n",
    "CTRL=bg%>%\n",
    "  filter((newid%in%unlist(ctrl$newid)))%>%\n",
    "  dplyr::select(newid2,seq)%>%mutate(newid2=paste0(\">\",newid2))\n",
    "nrow(UP)#49\n",
    "nrow(CTRL)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48d87d29",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "d3990f21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>newid2</th><th scope=col>seq</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>&gt;ENST00000263741.11:1216930:1216924(-)</td><td>TCGCTTCAGTCACTCTGCCGAAGGCGCTGACGGCATCGCGGCCGGAACCTCTGGGCCCGGCCCCTCCCAGGGCCGCCGCTCCGTGGGAAAAAACAGCTCCTCCATTTCCTTGAAAACTGAACGATTATTAAAAATAGATTAAACTTCGCTGGAAATGAGTAGCCAGGAAGTTCAGGGGAGGGTGCCGGGTCCTTCCCGGGCCTGGCGTGTCGGAGCCACCCAGGTCCCGCAGCTGCCGCTGAGAAAATGCAAATATTTGTTGTGACAAGAATCACATACATTTACTTTAAATATAGTTGCC</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>&gt;ENST00000492508.1:1402181:1402045(-) </td><td>CTGGGTTCTTCCTGAGCCAGCGCCGCAACCACGCGGAGCCGGAGTGTGTGTCTCACACTGACAGCAGGTTCTGATTCATTCTTTTTTCAGTGCCAGGTGGAGCTCAACAGGAAAGTCCTAGCGGATCTGGCCATCTACGAGCCAAAGACTTTCAAATCTTTGGCTGCCTTGGCCAGTAGGAGGCGACACGAAGGATTTGCTGCTGCCTTGGGGGATGGGAAGGAACCTGAAGGCATTTTTTCCAGAGTGGTGCAGTACCACTGAGGACTGTTGCTGTATTGATTAGGAAAAGAGACAGAGT</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>&gt;ENST00000291386.3:1541672:1541680(-) </td><td>TTCTGCTGGTGACGGTAGCATTACCCAAATAAACTGTGCATATGAAATGGGAGAGGAGATGCCAAAACGCCAGATGAAAGCAATCAAGTTTCTTCTTTTCCACTTTTACTTATGAGCAGGATATTGATTACAAAGTTTTTCTTCTTTAACCAAAAAGGAAAGACAACGGTTTGTGTGCACTTCCCGACATACCTGTGTCTTCGTGTGCCTGCCTTCCCTCCCTCCTCCCCACCGGGCCGGACTGTACAGAGCCCTGCTGCGGCGTGTTAGGAATGACCTGGAATTGTCAATAAACAGATGC</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>&gt;ENST00000422725.3:1598011:1598014(-) </td><td>CTGCTGCAGACCCGGGCCCCGGGCCGAGGTTGCTTTCTGTGACGACCGCCTGGTGACGCTGGACACCAGGAGGGTTATGGCACTGCCCTCCCCTCCTGGAGCCCCTGGCACTGACCACCTGGACTTGACCTGACTCCTGGTCACTGGCCAAACACCCATCCTCACCCCAGCCACCCCGGGCCCTGGCTGGTGCCGTGTGACACCTGTAGCCCACCTGTGTGTCTTTGCTCCGCCCACCCCTGCGTGGCGCCGGTCAGGCGCAAACCTGGCGCCAGTGTCTAAATGTCCATCATTCTCAGAC</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>&gt;ENST00000342348.9:1752470:1752512(-) </td><td>CTCAGTGGCACGGCCACTTCCTTTCTGTAGCTGGGTTAGAGCCTGGGTCTGCCTTTTGTCCAGATCAGCTGTTTTTTTAAAATGTCTGACTTTTTTTGCATTTCTAAAGAAGCGTGAGAAATGGGCTGGGAGTGCTTCTGTCCTGCTGACACCCCGCGGTGGGTCCCTGGAGCGCGGCCTCCAGCTGCCGCAATTTCCATGCCAGGATATTTTTCCGCAAATCAGTCGGTTGAAATTCAGAGGAGTCAGAATGACTCGACCTGTCCTTCAATGTTGATAATAAATGTCTCAGCCAAAAACC</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>&gt;ENST00000378513.7:2405442:2405433(+) </td><td>TGACGTGTATTGAAGGTGTGTGTGCCAAATGCTTCCGACGGAGGTGCTGGCCTTGGTTGGTTTCTCTCTGCCCCGTGTGGTCATCAAGTCCTGGGGGATGTGCTCTGCCCAGCCGCCCTCGGGGAGAGCAGCGCCGCCTCCCATGGGGCCGTGGGGCTGCTGTTCTCACTGCACTGGCTGAAGCAACCCGCCAGCCTCCGTGCCCCACCCCACCCAGCACGCACTCATTCAGTCCATTGCCTTAACACAAGCCTGATGGGGCTGTTTTCTCACAATATAAACGAATAAAGTGTCTTCTGGC</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & newid2 & seq\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & >ENST00000263741.11:1216930:1216924(-) & TCGCTTCAGTCACTCTGCCGAAGGCGCTGACGGCATCGCGGCCGGAACCTCTGGGCCCGGCCCCTCCCAGGGCCGCCGCTCCGTGGGAAAAAACAGCTCCTCCATTTCCTTGAAAACTGAACGATTATTAAAAATAGATTAAACTTCGCTGGAAATGAGTAGCCAGGAAGTTCAGGGGAGGGTGCCGGGTCCTTCCCGGGCCTGGCGTGTCGGAGCCACCCAGGTCCCGCAGCTGCCGCTGAGAAAATGCAAATATTTGTTGTGACAAGAATCACATACATTTACTTTAAATATAGTTGCC\\\\\n",
       "\t2 & >ENST00000492508.1:1402181:1402045(-)  & CTGGGTTCTTCCTGAGCCAGCGCCGCAACCACGCGGAGCCGGAGTGTGTGTCTCACACTGACAGCAGGTTCTGATTCATTCTTTTTTCAGTGCCAGGTGGAGCTCAACAGGAAAGTCCTAGCGGATCTGGCCATCTACGAGCCAAAGACTTTCAAATCTTTGGCTGCCTTGGCCAGTAGGAGGCGACACGAAGGATTTGCTGCTGCCTTGGGGGATGGGAAGGAACCTGAAGGCATTTTTTCCAGAGTGGTGCAGTACCACTGAGGACTGTTGCTGTATTGATTAGGAAAAGAGACAGAGT\\\\\n",
       "\t3 & >ENST00000291386.3:1541672:1541680(-)  & TTCTGCTGGTGACGGTAGCATTACCCAAATAAACTGTGCATATGAAATGGGAGAGGAGATGCCAAAACGCCAGATGAAAGCAATCAAGTTTCTTCTTTTCCACTTTTACTTATGAGCAGGATATTGATTACAAAGTTTTTCTTCTTTAACCAAAAAGGAAAGACAACGGTTTGTGTGCACTTCCCGACATACCTGTGTCTTCGTGTGCCTGCCTTCCCTCCCTCCTCCCCACCGGGCCGGACTGTACAGAGCCCTGCTGCGGCGTGTTAGGAATGACCTGGAATTGTCAATAAACAGATGC\\\\\n",
       "\t4 & >ENST00000422725.3:1598011:1598014(-)  & CTGCTGCAGACCCGGGCCCCGGGCCGAGGTTGCTTTCTGTGACGACCGCCTGGTGACGCTGGACACCAGGAGGGTTATGGCACTGCCCTCCCCTCCTGGAGCCCCTGGCACTGACCACCTGGACTTGACCTGACTCCTGGTCACTGGCCAAACACCCATCCTCACCCCAGCCACCCCGGGCCCTGGCTGGTGCCGTGTGACACCTGTAGCCCACCTGTGTGTCTTTGCTCCGCCCACCCCTGCGTGGCGCCGGTCAGGCGCAAACCTGGCGCCAGTGTCTAAATGTCCATCATTCTCAGAC\\\\\n",
       "\t5 & >ENST00000342348.9:1752470:1752512(-)  & CTCAGTGGCACGGCCACTTCCTTTCTGTAGCTGGGTTAGAGCCTGGGTCTGCCTTTTGTCCAGATCAGCTGTTTTTTTAAAATGTCTGACTTTTTTTGCATTTCTAAAGAAGCGTGAGAAATGGGCTGGGAGTGCTTCTGTCCTGCTGACACCCCGCGGTGGGTCCCTGGAGCGCGGCCTCCAGCTGCCGCAATTTCCATGCCAGGATATTTTTCCGCAAATCAGTCGGTTGAAATTCAGAGGAGTCAGAATGACTCGACCTGTCCTTCAATGTTGATAATAAATGTCTCAGCCAAAAACC\\\\\n",
       "\t6 & >ENST00000378513.7:2405442:2405433(+)  & TGACGTGTATTGAAGGTGTGTGTGCCAAATGCTTCCGACGGAGGTGCTGGCCTTGGTTGGTTTCTCTCTGCCCCGTGTGGTCATCAAGTCCTGGGGGATGTGCTCTGCCCAGCCGCCCTCGGGGAGAGCAGCGCCGCCTCCCATGGGGCCGTGGGGCTGCTGTTCTCACTGCACTGGCTGAAGCAACCCGCCAGCCTCCGTGCCCCACCCCACCCAGCACGCACTCATTCAGTCCATTGCCTTAACACAAGCCTGATGGGGCTGTTTTCTCACAATATAAACGAATAAAGTGTCTTCTGGC\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | newid2 &lt;chr&gt; | seq &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | &gt;ENST00000263741.11:1216930:1216924(-) | TCGCTTCAGTCACTCTGCCGAAGGCGCTGACGGCATCGCGGCCGGAACCTCTGGGCCCGGCCCCTCCCAGGGCCGCCGCTCCGTGGGAAAAAACAGCTCCTCCATTTCCTTGAAAACTGAACGATTATTAAAAATAGATTAAACTTCGCTGGAAATGAGTAGCCAGGAAGTTCAGGGGAGGGTGCCGGGTCCTTCCCGGGCCTGGCGTGTCGGAGCCACCCAGGTCCCGCAGCTGCCGCTGAGAAAATGCAAATATTTGTTGTGACAAGAATCACATACATTTACTTTAAATATAGTTGCC |\n",
       "| 2 | &gt;ENST00000492508.1:1402181:1402045(-)  | CTGGGTTCTTCCTGAGCCAGCGCCGCAACCACGCGGAGCCGGAGTGTGTGTCTCACACTGACAGCAGGTTCTGATTCATTCTTTTTTCAGTGCCAGGTGGAGCTCAACAGGAAAGTCCTAGCGGATCTGGCCATCTACGAGCCAAAGACTTTCAAATCTTTGGCTGCCTTGGCCAGTAGGAGGCGACACGAAGGATTTGCTGCTGCCTTGGGGGATGGGAAGGAACCTGAAGGCATTTTTTCCAGAGTGGTGCAGTACCACTGAGGACTGTTGCTGTATTGATTAGGAAAAGAGACAGAGT |\n",
       "| 3 | &gt;ENST00000291386.3:1541672:1541680(-)  | TTCTGCTGGTGACGGTAGCATTACCCAAATAAACTGTGCATATGAAATGGGAGAGGAGATGCCAAAACGCCAGATGAAAGCAATCAAGTTTCTTCTTTTCCACTTTTACTTATGAGCAGGATATTGATTACAAAGTTTTTCTTCTTTAACCAAAAAGGAAAGACAACGGTTTGTGTGCACTTCCCGACATACCTGTGTCTTCGTGTGCCTGCCTTCCCTCCCTCCTCCCCACCGGGCCGGACTGTACAGAGCCCTGCTGCGGCGTGTTAGGAATGACCTGGAATTGTCAATAAACAGATGC |\n",
       "| 4 | &gt;ENST00000422725.3:1598011:1598014(-)  | CTGCTGCAGACCCGGGCCCCGGGCCGAGGTTGCTTTCTGTGACGACCGCCTGGTGACGCTGGACACCAGGAGGGTTATGGCACTGCCCTCCCCTCCTGGAGCCCCTGGCACTGACCACCTGGACTTGACCTGACTCCTGGTCACTGGCCAAACACCCATCCTCACCCCAGCCACCCCGGGCCCTGGCTGGTGCCGTGTGACACCTGTAGCCCACCTGTGTGTCTTTGCTCCGCCCACCCCTGCGTGGCGCCGGTCAGGCGCAAACCTGGCGCCAGTGTCTAAATGTCCATCATTCTCAGAC |\n",
       "| 5 | &gt;ENST00000342348.9:1752470:1752512(-)  | CTCAGTGGCACGGCCACTTCCTTTCTGTAGCTGGGTTAGAGCCTGGGTCTGCCTTTTGTCCAGATCAGCTGTTTTTTTAAAATGTCTGACTTTTTTTGCATTTCTAAAGAAGCGTGAGAAATGGGCTGGGAGTGCTTCTGTCCTGCTGACACCCCGCGGTGGGTCCCTGGAGCGCGGCCTCCAGCTGCCGCAATTTCCATGCCAGGATATTTTTCCGCAAATCAGTCGGTTGAAATTCAGAGGAGTCAGAATGACTCGACCTGTCCTTCAATGTTGATAATAAATGTCTCAGCCAAAAACC |\n",
       "| 6 | &gt;ENST00000378513.7:2405442:2405433(+)  | TGACGTGTATTGAAGGTGTGTGTGCCAAATGCTTCCGACGGAGGTGCTGGCCTTGGTTGGTTTCTCTCTGCCCCGTGTGGTCATCAAGTCCTGGGGGATGTGCTCTGCCCAGCCGCCCTCGGGGAGAGCAGCGCCGCCTCCCATGGGGCCGTGGGGCTGCTGTTCTCACTGCACTGGCTGAAGCAACCCGCCAGCCTCCGTGCCCCACCCCACCCAGCACGCACTCATTCAGTCCATTGCCTTAACACAAGCCTGATGGGGCTGTTTTCTCACAATATAAACGAATAAAGTGTCTTCTGGC |\n",
       "\n"
      ],
      "text/plain": [
       "  newid2                                \n",
       "1 >ENST00000263741.11:1216930:1216924(-)\n",
       "2 >ENST00000492508.1:1402181:1402045(-) \n",
       "3 >ENST00000291386.3:1541672:1541680(-) \n",
       "4 >ENST00000422725.3:1598011:1598014(-) \n",
       "5 >ENST00000342348.9:1752470:1752512(-) \n",
       "6 >ENST00000378513.7:2405442:2405433(+) \n",
       "  seq                                                                                                                                                                                                                                                                                                          \n",
       "1 TCGCTTCAGTCACTCTGCCGAAGGCGCTGACGGCATCGCGGCCGGAACCTCTGGGCCCGGCCCCTCCCAGGGCCGCCGCTCCGTGGGAAAAAACAGCTCCTCCATTTCCTTGAAAACTGAACGATTATTAAAAATAGATTAAACTTCGCTGGAAATGAGTAGCCAGGAAGTTCAGGGGAGGGTGCCGGGTCCTTCCCGGGCCTGGCGTGTCGGAGCCACCCAGGTCCCGCAGCTGCCGCTGAGAAAATGCAAATATTTGTTGTGACAAGAATCACATACATTTACTTTAAATATAGTTGCC\n",
       "2 CTGGGTTCTTCCTGAGCCAGCGCCGCAACCACGCGGAGCCGGAGTGTGTGTCTCACACTGACAGCAGGTTCTGATTCATTCTTTTTTCAGTGCCAGGTGGAGCTCAACAGGAAAGTCCTAGCGGATCTGGCCATCTACGAGCCAAAGACTTTCAAATCTTTGGCTGCCTTGGCCAGTAGGAGGCGACACGAAGGATTTGCTGCTGCCTTGGGGGATGGGAAGGAACCTGAAGGCATTTTTTCCAGAGTGGTGCAGTACCACTGAGGACTGTTGCTGTATTGATTAGGAAAAGAGACAGAGT\n",
       "3 TTCTGCTGGTGACGGTAGCATTACCCAAATAAACTGTGCATATGAAATGGGAGAGGAGATGCCAAAACGCCAGATGAAAGCAATCAAGTTTCTTCTTTTCCACTTTTACTTATGAGCAGGATATTGATTACAAAGTTTTTCTTCTTTAACCAAAAAGGAAAGACAACGGTTTGTGTGCACTTCCCGACATACCTGTGTCTTCGTGTGCCTGCCTTCCCTCCCTCCTCCCCACCGGGCCGGACTGTACAGAGCCCTGCTGCGGCGTGTTAGGAATGACCTGGAATTGTCAATAAACAGATGC\n",
       "4 CTGCTGCAGACCCGGGCCCCGGGCCGAGGTTGCTTTCTGTGACGACCGCCTGGTGACGCTGGACACCAGGAGGGTTATGGCACTGCCCTCCCCTCCTGGAGCCCCTGGCACTGACCACCTGGACTTGACCTGACTCCTGGTCACTGGCCAAACACCCATCCTCACCCCAGCCACCCCGGGCCCTGGCTGGTGCCGTGTGACACCTGTAGCCCACCTGTGTGTCTTTGCTCCGCCCACCCCTGCGTGGCGCCGGTCAGGCGCAAACCTGGCGCCAGTGTCTAAATGTCCATCATTCTCAGAC\n",
       "5 CTCAGTGGCACGGCCACTTCCTTTCTGTAGCTGGGTTAGAGCCTGGGTCTGCCTTTTGTCCAGATCAGCTGTTTTTTTAAAATGTCTGACTTTTTTTGCATTTCTAAAGAAGCGTGAGAAATGGGCTGGGAGTGCTTCTGTCCTGCTGACACCCCGCGGTGGGTCCCTGGAGCGCGGCCTCCAGCTGCCGCAATTTCCATGCCAGGATATTTTTCCGCAAATCAGTCGGTTGAAATTCAGAGGAGTCAGAATGACTCGACCTGTCCTTCAATGTTGATAATAAATGTCTCAGCCAAAAACC\n",
       "6 TGACGTGTATTGAAGGTGTGTGTGCCAAATGCTTCCGACGGAGGTGCTGGCCTTGGTTGGTTTCTCTCTGCCCCGTGTGGTCATCAAGTCCTGGGGGATGTGCTCTGCCCAGCCGCCCTCGGGGAGAGCAGCGCCGCCTCCCATGGGGCCGTGGGGCTGCTGTTCTCACTGCACTGGCTGAAGCAACCCGCCAGCCTCCGTGCCCCACCCCACCCAGCACGCACTCATTCAGTCCATTGCCTTAACACAAGCCTGATGGGGCTGTTTTCTCACAATATAAACGAATAAAGTGTCTTCTGGC"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "1993"
      ],
      "text/latex": [
       "1993"
      ],
      "text/markdown": [
       "1993"
      ],
      "text/plain": [
       "[1] 1993"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "write.table(UP,\"up.fa\",col.names = F,row.names = F, sep=\"\\n\",quote=F)\n",
    "\n",
    "\n",
    "\n",
    "head(CTRL)\n",
    "write.table(CTRL,\"all.fa\",col.names = F,row.names = F, sep=\"\\n\",quote=F)\n",
    "nrow(CTRL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "eb3cbc17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "! Difference between GC content in negative and positive TREs (use 'gc.correction.pdf' to see pdf figure):\n",
      "  p-value (Wilcoxon-Mann-Whitney test) = 4.40451e-19 \n",
      "  median/sample size of GC positive = 0.3671096 / 374 \n",
      "  median/sample size of GC negative = 0.4318937 / 1993 \n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "t.xcomp <- rbpbs.enrichmentTest( tfs1, \n",
    "                                 \"up.fa\", \n",
    "                                 \"all.fa\",\n",
    "                                 threshold = 6,\n",
    "                                 gc.correction=F,\n",
    "                                 file.prefix=NA,\n",
    "                                 ncores = 6,\n",
    "                                 #use.cluster = T,\n",
    "                                 background.order = 2,\n",
    "                                 background.length = 5000); \n",
    "tmp=t.xcomp$result\n",
    "tmp=tmp%>%filter(rbp.name %in%unlist(x$hgnc))%>%arrange(motif.id,rbp.name)%>%distinct()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "22d27c23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.0501187233627272"
      ],
      "text/latex": [
       "0.0501187233627272"
      ],
      "text/markdown": [
       "0.0501187233627272"
      ],
      "text/plain": [
       "[1] 0.05011872"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.698970004336019"
      ],
      "text/latex": [
       "0.698970004336019"
      ],
      "text/markdown": [
       "0.698970004336019"
      ],
      "text/plain": [
       "[1] 0.69897"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.ret <- tmp[,c('motif.id','rbp.name', 'fe.ratio', 'pvalue','pv.adj' ) ];\n",
    "\tdf.ret$y.log <- -log10( df.ret$pvalue );\n",
    "\tdf.ret$adj.y.log <- -log10(p.adjust(df.ret$pvalue,method=\"fdr\"))\n",
    "\t#df.ret$adj.y.log <- -log10( df.ret$pv.adj )\n",
    "\t#df.ret$y.log [ is.infinite(df.ret$y.log) | (df.ret$y.log >= options$y.max) ] <- options$y.max;\n",
    "idx.enrich  <- which( df.ret$fe.ratio >= 1 );\n",
    "idx.deple   <- which( df.ret$fe.ratio < 1 )\n",
    "df.ret$y.log[idx.deple] <- - df.ret$y.log[idx.deple]\n",
    "\n",
    "df.ret <- df.ret[order( df.ret$y.log, decreasing = F ),];\n",
    "df.ret$order=1:nrow(df.ret)\n",
    "10^(-1.3)\n",
    "-log10(0.2)\n",
    "\n",
    "#df.ret=df.ret%>%distinct_at(vars(rbp.name),.keep_all = T)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "c0bc5842",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.ret=df.ret%>%arrange(-order)%>%distinct_at(vars(rbp.name),.keep_all = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "20e06c95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"300nt_RBP_enrichment.Nosample.0h.1h.PALincrease.compared to alln7100_Last500_7nt.pdf\",width=5,height=4)\n",
    "ggplot(df.ret,aes(x=order,y=y.log,label=rbp.name))+\n",
    "  #geom_point(aes(colour=fe.ratio))+\n",
    "  #scale_colour_gradientn(colours = yk.col(10)[1:10])\n",
    "  #scale_colour_gradient( low = \"#132B43\", high = \"#56B1F7\",space = \"Lab\", na.value = \"grey50\", guide = \"colourbar\")\n",
    "\n",
    "  geom_point(color = dplyr::case_when(abs(df.ret$adj.y.log) >0.7~ \"#EE7447\", \n",
    "                                      abs(df.ret$adj.y.log)<=0.7 ~ \"black\"), \n",
    "             size = 1, alpha = 0.8) +theme_bw()+\n",
    "  geom_text_repel(data        = subset(df.ret, abs(y.log)>1.5),\n",
    "                  #nudge_y       = -(5 - subset(df.ret, abs(y.log)>2)$y.log),\n",
    "                  #nudge_x       = 150 - subset(df.ret, abs(adj.y.log)> 4)$order,\n",
    "                  #nudge_x = 4.5 - subset(df.ret, abs(adj.y.log)> 4)$y.log,\n",
    "                  size          = 6,\n",
    "                  seed=40,\n",
    "                  box.padding   = 1,\n",
    "                  point.padding = 0.2,\n",
    "                  #force.pull   = 1,\n",
    "                  segment.size  = 0.2,\n",
    "                  segment.color = \"grey50\",\n",
    "                  direction     = \"both\")+xlab(\"Rank order\")+labs(title = expression(\"0h\"%->%\"1h\"))+ylab(\"-log10(p-value)\")+theme(legend.position = \"right\")#+xlim(c(0,200))#+ylim(c(-3,16))\n",
    "dev.off() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "f2518181",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "! Difference between GC content in negative and positive TREs (use 'gc.correction.pdf' to see pdf figure):\n",
      "  p-value (Wilcoxon-Mann-Whitney test) = 1.207435e-10 \n",
      "  median/sample size of GC positive = 0.3671096 / 374 \n",
      "  median/sample size of GC negative = 0.4119601 / 5047 \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 54 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>motif.id</th><th scope=col>rbp.name</th><th scope=col>fe.ratio</th><th scope=col>pvalue</th><th scope=col>pv.adj</th><th scope=col>y.log</th><th scope=col>adj.y.log</th><th scope=col>order</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>M031_0.6</td><td>ELAVL1 </td><td> 1.4829288</td><td>6.137493e-05</td><td>0.01239773</td><td> 4.212009e+00</td><td>2.26405544</td><td>90</td></tr>\n",
       "\t<tr><td>M025_0.6</td><td>HNRNPC </td><td> 1.5273317</td><td>2.243182e-04</td><td>0.04531227</td><td> 3.649136e+00</td><td>2.17201430</td><td>88</td></tr>\n",
       "\t<tr><td>M150_0.6</td><td>RALY   </td><td> 1.5876062</td><td>2.264581e-03</td><td>0.45744540</td><td> 2.645012e+00</td><td>1.64501210</td><td>86</td></tr>\n",
       "\t<tr><td>M350_0.6</td><td>ZFP36  </td><td> 1.3799526</td><td>2.709075e-03</td><td>0.54723310</td><td> 2.567179e+00</td><td>1.61293650</td><td>85</td></tr>\n",
       "\t<tr><td>M146_0.6</td><td>PABPC1 </td><td> 1.4394296</td><td>1.547272e-02</td><td>1.00000000</td><td> 1.810433e+00</td><td>0.97013423</td><td>84</td></tr>\n",
       "\t<tr><td>M062_0.6</td><td>SART3  </td><td> 1.4394296</td><td>1.547272e-02</td><td>1.00000000</td><td> 1.810433e+00</td><td>0.97013423</td><td>83</td></tr>\n",
       "\t<tr><td>M036_0.6</td><td>MATR3  </td><td> 1.3222426</td><td>2.055724e-02</td><td>1.00000000</td><td> 1.687035e+00</td><td>0.89522058</td><td>82</td></tr>\n",
       "\t<tr><td>M077_0.6</td><td>U2AF2  </td><td> 1.3923054</td><td>2.121427e-02</td><td>1.00000000</td><td> 1.673372e+00</td><td>0.89522058</td><td>81</td></tr>\n",
       "\t<tr><td>M353_0.6</td><td>SRSF7  </td><td> 1.4638267</td><td>4.186998e-02</td><td>1.00000000</td><td> 1.378097e+00</td><td>0.62797468</td><td>80</td></tr>\n",
       "\t<tr><td>M234_0.6</td><td>RBM47  </td><td> 1.2378816</td><td>6.468847e-02</td><td>1.00000000</td><td> 1.189173e+00</td><td>0.53536903</td><td>79</td></tr>\n",
       "\t<tr><td>M142_0.6</td><td>RBM42  </td><td> 1.1245544</td><td>3.833472e-01</td><td>1.00000000</td><td> 4.164077e-01</td><td>0.07167153</td><td>77</td></tr>\n",
       "\t<tr><td>M012_0.6</td><td>CPEB2  </td><td> 1.1891839</td><td>3.996302e-01</td><td>1.00000000</td><td> 3.983417e-01</td><td>0.06734851</td><td>76</td></tr>\n",
       "\t<tr><td>M152_0.6</td><td>FXR1   </td><td> 1.1387892</td><td>4.100305e-01</td><td>1.00000000</td><td> 3.871838e-01</td><td>0.06640979</td><td>75</td></tr>\n",
       "\t<tr><td>M348_0.6</td><td>SNRPA  </td><td> 1.1061190</td><td>4.923351e-01</td><td>1.00000000</td><td> 3.077392e-01</td><td>0.04369279</td><td>74</td></tr>\n",
       "\t<tr><td>M027_0.6</td><td>HNRNPL </td><td> 1.0764217</td><td>5.967982e-01</td><td>1.00000000</td><td> 2.241725e-01</td><td>0.01024884</td><td>72</td></tr>\n",
       "\t<tr><td>M074_0.6</td><td>TARDBP </td><td> 1.0564499</td><td>6.945445e-01</td><td>1.00000000</td><td> 1.582999e-01</td><td>0.00000000</td><td>71</td></tr>\n",
       "\t<tr><td>M242_0.6</td><td>HNRNPR </td><td> 1.0724227</td><td>7.348285e-01</td><td>1.00000000</td><td> 1.338140e-01</td><td>0.00000000</td><td>70</td></tr>\n",
       "\t<tr><td>M042_0.6</td><td>PABPC4 </td><td> 1.0744150</td><td>7.393874e-01</td><td>1.00000000</td><td> 1.311280e-01</td><td>0.00000000</td><td>69</td></tr>\n",
       "\t<tr><td>M056_0.6</td><td>SRSF3  </td><td> 1.0267670</td><td>8.637077e-01</td><td>1.00000000</td><td> 6.363320e-02</td><td>0.00000000</td><td>68</td></tr>\n",
       "\t<tr><td>M332_0.6</td><td>SRSF2  </td><td> 1.0232912</td><td>8.935065e-01</td><td>1.00000000</td><td> 4.890229e-02</td><td>0.00000000</td><td>67</td></tr>\n",
       "\t<tr><td>M155_0.6</td><td>ZNF638 </td><td> 1.0160066</td><td>9.296218e-01</td><td>1.00000000</td><td> 3.169369e-02</td><td>0.00000000</td><td>65</td></tr>\n",
       "\t<tr><td>M068_0.6</td><td>SNRNP70</td><td> 1.0105200</td><td>9.358803e-01</td><td>1.00000000</td><td> 2.877968e-02</td><td>0.00000000</td><td>64</td></tr>\n",
       "\t<tr><td>M354_0.6</td><td>YTHDC1 </td><td> 0.9072035</td><td>1.000000e+00</td><td>1.00000000</td><td> 0.000000e+00</td><td>0.00000000</td><td>63</td></tr>\n",
       "\t<tr><td>M344_0.6</td><td>RBMX   </td><td>13.4946524</td><td>1.000000e+00</td><td>1.00000000</td><td> 0.000000e+00</td><td>0.00000000</td><td>61</td></tr>\n",
       "\t<tr><td>M292_0.6</td><td>EIF4B  </td><td> 0.9778734</td><td>1.000000e+00</td><td>1.00000000</td><td> 0.000000e+00</td><td>0.00000000</td><td>60</td></tr>\n",
       "\t<tr><td>M271_0.6</td><td>HNRNPA1</td><td> 0.9887171</td><td>1.000000e+00</td><td>1.00000000</td><td> 0.000000e+00</td><td>0.00000000</td><td>58</td></tr>\n",
       "\t<tr><td>M262_0.6</td><td>QKI    </td><td> 0.9943428</td><td>1.000000e+00</td><td>1.00000000</td><td> 0.000000e+00</td><td>0.00000000</td><td>57</td></tr>\n",
       "\t<tr><td>M256_0.6</td><td>ACO1   </td><td>13.4946524</td><td>1.000000e+00</td><td>1.00000000</td><td> 0.000000e+00</td><td>0.00000000</td><td>56</td></tr>\n",
       "\t<tr><td>M236_0.6</td><td>SYNCRIP</td><td> 1.0037345</td><td>1.000000e+00</td><td>1.00000000</td><td> 0.000000e+00</td><td>0.00000000</td><td>55</td></tr>\n",
       "\t<tr><td>M048_0.6</td><td>CIRBP  </td><td> 0.9710792</td><td>1.000000e+00</td><td>1.00000000</td><td> 0.000000e+00</td><td>0.00000000</td><td>53</td></tr>\n",
       "\t<tr><td>M228_0.6</td><td>PTBP1  </td><td> 0.9673586</td><td>1.000000e+00</td><td>1.00000000</td><td>-4.821637e-17</td><td>0.00000000</td><td>52</td></tr>\n",
       "\t<tr><td>M002_0.6</td><td>ANKHD1 </td><td> 0.9493725</td><td>8.425994e-01</td><td>1.00000000</td><td>-7.437888e-02</td><td>0.00000000</td><td>46</td></tr>\n",
       "\t<tr><td>M160_0.6</td><td>KHDRBS1</td><td> 0.9628923</td><td>7.841394e-01</td><td>1.00000000</td><td>-1.056067e-01</td><td>0.00000000</td><td>43</td></tr>\n",
       "\t<tr><td>M273_0.6</td><td>SRSF1  </td><td> 0.8687544</td><td>7.002369e-01</td><td>1.00000000</td><td>-1.547550e-01</td><td>0.00000000</td><td>42</td></tr>\n",
       "\t<tr><td>M053_0.6</td><td>RBM5   </td><td> 0.9306657</td><td>6.916618e-01</td><td>1.00000000</td><td>-1.601062e-01</td><td>0.00000000</td><td>41</td></tr>\n",
       "\t<tr><td>M043_0.6</td><td>PCBP2  </td><td> 0.9371286</td><td>6.889702e-01</td><td>1.00000000</td><td>-1.617996e-01</td><td>0.00000000</td><td>40</td></tr>\n",
       "\t<tr><td>M231_0.6</td><td>EIF2S1 </td><td> 0.9338860</td><td>6.185624e-01</td><td>1.00000000</td><td>-2.086165e-01</td><td>0.01024884</td><td>38</td></tr>\n",
       "\t<tr><td>M047_0.6</td><td>RBM28  </td><td> 0.9118008</td><td>6.162987e-01</td><td>1.00000000</td><td>-2.102088e-01</td><td>0.01024884</td><td>37</td></tr>\n",
       "\t<tr><td>M333_0.6</td><td>SRSF9  </td><td> 0.8178577</td><td>5.791590e-01</td><td>1.00000000</td><td>-2.372022e-01</td><td>0.01024884</td><td>35</td></tr>\n",
       "\t<tr><td>M085_0.6</td><td>ZCRB1  </td><td> 0.9106821</td><td>5.750795e-01</td><td>1.00000000</td><td>-2.402721e-01</td><td>0.01024884</td><td>34</td></tr>\n",
       "\t<tr><td>M177_0.6</td><td>PCBP1  </td><td> 0.9156549</td><td>5.274031e-01</td><td>1.00000000</td><td>-2.778573e-01</td><td>0.02258484</td><td>33</td></tr>\n",
       "\t<tr><td>M032_0.6</td><td>IGF2BP2</td><td> 0.8480246</td><td>4.686091e-01</td><td>1.00000000</td><td>-3.291893e-01</td><td>0.04369279</td><td>30</td></tr>\n",
       "\t<tr><td>M168_0.6</td><td>SFPQ   </td><td> 0.7774697</td><td>4.479276e-01</td><td>1.00000000</td><td>-3.487922e-01</td><td>0.04369279</td><td>28</td></tr>\n",
       "\t<tr><td>M161_0.6</td><td>RBM6   </td><td> 0.8484662</td><td>3.016420e-01</td><td>1.00000000</td><td>-5.205082e-01</td><td>0.13446743</td><td>24</td></tr>\n",
       "\t<tr><td>M316_0.6</td><td>FUS    </td><td> 0.7828390</td><td>2.955792e-01</td><td>1.00000000</td><td>-5.293261e-01</td><td>0.13446743</td><td>23</td></tr>\n",
       "\t<tr><td>M021_0.6</td><td>G3BP2  </td><td> 0.8217897</td><td>2.204943e-01</td><td>1.00000000</td><td>-6.566026e-01</td><td>0.22087398</td><td>20</td></tr>\n",
       "\t<tr><td>M136_0.6</td><td>SNRPB2 </td><td> 0.7990255</td><td>1.847532e-01</td><td>1.00000000</td><td>-7.334079e-01</td><td>0.27052710</td><td>18</td></tr>\n",
       "\t<tr><td>M188_0.6</td><td>PCBP4  </td><td> 0.7067025</td><td>1.098550e-01</td><td>1.00000000</td><td>-9.591801e-01</td><td>0.42083788</td><td>13</td></tr>\n",
       "\t<tr><td>M320_0.6</td><td>MBNL1  </td><td> 0.6649539</td><td>1.061240e-01</td><td>1.00000000</td><td>-9.741864e-01</td><td>0.42083788</td><td>12</td></tr>\n",
       "\t<tr><td>M054_0.6</td><td>RBM8A  </td><td> 0.7102449</td><td>1.017316e-01</td><td>1.00000000</td><td>-9.925440e-01</td><td>0.42083788</td><td>11</td></tr>\n",
       "\t<tr><td>M013_0.6</td><td>DAZAP1 </td><td> 0.7509299</td><td>8.263547e-02</td><td>1.00000000</td><td>-1.082833e+00</td><td>0.49031880</td><td>10</td></tr>\n",
       "\t<tr><td>M109_0.6</td><td>RBM4   </td><td> 0.6779610</td><td>7.630974e-02</td><td>1.00000000</td><td>-1.117420e+00</td><td>0.49788858</td><td> 9</td></tr>\n",
       "\t<tr><td>M147_0.6</td><td>CNOT4  </td><td> 0.7294407</td><td>5.953726e-02</td><td>1.00000000</td><td>-1.225211e+00</td><td>0.53536903</td><td> 7</td></tr>\n",
       "\t<tr><td>M307_0.6</td><td>NONO   </td><td> 0.7333078</td><td>9.606665e-04</td><td>0.19405463</td><td>-3.017427e+00</td><td>1.84133610</td><td> 3</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 54 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " motif.id & rbp.name & fe.ratio & pvalue & pv.adj & y.log & adj.y.log & order\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t M031\\_0.6 & ELAVL1  &  1.4829288 & 6.137493e-05 & 0.01239773 &  4.212009e+00 & 2.26405544 & 90\\\\\n",
       "\t M025\\_0.6 & HNRNPC  &  1.5273317 & 2.243182e-04 & 0.04531227 &  3.649136e+00 & 2.17201430 & 88\\\\\n",
       "\t M150\\_0.6 & RALY    &  1.5876062 & 2.264581e-03 & 0.45744540 &  2.645012e+00 & 1.64501210 & 86\\\\\n",
       "\t M350\\_0.6 & ZFP36   &  1.3799526 & 2.709075e-03 & 0.54723310 &  2.567179e+00 & 1.61293650 & 85\\\\\n",
       "\t M146\\_0.6 & PABPC1  &  1.4394296 & 1.547272e-02 & 1.00000000 &  1.810433e+00 & 0.97013423 & 84\\\\\n",
       "\t M062\\_0.6 & SART3   &  1.4394296 & 1.547272e-02 & 1.00000000 &  1.810433e+00 & 0.97013423 & 83\\\\\n",
       "\t M036\\_0.6 & MATR3   &  1.3222426 & 2.055724e-02 & 1.00000000 &  1.687035e+00 & 0.89522058 & 82\\\\\n",
       "\t M077\\_0.6 & U2AF2   &  1.3923054 & 2.121427e-02 & 1.00000000 &  1.673372e+00 & 0.89522058 & 81\\\\\n",
       "\t M353\\_0.6 & SRSF7   &  1.4638267 & 4.186998e-02 & 1.00000000 &  1.378097e+00 & 0.62797468 & 80\\\\\n",
       "\t M234\\_0.6 & RBM47   &  1.2378816 & 6.468847e-02 & 1.00000000 &  1.189173e+00 & 0.53536903 & 79\\\\\n",
       "\t M142\\_0.6 & RBM42   &  1.1245544 & 3.833472e-01 & 1.00000000 &  4.164077e-01 & 0.07167153 & 77\\\\\n",
       "\t M012\\_0.6 & CPEB2   &  1.1891839 & 3.996302e-01 & 1.00000000 &  3.983417e-01 & 0.06734851 & 76\\\\\n",
       "\t M152\\_0.6 & FXR1    &  1.1387892 & 4.100305e-01 & 1.00000000 &  3.871838e-01 & 0.06640979 & 75\\\\\n",
       "\t M348\\_0.6 & SNRPA   &  1.1061190 & 4.923351e-01 & 1.00000000 &  3.077392e-01 & 0.04369279 & 74\\\\\n",
       "\t M027\\_0.6 & HNRNPL  &  1.0764217 & 5.967982e-01 & 1.00000000 &  2.241725e-01 & 0.01024884 & 72\\\\\n",
       "\t M074\\_0.6 & TARDBP  &  1.0564499 & 6.945445e-01 & 1.00000000 &  1.582999e-01 & 0.00000000 & 71\\\\\n",
       "\t M242\\_0.6 & HNRNPR  &  1.0724227 & 7.348285e-01 & 1.00000000 &  1.338140e-01 & 0.00000000 & 70\\\\\n",
       "\t M042\\_0.6 & PABPC4  &  1.0744150 & 7.393874e-01 & 1.00000000 &  1.311280e-01 & 0.00000000 & 69\\\\\n",
       "\t M056\\_0.6 & SRSF3   &  1.0267670 & 8.637077e-01 & 1.00000000 &  6.363320e-02 & 0.00000000 & 68\\\\\n",
       "\t M332\\_0.6 & SRSF2   &  1.0232912 & 8.935065e-01 & 1.00000000 &  4.890229e-02 & 0.00000000 & 67\\\\\n",
       "\t M155\\_0.6 & ZNF638  &  1.0160066 & 9.296218e-01 & 1.00000000 &  3.169369e-02 & 0.00000000 & 65\\\\\n",
       "\t M068\\_0.6 & SNRNP70 &  1.0105200 & 9.358803e-01 & 1.00000000 &  2.877968e-02 & 0.00000000 & 64\\\\\n",
       "\t M354\\_0.6 & YTHDC1  &  0.9072035 & 1.000000e+00 & 1.00000000 &  0.000000e+00 & 0.00000000 & 63\\\\\n",
       "\t M344\\_0.6 & RBMX    & 13.4946524 & 1.000000e+00 & 1.00000000 &  0.000000e+00 & 0.00000000 & 61\\\\\n",
       "\t M292\\_0.6 & EIF4B   &  0.9778734 & 1.000000e+00 & 1.00000000 &  0.000000e+00 & 0.00000000 & 60\\\\\n",
       "\t M271\\_0.6 & HNRNPA1 &  0.9887171 & 1.000000e+00 & 1.00000000 &  0.000000e+00 & 0.00000000 & 58\\\\\n",
       "\t M262\\_0.6 & QKI     &  0.9943428 & 1.000000e+00 & 1.00000000 &  0.000000e+00 & 0.00000000 & 57\\\\\n",
       "\t M256\\_0.6 & ACO1    & 13.4946524 & 1.000000e+00 & 1.00000000 &  0.000000e+00 & 0.00000000 & 56\\\\\n",
       "\t M236\\_0.6 & SYNCRIP &  1.0037345 & 1.000000e+00 & 1.00000000 &  0.000000e+00 & 0.00000000 & 55\\\\\n",
       "\t M048\\_0.6 & CIRBP   &  0.9710792 & 1.000000e+00 & 1.00000000 &  0.000000e+00 & 0.00000000 & 53\\\\\n",
       "\t M228\\_0.6 & PTBP1   &  0.9673586 & 1.000000e+00 & 1.00000000 & -4.821637e-17 & 0.00000000 & 52\\\\\n",
       "\t M002\\_0.6 & ANKHD1  &  0.9493725 & 8.425994e-01 & 1.00000000 & -7.437888e-02 & 0.00000000 & 46\\\\\n",
       "\t M160\\_0.6 & KHDRBS1 &  0.9628923 & 7.841394e-01 & 1.00000000 & -1.056067e-01 & 0.00000000 & 43\\\\\n",
       "\t M273\\_0.6 & SRSF1   &  0.8687544 & 7.002369e-01 & 1.00000000 & -1.547550e-01 & 0.00000000 & 42\\\\\n",
       "\t M053\\_0.6 & RBM5    &  0.9306657 & 6.916618e-01 & 1.00000000 & -1.601062e-01 & 0.00000000 & 41\\\\\n",
       "\t M043\\_0.6 & PCBP2   &  0.9371286 & 6.889702e-01 & 1.00000000 & -1.617996e-01 & 0.00000000 & 40\\\\\n",
       "\t M231\\_0.6 & EIF2S1  &  0.9338860 & 6.185624e-01 & 1.00000000 & -2.086165e-01 & 0.01024884 & 38\\\\\n",
       "\t M047\\_0.6 & RBM28   &  0.9118008 & 6.162987e-01 & 1.00000000 & -2.102088e-01 & 0.01024884 & 37\\\\\n",
       "\t M333\\_0.6 & SRSF9   &  0.8178577 & 5.791590e-01 & 1.00000000 & -2.372022e-01 & 0.01024884 & 35\\\\\n",
       "\t M085\\_0.6 & ZCRB1   &  0.9106821 & 5.750795e-01 & 1.00000000 & -2.402721e-01 & 0.01024884 & 34\\\\\n",
       "\t M177\\_0.6 & PCBP1   &  0.9156549 & 5.274031e-01 & 1.00000000 & -2.778573e-01 & 0.02258484 & 33\\\\\n",
       "\t M032\\_0.6 & IGF2BP2 &  0.8480246 & 4.686091e-01 & 1.00000000 & -3.291893e-01 & 0.04369279 & 30\\\\\n",
       "\t M168\\_0.6 & SFPQ    &  0.7774697 & 4.479276e-01 & 1.00000000 & -3.487922e-01 & 0.04369279 & 28\\\\\n",
       "\t M161\\_0.6 & RBM6    &  0.8484662 & 3.016420e-01 & 1.00000000 & -5.205082e-01 & 0.13446743 & 24\\\\\n",
       "\t M316\\_0.6 & FUS     &  0.7828390 & 2.955792e-01 & 1.00000000 & -5.293261e-01 & 0.13446743 & 23\\\\\n",
       "\t M021\\_0.6 & G3BP2   &  0.8217897 & 2.204943e-01 & 1.00000000 & -6.566026e-01 & 0.22087398 & 20\\\\\n",
       "\t M136\\_0.6 & SNRPB2  &  0.7990255 & 1.847532e-01 & 1.00000000 & -7.334079e-01 & 0.27052710 & 18\\\\\n",
       "\t M188\\_0.6 & PCBP4   &  0.7067025 & 1.098550e-01 & 1.00000000 & -9.591801e-01 & 0.42083788 & 13\\\\\n",
       "\t M320\\_0.6 & MBNL1   &  0.6649539 & 1.061240e-01 & 1.00000000 & -9.741864e-01 & 0.42083788 & 12\\\\\n",
       "\t M054\\_0.6 & RBM8A   &  0.7102449 & 1.017316e-01 & 1.00000000 & -9.925440e-01 & 0.42083788 & 11\\\\\n",
       "\t M013\\_0.6 & DAZAP1  &  0.7509299 & 8.263547e-02 & 1.00000000 & -1.082833e+00 & 0.49031880 & 10\\\\\n",
       "\t M109\\_0.6 & RBM4    &  0.6779610 & 7.630974e-02 & 1.00000000 & -1.117420e+00 & 0.49788858 &  9\\\\\n",
       "\t M147\\_0.6 & CNOT4   &  0.7294407 & 5.953726e-02 & 1.00000000 & -1.225211e+00 & 0.53536903 &  7\\\\\n",
       "\t M307\\_0.6 & NONO    &  0.7333078 & 9.606665e-04 & 0.19405463 & -3.017427e+00 & 1.84133610 &  3\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 54 × 8\n",
       "\n",
       "| motif.id &lt;chr&gt; | rbp.name &lt;chr&gt; | fe.ratio &lt;dbl&gt; | pvalue &lt;dbl&gt; | pv.adj &lt;dbl&gt; | y.log &lt;dbl&gt; | adj.y.log &lt;dbl&gt; | order &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| M031_0.6 | ELAVL1  |  1.4829288 | 6.137493e-05 | 0.01239773 |  4.212009e+00 | 2.26405544 | 90 |\n",
       "| M025_0.6 | HNRNPC  |  1.5273317 | 2.243182e-04 | 0.04531227 |  3.649136e+00 | 2.17201430 | 88 |\n",
       "| M150_0.6 | RALY    |  1.5876062 | 2.264581e-03 | 0.45744540 |  2.645012e+00 | 1.64501210 | 86 |\n",
       "| M350_0.6 | ZFP36   |  1.3799526 | 2.709075e-03 | 0.54723310 |  2.567179e+00 | 1.61293650 | 85 |\n",
       "| M146_0.6 | PABPC1  |  1.4394296 | 1.547272e-02 | 1.00000000 |  1.810433e+00 | 0.97013423 | 84 |\n",
       "| M062_0.6 | SART3   |  1.4394296 | 1.547272e-02 | 1.00000000 |  1.810433e+00 | 0.97013423 | 83 |\n",
       "| M036_0.6 | MATR3   |  1.3222426 | 2.055724e-02 | 1.00000000 |  1.687035e+00 | 0.89522058 | 82 |\n",
       "| M077_0.6 | U2AF2   |  1.3923054 | 2.121427e-02 | 1.00000000 |  1.673372e+00 | 0.89522058 | 81 |\n",
       "| M353_0.6 | SRSF7   |  1.4638267 | 4.186998e-02 | 1.00000000 |  1.378097e+00 | 0.62797468 | 80 |\n",
       "| M234_0.6 | RBM47   |  1.2378816 | 6.468847e-02 | 1.00000000 |  1.189173e+00 | 0.53536903 | 79 |\n",
       "| M142_0.6 | RBM42   |  1.1245544 | 3.833472e-01 | 1.00000000 |  4.164077e-01 | 0.07167153 | 77 |\n",
       "| M012_0.6 | CPEB2   |  1.1891839 | 3.996302e-01 | 1.00000000 |  3.983417e-01 | 0.06734851 | 76 |\n",
       "| M152_0.6 | FXR1    |  1.1387892 | 4.100305e-01 | 1.00000000 |  3.871838e-01 | 0.06640979 | 75 |\n",
       "| M348_0.6 | SNRPA   |  1.1061190 | 4.923351e-01 | 1.00000000 |  3.077392e-01 | 0.04369279 | 74 |\n",
       "| M027_0.6 | HNRNPL  |  1.0764217 | 5.967982e-01 | 1.00000000 |  2.241725e-01 | 0.01024884 | 72 |\n",
       "| M074_0.6 | TARDBP  |  1.0564499 | 6.945445e-01 | 1.00000000 |  1.582999e-01 | 0.00000000 | 71 |\n",
       "| M242_0.6 | HNRNPR  |  1.0724227 | 7.348285e-01 | 1.00000000 |  1.338140e-01 | 0.00000000 | 70 |\n",
       "| M042_0.6 | PABPC4  |  1.0744150 | 7.393874e-01 | 1.00000000 |  1.311280e-01 | 0.00000000 | 69 |\n",
       "| M056_0.6 | SRSF3   |  1.0267670 | 8.637077e-01 | 1.00000000 |  6.363320e-02 | 0.00000000 | 68 |\n",
       "| M332_0.6 | SRSF2   |  1.0232912 | 8.935065e-01 | 1.00000000 |  4.890229e-02 | 0.00000000 | 67 |\n",
       "| M155_0.6 | ZNF638  |  1.0160066 | 9.296218e-01 | 1.00000000 |  3.169369e-02 | 0.00000000 | 65 |\n",
       "| M068_0.6 | SNRNP70 |  1.0105200 | 9.358803e-01 | 1.00000000 |  2.877968e-02 | 0.00000000 | 64 |\n",
       "| M354_0.6 | YTHDC1  |  0.9072035 | 1.000000e+00 | 1.00000000 |  0.000000e+00 | 0.00000000 | 63 |\n",
       "| M344_0.6 | RBMX    | 13.4946524 | 1.000000e+00 | 1.00000000 |  0.000000e+00 | 0.00000000 | 61 |\n",
       "| M292_0.6 | EIF4B   |  0.9778734 | 1.000000e+00 | 1.00000000 |  0.000000e+00 | 0.00000000 | 60 |\n",
       "| M271_0.6 | HNRNPA1 |  0.9887171 | 1.000000e+00 | 1.00000000 |  0.000000e+00 | 0.00000000 | 58 |\n",
       "| M262_0.6 | QKI     |  0.9943428 | 1.000000e+00 | 1.00000000 |  0.000000e+00 | 0.00000000 | 57 |\n",
       "| M256_0.6 | ACO1    | 13.4946524 | 1.000000e+00 | 1.00000000 |  0.000000e+00 | 0.00000000 | 56 |\n",
       "| M236_0.6 | SYNCRIP |  1.0037345 | 1.000000e+00 | 1.00000000 |  0.000000e+00 | 0.00000000 | 55 |\n",
       "| M048_0.6 | CIRBP   |  0.9710792 | 1.000000e+00 | 1.00000000 |  0.000000e+00 | 0.00000000 | 53 |\n",
       "| M228_0.6 | PTBP1   |  0.9673586 | 1.000000e+00 | 1.00000000 | -4.821637e-17 | 0.00000000 | 52 |\n",
       "| M002_0.6 | ANKHD1  |  0.9493725 | 8.425994e-01 | 1.00000000 | -7.437888e-02 | 0.00000000 | 46 |\n",
       "| M160_0.6 | KHDRBS1 |  0.9628923 | 7.841394e-01 | 1.00000000 | -1.056067e-01 | 0.00000000 | 43 |\n",
       "| M273_0.6 | SRSF1   |  0.8687544 | 7.002369e-01 | 1.00000000 | -1.547550e-01 | 0.00000000 | 42 |\n",
       "| M053_0.6 | RBM5    |  0.9306657 | 6.916618e-01 | 1.00000000 | -1.601062e-01 | 0.00000000 | 41 |\n",
       "| M043_0.6 | PCBP2   |  0.9371286 | 6.889702e-01 | 1.00000000 | -1.617996e-01 | 0.00000000 | 40 |\n",
       "| M231_0.6 | EIF2S1  |  0.9338860 | 6.185624e-01 | 1.00000000 | -2.086165e-01 | 0.01024884 | 38 |\n",
       "| M047_0.6 | RBM28   |  0.9118008 | 6.162987e-01 | 1.00000000 | -2.102088e-01 | 0.01024884 | 37 |\n",
       "| M333_0.6 | SRSF9   |  0.8178577 | 5.791590e-01 | 1.00000000 | -2.372022e-01 | 0.01024884 | 35 |\n",
       "| M085_0.6 | ZCRB1   |  0.9106821 | 5.750795e-01 | 1.00000000 | -2.402721e-01 | 0.01024884 | 34 |\n",
       "| M177_0.6 | PCBP1   |  0.9156549 | 5.274031e-01 | 1.00000000 | -2.778573e-01 | 0.02258484 | 33 |\n",
       "| M032_0.6 | IGF2BP2 |  0.8480246 | 4.686091e-01 | 1.00000000 | -3.291893e-01 | 0.04369279 | 30 |\n",
       "| M168_0.6 | SFPQ    |  0.7774697 | 4.479276e-01 | 1.00000000 | -3.487922e-01 | 0.04369279 | 28 |\n",
       "| M161_0.6 | RBM6    |  0.8484662 | 3.016420e-01 | 1.00000000 | -5.205082e-01 | 0.13446743 | 24 |\n",
       "| M316_0.6 | FUS     |  0.7828390 | 2.955792e-01 | 1.00000000 | -5.293261e-01 | 0.13446743 | 23 |\n",
       "| M021_0.6 | G3BP2   |  0.8217897 | 2.204943e-01 | 1.00000000 | -6.566026e-01 | 0.22087398 | 20 |\n",
       "| M136_0.6 | SNRPB2  |  0.7990255 | 1.847532e-01 | 1.00000000 | -7.334079e-01 | 0.27052710 | 18 |\n",
       "| M188_0.6 | PCBP4   |  0.7067025 | 1.098550e-01 | 1.00000000 | -9.591801e-01 | 0.42083788 | 13 |\n",
       "| M320_0.6 | MBNL1   |  0.6649539 | 1.061240e-01 | 1.00000000 | -9.741864e-01 | 0.42083788 | 12 |\n",
       "| M054_0.6 | RBM8A   |  0.7102449 | 1.017316e-01 | 1.00000000 | -9.925440e-01 | 0.42083788 | 11 |\n",
       "| M013_0.6 | DAZAP1  |  0.7509299 | 8.263547e-02 | 1.00000000 | -1.082833e+00 | 0.49031880 | 10 |\n",
       "| M109_0.6 | RBM4    |  0.6779610 | 7.630974e-02 | 1.00000000 | -1.117420e+00 | 0.49788858 |  9 |\n",
       "| M147_0.6 | CNOT4   |  0.7294407 | 5.953726e-02 | 1.00000000 | -1.225211e+00 | 0.53536903 |  7 |\n",
       "| M307_0.6 | NONO    |  0.7333078 | 9.606665e-04 | 0.19405463 | -3.017427e+00 | 1.84133610 |  3 |\n",
       "\n"
      ],
      "text/plain": [
       "   motif.id rbp.name fe.ratio   pvalue       pv.adj     y.log        \n",
       "1  M031_0.6 ELAVL1    1.4829288 6.137493e-05 0.01239773  4.212009e+00\n",
       "2  M025_0.6 HNRNPC    1.5273317 2.243182e-04 0.04531227  3.649136e+00\n",
       "3  M150_0.6 RALY      1.5876062 2.264581e-03 0.45744540  2.645012e+00\n",
       "4  M350_0.6 ZFP36     1.3799526 2.709075e-03 0.54723310  2.567179e+00\n",
       "5  M146_0.6 PABPC1    1.4394296 1.547272e-02 1.00000000  1.810433e+00\n",
       "6  M062_0.6 SART3     1.4394296 1.547272e-02 1.00000000  1.810433e+00\n",
       "7  M036_0.6 MATR3     1.3222426 2.055724e-02 1.00000000  1.687035e+00\n",
       "8  M077_0.6 U2AF2     1.3923054 2.121427e-02 1.00000000  1.673372e+00\n",
       "9  M353_0.6 SRSF7     1.4638267 4.186998e-02 1.00000000  1.378097e+00\n",
       "10 M234_0.6 RBM47     1.2378816 6.468847e-02 1.00000000  1.189173e+00\n",
       "11 M142_0.6 RBM42     1.1245544 3.833472e-01 1.00000000  4.164077e-01\n",
       "12 M012_0.6 CPEB2     1.1891839 3.996302e-01 1.00000000  3.983417e-01\n",
       "13 M152_0.6 FXR1      1.1387892 4.100305e-01 1.00000000  3.871838e-01\n",
       "14 M348_0.6 SNRPA     1.1061190 4.923351e-01 1.00000000  3.077392e-01\n",
       "15 M027_0.6 HNRNPL    1.0764217 5.967982e-01 1.00000000  2.241725e-01\n",
       "16 M074_0.6 TARDBP    1.0564499 6.945445e-01 1.00000000  1.582999e-01\n",
       "17 M242_0.6 HNRNPR    1.0724227 7.348285e-01 1.00000000  1.338140e-01\n",
       "18 M042_0.6 PABPC4    1.0744150 7.393874e-01 1.00000000  1.311280e-01\n",
       "19 M056_0.6 SRSF3     1.0267670 8.637077e-01 1.00000000  6.363320e-02\n",
       "20 M332_0.6 SRSF2     1.0232912 8.935065e-01 1.00000000  4.890229e-02\n",
       "21 M155_0.6 ZNF638    1.0160066 9.296218e-01 1.00000000  3.169369e-02\n",
       "22 M068_0.6 SNRNP70   1.0105200 9.358803e-01 1.00000000  2.877968e-02\n",
       "23 M354_0.6 YTHDC1    0.9072035 1.000000e+00 1.00000000  0.000000e+00\n",
       "24 M344_0.6 RBMX     13.4946524 1.000000e+00 1.00000000  0.000000e+00\n",
       "25 M292_0.6 EIF4B     0.9778734 1.000000e+00 1.00000000  0.000000e+00\n",
       "26 M271_0.6 HNRNPA1   0.9887171 1.000000e+00 1.00000000  0.000000e+00\n",
       "27 M262_0.6 QKI       0.9943428 1.000000e+00 1.00000000  0.000000e+00\n",
       "28 M256_0.6 ACO1     13.4946524 1.000000e+00 1.00000000  0.000000e+00\n",
       "29 M236_0.6 SYNCRIP   1.0037345 1.000000e+00 1.00000000  0.000000e+00\n",
       "30 M048_0.6 CIRBP     0.9710792 1.000000e+00 1.00000000  0.000000e+00\n",
       "31 M228_0.6 PTBP1     0.9673586 1.000000e+00 1.00000000 -4.821637e-17\n",
       "32 M002_0.6 ANKHD1    0.9493725 8.425994e-01 1.00000000 -7.437888e-02\n",
       "33 M160_0.6 KHDRBS1   0.9628923 7.841394e-01 1.00000000 -1.056067e-01\n",
       "34 M273_0.6 SRSF1     0.8687544 7.002369e-01 1.00000000 -1.547550e-01\n",
       "35 M053_0.6 RBM5      0.9306657 6.916618e-01 1.00000000 -1.601062e-01\n",
       "36 M043_0.6 PCBP2     0.9371286 6.889702e-01 1.00000000 -1.617996e-01\n",
       "37 M231_0.6 EIF2S1    0.9338860 6.185624e-01 1.00000000 -2.086165e-01\n",
       "38 M047_0.6 RBM28     0.9118008 6.162987e-01 1.00000000 -2.102088e-01\n",
       "39 M333_0.6 SRSF9     0.8178577 5.791590e-01 1.00000000 -2.372022e-01\n",
       "40 M085_0.6 ZCRB1     0.9106821 5.750795e-01 1.00000000 -2.402721e-01\n",
       "41 M177_0.6 PCBP1     0.9156549 5.274031e-01 1.00000000 -2.778573e-01\n",
       "42 M032_0.6 IGF2BP2   0.8480246 4.686091e-01 1.00000000 -3.291893e-01\n",
       "43 M168_0.6 SFPQ      0.7774697 4.479276e-01 1.00000000 -3.487922e-01\n",
       "44 M161_0.6 RBM6      0.8484662 3.016420e-01 1.00000000 -5.205082e-01\n",
       "45 M316_0.6 FUS       0.7828390 2.955792e-01 1.00000000 -5.293261e-01\n",
       "46 M021_0.6 G3BP2     0.8217897 2.204943e-01 1.00000000 -6.566026e-01\n",
       "47 M136_0.6 SNRPB2    0.7990255 1.847532e-01 1.00000000 -7.334079e-01\n",
       "48 M188_0.6 PCBP4     0.7067025 1.098550e-01 1.00000000 -9.591801e-01\n",
       "49 M320_0.6 MBNL1     0.6649539 1.061240e-01 1.00000000 -9.741864e-01\n",
       "50 M054_0.6 RBM8A     0.7102449 1.017316e-01 1.00000000 -9.925440e-01\n",
       "51 M013_0.6 DAZAP1    0.7509299 8.263547e-02 1.00000000 -1.082833e+00\n",
       "52 M109_0.6 RBM4      0.6779610 7.630974e-02 1.00000000 -1.117420e+00\n",
       "53 M147_0.6 CNOT4     0.7294407 5.953726e-02 1.00000000 -1.225211e+00\n",
       "54 M307_0.6 NONO      0.7333078 9.606665e-04 0.19405463 -3.017427e+00\n",
       "   adj.y.log  order\n",
       "1  2.26405544 90   \n",
       "2  2.17201430 88   \n",
       "3  1.64501210 86   \n",
       "4  1.61293650 85   \n",
       "5  0.97013423 84   \n",
       "6  0.97013423 83   \n",
       "7  0.89522058 82   \n",
       "8  0.89522058 81   \n",
       "9  0.62797468 80   \n",
       "10 0.53536903 79   \n",
       "11 0.07167153 77   \n",
       "12 0.06734851 76   \n",
       "13 0.06640979 75   \n",
       "14 0.04369279 74   \n",
       "15 0.01024884 72   \n",
       "16 0.00000000 71   \n",
       "17 0.00000000 70   \n",
       "18 0.00000000 69   \n",
       "19 0.00000000 68   \n",
       "20 0.00000000 67   \n",
       "21 0.00000000 65   \n",
       "22 0.00000000 64   \n",
       "23 0.00000000 63   \n",
       "24 0.00000000 61   \n",
       "25 0.00000000 60   \n",
       "26 0.00000000 58   \n",
       "27 0.00000000 57   \n",
       "28 0.00000000 56   \n",
       "29 0.00000000 55   \n",
       "30 0.00000000 53   \n",
       "31 0.00000000 52   \n",
       "32 0.00000000 46   \n",
       "33 0.00000000 43   \n",
       "34 0.00000000 42   \n",
       "35 0.00000000 41   \n",
       "36 0.00000000 40   \n",
       "37 0.01024884 38   \n",
       "38 0.01024884 37   \n",
       "39 0.01024884 35   \n",
       "40 0.01024884 34   \n",
       "41 0.02258484 33   \n",
       "42 0.04369279 30   \n",
       "43 0.04369279 28   \n",
       "44 0.13446743 24   \n",
       "45 0.13446743 23   \n",
       "46 0.22087398 20   \n",
       "47 0.27052710 18   \n",
       "48 0.42083788 13   \n",
       "49 0.42083788 12   \n",
       "50 0.42083788 11   \n",
       "51 0.49031880 10   \n",
       "52 0.49788858  9   \n",
       "53 0.53536903  7   \n",
       "54 1.84133610  3   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.0501187233627272"
      ],
      "text/latex": [
       "0.0501187233627272"
      ],
      "text/markdown": [
       "0.0501187233627272"
      ],
      "text/plain": [
       "[1] 0.05011872"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.698970004336019"
      ],
      "text/latex": [
       "0.698970004336019"
      ],
      "text/markdown": [
       "0.698970004336019"
      ],
      "text/plain": [
       "[1] 0.69897"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "\n",
    "t.xcomp <- rbpbs.enrichmentTest( tfs1, \n",
    "                                 \"up.fa\", \n",
    "                                 \"all.fa\",\n",
    "                                 threshold = 6,\n",
    "                                 gc.correction=F,\n",
    "                                 file.prefix=NA,\n",
    "                                 ncores = 6,\n",
    "                                 #use.cluster = T,\n",
    "                                 background.order = 2,\n",
    "                                 background.length = 1000); \n",
    "tmp=t.xcomp$result\n",
    "tmp=tmp%>%filter(rbp.name %in%unlist(x$hgnc))%>%arrange(motif.id,rbp.name)%>%distinct()\n",
    "\n",
    "\n",
    "df.ret\n",
    "df.ret <- tmp[,c('motif.id','rbp.name', 'fe.ratio', 'pvalue','pv.adj' ) ];\n",
    "\tdf.ret$y.log <- -log10( df.ret$pvalue );\n",
    "\tdf.ret$adj.y.log <- -log10(p.adjust(df.ret$pvalue,method=\"fdr\"))\n",
    "\t#df.ret$adj.y.log <- -log10( df.ret$pv.adj )\n",
    "\t#df.ret$y.log [ is.infinite(df.ret$y.log) | (df.ret$y.log >= options$y.max) ] <- options$y.max;\n",
    "idx.enrich  <- which( df.ret$fe.ratio >= 1 );\n",
    "idx.deple   <- which( df.ret$fe.ratio < 1 )\n",
    "df.ret$y.log[idx.deple] <- - df.ret$y.log[idx.deple]\n",
    "\n",
    "df.ret <- df.ret[order( df.ret$y.log, decreasing = F ),];\n",
    "df.ret$order=1:nrow(df.ret)\n",
    "10^(-1.3)\n",
    "-log10(0.2)\n",
    "pdf(\"RBP_enrichment.Nosample.0h.1h.PALincrease.compared to alln7100_Last500_7nt.pdf\",width=6,height=4.5)\n",
    "ggplot(df.ret,aes(x=order,y=y.log,label=rbp.name))+\n",
    "  #geom_point(aes(colour=fe.ratio))+\n",
    "  #scale_colour_gradientn(colours = yk.col(10)[1:10])\n",
    "  #scale_colour_gradient( low = \"#132B43\", high = \"#56B1F7\",space = \"Lab\", na.value = \"grey50\", guide = \"colourbar\")\n",
    "\n",
    "  geom_point(color = dplyr::case_when(abs(df.ret$adj.y.log) >0.7~ \"#EE7447\", \n",
    "                                      abs(df.ret$adj.y.log)<=0.7 ~ \"black\"), \n",
    "             size = 1, alpha = 0.8) +theme_bw()+\n",
    "  geom_text_repel(data        = subset(df.ret, abs(y.log)>1.5),\n",
    "                  #nudge_y       = -(5 - subset(df.ret, abs(y.log)>2)$y.log),\n",
    "                  #nudge_x       = 150 - subset(df.ret, abs(adj.y.log)> 4)$order,\n",
    "                  #nudge_x = 4.5 - subset(df.ret, abs(adj.y.log)> 4)$y.log,\n",
    "                  size          = 4.5,\n",
    "                  seed=40,\n",
    "                  box.padding   = 1,\n",
    "                  point.padding = 0.2,\n",
    "                  #force.pull   = 1,\n",
    "                  segment.size  = 0.2,\n",
    "                  segment.color = \"grey50\",\n",
    "                  direction     = \"both\")+xlab(\"Rank order\")+labs(title = expression(\"0h\"%->%\"1h\"))+ylab(\"-log10(p-value)\")+theme(legend.position = \"right\")#+xlim(c(0,200))#+ylim(c(-3,16))\n",
    " dev.off() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a2b275b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "a6e699f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "3665"
      ],
      "text/latex": [
       "3665"
      ],
      "text/markdown": [
       "3665"
      ],
      "text/plain": [
       "[1] 3665"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>hgnc</th><th scope=col>newid</th><th scope=col>cluster</th><th scope=col>ALE</th><th scope=col>tandem.new</th><th scope=col>X0</th><th scope=col>X1</th><th scope=col>X2</th><th scope=col>X4</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>NOC2L  </td><td>ENST00000327044.6:944210  </td><td>NOC2L:1:1   </td><td>1</td><td>2</td><td>110.66372</td><td>104.58583</td><td>101.91340</td><td> 95.66622</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ISG15  </td><td>ENST00000379389.4:1014538 </td><td>ISG15:1:1   </td><td>1</td><td>1</td><td> 63.72473</td><td> 61.79076</td><td> 91.34486</td><td>103.15974</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>SDF4   </td><td>ENST00000360001.10:1216879</td><td>SDF4:3:1    </td><td>3</td><td>1</td><td> 81.95110</td><td> 76.16387</td><td> 75.88935</td><td> 76.20183</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>B3GALT6</td><td>ENST00000379198.3:1233725 </td><td>B3GALT6:1:1 </td><td>1</td><td>1</td><td>119.11944</td><td>106.73089</td><td>111.88287</td><td>111.63241</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>B3GALT6</td><td>ENST00000379198.3:1234175 </td><td>B3GALT6:1:2 </td><td>1</td><td>2</td><td> 97.47035</td><td> 93.26731</td><td> 97.35037</td><td> 88.66203</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>B3GALT6</td><td>ENST00000379198.3:1235043 </td><td>B3GALT6:1:NA</td><td>1</td><td>3</td><td>148.36430</td><td>146.46115</td><td>142.44348</td><td>145.24182</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & hgnc & newid & cluster & ALE & tandem.new & X0 & X1 & X2 & X4\\\\\n",
       "  & <chr> & <chr> & <chr> & <int> & <int> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & NOC2L   & ENST00000327044.6:944210   & NOC2L:1:1    & 1 & 2 & 110.66372 & 104.58583 & 101.91340 &  95.66622\\\\\n",
       "\t2 & ISG15   & ENST00000379389.4:1014538  & ISG15:1:1    & 1 & 1 &  63.72473 &  61.79076 &  91.34486 & 103.15974\\\\\n",
       "\t3 & SDF4    & ENST00000360001.10:1216879 & SDF4:3:1     & 3 & 1 &  81.95110 &  76.16387 &  75.88935 &  76.20183\\\\\n",
       "\t4 & B3GALT6 & ENST00000379198.3:1233725  & B3GALT6:1:1  & 1 & 1 & 119.11944 & 106.73089 & 111.88287 & 111.63241\\\\\n",
       "\t5 & B3GALT6 & ENST00000379198.3:1234175  & B3GALT6:1:2  & 1 & 2 &  97.47035 &  93.26731 &  97.35037 &  88.66203\\\\\n",
       "\t6 & B3GALT6 & ENST00000379198.3:1235043  & B3GALT6:1:NA & 1 & 3 & 148.36430 & 146.46115 & 142.44348 & 145.24182\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | hgnc &lt;chr&gt; | newid &lt;chr&gt; | cluster &lt;chr&gt; | ALE &lt;int&gt; | tandem.new &lt;int&gt; | X0 &lt;dbl&gt; | X1 &lt;dbl&gt; | X2 &lt;dbl&gt; | X4 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | NOC2L   | ENST00000327044.6:944210   | NOC2L:1:1    | 1 | 2 | 110.66372 | 104.58583 | 101.91340 |  95.66622 |\n",
       "| 2 | ISG15   | ENST00000379389.4:1014538  | ISG15:1:1    | 1 | 1 |  63.72473 |  61.79076 |  91.34486 | 103.15974 |\n",
       "| 3 | SDF4    | ENST00000360001.10:1216879 | SDF4:3:1     | 3 | 1 |  81.95110 |  76.16387 |  75.88935 |  76.20183 |\n",
       "| 4 | B3GALT6 | ENST00000379198.3:1233725  | B3GALT6:1:1  | 1 | 1 | 119.11944 | 106.73089 | 111.88287 | 111.63241 |\n",
       "| 5 | B3GALT6 | ENST00000379198.3:1234175  | B3GALT6:1:2  | 1 | 2 |  97.47035 |  93.26731 |  97.35037 |  88.66203 |\n",
       "| 6 | B3GALT6 | ENST00000379198.3:1235043  | B3GALT6:1:NA | 1 | 3 | 148.36430 | 146.46115 | 142.44348 | 145.24182 |\n",
       "\n"
      ],
      "text/plain": [
       "  hgnc    newid                      cluster      ALE tandem.new X0       \n",
       "1 NOC2L   ENST00000327044.6:944210   NOC2L:1:1    1   2          110.66372\n",
       "2 ISG15   ENST00000379389.4:1014538  ISG15:1:1    1   1           63.72473\n",
       "3 SDF4    ENST00000360001.10:1216879 SDF4:3:1     3   1           81.95110\n",
       "4 B3GALT6 ENST00000379198.3:1233725  B3GALT6:1:1  1   1          119.11944\n",
       "5 B3GALT6 ENST00000379198.3:1234175  B3GALT6:1:2  1   2           97.47035\n",
       "6 B3GALT6 ENST00000379198.3:1235043  B3GALT6:1:NA 1   3          148.36430\n",
       "  X1        X2        X4       \n",
       "1 104.58583 101.91340  95.66622\n",
       "2  61.79076  91.34486 103.15974\n",
       "3  76.16387  75.88935  76.20183\n",
       "4 106.73089 111.88287 111.63241\n",
       "5  93.26731  97.35037  88.66203\n",
       "6 146.46115 142.44348 145.24182"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#pos<-read.ms(\"negative.fa\")\n",
    "x<-read.table(\"custom.meanPAL.acrossalltimepoints.txt\",header=T,stringsAsFactors = F)\n",
    "bg=x%>%\n",
    "  #separate(newid,c(\"enst\",\"ref\",\"cps\"),sep=\":\")%>%unite(\"newid\",c(enst,cps),sep=\":\")%>%\n",
    "  inner_join(utr.id,by=\"newid\")%>%rename(seq=V2)\n",
    "#bg=unlist(bg$newid)\n",
    "nrow(bg) #6269\n",
    "head(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "0dec3bfa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "50"
      ],
      "text/latex": [
       "50"
      ],
      "text/markdown": [
       "50"
      ],
      "text/plain": [
       "[1] 50"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "3080"
      ],
      "text/latex": [
       "3080"
      ],
      "text/markdown": [
       "3080"
      ],
      "text/plain": [
       "[1] 3080"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "UP=bg%>%filter(X1-X0>(10))%>%dplyr::select(newid2,seq)#%>%mutate(newid2=paste0(\">\",newid2))\n",
    "write.table(UP,\"up.fa\",col.names = F,row.names = F, sep=\"\\n\",quote=F)\n",
    "nrow(UP)\n",
    "\n",
    "CTRL=bg%>%\n",
    "  filter(X1-X0<(0))%>%\n",
    "  dplyr::select(newid2,seq)#%>%mutate(newid2=paste0(\">\",newid2))\n",
    "write.table(CTRL,\"all.fa\",col.names = F,row.names = F, sep=\"\\n\",quote=F)\n",
    "nrow(CTRL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfe67b4d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "81582f2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "t.xcomp <- rbpbs.enrichmentTest( tfs1, \n",
    "                                 \"up.fa\", \n",
    "                                 \"all.fa\",\n",
    "                                 threshold = 7,\n",
    "                                 gc.correction=T,\n",
    "                                 file.prefix=NA,\n",
    "                                 ncores = 6,\n",
    "                                 #use.cluster = T,\n",
    "                                 background.order = 2,\n",
    "                                 background.length = 6000); \n",
    "tmp=t.xcomp$result\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "c18c603e",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp=tmp%>%\n",
    "filter(rbp.name %in%unlist(x$hgnc))%>%\n",
    "arrange(motif.id,rbp.name)%>%distinct()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "18eebb00",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "df.ret <- tmp[,c('motif.id','rbp.name', 'fe.ratio', 'pvalue','pv.adj' ) ];\n",
    "\tdf.ret$y.log <- -log10( df.ret$pvalue );\n",
    "\tdf.ret$adj.y.log <- -log10(p.adjust(df.ret$pvalue,method=\"fdr\"))\n",
    "\t#df.ret$adj.y.log <- -log10( df.ret$pv.adj )\n",
    "\t#df.ret$y.log [ is.infinite(df.ret$y.log) | (df.ret$y.log >= options$y.max) ] <- options$y.max;\n",
    "idx.enrich  <- which( df.ret$fe.ratio >= 1 );\n",
    "idx.deple   <- which( df.ret$fe.ratio < 1 )\n",
    "df.ret$y.log[idx.deple] <- - df.ret$y.log[idx.deple]\n",
    "\n",
    "df.ret <- df.ret[order( df.ret$y.log, decreasing = F ),];\n",
    "df.ret$order=1:nrow(df.ret)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "09ae2222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "-0.698970004336019"
      ],
      "text/latex": [
       "-0.698970004336019"
      ],
      "text/markdown": [
       "-0.698970004336019"
      ],
      "text/plain": [
       "[1] -0.69897"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "log10(0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "c0d944db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>null device:</strong> 1"
      ],
      "text/latex": [
       "\\textbf{null device:} 1"
      ],
      "text/markdown": [
       "**null device:** 1"
      ],
      "text/plain": [
       "null device \n",
       "          1 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAYAAAD958/bAAAEGWlDQ1BrQ0dDb2xvclNwYWNl\nR2VuZXJpY1JHQgAAOI2NVV1oHFUUPrtzZyMkzlNsNIV0qD8NJQ2TVjShtLp/3d02bpZJNtoi\n6GT27s6Yyc44M7v9oU9FUHwx6psUxL+3gCAo9Q/bPrQvlQol2tQgKD60+INQ6Ium65k7M5lp\nurHeZe58853vnnvuuWfvBei5qliWkRQBFpquLRcy4nOHj4g9K5CEh6AXBqFXUR0rXalMAjZP\nC3e1W99Dwntf2dXd/p+tt0YdFSBxH2Kz5qgLiI8B8KdVy3YBevqRHz/qWh72Yui3MUDEL3q4\n4WPXw3M+fo1pZuQs4tOIBVVTaoiXEI/MxfhGDPsxsNZfoE1q66ro5aJim3XdoLFw72H+n23B\naIXzbcOnz5mfPoTvYVz7KzUl5+FRxEuqkp9G/Ajia219thzg25abkRE/BpDc3pqvphHvRFys\n2weqvp+krbWKIX7nhDbzLOItiM8358pTwdirqpPFnMF2xLc1WvLyOwTAibpbmvHHcvttU57y\n5+XqNZrLe3lE/Pq8eUj2fXKfOe3pfOjzhJYtB/yll5SDFcSDiH+hRkH25+L+sdxKEAMZahrl\nSX8ukqMOWy/jXW2m6M9LDBc31B9LFuv6gVKg/0Szi3KAr1kGq1GMjU/aLbnq6/lRxc4XfJ98\nhTargX++DbMJBSiYMIe9Ck1YAxFkKEAG3xbYaKmDDgYyFK0UGYpfoWYXG+fAPPI6tJnNwb7C\nlP7IyF+D+bjOtCpkhz6CFrIa/I6sFtNl8auFXGMTP34sNwI/JhkgEtmDz14ySfaRcTIBInmK\nPE32kxyyE2Tv+thKbEVePDfW/byMM1Kmm0XdObS7oGD/MypMXFPXrCwOtoYjyyn7BV29/MZf\nsVzpLDdRtuIZnbpXzvlf+ev8MvYr/Gqk4H/kV/G3csdazLuyTMPsbFhzd1UabQbjFvDRmcWJ\nxR3zcfHkVw9GfpbJmeev9F08WW8uDkaslwX6avlWGU6NRKz0g/SHtCy9J30o/ca9zX3Kfc19\nzn3BXQKRO8ud477hLnAfc1/G9mrzGlrfexZ5GLdn6ZZrrEohI2wVHhZywjbhUWEy8icMCGNC\nUdiBlq3r+xafL549HQ5jH+an+1y+LlYBifuxAvRN/lVVVOlwlCkdVm9NOL5BE4wkQ2SMlDZU\n97hX86EilU/lUmkQUztTE6mx1EEPh7OmdqBtAvv8HdWpbrJS6tJj3n0CWdM6busNzRV3S9KT\nYhqvNiqWmuroiKgYhshMjmhTh9ptWhsF7970j/SbMrsPE1suR5z7DMC+P/Hs+y7ijrQAlhyA\ngccjbhjPygfeBTjzhNqy28EdkUh8C+DU9+z2v/oyeH791OncxHOs5y2AtTc7nb/f73TWPkD/\nqwBnjX8BoJ98VQNcC+8AAEAASURBVHgB7N0HfFRV+v/xJ4U0aqiCFFEURVFABQsIFpQFFRRF\nURH96dp7WVaxo+yi2Fi7YvujiF1si8qCyiLNRWmKUhQB6T0BUv/5HpxxJgkkM7lJpnyOrzG3\nnnvP+15m5plTbkJhUTISAggggAACCCCAAAIIIICAJWKAAAIIIIAAAggggAACCCCwS4AAiTsB\nAQQQQAABBBBAAAEEEPhDgACJWwEBBBBAAAEEEEAAAQQQ+EOAAIlbAQEEEEAAAQQQQAABBBD4\nQ4AAiVsBAQQQQAABBBBAAAEEEPhDgACJWwEBBBBAAAEEEEAAAQQQ+EOAAIlbAQEEEEAAAQQQ\nQAABBBD4Q4AAiVsBAQQQQAABBBBAAAEEEPhDIKYDpFWrVtnf/vY369mzp/Xp08eGDx9uixcv\nLnHxn376aTv++OPt008/LbEuWhZ88MEH1qFDB/vxxx9LnPL27dtd+c4555wS61iAAAIIIIAA\nAggggAACfwok/zkZW1PLly+39u3b26ZNm6xJkyaucJ988ok98sgjLhA68sgj/QVW0DR58mQb\nNGiQf1k0TcyfP98uuOAC27Ztm2VnZ5c49fz8fFe+Fi1alFjHAgQQQAABBBBAAAEEEPhTIGZr\nkK644grbunWrvfjii6aaJL1ef/1127Jli51yyikumPiTIXqnpkyZYr17946Z8kTvleDMEUAA\nAQQQQAABBGJBICYDpHHjxtnHH39sp512ml188cX+6zRw4EC7+uqrbePGjfbaa6/5l0fjhGqL\nrrnmGjvuuONMtWV169aNxmJwzggggAACCCCAAAIIRJRATAZIb775pkMurcnc+eef79a98MIL\npV6IDRs22PPPP2+qgbr77rttwoQJpiZqkZbUt+rJJ5+0vffe2yZNmmRHHHFEuU/xq6++snvv\nvdcuv/xyk8Mvv/xS7n3ZEAEEEEAAAQQQQACBWBaIyT5IM2bMcNesS5cuJa7d4YcfbomJiTZn\nzhzLy8uz5OQ/CWbNmuUCh2XLlgXtp/49r7zyitsvaEUFZxSM1a9fP6xcMjMz7Z///KerEatV\nq1a58lB5r7vuOvvXv/7l3/65555ztU9ffPFFSEGWPwMmEEAAAQQQQAABBBCIIYGYrEH6/fff\n3SVq0KBBiUuVkJBg9erVs5ycHFu9enXQeo1mt//++7sBDbRO840aNbIxY8ZUSpO8M844ww47\n7DBXY1Xa4ApBJ1ds5oEHHrAhQ4ZYeYMj7S6XZ5991p544gn77bff7JtvvrGzzz7bNm/ebIxw\nVwyYWQQQQAABBBBAAIG4FIi5ACkrK8s1iatZs6alpaWVelF9tTbaNjAdeOCB9v7771v37t2t\ncePGrpmd+i0pzZ07N3BTT6bVFPDcc891w483b97cbrnlFluyZIknee8uk6eeesrVOul4Rx11\nlGleNWpLly614h67y4PlCCCAAAIIIIAAAgjEqsCf7ctipIS+/kK7C45UTN86NTkLTCeddFKJ\nGhmNEDdq1ChbuHBh4KYlptevX28zZ84ssbysBR07dnR9if7973/b6NGj7dFHH3XPbBo/fnxZ\nu4a1vl+/fkH7NWzY0Dp37mzTpk2zn376yXQ+JAQQQAABBBBAAAEE4lUg5gKkOnXqWEZGhnv+\nUWFhoalJXfGkUeyUtG1gOuCAAwJn3bTvGUoaNW5PSTVMaq5WkaRmfwUFBabnNVVGUs1Zac0O\ny1vGyjgn8kQAAQQQQAABBBBAIJIEYi5AEm6zZs1s0aJF7plHpQ1/rcERlNQXKTCpWV64qUeP\nHu65S+Hsr0ElnnnmGdPw5Iceeqhde+214WRT5j4VKV+ZmbMBAggggAACCCCAAAIxIBBzfZB0\nTRQgKen5QMWTmsJt377d9tlnnxLN6YpvW9nzL7/8snXq1Mm6devmBo2YOHGia6Z34YUXVvah\nyR8BBBBAAAEEEEAAAQRKEYjJAOnEE090RX3nnXdKFNm3rGvXriXWVfWCBQsWmEay07DiGilP\ngyaQEEAAAQQQQAABBBBAoPoEYjJAuuqqqyw9Pd0NerBq1Sq/roaz1hDXWnfPPff4l1fXxIMP\nPmh33nmn+foAVdd5cFwEEEAAAQQQQAABBBDYJRCTfZA0Mpv68SgAOfbYY+2yyy6z1NRUFzDN\nmzfPHnroIdtvv/24BxBAAAEEEEAAAQQQQACBIIGYDJBUwhEjRrhnGamm6O9//7srdEpKit14\n443uFaTADAIIIIAAAggggAACCCBQJJBQNBR2YSxLaNhsPd9HD0HVMN61a9eO5eJSNgQQQAAB\nBBBAAAEEEKiAQMwHSBWwYVcEEEAAAQQQQAABBBCIM4GYHKQhzq4hxUUAAQQQQAABBBBAAAGP\nBAiQPIIkGwQQQAABBBBAAAEEEIh+AQKk6L+GlAABBBBAAAEEEEAAAQQ8EiBA8giSbBBAAAEE\nEEAAAQQQQCD6BQiQov8aUgIEEEAAAQQQQAABBBDwSIAAySNIskEAAQQQQAABBBBAAIHoFyBA\niv5rSAkQQAABBBBAAAEEEEDAIwECJI8gyQYBBBBAAAEEEEAAAQSiX4AAKfqvISVAAAEEEEAA\nAQQQQAABjwQIkDyCJBsEEEAAAQQQQAABBBCIfgECpOi/hpQAAQQQQAABBBBAAAEEPBIgQPII\nkmwQQAABBBBAAAEEEEAg+gWSo78IwSUYMGCArV69OnhhJcwVFBS4XBMTiTG95JUrpt6JFhYW\nml4JCQnu5V3O8Z2TTJXkSvJGgPdUbxyL58J7anGRis/LlPfUijsG5sB7aqCGN9O+z3++UwV7\n7r333vb6668HLyxlLuYCpDVr1tjkyZMr/YuLgjC9QTZu3LgUVhaFI5Cbm2tbtmyxBg0ahLM7\n+5QisGPHDtu4caPVrl3batWqVcoWLApHYPPmzZaammppaWnh7M4+pQisX7/ecnJyrEmTJvxI\nUopPuIv0WSVTkjcCeXl5tnbtWvdvPzMz05tMycWys7MtPz/ffVbB4Y2Avk9lZWW571QpKSne\nZBrlueh7Zq9evcpVioiv/pg1a5Z98cUX5SoMGyGAAAIIIIAAAggggAACFRGI6ABJv3zdcccd\n9vnnn1ekjOyLAAIIIIAAAggggAACCJRLIGIDJLXxHTZsWKU3lSuXEhshgAACCCCAAAIIIIBA\nXAhEbB+ksWPHuuDohBNOsHXr1pV6MdReffv27UHr1CnN14EyaEUlzPiOVQlZx2WWum6Yenvp\nZaqEq7euPk+fr7e5x3dumHp7/X33qre5xm9u8vQl7lWfRMX/ypV7teKOgTn47lXdp9yru2RC\ncYjIAGnhwoWmAOmFF16wMWPGBF7voGmtGzFiRNCyhg0bulHsqmp0qaoYMS+ogHEwg6n3F3nb\ntm2mF8k7Af04o8EaSN4KqAM8yVsB3lO99VRuGgBHL5K3AhpUgOStgAZqIu0S0CAr5Q2SIi5A\n2rlzp2tad/XVV9tee+21x2vaokULO+6444K2+eGHH9zoUpUdIOk8lTSSFckbAd20unkZbcUb\nT+WiUYFkmpSUZMnJEffP3buCVnFOGglHQ6fKleSNgFoE6BdP/fuv7Pdvb844OnLRZxWfU95d\nK92julf1779GjRreZRznOemzSrZ8Tnl3I+izX666Txnqe5erPrPL+/kScd+YnnzySWvVqpX9\n5S9/KfMu6dmzp+kVmHr06GH169cvN0DgvqFM6xc5IetYJG8EfMN8Y+qNp3LxDfOdkZHBMN/e\nsbqaI4b59hC0KCvfMN8aOpkPc+9s9VnFe6p3nvrSqVpOBfIM8+2dK8N8e2fpy8k3zHedOnX4\n4fkPFH3PjMoASW/k7733nrVv396GDBniirN48WL3a43mb7vtNqtXr57v2vMXAQQQQAABBBBA\nAAEEEPBUIKJqkNLT0+3SSy8NKuCGDRvcg67atWtHdXaQDDMIIIAAAggggAACCCDgtUBEBUiq\nBhw8eHBQGVWVrVfx5UEbMYMAAggggAACCCCAAAIIeCAQsc9B8qBsZIEAAggggAACCCCAAAII\nhCQQUTVIpZ35LbfcUtpiliGAAAIIIIAAAggggAACngtQg+Q5KRkigAACCCCAAAIIIIBAtAoQ\nIEXrleO8EUAAAQQQQAABBBBAwHMBAiTPSckQAQQQQAABBBBAAAEEolWAAClarxznjQACCCCA\nAAIIIIAAAp4LECB5TkqGCCCAAAIIIIAAAgggEK0CBEjReuU4bwQQQAABBBBAAAEEEPBcgADJ\nc1IyRAABBBBAAAEEEEAAgWgVIECK1ivHeSOAAAIIIIAAAggggIDnAgRInpOSIQIIIIAAAggg\ngAACCESrAAFStF45zhsBBBBAAAEEEEAAAQQ8F0j2PEcyRAABBBBAAAEEEEAgDgXWrFlj3333\nndWoUcMOPfRQa9CggV8hOzvbfv75Z/+8JtLT061ly5aWlpYWtFwzBQUFtmjRIvvxxx+tRYsW\n1q5dO0tNTfVvt3PnTrfOvyBgIiEhwRo1ahSwxGzz5s22du1aa9OmTdByZkoKECCVNGEJAggg\ngAACCCCAAALlFsjPz7chQ4bYww8/bHXq1LHCwkLbunWr3XLLLfbPf/7TkpKSbN68edalSxer\nVauWJSbuasS1Y8cOy83Nte7du9tbb71lDRs2dMdcvny59e/f32bMmGFNmza11atXu3zHjh1r\nvXr1ctssXrzYOnToUOo5nnzyyfbyyy/71+Xl5Vm/fv1coPXqq6/6lzNRugBN7Ep3YSkCCCCA\nAAIIIIAAAuUSePPNN23UqFE2efJk27Bhg61fv94+/vhje+SRR2z06NFBecyZM8fV5qhGR0HU\n559/bvPnz7dbb73Vv911111nqgX65ZdfbOXKla7mZ/DgwXbqqae6af+GRRPffvutC7IUaPle\nb7zxhn+TZcuWWe/evd25+RcysUcBAqQ98rASAQQQQAABBBBAAIE9C0yfPt06derkaoJUW6Qm\ndgpKbrrpJlu3bt1ud05JSbETTzzRzjvvPPvmm2/82ym/s88+21q1auWW1a9f3+677z47/vjj\nbeHChf7tNJGcnFzipXNQUpO6gw8+2OrWrWtnnHGGW8b/yhYgQCrbiC0QQAABBBBAAAEEENit\nwAknnOCaww0dOtQ1pfNt+NBDD9ntt9/umy31r2qRpkyZYieddJJ/vQKhRx991F566SVTvyYl\nNd1TbVPXrl3925U1oQDs008/dc33MjMzd7u5+keR/hSgD9KfFkwhgAACCCCAAAIIIBCywOmn\nn27PP/+83XbbbTZ8+HBr0qSJC3iuuuoqO+aYY4Lyu/zyyy0jI8Mt27Rpky1YsMAN6HDjjTf6\nt1Ne2lfbqv+QaoFUI3XzzTdb48aN/dtpQn2VAgdv0MAPEydOdNuo5qg8AdXs2bPthx9+cHkr\nfw3woL+1a9cOOla8zBAgxcuVppwIIIAAAggggAAClSZwySWX2EUXXeRqkhSgvP/++67J3VNP\nPWV//etf/cc97LDDrF69erZixQqbMGGCKTBSUBWYFOSo9uixxx6zSZMmuYDnlVdesddff90+\n++wzO+igg/ybq2+SRrnzJTW5CzUde+yxdsQRR7jaKtVY/fTTT/bf//7XjaQXGDBpWucW6yl0\nwVgXoXwIIIAAAggggAACCIQgoGDo8MMPd4HK0UcfbXrdcccdrg/SvffeGxQgqWaodevWLnfV\nPPXp08f2228/U4ClpBHrvvrqK9cHSTVAGn1Or2HDhrnaoGeffdYFTm7jov8pDw0pHpi2bNkS\nOFuuadVCKdAKDLaysrL8QZMGl1CfJvWvUqDke6m2SU35iqeCDWss67M3LHf5EkvZp61l9DzH\nEuvWL75ZRM4TIEXkZeGkEEAAAQQQQAABBKJFQDVAxx13nI0cOTLolFUzo1Hs1EyutKQhu1WD\npKDpyCOPdIHO9u3bbcCAAaaBGjp37uzfTbVOhxxyiKlZXlWlmjVrumDOF9DpuBp9T7VMes2c\nOdMNQqHtfAFT27ZtrUZ+rm16cqjlFwVJRSNI2PYVSy1n0TzLvPkRS6jx57OcqqocoR6HAClU\nMbZHAAEEEEAAAQQQQCBA4NJLL3VBjvoWnXXWWbbXXnu54bf//ve/u2BnT83e7r//fvvggw9M\neUybNs322Wcf13+pb9++bqAGBVkKsFRLpe3ee++9gCNX/aRqtfTaf//93cH1QNuNGze62iUF\nTQqg6iybbwWb1llinXpF2yS450Llr1lhuT99bykH/xn0Vf3Zl++IBEjlc2IrBBBAAAEEEEAA\nAQRKFbjsssvcwAtqTvfAAw+4vjsamnvQoEElapWKZ5CWlmbPPfecaSQ8PUvphhtusI8++sj0\nLCS91KxNad9997XAB8UWz6e65vXQ2wYNGrjXgQce6E5jx6IcKywKjPSfkp7pVKhXzk43H+n/\nI0CK9CvE+SGAAAIIIIAAAghEvMAFF1xgemnIbD0oNrAvj05ezeUKCwtLLYeG9Q5cp/5A6muk\nl2plVAOlgCswtWvXLmifwHWlTRd/YG1p23i1rEbbDpZQdM4F2zZbQmq6Fe7ItoS0DKvRpr1X\nh6jUfHgOUqXykjkCCCCAAAIIIIBAPAmomV3x4Kgi5VffnuLBUUXyq4p9k+o3tjr/d7sl1W9i\nhTt3WFLj5lb3r3daYm01uYv8RA1S5F8jzhABBBBAAAEEEEAAgagSSCmqRao/9BnXrC4hJfIH\nZgjEpQYpUINpBBBAAAEEEEAAAQQQ8Ewg2oIjFZwAybPLT0YIIIAAAggggAACCCAQ7QIESNF+\nBTl/BBBAAAEEEEAAAQQQ8EyAAMkzSjJCAAEEEEAAAQQQQACBaBcgQIr2K8j5I4AAAggggAAC\nCCCAgGcCBEieUZIRAggggAACCCCAAAIIRLsAAVK0X0HOHwEEEEAAAQQQQAABBDwTIEDyjJKM\nEEAAAQQQQAABBBBAINoFCJCi/Qpy/ggggAACCCCAAAIIIOCZAAGSZ5RkhAACCCCAAAIIIIAA\nAtEuQIAU7VeQ80cAAQQQQAABBBBAAAHPBAiQPKMkIwQQQAABBBBAAAEEEIh2AQKkaL+CnD8C\nCCCAAAIIIIAAAgh4JkCA5BklGSGAAAIIIIAAAggggEC0CxAgRfsV5PwRQAABBBBAAAEEEEDA\nMwECJM8oyQgBBBBAAAEEEEAAAQSiXYAAKdqvIOePAAIIIIAAAggggAACngkQIHlGSUYIIIAA\nAggggAACCCAQ7QIESNF+BTl/BBBAAAEEEEAAAQQQ8EyAAMkzSjJCAAEEEEAAAQQQQACBaBcg\nQIr2K8j5I4AAAggggAACCCCAgGcCBEieUZIRAggggAACCCCAAAIIRLsAAVK0X0HOHwEEEEAA\nAQQQQAABBDwTIEDyjJKMEEAAAQQQQAABBBBAINoFCJCi/Qpy/ggggAACCCCAAAIIIOCZAAGS\nZ5RkhAACCCCAAAIIIIAAAtEuQIAU7VeQ80cAAQQQQAABBBBAAAHPBAiQPKMkIwQQQAABBBBA\nAAEEEIh2AQKkaL+CnD8CCCCAAAIIIIAAAgh4JkCA5BklGSGAAAIIIIAAAggggEC0CxAgRfsV\n5PwRQAABBBBAAAEEEEDAMwECJM8oyQgBBBBAAAEEEEAAAQSiXYAAKdqvIOePAAIIIIAAAggg\ngAACngkQIHlGSUYIIIAAAggggAACCCAQ7QIESNF+BTl/BBBAAAEEEEAAAQQQ8EyAAMkzSjJC\nAAEEEEAAAQQQQACBaBcgQIr2K8j5I4AAAggggAACCCCAgGcCBEieUZIRAggggAACCCCAAAII\nRLsAAVK0X0HOHwEEEEAAAQQQQAABBDwTIEDyjJKMEEAAAQQQQAABBBBAINoFkqO9AMXPv7Cw\n0DZu3GgJCQnFV3k6X1BQ4I6hY5G8EZBpXl6eu37e5Egu+fn5DmH79u2Wm5sLiEcCstRLriRv\nBPRvX2nTpk2V/v7tzRlHRy56X+VzyrtrJU+lnJwcXL1jdZ/9ys73PuBh1nGble8zf+vWrZaY\nSH2I7/5SnFCeFHMBkgqdkZFR6R+wO3fudMfQsUjeCOjLfFZWlrt+3uRILvoQ15tkjRo1LD09\nHRCPBHSfyjQlJcWjHMlG//715bMq3r/jSVufVXxOeXfFdZ/qfTU5ORlX71hN96n+/fM55R1q\ndna26X5NTU11n1fe5VwsJwUclVwpUeyIYc/q+1B5K1BiLkBSwXUzlBcgXGXl7ztWuHmwX7CA\nblz9yqHrR/JGwPdLiT7McfXGVLns2LHDfeBg6p3ptm3bXGYKOvm10ztXPqe8s1ROvhoOPqu8\nddUXed+XeW9zjt/cFHQq6T21Mn7MK8zZaVnjX7QdsyabJSZZetfeltFroCUUTUdqCuWzhTq3\nSL2KnBcCCCCAAAIIIIAAAhEosG38S7Z9yqe7ao/y8yz787ds++QPIvBMwzslAqTw3NgLAQQQ\nQAABBBBAAIG4FMj535eWkFbUpSUlbdff5Bq2c9akmLEgQIqZS0lBEEAAAQQQQAABBBCofIHC\npGQrLNg1EJQ7WmHRACbJsdMvlwCp8u8hjoAAAggggAACCCCAQMwIpHc7zQpzc6wga6sVbNts\nVjRYQ3q3PjFTvpgbpCFmrgwFQQABBBBAAAEEEEAgAgVq9jzbEtMzbPv0iZZQNKprRtEgDamH\n94jAMw3vlAiQwnNjLwQQQAABBBBAAAEE4lOgaDTn9G6nWlrXPpU+cnR1ANPErjrUOSYCCCCA\nAAIIIIAAAlEs8N1339nUqVOjuAS7P3UCpN3bsAYBBBBAAAEEEEAAAQRKEVi7dq01atSolDXR\nv4gAKfqvISVAAAEEEEAAAQQQQKBKBQiQqpSbgyGAAAIIIIAAAggggECkCuTk5FhWVpbVq1cv\nUk+xQudFDVKF+NgZAQQQQAABBBBAAIH4Eli3bp01bNgwJgdo0JUkQIqv+5nSIoAAAggggAAC\nCCBQIQFfgFShTCJ4ZwKkCL44nBoCCCCAAAIIIIAAApEmEMv9j2RNgBRpdxzngwACCCCAAAII\nIIBABAtQgxTBF4dTQwABBBBAAAEEEEAAgaoTyMvLsy1btlj9+vWr7qBVfCRqkKoYnMMhgAAC\nCCCAAAIIIBCtAuvXr7fMzExLTIzdMCJ2Sxatdx3njQACCCCAAAIIIIBAhArEev8jsRMgRejN\nx2khgAACCCCAAAIIIBBpArHe/0jeBEiRdtdxPggggAACCCCAAAIIRKgANUgRemE4LQQQQAAB\nBBBAAAEEEKhagfz8fNu0aZM1aNCgag9cxUejBqmKwTkcAggggAACCCCAAALRKLBhwwarW7eu\nJSUlRePpl/ucCZDKTcWGCCCAAAIIIIAAAgjEr0A89D/S1SVAit97nJIjgAACCCCAAAIIIFBu\ngXjofyQMAqRy3xJsiAACCCCAAAIIIIBA/ApQgxS/156SI4AAAggggAACCCCAQIBAQUGB6SGx\nDRs2DFgam5PUIMXmdaVUCCCAAAIIIIAAAgh4JqDR62rVqmU1atTwLM9IzYgAKVKvDOeFAAII\nIIAAAggggECECKj/UTzUHombAClCbjpOAwEEEEAAAQQQQACBSBVQ/6NGjRpF6ul5el7JnuZG\nZggggAACCCCAAAIIIOAXyMrKskWLFvnnExMTrXbt2tasWTNLSUnxL1+wYIHl5ua6+YSEBEtP\nT7fGjRu75w75NsrOzraff/7ZN+v+aruWLVtaWlpa0PLAmZUrV9r3339vXbp0sfr16weuCpre\nvHmzqaaoTZs2Qcs1owBpn332KbE8FhcQIMXiVaVMCCCAAAIIIIAAAhEhMHv2bOvWrVuJc1Gg\ncuedd9oNN9zg1p100kn2+++/B22nB7L27dvXxowZ4wKmefPmuSBHfYEUaCnt2LHDBVbdu3e3\nt956K6gZ3NSpU+2yyy6zxYsXW6tWrey3336zY445xp555hk76KCDgo6Vl5dn/fr1sxYtWtir\nr74atE4zNLErQcICBBBAAAEEEEAAAQQQCFdg6dKlLpDJyclxtTHXXHON3XTTTTZ9+nR/lg89\n9JDbRjVJW7dutU8++cQmTZrkAin/RkUTc+bMMdX26KXtPv/8c5s/f77deuut/s1UY3T66afb\nOeec4wKvJUuW2K+//mo1a9a0iy66yPLz8/3bLlu2zHr37m2TJ0/2Lwuc0AANqqlKTU0NXByz\n0/RBitlLS8EQQAABBBBAAAEEIkUgOTnZ9NIocA0aNLC77rrLNbULDEpUY+TbTrVEJ598sqtB\nUpC0u6RmeieeeKKdd9559s033/g3Gzp0qPXo0cOGDRvmmupphZr1Pffcc7Zz505T7ZKSaoYO\nPvhg15TvjDPOcMuK/69u3bo2cODA4otjdp4AKWYvLQVDAAEEEEAAAQQQiFSBX375xdX+KPjY\nU5o7d25QP6TStlUt0pQpU0zN9JRUS6X9Bg0aVGLz5s2b23fffedv9qcA69NPP3XN8zIzM0ts\nrwXqE+Vr0lfqBjG2kD5IMXZBKQ4CCCCAAAIIIIBA5AmMHz/e1Rzpgavql+SbP/vss/0nq+Xj\nxo1z8+ov9MUXX9i3337rghf/RkUTl19+uWVkZLhFav6mAR4OPfRQu/HGG90yNcFTM70DDzww\ncLdSpxWgde3atdR18bqQAClerzzlRgABBBBAAAEEEKgygQcffNA1n1NNTJMmTVzzt9tvv90F\nTb6TUJ8jX9O3OnXqWOvWrW3ChAmuqZ1vG/097LDDrF69erZixQq3XoHR8OHD/ZtoXyXVLJFC\nFyBACt2MPRBAAAEEEEAAAQQQCElATeDUvG1PSf2GfLVAe9ruqquucsGTttFADH369LH99tvP\nLrnkErebhunWoAqqWdKodcWT+iF16NDBOnfuXHwV80UC9EHiNkAAAQQQQAABBBBAIEoFevXq\n5YIqBU1qWqekWioFRr4BGQKLpoEc1ERPI+CRShcgQCrdhaUIIIAAAggggAACCESFwP333+8e\nFnvppZea+jgpjRw50pYvX+4f3U59lfQ8JdUynXXWWdazZ8+oKFt1nCQBUnWoc0wEEEAAAQQQ\nQAABBDwSSEtLc7VFM2fOtFGjRrlc1c/pvffecyPQqRmeRqgbPHiwG9L7iSee8OjIsZkNfZBi\n87pSKgQQQAABBBBAAIEIENAIcYWFhWWeycqVK8vcRn2GdpfX8ccf71+3ZcsWl1fHjh3t7bff\ndsv1MFgFTQqmdpdGjx69u1VxtZwAKa4uN4VFAAEEEEAAAQQQiDcBPceoVatW8VbssMtLE7uw\n6dgRAQQQQAABBBBAAAEEYk2AACnWrijlQQABBBBAAAEEEEAAgbAFCJDCpmNHBBBAAAEEEEAA\nAQQQiDUBAqRYu6KUBwEEEEAAAQQQQAABBMIWIEAKm44dEUAAAQQQQAABBBBAINYECJBi7YpS\nHgQQQAABBBBAAAEEEAhbgAApbDp2RAABBBBAAAEEEEAAgVgTIECKtStKeRBAAAEEEEAAAQQQ\nQCBsAQKksOnYEQEEEEAAAQQQQAABBGJNgAAp1q4o5UEAAQQQQAABBBBAAIGwBQiQwqZjRwQQ\nQAABBBBAAAEEEIg1AQKkWLuilAcBBBBAAAEEEEAAAQTCFiBACpuOHRFAAAEEEEAAAQQQQCDW\nBAiQYu2KUh4EEEAAAQQQQAABBBAIW4AAKWw6dkQAAQQQQAABBBBAAIFYEyBAirUrSnkQQAAB\nBBBAAAEEEEAgbAECpLDp2BEBBBBAAAEEEEAAAQRiTYAAKdauKOVBAAEEEEAAAQQQQACBsAUI\nkMKmY0cEEEAAAQQQQAABBBCINQECpFi7opQHAQQQQAABBBBAAAEEwhYgQAqbjh0RQAABBBBA\nAAEEEEAg1gQIkGLtilIeBBBAAAEEEEAAAQQQCFuAAClsOnZEAAEEEEAAAQQQQACBWBMgQIq1\nK0p5EEAAAQQQQAABBBBAIGyB5LD3rOQdFy9ebNOmTbO9997bunTpYunp6ZV8RLJHAAEEEEAA\nAQQQQACBeBeIyBqke+65x2644QZbtmyZPfvss3bNNdfY5s2b4/1aUX4EEEAAAQQQQAABBBCo\nZIGIq0GaO3euTZ482caOHWtNmza1nJwcO+OMM+zTTz+1c889t5I5yB4BBBBAAAEEEEAAAQTi\nWSDiAqSGDRvaiBEjXHCkC5OcnGx16tSxDRs2lLhO27Zts61btwYtLywstPz8fEtISAhaXhkz\nvmNVRt7xmKeuG6beXvmCggKXof7Kl+SNgDwx9cayeC6+94Hiy5kPT4D31PDcdreX7z0V190J\nhbec99Tw3Pa0l+5RJb2n8vm/SyoUh4gLkFRrpJfSokWL7JNPPnHN60455ZRdpQv4/5tvvumC\nqYBFpgBrzZo1VRIg6bg6FslbAUy99VRuWVlZ7uV9zvGb444dO+K38JVY8nXr1lVi7vGZNe+p\n3l/3nTt38vnvPatlZ2dXQq7xneWmTZviGyCg9Hl5ee7HzYBFu52MuADJd6Zr1661q6++2v1j\nOfXUU61Fixa+Vf6/rVu3tl69evnnNfG///3P0tLSKj1A0pcj1VKlpqYGHZ+Z8AX0C5Ju3pSU\nlPAzYc8gAf1akpub62piVRtL8kZApomJiZaUlORNhuTimlPrPUDvqVXRAiBeyPVZpc9EkjcC\nukfV9F///vms8sZUueizSjUefE55Z6rPKbnqPtX9SjL3HbO8ny8JRTfkrjq4CJVTLdKwYcOs\nSZMm9uCDD5Z5lj169LBJkyZV+gfs6tWr3TEaN25c5jmxQfkE9I95y5Yt1qBBg/LtwFZlCujL\n0caNG6127dpWq1atMrdng/IJaNAYfZHni2f5vMqz1fr1690XT73X82FeHrHybaPPKpmSvBHQ\nj3j6AVf/9jMzM73JlFzcj+H6Mq/PKpI3Avo+pdYj+k5FML/LVN8zVbEyceLEMpEjPqRs06aN\nDRgwwGbMmEEToTIvJxsggAACCCCAAAIIIIBARQQiLkAaN26c3XjjjUFlUgSsau3yVosF7cwM\nAggggAACCCCAAAIIIFBOgYgLkNREbvbs2fbhhx+6toLff/+9vfPOO6blGRkZ5SwWmyGAAAII\nIIAAAggggAACoQtEXK9ttZW+/vrrbdSoUfbYY4+5Nuk9e/a0m266KfTSsQcCCCCAAAIIIIAA\nAgggEIJAxAVIOnc9GPa0006zVatWuc5l6enpIRSJTRFAAAEEEEAAAQQQQACB8AQiMkBSUTTU\nY/PmzcMrFXshgAACCCCAAAIIIIAAAmEIRFwfpDDKwC4IIIAAAggggAACCCCAgCcCBEieMJIJ\nAggggAACCCCAAAIIxIIAAVIsXEXKgAACCCCAAAIIIIAAAp4IECB5wkgmCCCAAAIIIIAAAggg\nEAsCBEixcBUpAwIIIIAAAggggAACCHgiQIDkCSOZIIAAAggggAACCCCAQCwIECDFwlWkDAgg\ngAACCCCAAAIIIOCJAAGSJ4xkggACCCCAAAIIIIAAArEgQIAUC1eRMiCAAAIIIIAAAggggIAn\nAgRInjCSCQIIIIAAAggggAACCMSCAAFSLFxFyoAAAggggAACCCCAAAKeCBAgecJIJggggAAC\nCCCAAAIIIBALAgRIsXAVKQMCCCCAAAIIIIAAAgh4IkCA5AkjmSCAAAIIIIAAAggggEAsCBAg\nxcJVpAwIIIAAAggggAACCCDgiQABkieMZIIAAggggAACCCCAAAKxIECAFAtXkTIggAACCCCA\nAAIIIICAJwIESJ4wkgkCCCCAAAIIIIAAAgjEggABUixcRcqAAAIIIIAAAggggAACnggQIHnC\nSCYIIIAAAggggAACCCAQCwIESLFwFSkDAggggAACCCCAAAIIeCJAgOQJI5kggAACCCCAAAII\nIIBALAgQIMXCVaQMCCCAAAIIIIAAAggg4IkAAZInjGSCAAIIIIAAAggggAACsSBAgBQLV5Ey\nIIAAAggggAACCCCAgCcCBEieMJIJAggggAACCCCAAAIIxIIAAVIsXEXKgAACCCCAAAIIIIAA\nAp4IECB5wkgmCCCAAAIIIIAAAgggEAsCBEixcBUpAwIIIIAAAggggAACCHgiQIDkCSOZIIAA\nAggggAACCCCAQCwIECDFwlWkDAgggAACCCCAAAIIIOCJAAGSJ4xkggACCCCAAAIIIIAAArEg\nQIAUC1eRMiCAAAIIIIAAAggggIAnAgRInjCSCQIIIIAAAggggAACCMSCQHIsFIIyIIAAAggg\ngAACCCCAQOkCubm5tmDBgtJXFi098MAD3boff/yx1G3q1KljrVu3tpUrV9ratWv929SoUcMy\nMzOtadOm/mWBE7/99pvVrl3b6tWrF7jYTRcWFtqiRYusWbNmVrNmzRLrq3MBAVJ16nNsBBBA\nAAEEEEAAAQQqWUCBTYcOHXZ7lHnz5llCQsJut+nTp4999NFHNmLECBs1alSJfA4++GB7+eWX\n7YgjjnDrZsyYYddee60tXbrUtmzZYqeddpr9v//3/ywtLc2tnzJlip1zzjlWUFBgWVlZ1rdv\nX3vllVcsMTEyGrdFxlmUYGYBAggggAACCCCAAAIIeCHQqlUrUy1S4GvDhg3Wvn17O+644+yA\nAw7wH+bbb78N2k77fPDBB/71Xbp08a/fvn27qwWqX7++nXXWWf5tFBx169bN1TipFmnmzJn2\nz3/+061XQNSvXz+7+uqr3fq5c+fal19+aW+++aZ//+qeoAapuq8Ax0cAAQQQQAABBBBAoJIF\nkpP//NqvmptBgwbZ5s2b7YsvvjA1lfMlbRe4rW+5769qmnzr9Xe//faz2267zXr37m1Lliyx\nunXrWteuXd0yrW/UqJGb/+6771wWqmlSszrto7wUvL3xxhu+7CPi759SEXE6nAQCCCCAAAII\nIIAAAghUpsBdd91l//nPf0xN3Ro3blzhQ6kWSEnBUYMGDezhhx928wrEdIx33nnHnnzySbds\n9uzZrinf999/b++//74lJSW5YG2fffZx6yPhfwRIkXAVOAcEEEAAAQQQQAABBKpA4N1337Xh\nw4fbmDFjrFOnTiWO2L9/f0tNTfUvT09Pd03kfAvWr19v48aNc7Pbtm2zr7/+2gU6AwYMcMGR\nbzv97dixo82ZM8fVLg0ePNitWr58uWm/U045xU499VQXQI0cOdLlc+ihhwbuXm3TBEjVRs+B\nEUAAAQQQQAABBBCoOgGNZKdA5ZZbbrHzzjuv1ANrfYsWLfzrfM3pfAvUp2jo0KFuNiUlxW2r\n+euuu863if/vhAkT7IcffrAhQ4ZYr1697PPPP3eDMkybNs1+/fVX23vvvU2j2R111FF2xx13\n2Pjx4/37VucEAVJ16nNsBBBAAAEEEEAAAQSqQED9jTQ4wrHHHmv/+Mc/dnvE008/3fZUk6PR\n8L755pvd7h+4Yq+99jK9VGPVs2dPW7Zsmet/1L17dxccaVv1Q1Kt1dNPPx24a7VOM4pdtfJz\ncAQQQAABBBBAAAEEKldAfYFUY6TamrFjx7p+P5V1RA0ZrhqhjRs3+g/hq4VS0zo9c2ndunX+\ndZrQcOD77rtv0LLqnCFAqk59jo0AAggggAACCCCAQCULPPDAA/bZZ5/Z3Xff7Uaa01Dega/A\nYKaip6JnImn4bx1LfxX83Hfffa6/k4KjK664wn766Sc3aIMCNzW30yAOqkWKlEQTu0i5EpwH\nAggggAACCCCAAAKVIDB16lTLy8tzo8WVlr2eQaTAxoukJnPPP/+8DRs2zDWjU9M+NanT4BB6\nEGzTpk1dX6OLL77Ybr/9dvew2CuvvNKuuuoqLw7vSR4ESJ4wkgkCCCCAAAIIIIAAApEp8Omn\nn5brxNQEb0/p8ccf39Nq/7rOnTvbhx9+aHoYrQKmzMxM/zpNqD+SRrNbsWKFG2Y88DlMQRtW\n0wwBUjXBc1gEEEAAAQQQQAABBGJZoH79+nssnkaxi8REH6RIvCqcEwIIIIAAAggggAACESqw\nY8cOV0MUoadX4dMiQKowIRkggAACCCCAAAIIIBA/Aps2bbLc3NyYLTABUsxeWgqGAAIIIIAA\nAggggID3Ahp4oW7dut5nHCE5EiBFyIXgNBBAAAEEEEAAAQQQiAYBAqRouEqcIwIIIIAAAggg\ngAACCFSJAAFSlTBzEAQQQAABBBBAAAEEEIgGAQVIderUiYZTDescaWIXFhs7IYAAAggggAAC\nCCAQnwJbtmyhD1J8XnpKjQACCCCAAAIIIIAAAoECGuJbKS0tLXBxTE1TgxRTl5PCIIAAAggg\ngAACCCBQeQKx3rxOcgRIlXf/kDMCCCCAAAIIIIAAAjElEOvN63SxCJBi6palMAgggAACCCCA\nAAIIVJ5ArI9gJzkCpMq7f8gZAQQQQAABBBBAAIGYEqCJXUxdTgqDAAIIIIAAAggggAACFRGg\niV1F9NgXAQQQQAABBBBAAAEEYkqAJnYxdTkpDAIIIIAAAggggAACCIQrkJOTY3l5eZaRkRFu\nFlGxH32QouIycZIIIIAAAggggAACCFSvQDzUHkk4uXqZvT96QUGBbdiwwRISErzPPCBHHUdJ\nxyJ5IyBT/SqBqTeeyiU/P99llp2dbfrVh+SNQG5urvOUK8kbAZkqbdy4sdLfv7054+jIxfeZ\nGB1nG/lnWVhY6E5S76d8Vnl3vfRZJVvf+4B3OcdvTvo+paT+QomJ3tWHLF++3GrUqBGV979M\nfP+Gy7ozYi5A0k1Qu3btSv+A1ZujgjAdi+SNgG7crKwsTL3hdLns3LnTtm7daqmpqTFfHe4h\nW5lZbdu2zX1AyJXkjYB+ldSXo1q1ann6Ye7N2UVvLjLlc8q766cv8vr8T05OxtU7VtuxY4cp\nmI/1ZlsekpWZlb5Pbd++3ZkqoPEq6btaw4YNo/L+1/theStQYi5A0g2gG6G8AOHeMMpfLy9v\nunDPJZb2w9Tbq+mrQUpKSuJe9ZBWP8ToCxL//r1D9b1ny9TLXzu9O8PozYn71Ltr57tPdY/i\n6p2rr+YIU+9Mfe+jXn9WKfBq0qRJzN//3tW5eXdNyQkBBBBAAAEEEEAAAQQiTCAenoEkcgKk\nCLvxOB0EEEAAAQQQQAABBCJRIF4GaSBAisS7j3NCAAEEEEAAAQQQQCCCBNQUUn2b1Vc01hMB\nUqxfYcqHAAIIIIAAAggggEAFBTQiXrwM+kKAVMGbhd0RQAABBBBAAAEEEIh1AQVIdevWjfVi\nuvIRIMXFZaaQCCCAAAIIIIAAAgiELxAv/Y8kRIAU/n3CnggggAACCCCAAAIIxIVAvIxgp4tJ\ngBQXtzSFRAABBBBAAAEEEEAgfAGa2IVvx54IIIAAAggggAACCCAQYwI0sYuxC0pxEEAAAQQQ\nQAABBBBAIDyB/Px8y8rKioshviVEE7vw7hP2QgABBBBAAAEEEEAgLgR8Q3wnJsZH6BAfpYyL\nW5dCIoAAAggggAACCCDgvUA8Na+THgGS9/cQOSKAAAIIIIAAAgggEDMC8TRAgy4aAVLM3LoU\nBAEEEEAAAQQQQAAB7wWoQfLelBwRQAABBBBAAAEEEEAgSgVUg1SnTp0oPfvQT5sapNDN2AMB\nBBBAAAEEEEAAgbgRoAYpbi41BUUAAQQQQAABBBBAAIE9CRQUFNjWrVupQdoTEusQQAABBBBA\nAAEEEEAgPgS2bdtmNWvWtHgZ4ltXlSZ28XFvU0oEEEAAAQQQQAABBEIWiLfmdQIiQAr5NmEH\nBBBAAAEEEEAAAQTiQ4AAKT6uM6VEAAEEEEAAAQQQQACBcggQIJUDiU0QQAABBBBAAAEEEEAg\nPgQUIMXTEN+6qjSxi497m1IigAACCCCAAAIIIBCygJ6BVLdu3ZD3i+YdCJCi+epx7ggggAAC\nCCCAAAIIVJJAYWGhESBVEi7ZIoAAAggggAACCCCAQHQJaIjvtLQ0S0pKiq4Tr+DZUoNUQUB2\nRwABBBBAAAEEEEAgFgXisfZI1zE5Fi8mZUIAAQQQQAABBBCIT4Hff//d1qxZ4y98QkKC1a9f\n35o1axb0sNOtW7fakiVL/NvVqFHD9t13X1dj4l9YNDFv3jzLz8+3du3ambYJTFqu9XqQaps2\nbQJXuXNIT0+32rVrBy33zegcda7t27cPOi/f+kj4G48j2MmdGqRIuPs4BwQQQAABBBBAAAFP\nBB5++GE7/PDD7bjjjnOvbt26WcuWLS0jI8PuvPNOU78apa+//to6dOjgfx188MEu0OnZs6et\nX7/efy7HHnus2+bjjz/2L/NNTJw40a278MILfYts4cKF1qtXLxdQtW7d2s477zwXCPk3+GPi\n//7v/6xjx442YcKE4qsiZr5x48Z22GGHRcz5VNWJECBVlTTHQQABBBBAAAEEEKgSAQU+qv3w\nvVavXm3333+/e/373/8OOgcFQ7m5uW7bb7/91lasWGE333xz0DYNGjSwt956K2iZZt544w2r\nVatW0PJ//OMfbtny5ctt0aJF9sMPP9itt94atM3KlSvts88+s8GDB9vTTz8dtC4SZgoL8i1n\n3gzLmD/Vaq5fEQmnVKXnQIBUpdwcDAEEEEAAAQQQQKCqBRo1auSCHjWzmzp1atDhk5OTTS89\n60eB1bnnnmuzZs0K2ubMM8+0Dz/80Hbs2OFfnpOTY++//76ddtpp/mWa+Oijj9wyDW5Qr149\n6927twuSAjd65ZVXXM3Mtddea5988oktW7YscHW1Thfm59mWFx6wzS8Ot6zxL9nmp++0rA9e\nrNZzquqDEyBVtTjHQwABBBBAAAEEEKhygdmzZ9uqVatMTeh2l9atW2dqSqegJjCpmZ5qigJr\nnzR90EEH2T777BO4qQuOVCs0f/58mz59uo0dO9b69esXtM1LL73kArFOnTrZAQccYM8991zQ\n+uqcyZk73Xb++D9LrFnbEmvVtYTUdMv+6kPLWxU5QVxl+xAgVbYw+SOAAAIIIIAAAghUqcDi\nxYtdUKLApG/fvnbkkUfaSSedZFdeeaUp2AlMClCaNm3qBnJQTVNBQYGpmVxg0jDXZ599tr35\n5pv+xWpeN3DgQP+8b+Kpp56y7OxsO+SQQ+yoo45yAdAdd9zhW+36Pi1dutTOP/98t+ziiy+2\n0aNHu2Z+/o3+mPjuu+/syy+/NG2vZoBVkfLX/W4JRf9Z4q6hvRNqpBQdNsHy1/5eFYePiGMQ\nIEXEZeAkEEAAAQQQQAABBLwS0IAMnTt3doM1qCncggULXBO5J554wjSqXWAaNWqUPf/88y5I\n+de//uX6Ip188sn+wRx8255zzjn+ZnYKgNQ0TkFTYMrLy3PH3G+//UzBzaRJk1x+aqLnSwqG\nVHOkUezmzJljhx56qKvZUnO94qlt27aumZ5GylOzvPHjx7t8N2zYUHxTz+aTmrYqyqtoIIui\npnZKhTk7isyKhr5u2tLNx8P/GOY7Hq4yZUQAAQQQQAABBOJIQDVCt99+u7/Ef/3rX11N0syZ\nM00jywUmjTin/ke+dOKJJ7oR6NQ8TjVAvnT00UdbZmamffrpp642R7VSTZo08a12fzUy3s8/\n/2wa3U7noDRy5Ejr2rWradCGunXrusEeUlNTXY2W26Dofwro1CyveMClYcI1ipxeqkHSABK/\n/fabC840Gp9G59OrefPmJYYg9+Ud6t/UdkdY6hHH285vJxdVHLm6JMvoM8iSGu4qT6j5ReP2\nBEjReNU4ZwQQQAABBBBAAIFyC6iWaMqUKda/f3+bMWOGG5Rhdzv7gqXAob61rWqeBgwYYG+/\n/bYbrKG05nVr1651z1FSUz1fatiwoZtU/yf1W1JzPQU6Cn586d1333Xn9uOPP9qBBx7oWxz0\nV89g2meffdxLzQQ3btzoBndQ7ZICMg3JrWCpVatW7hhBO4cyU1TOOuddb3nH/sXyN6y2pGb7\nWHKTFqHkEPXbEiBF/SWkAAgggAACCCCAAAJ7ElAwoqZtqsnRc5KGDBni31yDN2gABtXIKMB5\n8MEHXe1P8b5K2kHN7NSXKTEx0eXnz+SPiR49erianHvuucf0UpB19913m5rcqSmdRq0766yz\ngoIj7aqR8BRUPfvss/boo48Wz7bUedVm6RVYu6TR8DTIhJoVKj8FWwqqij/gNnfJAstZMMsS\nUlIttWM3S2rUrMQxklsdYHrFYyJAiserTpkRQAABBBBAAIE4EzjmmGPsiiuusHvvvdfV1viK\nr6BGSUGPamGOOOIIV9Pjq0lyK//4n5rV6ZlICnY0hHfxpP3Vl0gPgX388cdt586d1r59e7dM\nA0dMmzbNHnjggeK7uQBm0KBBptHthg8fXiKAKrFDsQWBtUtapWZ4Op76XmmQB51Xu3btrE2b\nNrZj5n9s6xv/sqLRKFwu2ZPft3pXD7fkvYObHhY7RFzNJhRFy7seJxwjxdZNrg5xxTvgeV08\nPXBMx9ANR/JGQG1rt2zZ4t54vMmRXPS8BlXB165du8SD7NAJX0APHlT7cT3jguSNgH5l1S+e\nas+vLykkbwT0WVW8j4Q3OcdnLuqA72tCpV/uSd4IaMCD/Px891nlTY6RkYseBpuSkmK+JnZV\neVb6PpWVleW+U+n7qpr0admhRcHa+nsutoLsbZaYsesBtwVbNlpqh65WZ3Dww2yr8nyr4lj6\nnqn+ZmqOWFaiBqksIdYjgAACCCCAAAIIIBCigB5KGwnJV7ukcynM2WkF27YUPdvozx/4Eooe\nkpu/YVUknGrEnAM/00XMpeBEEEAAAQQQQAABBBCoPAH1OarRso0Vbt/m+lwVFtXcFeblWsoB\nHSrvoFGYMwFSFF40ThkBBBBAAAEEEEAAgXAEap17rRuUoXDbJivM3mI12na09BP7h5NVzO5D\nE7uYvbQUDAEEEEAAAQQQQACBYAEN2V1/yBOWu3yJG8Uu2T0YNnibeJ8jQIr3O4DyI4AAAggg\ngAACCMSXQFKy1YjTIbzLc6FpYlceJbZBAAEEEEAAAQQQQACBuBAgQIqLy0whEUAAAQQQQAAB\nBBBAoDwCBEjlUWIbBBBAAAEEEEAAAQQQiAsBAqS4uMwUEgEEEEAAAQQQQAABBMojQIBUHiW2\nQQABBBBAAAEEEEAAgbgQIECKi8tMIRFAAAEEEEAAAQQQQKA8AgRI5VFiGwQQQAABBBBAAAEE\nEIgLAQKkuLjMFBIBBBBAAAEEEEAAAQTKI0CAVB4ltkEAAQQQQAABBBBAAIG4ECBAiovLTCER\nQAABBBBAAAEEEECgPAIESOVRYhsEEEAAAQQQQAABBBCICwECpLi4zBQSAQQQQAABBBBAAAEE\nyiNAgFQeJbZBAAEEEEAAAQQQQACBuBAgQIqLy0whEUAAAQQQQAABBBBAoDwCBEjlUWIbBBBA\nAAEEEEAAAQQQiAsBAqS4uMwUEgEEEEAAAQQQQAABBMojQIBUHiW2QQABBBBAAAEEEEAgCgQK\nd2Rb3vTPzCa9Y7nzppsVFkbBWUfWKSZH1ulwNggggAACCCCAAAIIIBCOgIKjjaOGWO7KXy2h\nIN+yZ31uBUvnW+3+V4STXdzuQw1S3F56Co4AAggggAACCCAQSwI7Zk22/FW/mdWsY1arnhWm\nZtiO//7b8tf9HkvFrPSyECBVOjEHQAABBBBAAAEEEECg8gUKNq8valFXYAkJCe5gCck13N+C\nTesq/+AxdAQCpBi6mBQFAQQQQAABBBBAIH4FklseUBQcFX29z8t1CIXZ2ywxNd2Sm+0Tvyhh\nlDxi+yCtXLnSvv76a0tKSrJjjjnGmjVrFkbx2AUBBBBAAAEEEEAAgdgVKCwahOHHH3+01atX\nW48ePSytax/L/upDs4KimqSMWlZr4HVFf2vHLkAllCwiA6Q777zTpk+fbt26dbOlS5fa008/\nbffff78dffTRlUBAlggggAACCCCAAAIIRJ/AqlWrXIVCjRo17MQTT3QFqH3mXy2/Uw/bvnqF\n1W7b3lLrNYi+glXzGUdcgLRw4UL76quv7K233rLGjRs7nnvvvddGjRpFgFRHWgPgAABAAElE\nQVTNNwuHRwABBBBAAAEEEKh+gaysLPvmm2/s999/d9+P27RpE3RSifWbmKXWskRqjoJcyjsT\ncQHSxo0b7ZJLLvEHRypIx44dbfLkyUWdzgr9nc7KW0C2QwABBBBAAAEEEEAgFgTy8/Pt+++/\nd6+DDz7Yunfvbqo9InkrEHEB0lFHHWV6BaaJEyfaQQcdVCI42rRpk61fvz5wUxdE5eXlldg2\naCMPZhSsKelYJG8EZClXTL3xVC56I1UqKGqHjKuj8OR/8pQtpp5wukwC31MTExk/yCtZ3lO9\nktyVj+89FVfvXfmcKtv0119/talTp1rDhg2tX79+Vrv2rn5FpX0WyVNJ63hP3WVbmtOuNSX/\nH3EBUvFTHDdunIuSn3322eKr7N1337URI0YELddNs3bt2koPkHRQvUHqWCRvBTD11lO5qSpe\nL5J3Ajt27PAuM3LyCxT/0cu/gomwBXhPDZtutzvu3LmTz//d6oS/Ijs7O/ydY3jPLVu22OzZ\ns00+nTp1siZNmpg+g8rzObR58+YYlgmtaAqQfIFjWXtGdID04osv2muvvWYPPPCAtW3btkRZ\n9t9/fzvzzDODlk+ZMsXS09MrPUDavn27O0ZaWlrQ8ZkJX0A3bW5urqWmpoafCXsGCejXzpyc\nHEtOTqYKPkimYjMy1QibepG8EdAXTr0H6D3V9/wOb3KO71z0WaXPRJI3ArpHda/q335KSoo3\nmZKLq+XQj840FQu+GfSdSM3plixZYu3bt3ffhctbG6R9FRDoPuWzaperPMr7+RKRAZLegB5+\n+GH74osvbOTIka4PUvAts2tOo9zpFZg0vGG9evXKDRC4byjTeoMUso5F8kZA/5j1Kwmm3ngq\nF/26pC/z+oJUq1Yt7zKO85z0i5wCeX4g8e5GUM2R7tW6devSHMQ7VvdlnvdU70D1BUs1cvoi\nj6t3rqoZ0Q96viZj3uUcvTlp2O5p06ZZ69atbdCgQSH/0KHvU7pfZUowv+s+0PfMqA6Qhg0b\n5iJmDe+97777Ru/dzZkjgAACCCCAAAIIIFBOAT3LyPcc0D59+lijRo3KuSebeSkQcTVIn376\nqas5uvXWW23r1q0uUPIV+JBDDqGa0IfBXwQQQAABBBBAAIGYEFAtmmqMli9f7obtVjcSUvUJ\nRFyA9PbbbzuNhx56qITKhAkTLCMjo8RyFiCAAAIIIIAAAgggEG0Calo4d+5cNwiDRmweOHAg\nfbEi4CJGXIA0evToCGDhFBBAAAEEEEAAAQQQqDyBZcuWmQYXy8zMtP79+1udOnUq72DkHJJA\nxAVIIZ09GyOAAAIIIIAAAgggEEUCGujnv//9r+mvBhtr0aJFFJ19fJwqAVJ8XGdKiQACCCCA\nAAIIIFCNAhqp89tvvzWNUKfnGfXq1YtRO6vxeuzp0ARIe9JhHQIIIIAAAggggAACFRRYuHCh\nTZ8+3dUWnXPOOfSpr6BnZe9OgFTZwuSPAAIIIIAAAgggEJcCa9ascf2MVHjVGDVu3DguHaKt\n0ARI0XbFOF8EEEAAAQQQQACBiBbQsN2qMfrtt9+sS5cu1rZt24g+X04uWIAAKdiDOQQQQAAB\nBBBAAAEEwhIoKChww3b/73//swMPPNDOPfdcS0lJCSsvdqo+AQKk6rPnyAgggAACCCCAAAIx\nIqDaIg3breG6zzzzTKtbt26MlCz+ikGAFH/XnBIjgAACCCCAAAIIeCSwZcsWN2z3xo0b7dhj\nj7VWrVp5lDPZVJdAYnUdmOMigAACCCCAAAIIRLeAhq7+/vvvTcFB8bRixQo3pHXx5ZrXUNfr\n1q1zq7Zu3eqapal5WvE0d+5c0whwSqtWrbKffvqp+CZu/ocffrC1a9e66aysLHdOOi+9fOsK\nCwtL3de3UENwh5Jyc3Nt2rRp9s4771jTpk1No9MRHIUiGLnbEiBF7rXhzBBAAAEEEEAAgYgW\nGDt2rHXs2NFuuummEuc5fPhwa9eunf3nP/8psa5v3742ZswYt1wPTVXNy7Zt24K2u/766+2o\no46yZcuWueWjRo2ys88+O2gb38wpp5xiL7/8spudPXu2dejQwf/SOWj0uPbt29snn3zi2yXo\nr/Lu2rVr0LI9zfz888+msmswBgVGOl5SUtKedmFdFAkQIEXRxeJUEUAAAQQQQACBSBJ48cUX\n7eKLL7Zx48aVWoukWptLLrmkRPBTVhmuu+46U94KaHr27FnW5qWuX7p0qamWR03g5s+f7wZN\n6N+/v5v27aAasLvvvrvUAM+3TeBf1VK99957rmZKQdkJJ5zAM40CgWJkmgApRi4kxUAAAQQQ\nQAABBKpSQLUoX3/9td1xxx3WrFkze+WVV0oc/phjjnFByi233BK0bk/N3RQcvfrqq/bZZ59Z\n9+7dg/YLZSY5Odn0ql27tqvJevPNN61z58527bXX+rNRwPT666/bXXfd5V9W2sT27dvtyy+/\ntI8//tgOOuggO+uss6xJkyalbcqyGBAIO0DKz883/cN4//337ZlnnrF3333XZs6caXl5eTHA\nQhEQQAABBBBAAAEE9iTw0ksv2ZFHHmmtW7e2Cy64wH0fLL59vXr17Pnnn7fnnnvOvvjii+Kr\ng+YVNCk4eu2112zixIl29NFHB62v6ExiYqKddNJJNmPGDPMFaDqeapcUOO0uqZbpjTfecMHW\neeed52qidrcty2NDIKRR7FRFqZtcN646vO3YsaOEQmZmpvXu3dtuvvlm1ya1xAYsQAABBBBA\nAAEEEIhqAf1Qrhoj1R4pDR482O677z7X30jNzgLTX/7yF7voootcU7t58+a5Gp3A9b7pG264\nwX3HTE9PN32fLC3NmTPHEhISSltVrmUK6DSIg/o1aUCF8jTf04///fr12+05levAbBRVAuWq\nQdI/ggcffNBatGhhQ4YMsfr165tu4ieffNI+/PBDF+U//fTTbpmeFjx58mQ74ogj7NJLL7XV\nq1dHFQgniwACCCCAAAIIILBngQkTJrhR5TQAgoIWjUR3yCGHlFqLpJweffRR0yh1xZvaBR5F\nTer0gNUDDjjAPWBV/YeKpzZt2tisWbNKvMrb3M3X0imUh7dmZGQQHBW/EDE+X2YNkjq4acQQ\n1R4pSFJbzYYNG5ZgCfy1QP8A1PRuxIgRtv/++7tfFBRQkRBAAAEEEEAAAQSiX2D06NGWlpYW\nNKqcgg8N363huPfaa6+gQuqhqWqFpFZG+i5ZWlKAdPDBB7vR7Tp16mS33XabjRw5MmhTBSuH\nH3540DLNlDfgUfcQ5VH8/EpkyIK4FiizBkk3kkYfWbBggV1++eWlBkfFBdXGU08Qnj59un3w\nwQf+8euLb8c8AggggAACCCCAQHQJaCQ3tSDSMN1r1qzxv5YvX+6CphdeeKHUAvXq1cuNeKcW\nRmrmVjyppZJS27Zt7eGHH7ZHHnlkt8NyF9+3PPPqS/TUU0+5AK0izfTKcyy2iW6BMmuQTj75\n5AqV8Pjjjze9SAgggAACCCCAAALRL6DASCPD9enTJ6gwqpk599xzXU2Ran9KSwp61BRv5cqV\npa32L7viiivso48+cn2b9LBXjZIXalLTP3X1UJ959TlSdxA90PbOO+8MNSu2jzOBMmuQ9uSh\nG+/tt982tUNV+vXXX/e0OesQQAABBBBAAAEEolxAzycaMGBAqc3a1OpIwcjuHsiqpna7q2Eq\nzqLjqFXS+eef7/ovFV9f1rwCOPWJ79Gjh/3tb39zfemnTJniun+UtS/r41sgoWiYw8JQCdTc\nTpG9xr5X0hOENfyhhnLUcIlDhw611NTUULP1ZHv9I5g0aVKFRjgpz4noFwlVz+rJzCRvBHwP\nc2vQoIE3GZKL+9VMv5bpl75atWoh4pHA5s2b3Xuc2t+TvBFYv369qfmLOlrrCxHJGwF9VpW3\n87o3R4ztXNTHRs3L9G9/d6OsxbZA5ZQuOzvbNCCYPqtI3gho7AA1Y9R3qvL2z/LmyJGbi75n\nqpmnhpAvK5XZxK54BgJXBzsdREN5T5061W2iG1sHHTZsmK1YscLUeY+EAAIIIIAAAggggAAC\nCESTQMg/0+lBX/r19JtvvnEjizRv3tyVNykpydUi3XTTTe7px6V1vosmGM4VAQQQQAABBBBA\nAAEE4k8g5ABp9uzZri1ny5YtS9VS5zxVQf/yyy+lrmchAggggAACCCCAAAIIIBCpAiEHSBqh\nRH2QdpfUjlSJfiS7E2I5AggggAACCCCAAAIIRKpAyAFS586d7aeffrL33nuvRJnUP+nee+91\nQzHyAK4SPCxAAAEEEEAAAQQQQACBCBcIeZCGiy++2NQPSQ+CPfroo01BUXp6uhuCUUHT9u3b\nbdy4cRFebE4PAQQQQAABBBBAAAEEECgpEHKAlJyc7Ma2//vf/24vv/yyf1z6WbNmWdOmTV3w\npLHxSQgggAACCCCAAAIIIIBAtAmEHCCpgI0aNXLDeD/88MP2888/27p162zfffd1rxo1akSb\nAeeLAAIIIIAAAggggAACCDiBsAIkn50eDHvkkUf6ZvmLAAIIIIAAAggggAACCES1QMgB0iOP\nPGKPP/54mYX+9ddfy9yGDRBAAAEEEEAAAQQQQACBSBIIOUBq2LChHXDAAUFlyM/Pt2XLlpmC\noszMTDdgQ9AGzCCAAAIIIIAAAggggAACUSAQcoB04YUXml6lpSVLltgpp5ziBmsobT3LEEAA\nAQQQQAABBBBAAIFIFgj5OUh7KowGahg6dKjdf//9plolEgIIIIAAAggggAACCCAQTQKeBkgq\neIsWLWzr1q1udLtoguBcEUAAAQQQQAABBBBAAAFPA6Ts7Gx76qmnLCkpyVq2bIkuAggggAAC\nCCCAAAIIIBBVAiH3QXr++efdM5CKlzI3N9cN0rB+/Xq76KKLLCMjo/gmzCOAAAIIIIAAAggg\ngAACES0QcoCUk5NjWVlZJQqlWqNDDjnEDdJw/fXXl1jPAgQQQAABBBBAAAEEEEAg0gVCDpCu\nvvpq04uEAAIIIIAAAggggAACCMSagKd9kGINh/IggAACCCCAAAIIIIBAfAkQIMXX9aa0CCCA\nAAIIIIAAAgggsAeBMpvYrVq1yvr167eHLEpfNW3atNJXsBQBBBBAAAEEEEAAAQQQiFCBMgOk\ngoKCUgdliNDycFoIIIAAAggggAACCCDwh4C+yy9atMh+/PFH97zSdu3aWWpqqt/n999/tzVr\n1vjnExISrH79+tasWTNLTCzZ2Gz79u02d+5cW716tR188MHWunVr0z6+tHDhQtuxY4dvNujv\noYceajt37jSNfl08KY9atWoVX1wt82UGSMIRAgkBBBBAAAEEEEAAAQSiR2D58uU2cOBAmzFj\nhjVt2tQFNXXq1LGxY8dar169XEEefvhhe+yxx6xmzZr+gm3dutVSUlLs1ltvtfvuu88fAH3y\nySd2/vnnm9Y3atTI1NKsS5cu9u6777qAShmceeaZtmDBAn9egRN6ZurFF19s48aNC1zspjMz\nM23Dhg0lllfHgpJhYQXPorCw0L7++usK5sLuCCCAAAIIIIAAAgggUBGBm266yQU3v/zyi61c\nudLWrl1rgwcPtlNPPdVN+/Lu0KGDbd682f9S7dD999/vXv/+97/dZtu2bXPBj0azViCjmifV\nSqm2SHkGpttvv93VEqmmKPCVnp5ujz76qKmWyff63//+Zw0bNrQrrrgiMItqnQ4rQHrxxRft\n8MMPd9V0e+21l+nVpEkTa9CggauyO+6446q1UBwcAQQQQAABBBBAAIF4F1DN0dlnn22tWrVy\nFGo6pxqh448/3gUou/NR7dDNN9/saoWmTp3qNvv+++9d87jLLrvMVAul1LZtWxs+fLjVqFHD\n8vLy3DL9T03zkpOTS7y0TjVZBxxwgP81evRoa9mypd17771aHREp5ABJtUOXXnqpzZkzx2Er\nwmzevLmrZlNkKZCnn346IgrHSSCAAAIIIIAAAgggEK8C3bt3dzU2L730kr+fkYKbzz//3Lp2\n7bpHltmzZ7smdD179nTbdezY0erVq+figC+++MIFS1rRu3dvU9M7BUShpkmTJtnzzz9vY8aM\ncUFWqPtX1vYhB0gfffSRC4KWLl1qU6ZMMXX0GjBggM2bN8/mz5/vapKSkpIq63zJFwEEEEAA\nAQQQQAABBMoh8Mwzz5gCnMsvv9y1+Grfvr0NGTLEHyz5sli8eLEbtVojV/ft29eOPPJIO+mk\nk+zKK6+0bt26uc0yMjLcd//169e7PNVn6MQTT7TnnnvO1MUmMD377LN2yCGHBL2++uqrwE3c\ntJrxnXHGGXbQQQeVWFedC0IO9QR49NFHu1ojnbiiSd+Q3m3atLERI0bY9ddfb3/961+rs1wc\nGwEEEEAAAQQQQACBuBZQnx/VHmkQBtXWTJw40V555RV7/fXX7bPPPvMHJgp+OnfubPn5+aYm\ndRpkQeuPPfbYID+NWvftt9/akiVLTLVIqolSn6Tx48fbBx98YL5KkqOOOsr69+8ftO++++4b\nNK+YQuf05ZdfBi2PhJmQAyRFi1u2bPGfu9oeqk+SLx1zzDEuKtWoGWp6R0IAAQQQQAABBBBA\nAIHKFZg+fbobiU41N0oakGHy5MluFLu6dev6a4iGDRvmmteplkeBk5L6BWlgBV9SRYdqkmbO\nnOmG8dZyDaagIb4VNCnYUV8kvRRQaZmOrzhA6bDDDisxcINbEfC/F154wdUw+WqoAlZV+2TI\nTewOPPBA++abb9wwgTp7NbHTyBjLli1zhVEzO/VDUmctEgIIIIAAAggggAACCFS+gMYFUCDk\nSxpd7rzzznNDfPuW6a/6ESmI2rRpU+DioOlRo0a58QVUC+QbfEG1T4MGDTI9VykwaeA21VRt\n3LgxcHGZ0+q2c84555S5XXVsEHKAdOGFFzqE/fff31WJnXDCCS5aFaBGsbjmmmtcEzyNakdC\nAAEEEEAAAQQQQACByhdQ3yCNKO1LLVq0MH1PV03QG2+8Yb/99ptpDAENs63mcOeee65v0xJ/\nFfBodLnvvvvO9JwkpbPOOssFQXp+kvoTqYZK6xU0KTDTscqbFHT99NNPpj5RkZhCDpA07N97\n773n+h4pMlWTO41aJ6ChQ4c6fPVBIiGAAAIIIIAAAggggEDlC2RlZbnnHakvUWB6//337fTT\nT7frrrvODaWtpnFPPPFE0INiA7cPnFZzOT2bSMNvL1q0yDW1UzM6NbNTkNS4cWP3kFgFSlqu\noKq8Sf2PcnJyXE1Wefepyu0SikadCB52IoSja9eEhAS3h6r1NBygOm8pYq2u1KNHD9fhy3de\nlXUeKq+OoZuD5I2AHiSm/m2Bv354k3P85qIfMVTlXbt2batVq1b8Qnhccj1MLzU11dLS0jzO\nOX6z0y+f+rBU6wM10yZ5I6DPKlp0eGOpXPSrt74M6t++fiAmeSOQnZ3tBgfQZxUpPAF1ddFz\nik477TSXgb5PKWjSd6qUlBS3bM2aNW4obj0LqaJJzexUI6W+S778K5pnZe+v75kK7NRUsKwU\n8iANapOoKPLiiy92tUi+A+gNWAclIYAAAggggAACCCCAQNUJbNiwwcoKfLz8UV8/ZPkePlt1\npay6I4X8M51+NVWTuk6dOlmHDh3s8ccft3Xr1lXdGXMkBBBAAAEEEEAAAQQQ8AsU73/kX8FE\nWAIhB0h60NSKFSvcsIB6Yu4NN9xgzZo1c2OdazQK30gXYZ0NOyGAAAIIIIAAAggggEBIAgRI\nIXGVuXHIAZJyVBWdBmKYNWuWe5DULbfc4qbV7lH9j/72t7+VeWA2QAABBBBAAAEEEEAAgYoJ\nqD+Qhuwuq4ldxY4SX3uHFSAFEh100EFueG89SOrSSy+1VatW2UMPPRS4CdMIIIAAAggggAAC\nCCBQCQIKjjTARVJSUiXkHp9ZhjxIQyDTtm3b7N1337UxY8bYf/7zH9OodqeccoobwCFwO6YR\nQAABBBBAAAEEEEDAewGa13lvGnKApD5GEyZMcEHR+PHjTUMztmnTxo2RPnjwYGvevLn3Z0mO\nCCCAAAIIIIAAAgggUEKAAKkESYUXhBwg3X///S4Yqlmzpg0YMMDVFh133HEVPhEyQAABBBBA\nAAEEEEAAgdAENMS3uryQvBMIOUBq166djR492gVHPHjSuwtBTggggAACCCCAAAIIhCpADVKo\nYmVvH3KApFqjwKShvadMmWL33HMPT5UPhGEaAQQQQAABBBBAAIFKFNi5c6fpVadOnUo8Svxl\nXeFR7CZOnGgjRoxwFyf++CgxAggggAACCCCAAALVI0DtUeW4VzhAqpzTIlcEEEAAAQQQQAAB\nBBDYk4D6H/H8oz0Jhbcu5CZ24R2m6vbSw7IUTSckJFTqQXUcJR2L5I2AhonXKImYeuOpXHz3\nqUabVBU8yRsB3ac5OTmWlZXlTYbkYrm5uU5BH/aV/f4dT9y+z8R4KnNlllWfU0r6989nlXfS\n+fn5flfvco2PnJYtW2b16tUrcT/qc0pp8+bNlphIfYgsZOL7N6z5PaWYC5B0E9StW7fSP2DX\nrVvnjqFjkbwR0I2rZ2th6o2nclFQtGXLFtc/MCMjw7uM4zynrVu3WkpKiqWmpsa5hHfF14MO\nFSSpHT0f5t656ks876neeeqLvIL4GjVq0OfDO1bbvn27+0FPIySTQhPQD6Dt27cv8e9c36fk\nKlN9XpHMfcaU9we4CgdIZ511lnsOUnp6esTYJycnV3qAJGC9dCySNwKK6jH1xtKXi+8XJH3h\n5F71qVT8rzz1xHJMK27py8H3oSVTAiSfijd/uU+9cQzMhc+qQI2KT+v9VIl7NXTLjRs3WpMm\nTUrY+d5HZYrrLtfy1h5p67Dr3HSQRYsW2cqVK92vqPPmzXNVzqFfWvZAAAEEEEAAAQQQQACB\nUAR8LUSoIQpFrXzbhlX98csvv9gZZ5xh3333XdBRVIukZyQNHDgwaDkzCCCAAAIIIIAAAggg\n4J2AmtA2aNDAuwzJyS8QcoC0YsUKO/zww12bxkceecQOPfRQU2CkTmKvvvqqnX/++a6j2DXX\nXOM/CBMIIIAAAggggAACCCDgnQABkneWxXMKOUAaP368a1I3c+ZM1+bRl+Exxxxj5557rl11\n1VX26KOPGgGST4a/CCCAAAIIIIAAAgh4K6AAab/99vM2U3JzAiH3Qfrqq6+sd+/eQcFRoOXl\nl19uS5Ysca/A5UwjgAACCCCAAAIIIICANwI8A8kbx9JyCTlA2n///W3hwoWl5eWW/f777260\njL322mu327ACAQQQQAABBBBAAAEEwhfo27cvD4kNn2+Pe4YcIP3f//2fLV682G699dYSD0n8\n4Ycf7LrrrnPN7Hjmyh7dWYkAAggggAACCCCAQEgCGkX6jTfesIsuusiGDh1aYsC0kDJj490K\nhNwHafr06da4cWMbOXKkG7GuXbt2lpmZacuXL7c5c+a4B33pAWqHHXaY/6BqkvePf/zDP88E\nAggggAACCCCAAAIIhCYwYsQIe/rpp91zI/Xg4o8//tjefPNN69ChQ2gZsfUeBUIOkDZv3uye\nyHvkkUe6jHNycmz16tXuqdIa3a60pICJhAACCCCAAAIIIIAAAuEJbN++3V588UU3krSvpZa+\ng7/88sv22GOPhZcpe5UqEHKAdNlll5leJAQQQAABBBBAAAEEEPBWYOXKlbZ27Vpr27atpaWl\n+TPPysoyVUzo8Tq+lJSUZBs3bvTN8tcjgTL7IOn5Rlu3bq3Q4TQkOAkBBBBAAAEEEEAAAQRK\nF1D/ojvvvNP06JzTTjvNjj32WJs2bZp/44YNG7pnkWr0utzcXMvOzjbtc8opp/i3YcIbgTID\npPnz55tGrnvyySdt586d5T6q2kWOGzfOOnXqxDORyq3GhggggAACCCCAAALxKKBnjb766qtW\nu3Zta9Sokek5R3qu6I4dO/wco0aNMnVzUZeXvLw80+N19BxSkrcCZTax+8tf/mJjxoxxzeru\nuusu69evn5199tl2xBFHmCJZX1IEq4EafvzxR/v+++/tmWeesVWrVtmQIUPslltu8W3GXwQQ\nQAABBBBAAAEEECgmMGPGDDfYWWpqqlvToEED189/0aJFdsghh7hle++9t7377ru2bt06Uz8k\nX1+kYlkxW0GBMgMk5X/SSSfZvHnzXNDz+OOPuw5iWq42kM2bN3cXRxdPbSOVEhIS7IILLnAj\n1+lCkhBAAAEEEEAAAQQQQGCXQEFBgQty9PxQVSh069bNmjZt6r5D+4zUcis5OTmoQsK3LrCS\nwreMv94JlCtA0uEUod50003uOUdTpkyxuXPnmprf6aULq+EFFSydeOKJduqpp7qL7N1pkhMC\nCCCAAAIIIIBAtAkoENCP6Gph1KJFC9PjYXw1JCrLihUrbM2aNdamTZsSRdM+CgT0Un+bn3/+\nOWgb/VDfsmXLoIEMtMGCBQtcHx1N60d7badH1NStW1eLSqSffvrJjcbcunXrEuu8WqDmcBpx\nTgGRXpquU6eO+7687777unM855xz7LXXXnMmapmlARguueQS22uvvbw6DfIpp0C5AyRffopk\ne/To4V6+ZfxFAAEEEEAAAQQQQCBQQF0v+vfvb2o6ptoRX1AwduxY69Wrl9t0+PDh9tRTT9mH\nH37ofmAP3L9v37525ZVX2g033OBaMnXp0sVq1apliYm7utCrb44GK+jevbu99dZb/poWtXxS\nEBKYFGwoP3UbCRwFTiPA6dmdNWvWdF1FAkeNC9w/1GnV/viCIf1VfyIFenJo3769nXzyyUGB\novJXvyM56LlGqnxQefv06RPqodneA4GQAyQPjkkWCCCAAAIIIIAAAjEucN1117kanF9++cVa\ntWplGn3tvvvuc4GQggYFBL509dVXux/fFQDtKc2ZM8d8NT0a8vrrr7+2gQMH2q233movvfSS\nf9eHHnrIBVZaoEBq6tSpbjADjRI3cuRI/3aqsTnqqKPs119/dUHWoEGD/OtCmVA3Ew3P7QuK\nNAK0an4UECn/Jk2auBqhsvJUEHXVVVeVtRnrK1kg5ABJVZyq8lTnMN1wirgzMzNdlalXUXcl\nl5nsEUAAAQQQ+P/s3QecFEXax/FngWXJOQkGggoIAgIiomTMCqjoq8LpKWAOZzw99NRTTzBg\nOIxgwICK4qGnKEngMIuI4SQYAEHykuMuMG//S3uc2V3YQM/uhF99PgMz3T3V1d+unelnKjQC\nCCAQY4HPPvvMDc9QcKRUo0YNFyBpeMb8+fPDAZJaSnRbGU3qpUm+/KRuZntLZcuWdUM7zjvv\nPHv//fejNlWLkXo9KSnoUouNWpCmTZsWtZ1uvKpZ4NSS9MQTT1hBAyRtr4BILT0KitSSVb9+\nfRcQNW/e3DTBgt/SFbVDXiSEQIEDJM1Mp4hbTZh5Tfednp5uvXv3Nk0/qApCQgABBBBAAAEE\nEEhdge7du9tDDz3kAiN1FdM4II27mTx5chRKtWrVXAuQZknu16+fmxwsaoO9vFBLjcbGq1td\nfknj5yPHIc2ZM8f00KxwagEaOnSom4lZXe5yJjUMaLyU30qkcVRqHdI1b7t27UzHQEoegQIF\nSGqWPO6441wzqfqStmjRwlV2VXJVTDWZzp0719566y3XkqS+pZoenIQAAggggAACCCCQmgIj\nR4503cV0rx5NUqDrx5NPPtmuv/56FyxFqqiF589//rOblEAzJ+teQHkl5eVPbb1+/Xo3IUOr\nVq3s2muvjdr8q6++cvfj1MIlS5bYlClT7Msvv3Q/9PsbPvPMM65bX8OGDd0i3cJGLVhqScqZ\nPvroI3ftq3uDasyTX4ac2/E6OQQKFCANGjTIVWpF/JGRd04CRd/6heC+++4jQMqJw2sEEEAA\nAQQQQCCFBDQZgsYFPfzww65r29SpU2306NE2ZswYmzRpkqkrWmRSa5Pu96Oudk899VTkqvBz\nte6otUatORMnTnSBkSZ6yJkmTJjgxh1puX7Q17glba9ATEm9oTT+SGOfNK5JqUuXLqagTtex\nOQM0dc8jpY5AvgGSZt1Q65D6du4tOBKZxiOpefLYY491UzaqKZWEAAIIIIAAAgggkFoCmrHu\nv//9r6nbnK4f+/bt6x533XWXu05UAKTAKTJpOwUoamVSj6W8kiYw8Cdp0NAO/TDfpEkT1/IU\nuf2QIUNytSpFrh8/frxt3LjRBWKRwZh+7NdMd5o9j5S6Ar/Nk7iX41f0X5hmRM0osmvXrjzH\nKe1lN6xCAAEEEEAAAQQQSBKBbdu22dlnn+2m+I48JLX+qJVI3ePySpr++8ILLzT1XlKwsrek\nbdW1TkGT3wq0t+0j16l7ncqnezBFPrQscqKIyPfwPHUE8g2QFBypAt58882uD+feaFTBVFE1\nW4luBkZCAAEEEEAAAQQQSD0BjevRxAnqmvbqq6+6a8iFCxe6SRs0Zl0zx+0pDR8+3DSDnSZE\nyC/dfffd7maxCqh0U9qCJM2Yp+5+ec1Yd9FFF7lgS+PvSakrkG8XO9EoOFITqSp7N+8msRpk\np3na1RTqT9Kgqb81AE53LFalIyGAAAIIIIAAAgikrsA777xjuheSHqtXr3YQjRs3tsgbxeal\no+vLUaNGhW8mm9c2/jLdYubpp5+2Hj16uJmUdVPZ/NLzzz/vphjXBGQ5U8+ePd0P/ZqooVOn\nTjlX8zpFBNK8CH3vk8z/DrF582ZTf05Var+SRxopYFIkfs0117iKFbmuOJ8rgNMc9wrUYpnU\nt1b7YJxVcMq6h4D6A+veAaRgBHSvMt2rQYNN87v5XjB7TI1cNmzY4O6Azr3fgjvfGu+qLtq6\nmSL3DgnOVd9VMiUFI6CZ2HQNpL993QOSVHAB9TLSfYl0L6ScSffY1PCMnBMj5NyO1wUX0PWU\nuijqmkr3iyKZu1eVesUVpCGnQC1IQtXF1SOPPOIemjlEX2aq0LoLsuaBL8w4JU4SAggggAAC\nCCCAQOoI8INy6pzrZDjSAgdIkQfboEED04OEAAIIIIAAAggggAACCCSTQL6TNCTTwXIsCCCA\nAAIIIIAAAggggMDeBAiQ9qbDOgQQQAABBBBAAAEEEEgpgXy72P3www92+umnFxrlu+++K/R7\neAMCCCCAAAIIIIAAAgggUJIC+QZImoBBA+s0M5wmYmjVqlVJlpd9I4AAAggggAACCCCAAAIx\nE8g3QNLsdZMmTbJLLrnENG+8bsilOeJJCCCAAAIIIIAAAggggECyCRRoDJLmrdcNs5o1a2YX\nXHCB6d4qJAQQQAABBBBAAAEEEEAg2QQKFCDpoHWTqWeeecaWL19uI0eOTDYHjgcBBBBAAAEE\nEEAAAQQQsHy72EUadezY0d3pOHIZzxFAAAEEEEAAAQQQQACBZBEocAtSshwwx4EAAggggAAC\nCCCAAAII7EmAAGlPMixHAAEEEEAAAQQQQACBlBMoVBc76dx11112zz337BEqLS3NKlasaLVq\n1bLOnTvbsGHDrEaNGnvcnhUIIIAAAggggAACCCAQG4Hdu3fbjz/+aPPmzbMDDjjADjvsMMvI\nyAjvTPMLrFq1Kvxa1/K6dq9fv76VKpW7LWXbtm327bff2sqVK61FixbWqFEj03v8NH/+/D1O\n6KbbBe3YscOys7P9zcP/Kw/Nnh0PqdAB0jHHHGOtW7e2zz//3Nq0aWNt27a18uXL288//2yT\nJ0+29PR069Kli61du9ZN6vDFF1/YlClTXMAUDwdMGRBAAAEEEEAAAQQQSAWBpUuX2rnnnuuu\n23VvUwU1VapUsVdeecVOPPFER/Dggw/aww8/7Bo4fJNNmza5CdpuvPFG+8c//hEOgCZMmGD9\n+/c3ra9du7atWLHCjjrqKHvzzTddQKX3n3HGGfb999/7WUX9v3XrVrvwwgvttddei1quF9Wr\nV3fxQ64VJbAgd1iYTyEUUSpqfOqpp+yrr75yQdCIESNMYFquyO+EE06w6dOn23//+18XOI0e\nPTqfXFmNAAIIIIAAAggggAACQQpcd911LrhZtGiRLVu2zFavXu1u2XPqqae65/6+1OixYcOG\n8EOBlO59qsf777/vNtu8ebMLfq644goXyKjlSa1Suv2PbgMUmf72t7+5ViK1FEU+1Kjy0EMP\nmVqZ/Mfs2bNdQ8qll14amUWJPi90gPTyyy+7VqOLL744V8F1n6Rrr73WFDApHXvssda9e3f7\n5JNPcm3LAgQQQAABBBBAAAEEEIidgHp8nXXWWXbQQQe5naihQy1Cuj5XgLKnpNah66+/3rUK\nffzxx26zr7/+2nWPUwygViilpk2b2j//+U/Xg2znzp1umf5R1zzdRzXnQ+vUknXooYeGH7qN\n0IEHHmh33nmnVsdFKnSApKY0jS/aU6pWrZotWbIkvPqQQw4xNe+REEAAAQQQQAABBBBAoPgE\nunbt6lpsnnvuufA4IwU3Ghajhoy9JfUU03X/cccd5zY74ogjTNf5gwYNcsNnNJZI6eSTT3Y9\nyRQMFTZNmzbN3V/1pZdeckFWYd8fq+0LHSD17NnTpk6dagsWLMhVJjWhPf/8826Mkr9yxowZ\n1q1bN/8l/yOAAAIIIIAAAggggECAArt27YpqoPCzfvLJJ12Ac8kll1i9evXs8MMPt7/+9a/h\nYMnf7qeffrK+ffu6R58+fezII4+0Xr162WWXXeYmXdN2FSpUsA8//NAyMzNdnhozpLjg6aef\ntlAo5Gfl/tdQnJYtW0Y9NPQmZ1IXvtNPP92aN2+ec1WJvi50qHfKKafY7bffbrpp7FVXXeUm\naihbtqwbayQM9UV89913TTNmnHTSSTZr1iy7//77S/Qg2TkCCCCAAAIIIIAAAskkoNnkFi9e\n7B7qraXuc5qlLjJpzI9ajzQJg1pr1MihuQHGjBljkyZNCgcmCn46dOhgCrTUpU6TLGi9JmeL\nTJq17ssvv3TX/ZqETS1RGpP09ttv21tvvWWlS5d2mytOOPPMMyPfao0bN456raBMZVJjSryl\nQgdI6pOooOecc85xfRgjD6hhw4b26quvukkaNBjso48+cv0XNasdCQEEEEAAAQQQQAABBIou\noFmidY2twEjPFRDp+lvX2gqGIpMmZNCkaZrFrmrVquEWIt2yR93r1LChwElJ44I0sYKfBg8e\nbGpJ0mzUmsZbSZMpKChT0KRgR2OR9FBApWWfffaZderUyW2rGa9zTtzgVkT8M2rUKNfCpNsC\nxVsqdICkA1CQpAh0zZo1biY7zZ1+8MEHm/omqjVJSSdMUwBGzovuVvAPAggggAACCCCAAAII\n5CugFh3NPqeASIGRkgKi9u3bW4MGDfK8T5HbyPtHs8udd9551qRJE9c65C/XOCJ1f1u/fr2/\nKNf/jz76qOtOp1YgTfSg8UW69n/iiSfcPZUi74/Url07F5ytW7cuVz57W/DOO++4Bpe9bVNS\n6wo9BskvqGaqmDNnjhuLpAhVEaXmNveTmtgIjnwN/kcAAQQQQAABBBBIdQGN1VGgs7fgRNfU\nGrIyceJE1z1OPbcqVqzoJkMYMGCAa/1RQ0RkkJKXq7bp0aOHawlSDy9NorZw4UI3aYO6w6k3\n2J6SWqM0u5yu9XWfJKV+/fqZgiDdP0njiXT9r/V/+tOfXAuV9lXQpDhC8xloTFQ8piK1IKnv\n4Z///Gf77rvvch2Tpvq75ZZbci1nAQIIIIAAAggggAACqSqgoEeTHmjsTXp6urth6pAhQ1yD\ngiY+8FuJFIT4Xec0C125cuWKTDZ+/Hi74YYb7Oqrrw7f90jd4yJvFLunzNVdTvcm0vTbaklS\nbzF1oxs4cKALkhTIqeeYttPynF389pSvlssgKyvLtWTtbbuSWpfmRbLR007kUxJFvGqWU+Sn\nWTB091zdHFbR8LPPPusGaA0fPtzdDymfrGKyWjPmacBXrFuvdAMt7aNOnToxOY5UzFSzIG7c\nuNFq1qyZiocfk2NW87o+aCtXruz+TmOykxTMVDfTy8jI2KcvrRRk2+sh6+JAX5Z169bN91fR\nvWbEyigBfVfJlBSMgK599Ku5Llg1gxcpGAH1QFJXMn1XJWvS5bZaWBQY6HY5miJbN17VNbNa\nZBTI6F5F6j5Xv379ff4c1PXUli1b3DWVP/xFQ2LUVU6TOexr0mRsapHS2CU//33NM9bv13Wm\nrNVVML9U6BakkSNHumZBDdTSTZ781KpVK+vdu7dpGkH1T9QNY0kIIIAAAggggAACCKS6gFqH\n/OBIXePU2qKx+prBTa0z/fv3jzlRkD/q6xj8m8/GvOAlsINCj0HSXXTVShMZHEWWW7NZ/PDD\nD25AWeRyniOAAAIIIIAAAgggkIoCmkVO3erUUu4ntSpp4jNS/AkUOkDS5AuRJzfnIfnr1FRK\nQgABBBBAAAEEEEAg1QXUJVPj99VqpK5u6v6q2eQ06QIp/gQK3cVO0wpqsJem/NMNpSKTIuH7\n7rvP9a3U4DISAggggAACCCCAAAIImN16663uxqyaAU7jkDT7m3+PIXziS6DQAdKgQYPcgDJ1\ns9NNpBQkValSxU3S8Pzzz7ubSGmyBhICCCCAAAIIIIAAAgj8JqDJvTRVth6k+BYodICkQWUf\nffSRm+JPN5GKTGo+fOyxx9y0hZHLeY4AAggggAACCCCAAAIIJIJAoQMkHZSmH3zvvfds6dKl\nNnfuXNP0rLpLb/PmzZlKOBHOOmVEAAEEEEAAAQQQQACBPAWKFCD5Oe2///6mBwkBBBBAAAEE\nEEAAAQQQSAaBfAOkFStWWN++fQt9rJ9++mmh38MbEEAAAQQQQAABBBBAAIGSFMg3QNKdcnUn\nXhICCCCAAAIIIIAAAgggkOwC+QZIGm/07bffJrsDx4cAAggggAACCCCAAAIIWKFvFJvT7J13\n3rGbb77Ztm/fnnMVrxFAAAEEEEAAAQQQQACBhBLY5wBp6tSpNmzYMNuxY0dCHTiFRQABBBBA\nAAEEEEAAAQRyCuxzgJQzQ14jgAACCCCAAAIIIIAAAokqENcB0q5du2z06NG2cePGRPWl3Agg\ngAACCCCAAAIpIPDFF1/YwIED7bTTTrNHH32U3lUJfM7znaShJI/t8ccft7Fjx9pxxx1nVapU\nKcmisG8EEEAAAQQQQAABBPIU+Oabb6x///4uKCpTpox9/fXX9tNPP9kjjzyS5/YsjG+BfQ6Q\n+vXrZwcffLCVL18+sCNduXKlPfDAAzZ79uzA8iQjBBBAAAEEEEAAAQRiIfDaa6/Ztm3brG7d\nui57jc1/66237LbbbrNatWrFYpfkGUOBfe5id8wxx9gVV1xhZcuWDayYQ4cOtVAo5CZ/CCxT\nMkIAAQQQQAABBBBAIAYCumdoWlpaOOfSpUub7iXKLM9hkoR6ss8tSLE4Wk0brgh88eLFe81+\nzZo1tmLFiqhtFFhlZ2dHVdKoDQJ6of0oaV+kYAR27tzpPkwwDcZTuchUSeP5cHUUgfyjLz3Z\nYhoIp8sk8jO1VKl9/u0uuIIleE7+d2KCH0bcFF+fpUr6DODvP7jTItdE/57ScJDx48e7cfNq\nNFi/fr21bdvWXc+WRF1RHVXSd1Vk4BbcWUu8nApzHuIyQPKbJ/Ojf/vtt3O1MqkZU4FTcVQG\nffFoX6RgBTAN1lO5bd261T2Czzl1c9Svgps2bUpdgBgd+dq1a2OUc+pmy2dq8Oc+KyuL7//g\nWV0XtRhkWyxZKhj6y1/+YiNHjnTfDR07drRbb721xOvJhg0biuX4E2En/g/xBSlrXAZIBSm4\ntmnevLkbEBe5/eTJk61ChQoxD5B0wakgLMixV5HHkYrP9WuHvnTKlSuXiocfk2PWh4FM09PT\n3SMmO0nBTGWq7hN6kIIRUMCpzwB9phbHD1zBlDr+c9F3lb4TScEIqI6qrupvPyMjI5hMycW1\ncuhHZ31XJWpaunSpnXrqqfbnP//Zfe+WdP3Q95SuAVQOvqt+q1WFaU1L6ADp6KOPNj0i0wcf\nfGBVq1aN+ResPiD1Ja59kYIRUNOnpnTHNBhP5aJ66gedlSpVCi7jFM9Jv8jpS4dgPriK4Afz\nmrGULnbBueozgM/U4DxVT2WqC3lcg3NVIK8udpUrVw4u02LMad68efa///3PNHFZvHx+6XpK\n9VXf/UHOE1CMrIHvSteZBf0Bjo7egfOTIQIIIIAAAggggEAqCGzevNk++eQT69mzZ9wER6ng\nHutjJECKtTD5I4AAAggggAACCCSlgHoutWnTxmrWrJmUx5eqB0WAlKpnnuNGAAEEEEAAAQQQ\nKLLAt99+67qxKUAiJZdAXI9BOuigg2zmzJnJJc7RIIAAAggggAACCCS0gMaizpo1y84444wC\nj2tJ6ANOscLTgpRiJ5zDRQABBBBAAAEEECi6gGYznDp1qh155JFM1lF0xrh+JwFSXJ8eCocA\nAggggAACCCAQTwJz5sxxMxm2bNkynopFWQIUIEAKEJOsEEAAAQQQQAABBJJXIDMz077++mvr\n0aNH8h4kR2YESFQCBBBAAAEEEEAAAQTyEdC9mqZMmWKdOnWyihUr5rM1qxNZgAApkc8eZUcA\nAQQQQAABBBAoFoEvvvjCjTlq2rRpseyPnZScAAFSydmzZwQQQAABBBBAAIEEEFixYoXNmzfP\nunbtmgClpYj7KkCAtK+CvB8BBBBAAAEEEEAgaQWys7NNN4RVcFS+fPmkPU4O7A8BAqQ/LHiG\nAAIIIIAAAggggECUwKeffmr16tWzRo0aRS3nRfIKxPWNYpOXnSNDAAEEEEAAAQQQiHeBpUuX\n2qJFi+z//u//7MMPP3Q3h61du7b16dPHKlWqFO/Fp3xFFCBAKiIcb0MAAQQQQAABBBBIXoGs\nrCzXtU5Tej/22GP2yCOPWCgUsrS0NHvuuefsjTfesGrVqiUvQAofGV3sUvjkc+gIIIAAAggg\ngAACeQvMnDnTdavLyMiwESNGuKm91XpUs2ZN++GHH2zs2LF5v5GlCS9AgJTwp5ADQAABBBBA\nAAEEEAhSYOHChbZy5Urr2LGjLV++3DRRgz9Bg1qQdu/ebUuWLAlyl+QVRwIESHF0MigKAggg\ngAACCCCAQMkKbNu2zWbMmGE9e/a09PR0a9KkietKt27dOlcwBUulS5e21q1bl2xB2XvMBAiQ\nYkZLxggggAACCCCAAAKJJjB9+nRr3ry51a1b1xW9YsWKNnz4cFNXu8zMTFu/fr317t3b+vbt\nm2iHRnkLKMAkDQWEYjMEEEAAAQQQQACB5BaYP3++bdq0yY4//vioA+3Vq5d99NFH9u2335rG\nISmAIiWvAAFS8p5bjgwBBBBAAAEEEECggAJbtmyxjz/+2LUOqQtdzlS9enXr0qVLzsW8TkIB\nutgl4UnlkBBAAAEEEEAAAQQKJ/DBBx+4cUWapY6U2gIESKl9/jl6BBBAAAEEEEAg5QU0bXe5\ncuXsiCOOSHkLAMzoYkctQAABBBBAAAEEEEhpgUMOOcT0ICEgAVqQqAcIIIAAAggggAACCCCA\nwO8CBEhUBQQQQAABBBBAAAEEEEDgdwG62FEVEEAAAQQQQAABBEpcYNWqVTZnzhx3c9ZWrVpZ\nzskS/ve//7ll9erViyrr9u3bTdNzH3bYYe69c+fOtaysrPA2mpGuQYMGplnoItOyZcts9erV\n4UW6Kay22W+//cLL9ERTe+/evdstS0tLs/r161utWrWitvFfaL8LFiywRo0ame6fREpMAQKk\nxDxvlBoBBBBAAAEEEEgKgV27dtmQIUPsX//6l1WpUsVCoZC7F9ENN9xgQ4cONX/K7e7du7ug\nQwFLpUqVwseu4KhNmza2dOlSFwideOKJpmCrbNmybhvlrym8FVi98sor1q1bN7d82LBh9uij\nj4bz8Z+0aNHCnn/+eWvfvr1b1LFjR9u6dau/2v2///7725gxY6xz587ute6ddMUVV9i7775r\nZcqUceW/99577Zprrol6Hy8SQ4AudolxniglAggggAACCCCQlALjxo2zJ554wqZPn25r1661\nzMxMF2gMHz7cnnnmmahjXrRokd14441Ry/J6cccdd9iGDRvcY/Pmzfbdd9+5gKdv3762c+fO\n8FuOOuooy87Odo9t27bZjz/+aDVq1LB+/fqFt9GTp556KrzN4sWLTUHYaaed5pZp/ciRI11L\n09dff20rVqywJ5980q699lr74osvtJqUYAIESAl2wiguAggggAACCCCQTAIKItQC1LVrV9da\npK5uJ598sl133XW2Zs2aqEM966yzXLAyderUqOX5vVCr0M033+wCJnXB85O6zKnFRw9N892k\nSRO75ZZbTEHQzz//7G/myuVvc+CBB9r111/v8lJApaRj+Mtf/mJqWVKe559/vtWpU8c+/fTT\ncB48SRwButglzrmipAgggAACCCCAQNIJqMvb448/7rrZnXvuudayZUt3jPfff3+uY1XgpO51\nAwcOdC02lStXzrXNnha89957bvyQxirtLakLn1LVqlXz3EzjkV588UVr3LixNW3a1G2jrnuR\n6fvvv7eVK1fawQcfHLmY5wkiQICUICeKYiKAAAIIIIAAAskocMopp9iIESPszjvvtH/+859W\nt25d69Wrl11++eXWqVOnXIf80EMPmVqENEZJXd/ySi+88IJ98sknbtWOHTtca5AmUNB7/TFN\nWqnufK+99prbTl3xZs6caePHj7ezzz47apIItWbdeuutbrIGdQMsVaqUzZ492/3v3hzxj/Z3\n5ZVXuvFJ6opHSjwBAqTEO2eUGAEEEEAAAQQQSCoBdUm79NJL7fPPPzd1n1OQoi53alkaPHhw\n1LGqZUdjfhRYaayQurLlTAcccIB16NDBTa6g1p1q1arZtGnT3CQQkdsuWbLEtVxpmSZ10Ps0\nYcTVV18duZldcMEFdvzxx7tlmgBCkzEoiJs8eXK4xUsrN27caGeccYYLvCZOnOi620VlxIuE\nEGAMUkKcJgqJAAIIIIAAAggkp8B//vMfNwOdWnaOPvpo11Iza9Ysu+qqq1yrUl5HfdJJJ9mF\nF15ogwYNckFJzm00493f/vY3u/vuu+2zzz5z44XUfc+frtvfXmOfNI5ID3WLU1CjSSAyMjL8\nTdz/rVu3tlNPPdU9LrroItPEEprq+9lnnw1vpy51mtVOkz3MmDHDzZoXXsmThBIgQEqo00Vh\nEUAAAQQQQACB5BK477773Cx2OY/qmGOOcdNlR846F7mNZrlTwHPTTTdFLs71XIGMxgwp+Lnn\nnntyrS/qAo1/Unc7pV9//dUFR4cccohNmTLFtVgVNV/eV/ICdLEr+XNACRBAAAEEEEAAgZQV\nUEuQ7hekbnDqMqf7FX355Zdu1jmNBdLscXklv6udWpPyS2qZUrc5tShpH82bN8/vLVHrNb24\nyqSkFqK3337bjXFSdzylSy65xI1tUrCmlig/6aayOW8866/j//gVoAUpfs8NJUMAAQQQQAAB\nBJJeQF3WdN8gjRU64ogj3CQNAwYMcGOMdH+kvSVNgqAZ7QqS7rrrLhesaEyTbkZbmKTASjeO\n1UNd7TROavTo0a6MP/zwgxuTNG/ePNN9lfzt9L/GSpESTyDvkDzxjoMSI4AAAggggAACCCSo\nwDnnnOMmY9i6daub4ECTJeRMmhwhrzRq1CjTw0+6h1FeqWLFiqaWID898sgj/tO9/r9ly5a9\nrle3usIGXHvNkJUlLkCAVOKngAIggAACCCCAAAIISKBChQruUdwa69atc61AmzZtcmOJ/Hsx\nFXc52F98CBAgxcd5oBQIIIAAAggggAACJSCgCRbOPPNMW7Fihdu7blD74IMP2umnn14CpWGX\n8SDAGKR4OAuUAQEEEEAAAQQQQKBEBHSz2WXLlrlpuzXjnaYb101r9zR7XokUkp0WqwABUrFy\nszMEEEAAAQQQQACBkhDQfYqysrJy7Vr3QFJQ5CeNVVq/fr2p2x0pNQUIkFLzvHPUCCCAAAII\nIIBAUgh8/PHHdtlll9n5559vr7/+eq5j2s8OxgAAQABJREFUmjNnjnXr1s06dOhgbdu2tZdf\nfjlqG802t2vXLvfQCgVGBx10kGtRitqQFykjwBiklDnVHCgCCCCAAAIIIJBcAjNnzjTdR0kt\nQ6VKlbIZM2a4sURXXXWVO1DNijdo0CBbvXq1Va9e3d3DSPcu0sxzCpiULr74YlOQ9cUXX7g8\ntJ3GIKWlpbn1/JN6AgRIqXfOOWIEEEAAAQQQQCApBDS9d3Z2ttWpU8cdz+bNm01jii6//HLX\nbe6bb75xwVHt2rXd+sqVK9v27dtNgZUfIFWqVMnGjh1rs2fPNr1frUxVqlRJCh8OomgCBEhF\nc+NdCCCAAAIIIIAAAsUkoDFBClrUShSZtLxMmT8uZ/VcAZAmWNC4Ir1H9yjS68jtcgZAyldd\n7UgISCC6lmGCAAIIIIAAAggggECcCHz55ZfWo0cPO+KII+yoo46yCRMmRJXslFNOcS1I27Zt\nc93sNmzYYN288UYZGRluu8MOO8y9f+3ate4GtJqoQa1Jffr0icqHFwhECvwRckcu5TkCCCCA\nAAIIIIAAAiUooGBn4MCBbtKEatWquf81tqhx48bWrFkzV7KLLrrITdGtiRfUSqRgaujQoVGl\nfuKJJ+z5559344waNmxogwcPDnfJi9qQFwj8LkCARFVAAAEEEEAAAQQQiDuBWbNmmYIkf/yQ\ngqTMzEw3fsgPkNRt7o477rC//vWvrgWpatWquY6jfPnybpY7zXRHQqAgAgRIBVFiGwQQQAAB\nBBBAAIFiFdD9iHbv3u0e/tgjjSeqUKFCrnIoCNKDhEAQAoxBCkKRPBBAAAEEEEAAAQQCFdCk\nCa1bt7Y1a9a47nWrVq1yXeNOOumkQPdDZgjkFKAFKacIrxFAAAEEEEAAAQRKXEDd51544QXT\nGCJ1tzv00EPd9N01atQo8bJRgOQWIEBK7vPL0SGAAAIIIIAAAgkroHFHt9xyS8KWn4InpgBd\n7BLzvFFqBBBAAAEEEEAAAQQQiIEAAVIMUMkSAQQQQAABBBBAAAEEElOAACkxzxulRgABBBBA\nAAEEEEAAgRgIECDFAJUsEUAAAQQQQAABBBBAIDEFCJAS87xRagQQQAABBBBAAAEEEIiBAAFS\nDFDJEgEEEEAAAQQQQAABBBJTgAApMc8bpUYAAQQQQAABBBBAAIEYCBAgxQCVLBFAAAEEEEAA\nAQQQQCAxBQiQEvO8UWoEEEAAAQQQQAABBBCIgQABUgxQyRIBBBBAAAEEEEAAAQQSU4AAKTHP\nG6VGAAEEEEAAAQQQQACBGAgQIMUAlSwRQAABBBBAAAEEEEAgMQUIkBLzvFFqBBBAAAEEEEAA\nAQQQiIEAAVIMUMkSAQQQQAABBBBAAAEEElOAACkxzxulRgABBBBAAAEEEEAAgRgIECDFAJUs\nEUAAAQQQQAABBBBAIDEFCJAS87xRagQQQAABBBBAAAEEEIiBQJkY5FmiWe7evdvWrFljaWlp\nMS2H9qOkfZGCEQiFQrZr1y5Mg+F0ufj1dMuWLbZ9+/YAc07trFRPd+zYYZs3b05tiACPfufO\nnS63zMzMmH9+B1jsuM/K/06M+4ImSAH1PaWkv3++/4M7aaqnspUrKRgBfU8prV+/3kqVoj1E\nFvqe8a+L9HpvKekCJFWC6tWrx/wL1v9g1L5IwQio4uqCs1q1asFkSC7uy2bDhg1Wvnx5q1ix\nIiIBCWzatMnKli1rGRkZAeVINvoSz8rKcn//fJkHVx/0XcX3VHCeuuhUEK+//6pVqwaXcYrn\ntG3bNnfhyvdUcBVB11Nbt261ypUru/oaXM6Jm1N2dnaBg8WkC5B02kqXLh3zAEn7USuV9kUK\nRsCP6jENxlO5+Bea+h/X4Fz1t49pcJ6ROame+vU2cjnPiy7A337R7XK+029B4vs/p8y+vdbf\nvGypq/vmGPluvyeVTHH9Tca/zox02tNz2tz2JMNyBBBAAAEEEEAAAQQQSDkBAqSUO+UcMAII\nIIAAAggggAACCOxJgABpTzIsRwABBBBAAAEEEEAAgZQTIEBKuVPOASOAAAIIIIAAAggggMCe\nBJJykoY9HSzLEUAAAQQQQAABBOJLQLMtahZLzbhGQiAeBGhBioezQBkQQAABBBBAAIEUE9A0\n1Jdddpl17tzZevToYeedd56bRj3FGDjcOBQgQIrDk0KREEAAAQQQQACBZBd46KGH7N1333Ut\nR7qv1MyZM+32229P9sPm+BJAgC52CXCSKCICCCCAAAIIIJBsApMmTbJy5cq5G5nqPkjqYjd1\n6tRkO0yOJwEFaEFKwJNGkRFAAAEEEEAAgUQXqFWrlht75B+HxiFVq1bNf8n/CJSYAAFSidGz\nYwQQQAABBBBAIHUFrrzySitTpoytWrXKVq9ebbt27bJrr702dUE48rgRoItd3JwKCoIAAggg\ngAACCKSOQPfu3e3ll1+2V155xXbs2GH9+vWznj17pg4ARxq3AgRIcXtqKBgCCCCAAAIIIJDc\nAkcddZQdfvjhrvWIab6T+1wn0tHRxS6RzhZlRQABBBBAAAEEEEAAgZgKECDFlJfMEUAAAQQQ\nQAABBBBAIJEECJAS6WxRVgQQQAABBBBAAAEEEIipAAFSTHnJHAEEEEAAAQQQSF2BX375xT74\n4ANbtGhR6iJw5AknwCQNCXfKKDACCCCAAAIIIBD/Ak8++aTdd999ppvAKmla7+uvvz7+C04J\nU16AFqSUrwIAIIAAAggggAAC0QK6J9H7779vCnLUAuQHOZFbvfHGG25qbk3PPW7cuMhVNn/+\nfBs2bJiVK1fOatasaRUrVrQRI0bYV199FbUdLxCIRwFakOLxrFAmBBBAAAEEEECghAR2795t\nl1xyiU2ZMiVcgjPOOMOGDx8efv3SSy/ZrbfeamlpaW7ZrFmzbPv27da/f3/3+rvvvnPrKlSo\n4F4rUNq6datp+RFHHBHOhycIxKMALUjxeFYoEwIIIIAAAgggUEICM2bMcMFRtWrVrFatWla1\nalV78803TUGQn0aOHGllypRxrUNqIdLzUaNG+att//33NwVaO3fudMvUIqXXWk5CIN4FaEGK\n9zNE+RBAAAEEEEAAgUIIKBCZNm2aLVmyxFq0aGFHHnlkrnd/+OGH9t5777nA5swzz7RWrVqF\nt9HECqVKlXLrtDA9Pd29Vn7t27d3223bts1Kly4dfo+ea5mfdAPYk08+2SZMmOAvsm7dulnX\nrl3Dr3mCQLwKECDF65mhXAgggAACCCCAQCEF1GIzaNAgmz59uuviprFD6i53yy23hHNSa5Am\nS1CrjtaPGTPGRo8ebZ06dXLbKKhSkKUuc+oap8BH2zVv3jycR+/eve3pp58OB0nqPnfhhReG\n1+uJxhxNmjTJFixYYI0bN7YTTzzRBVpRG/ECgTgUoItdHJ4UioQAAggggAACCBRFQAGJgqPq\n1au77m+VKlVygcyPP/4Yzu6BBx5wgU2dOnWsbt26lp2dbY8++mh4vVqJBg8ebFu2bLG1a9e6\nAOm6666zZs2ahbe54YYb7Pzzz3dBlgKtCy64wK699trwej1RK5SCoquvvtpOPfXUcItU1Ea8\nQCAOBWhBisOTQpEQQAABBBBAIHUF5s6da8uXL3fd4xTARCa15gwdOtTeeustNzOcWocUqPhp\n4cKF7qnf/c2fHEHLDz74YLcuMzPTMjIy/LdY2bJlbcWKFeHXejJkyBA766yz7Oeff7ZDDz3U\ntQBFbqB87777brvjjjvcYo1BIiGQLALU5mQ5kxwHAggggAACCCS0gLq1qWVGXeA0O5zG/ug+\nQn379g0f1+23326vvPKKC47UwvP3v//dPdc4IqWmTZu6/9UqpPer65vyOuSQQ9xy/dO5c2eb\nOHGiC4y0TwVdPXv2DK/3nygw0mNvicBobzqsS1QButgl6pmj3AgggAACCCCQVAKaNEHBkWaN\n08xwSn/9619dNzc9z8rKsvHjx1uVKlVMXee0nYKf1157TatdUqCjgGrDhg3ufRpHpKCrYcOG\nv29hruWnXbt2ppak9evXW69evXJ1jwtvzBMEUlCAFqQUPOkcMgIIIIAAAgjEn4DuEaQWHbX8\nKCkQWr16tbvp6tFHH+2CIQVEkTdt1XO/O53eo/UPP/yw/elPfwrPYhfZeqRt6tWr5wItdbvT\nvph6WyokBP4QIED6w4JnCCCAAAIIIIBAiQkoUFGAoyBJExyo9Uf/+wGMghmNC3rhhRfc/YU0\nOYK2P++883KVWS1EeuwtNWrUaG+rWYdAygrQxS5lTz0HjgACCCCAAALFKaDxQHfddZebTlv3\nCHrnnXeidn/66ae7+xGp69uaNWts8+bNdvHFF9sBBxwQ3u62226zK6+80mrXrm1NmjSx+++/\n30477bTwep4ggMC+C9CCtO+G5IAAAggggAACCLig5o033rCVK1eabpSqKa4j080332z//ve/\nrUKFCm7WOE1/XbFiRevevbvbTMtff/11FzhpFrs2bdq4CRUi89DsczfddJN7RC7nOQIIBCdA\ngBScJTkhgAACCCCAQIoKqMVHN0/99ddfXRe55557zt2w9dZbb3UiGzdudIFPtWrVwlNsr1q1\nysaOHRsOkLShps/u169fiipy2AjEhwBd7OLjPFAKBBBAAAEEEIhzgalTp9pDDz1ko0aNcq1E\nkcV99dVXbenSpa7rm+5dpFnmnn322fB2GlcUObmC3qvxQxpHREIAgfgSoAUpvs4HpUEAAQQQ\nQACBOBR45JFHbPjw4W5yBAU26gqnKbfr16/vSrts2bKoAEhd4TTVtmahU8CklqPjjz/eJkyY\n4IIn3adIEzCcccYZcXi0FAmB1BagBSm1zz9HjwACCCCAAAL5CCjQGTFihBs7VKtWLdNjxYoV\n9uKLL4bf2aFDBxfw7Nixwy3TRAt16tSxgw8+OLzNAw884GacK1u2rO233342bNiwXOOUwhvz\nBAEESkyAFqQSo2fHCCCAAAIIIJAIAgp2NOW2AiN1lfOTutT5SeOPZs2aZS+99JJrOdK2Cqo0\npshPlStXdkGRAiMSAgjErwABUvyeG0qGAAIIIIAAAnEgcOCBB7qpthUQqavczp073fgh3bzV\nT+oud/fdd9ull17qZrNr2rSplS9f3l/N/wggkEACdLFLoJNFURFAAAEEEECg+AXKlCnjWoPU\nZU6tSZqR7swzz7Szzz47V2F0U1dNz01wlIuGBQgkjAAtSAlzqigoAggggAACCJSUQNu2bW36\n9On2ySefuEkXWrVqVVJFYb8IIBBjAQKkGAOTPQIIIIAAAggkh4DGE7Vs2TJqXFFyHBlHgQAC\nkQJ0sYvU4DkCCCCAAAIIIIAAAgiktAABUkqffg4eAQQQQAABBBBAAAEEIgUIkCI1eI4AAggg\ngAACKSuwfv16W7RoUdQNX1MWgwNHIIUFCJBS+ORz6AgggAACCCBgLiC66667rF27dtalSxfr\n1auXzZ8/HxoEEEhRAQKkFD3xHDYCCCCAAAII/CYwbtw4GzVqlJt8oWbNmrZgwQK7/PLLaUmi\ngiCQogIESCl64jlsBBBAAAEEEPhNQNN3K+neRaVLl7ZatWrZDz/8YL/88otbzj8IIJBaAgRI\nqXW+OVoEEEAAAQQQyCFQu3Zt2717d3hpVlaWpaenW9WqVcPLeIIAAqkjQICUOueaI0UAAQQQ\nQACBPAQGDBhg1apVs5UrV9qqVats8+bNdtFFF7lleWzOIgQQSHIBbhSb5CeYw0MAAQQQQACB\nvQs0adLExo8fby+99JKtWbPGunbtameeeebe38RaBBBIWgECpKQ9tRwYAggggAACCPgCahWa\nN2+eG1/UsGFDf3H4fwVJt99+e/g1TxBAIHUF6GKXuueeI0cAAQQQQCAlBGbOnGnHHHOMaxXq\n1q2b3XLLLVFjjlICgYNEAIECCxAgFZiKDRFAAAEEEEAg0QS2bdtmV111lW3cuNE0GUOVKlVs\nzJgx9vbbbyfaoVBeBBAoJgECpGKCZjcIIIAAAgggUDSBUChky5cvt+3bt+eZwauvvmqdO3e2\n1q1bu9Yhdafzk+5ptHbtWqtRo4ZbVLZsWdd6NHv2bH8T/kcAAQSiBAiQojh4gQACCCCAAALx\nJKBARpMmdOzY0dq1a+cmUogs34QJE1xQpABqx44dbv2NN94Y3qRu3bru3kZap6RgKy0tzerX\nrx/ehicIIIBApAABUqQGzxFAAAEEEEAgbgS2bNligwcPtsWLF1v16tVt586ddtttt9nnn38e\nLuO4ceNci5Cm6a5UqZKbmnvixImuS502qlevng0aNMhN3a0pvDVLnSZpOPvss8N58AQBBBCI\nFGAWu0gNniOAAAIIIIBA3Ah8++23LqDR2CGlypUr2+rVq+3DDz+0Dh06uGWlS5d2rULuhfeP\n30JUqtQfvwEPGTLEjjzySPvyyy/dOCRN4a2Ai4QAAgjkJUCAlJcKyxBAAAEEEECg2ATUUpSV\nlZUraNGECkpqOSpT5o9LFn+51qklaPLkyW6ckbbRpAz9+vVzrUla76fjjz/e9CAhgAAC+Qn8\n8fNKfluyHgEEEEAAAQQQCFBAgc+tt97qJldo27atDRgwwLUY+bs47LDDrHv37i74Udc4dZHT\nmKI+ffr4m1ivXr3s0UcftaZNm7p7HF1yySX2z3/+M7yeJwgggEBhBf74Oaaw72R7BBBAAAEE\nEEBgHwRGjhxpL7zwgus6p65y06dPt5tvvtlGjRoVzvXJJ5+00aNH2yeffGKNGjWygQMHum5y\n4Q28J6eddpp7RC7jOQIIIFBUAQKkosrxPgQQQAABBBDYJ4H33nvP0tPTrXz58i4fTbSgIEkz\nzmVkZLhl5cqVM7UK6UFCAAEEikOALnbFocw+EEAAAQQQQCCXgD8znb8iOzvbBUsKmkgIIIBA\nSQkQIJWUPPtFAAEEEEAgxQU0hbcmVvCn31bL0eWXX26RM9ClOBGHjwACJSBAF7sSQGeXCCCA\nAAIIIGB27LHH2ssvv+xu7qqZ7DSW6IwzzoAGAQQQKFEBAqQS5WfnCCCAAAIIpLZAx44dTQ8S\nAgggEC8CdLGLlzNBORBAAAEEEEAAAQQQQKDEBeK2BemXX36xjz/+2GrUqGGdOnXKdcO3Epej\nAAgggAACCCCAAAIIIJB0AnHZgvTiiy/an/70J/v+++9t7Nixdtlll9m6deuSDp8DQgABBBBA\nAAEEEEAAgfgSiLsASS1Hzz33nD3yyCP2j3/8w3SDON0L4bXXXosvOUqDwD4I6O7xX375pc2e\nPdv0PK+0YsUK++ijj2zZsmV5rbZdu3bZnDlzbNasWaapcfNKq1evts8//9yWLl2a12oLhUL2\nzTffuG00e1Reae3ata4cixYtymu1W/bdd9/Zp59+atu2bctzm/Xr17sW4R9//DHP9Vo4b948\nl8fmzZvz3EbLdaNIbben9MMPP7j9bNy4Mc9Ntm7d6vbxv//9L8/1Wvjzzz+7PPb0o8z27dud\nuY5ZfnmlxYsXO7M1a9bktdqysrLsiy++cOdv9+7deW6jc6bzv3LlyjzXF2cdUjmWLFmSZzki\n65Bs8kp+HVq4cGFeq92yuXPnur+JPdWhDRs2JFQdktne6tBnn31m33777T7VIf3t6zNgT3Xo\n119/dX/bQdQh5ZVXKujnkDz0/U5CAAEEEkLA+3KLq/T666+H+vfvH1UmLfNmtYlatqcXXbt2\nDXlfFntaHdhy7+I15H3pBJYfGYVC3kVjyLugTHoK78I31LNnz9CBBx7oHieccEJo+fLlUcft\n3UU+1KRJk9BBBx0Uaty4cejxxx+PWu9NiRs65ZRTwnl069Yt5AUwUdu8+uqrLo/999/fbXff\nffdFrfcu3tzflV8OrytraP78+VHb/Oc//wk1a9bMlUNl+fvf/x613gta3N+rn8eRRx4Z+vrr\nr6O2mTJlSuiwww4L53HdddeFvIuq8DZeYBYaNGiQK6P20aZNm5B38Rheryded9tQ69atXR7a\n16WXXurqi7+RFyCGrr766nAeLVu2DHk3m/RXu/+9YDTUrl27sNkFF1wQ8i7Gw9voc+Nvf/tb\nuJzNmzcPeTexDK/XE69VO+QNJg/J9IADDgidffbZIS8Yi9rmnnvuCTVs2NDlc8ghh4TeeOON\nqPVeABbq0qVLuBy9e/fOVe8ffvjhUKNGjVweBx98cOiFF16IykN1qFevXuE88qpDzzzzTLgO\nKa8RI0ZE5eEFz/nWIe+HqZCO4SDvvOiYhg0bFpWHF/hG1aGjjz463zp02223ReXhzVwWGjBg\nQKhBgwbu0b59+1x1aOrUqaEWLVq4cqgsqkNegBjOR3XImy46fP5Vh7yAPbxeT/KrQ8rvmmuu\nCeehOjRt2rSoPL766itXh1QG1cPzzz8/5AXd4W3yqkMTJkwIr9cTLxB0dcj/mwmiDnkzv+Wq\nQ96PjK4Oqa6qDo0ePTqqHH4dUj1WWY4//viQ92NM1DYFqUOnnnqqe7/y0OeQFwRH5eH1Atlr\nHYraOAFe6LNGTl7QnwClTZwi6nMg52dp4pQ+Pkvq/ajk6qo+H0m/Ceg6s0ePHgXiSNNW8RTJ\n/etf/3K/dntfxOFiaSzSLbfcYt6XVdS9EfQLe85fpG644QabNGmSpaWlhd8fiyf6ZVD70F2/\nScEI6JdITfNapUqVYDKM01x0j4+JEyda7dq13a/HmZmZbmrbhx56yJVYrSCa6lYtpxUqVDC1\neuiX+TfffNO8i0S3zV//+lcbN26c1axZ09VDtVR4Pw6Yd0Hj1qv1wbvgcX8vykOtQ/plXt1X\njzrqKLfNvffe67bXOD/dc0R5HHHEEeb9IOHWq1zdu3d3LVxVq1Z1eagV54knnjAvwHPbeBfd\n5l3Mu7GCpUuXdnl4F2PmBRZuvbb3LppM/+uGkGrpUkvAgw8+aF5g4LbxLtzsrrvucut1PxTt\nt169eu7vXXnqPcpDrWE6Xj+PO++808477zyXhyxuvvlmUzl1g0m1WOj5f//7XytXrpxzPu64\n40wtO7Vq1XLHpL/hm266yS6++GKXh87JlVde6epf2bJlXTn1//Tp08N/53379jW1HMlMH53K\nQ12Ar7/+epfHhx9+aBdddJEbM6nzt2nTJredd4HvjkkbqfuwWsJUDuWh4/V+FDIdj5JaBP7v\n//7P3SyzfPny7m9Cx/zuu++aFyy7ba644gp7//33o+qQFzC7lndtoJY678I1qg7p/P/73/8O\n1yF5ecFbVB3yAjd79tln3T7UYiAz1Y3KlSu7Oqi/z/zqkBecuHyVic6Dzp1au/ZUhx577DFT\n3ddnqfYlU+/HAXd8ykP7VN3OWYceeOAB69OnjzYxL4B0dUh5RNahDz74wL3W/pWHX4f0Wq2a\nd9xxh7NXHvr7kok+fyLr0IwZM9y50LnKqw7deOONdskllygL93d91VVXOa891aHTTz/dtRzp\n/KvlR0ZewG/67lJSS8uFF1641zrkBWZuO/8zRH+7qkPqdaHk/UhhXuDlyq1y6PMjZx1SXdff\nqcqhpHqYVx3S+ytWrBj+HNLfmhc8uvd4Pyi4bvD+55B8VYfUC0RpT3VI5ytRZ63T95Q+w+RS\nqVIld5z8s+8C+o7S34M+80jBCPjXDvr85sbLv5nqc1DfG/pOzi/F3SQNCnpyXiDr5OoPRx9K\nusjyk/fLnEUGUlquD3t94cQ6QNK+9IWpfZGCFUh2U10c60tAdVpJF9K6sPaPW+u1Tsv1Zaz/\ndXGrHwr2228/9x491xe06qAeyk9d3Pw81H1HHwT+34vy0IWmtvFaBFweCh50Mamk/ekiSBdW\nXmuW26e65ukDVsGAyqFtdWGpPBRIKSkP/29N2+hvVQGeLtD1PgUTKpP+LrVeF8DKQ+XX/U+U\nZs6c6Y5B+Wgb/f0rwFM3OO/Xb/N+kXbdDHUR5ueh/2V24oknujyUn/JV/n4euljzfvG3pk2b\nuptQKh9dpGu99qWHLn779evn8tBxKQ8FZdpGFz+6aFRXOK/lyfmpTP5Fkd6vbad7AZQuaJWU\nhyxlpTwUnOqCX8t14ahzpfy03D//+uKSo3/udP61nc6v8lCApwtcLfd/kMmrDsnAz0PP86tD\nuhDPWYdUb/w8VGa/DqkcKqfy1L4j65AMlLROdUhdNtUtVOVWfvnVIZ0DWercKakOqf6oHumc\n76kOqRydO3d271Ed0v6Vj8oaWYe8FpJcdcjfTnXopJNOcnnITMebVx3yWlFdcJVXHdK5O+us\ns1we+dUhWeh4VIdUXiX5yUCBtZKOS+v2VIe0jeqQrP08dB5l4J87//xruVJedUjb6HNDdU1J\nnxFa5ufhl0Pvlan/OaTl9evXd+/RPlUvlIceqtf63PDz2FMd0vJDDz3U5ZGo/6ibrH+ciXoM\n8VhufdeRghXQD3Wk3wT0He9/buZnEncBkj5sdQCRyX+tD9/I1KpVq/Cvv/7y8ePHuy8OfQHG\nMumLTilnmWK5z2TPW5VWvyIl+y9IumDTeAv/QltftF4XNFdvdY697jvuYiPyIkkXHw0bNoza\nRoGMf1EpN7Uu6KLJz0N/A/rb0YWW3q/Xytvfxut2ZQsWLIjKo06dOi6wUR7an5+HfyGt/Xld\nacJ56Jd+jaWKLIeCEAVy2q/y0N+0jlEXWiqHLgi13C+H8ohsHdYXpNYpONL/2p/+zpSH//em\ncul9fh7KT2XQcj2Uhy7o/OPVBZ0uvOXkXzTKV61dfh7aj8rs56FtlaectI32rUBPLQ/KQ9vJ\nN2ceWq6k9+qCW8nPQ8+9rmQuAPSPRdvkPHfazrfSPnKef9UhTWITWYe8LoHhY2noeeg9OeuQ\n76H8tU3OOhRZTq1XigyOdGyReWh7BTKR51+tGgpslLRe63QMfh3S+Y+sQ/JTHfKT3FWHdM50\nPpRHXnUoshyqC2ot8svh1yE56dxp26LWITkoD5UhrzoUWQ+1Hx1fzjrk5xFZh/T3oCSbnOc/\nrzrk56H3qA5pTJhfh/S3EXnuVA4l1QHlpXqQsw7pHBSlDkWeO+0zZx2KLKfWK+WsQ5F5uA0S\n6B9Zqn7pPPvnMIGKH7dF1d+BbP3P57gtaAIVTJ8L+n5RPVV9Jf32eet/vubr4X1oxlXyutiF\nvO4qUWWa7o0lUB/rgiSvGwVjkAoCFYfbeH/MufrRx2Ex97lI3q+urk++d5Hg+u57v6RGjZfw\nviRCXotEyLvIceNctJ3G5XlfIOF9exeUIY2R8fPwLtJyjZfwuvq49d6FpsvH69oT8i4+w3lo\nLITGWfh5eBczIY05ikz+mBzvQtNt57XYhLyuTuFNvF/UQ15rUjgP7+IopLFPkWno0KFRx9LN\nG6fgtaqEN/EusEJet79wHjpujcGKTBqDpeUqq/7XWJfIMYBeS094XI/Kqm28bnyRWYS8rmFu\nuX+8GuviddENb+P9yhbyulC5bfw8br/99vB6PfG6qLmxOP4YJO9HmpAXZIa38S6c3GeV9u+b\nabxMZPK6ALuxQX45ND4rctyW94Xmxjb5eeh/rxtgZBYhr+UjVx3yftkPb6M65LVIRB2L1x0x\nqg5pTFbOOuQFGeE89ETjuiLLkVcdOvzww8PnLq86NGTIkHAeOuY91SGZqq5qfznrkMY+ablv\nWpA6NHLkyKhjiaxDKkdR6tBLL73kyuGfO41pK0odkpOfR151SOPStN6vh9dee23UsUyePDnf\nOnTOOee4svqmOeuQ14KYbx0aOHBgVDly1iGNycpZhzReLDJpXNdBv587/a865LWIRm6SUM/1\n98kYpOBPmdfDgTFIAbMyBik3qK4zE3YMkvrZa4yDWoL066GS98Xm+mxreX7J++J0v0YXOELM\nL8M9rNesQNqHfnEnBSOgXzo0+5j/63MwucZnLmq50ZgX1SF1E9Ov6JFJv6a988477td5/bqs\n8Tr6BTsyLfJmldPfi36d1XgjdQOKTPo1Tn9H6mam/DWuJecvnurKpv2oC5f3oWFqlY1M3seL\neRdjruudftHXGBy/xcXfTt1i3377bTdGRN3IvMDDXxX+Xy1EaiXQ34vyyNmNVmMo3nrrLdc6\no/ueeRev4ff6T9QdSl1z1HVPeej/yKSWHR2v8vImi3BjTiLX67m6/ygf7V+mOf9+Nc5FeeiY\nNJbGmwghZxZujJC696o14cwzzwx3N/I31K/LOhb9wq+xGt4ECuGWDX8bddXzJq9w51TjPrwL\nR3+V+1+//MlUXbrURfDkk08Ofx76G6rlRuOQYl2HNE5FLQ3eBbvru13YOqTyalyoum/uqQ7p\n8/Tll192f/861g4dOviHGf7fr0NqodI4npx1SN0hde5UD/ZUh9StUF3EClKHVI/1fZIzqXub\nupbp/Ksve1HrkPfDn/t7VD2US2SKrEMad6jPCL91zN9O50R/m/pc2FsdUpdHdRHVNv53qp9H\nkHVIn1n6HFJLZmTS55Bfh7yAz/3t5qxDkdvH+3Mdp7rv6hj8LszxXuZEKJ965ei7TH9XpGAE\ndD2lrvW6pqJl7jdTXWfq87QgY5DibpIGda/QRYf6dGswsy4CvV8xTYNB9aWXXyJAyk8often\nUoBUXGdBgY/GwOhLx++OVVz7Tub9aDykuvAl8oVevJ0fBTgKDOvWrZsrGIi3siZSeRR8ypQU\njAABUjCOOXMhQMopsu+vCZByGxYmQIq7MUi66NCMVprRSb8oajyKN8V3gYKj3BQsQQABBBBA\nAAEEEEAAAQQKLhB3AZKKrhmy1FVCv3ypO0XOrgUFPzy2RAABBBBAAAEEEEAAAQQKLhCXAZJf\nfLoF+BL8jwACCCCAAAIIIIAAAsUh8NuNJ4pjT+wDAQQQQAABBBBAAAEEEIhzAQKkOD9BFA8B\nBBBAAAEEEEAAAQSKTyCuu9gVHwN7QqDgAppGWHec17SZxxxzTJ43tp0zZ46bgVFTNOec9rbg\ne2JLBBBAAAEEEEAAgeIWIEAqbnH2F/cCmm501apV5t24Mtd9Q3RPkQsuuMDda0X3CNL9a8aM\nGePubO8fmKakf+WVV9z9abTNlVdead7Nj/3V/I8AAggggAACCCAQxwJ0sYvjk0PRil/gpZde\ncjdV7Ny5s7tZqW4qGpmGDBlia9eudTde083XdCPPBx98MLyJWpYUMOm+Q7oZpW6qOmLECJs3\nb154G54ggAACCCCAAAIIxK8AAVL8nhtKVswCs2fPtltvvdXU6qPp5desWWOXXXaZay1SUbR8\n7ty54Tt9p6WluRuF6n1+mj9/vpuW3r9rtW4kqu0IkHwh/kcAAQQQQAABBOJbgAApvs8PpStG\ngU8++cQFM5UqVXJBjlqINm/ebF999ZUrhQKdJk2a2JYtW8Kl2r59e9QYo0aNGtnu3btNd1tX\nysrKcq8bN24cfg9PEEAAAQQQQAABBOJXgAApfs8NJStmgWrVqrlWIrUUKSnI0fPq1auHS3Ln\nnXe6SRkyMzNND92r64Ybbgiv79q1q5188smu1Uld8TZu3Gj9+/e3Vq1ahbfhCQIIIIAAAggg\ngED8CjBJQ/yeG0pWzAKnnnqqPfnkk7Z48eLwBAvdunWzNm3ahEvSsWNHmzx5ss2YMcPNYtez\nZ8+oAEqtTI899phbv2jRImvWrJnpPSQEEEAAAQQQQACBxBAgQEqM80QpAxRYtmyZawWKbBlS\n9lWrVrU333zTRo8ebT/++KO1b9/eBgwYkGvPDRo0sPPOOy/Xcn+BgiQFViQEEEAAAQQQQACB\nxBMgQEq8c0aJiyigwOiKK64wTapQqlQpO+uss+yee+6x9PT0cI6anCGyy1x4BU8QQAABBBBA\nAAEEUkKAMUgpcZo5SAnoXkSzZs1yXeI0/farr75qzzzzDDgIIIAAAggggAACCIQFCJDCFDxJ\nZgHNPPfZZ5+ZJmIoU6aMm55b/0+aNCmZD5tjQwABBBBAAAEEECikAAFSIcHYPDEFMjIy3E1b\nNe22n3bt2uXud+S/5n8EEEAAAQQQQAABBAiQqAMpIaDWoiuvvNJ27Nhhq1atspUrV5qCpsGD\nB6fE8XOQCCCAAAIIIIAAAgUTYJKGgjmxVRIIXHLJJaYZ6N5//33TzWA1Ex33J0qCE8shIIAA\nAggggAACAQoQIAWISVbxL6B7HelBQgABBBBAAAEEEEAgLwG62OWlwjIEEEAAAQQQQAABBBBI\nSQECpJQ87Rw0AggggAACCCCAAAII5CVAgJSXCsviUmDBggX2/PPP27hx42zz5s25yjhlyhTr\n0aOHNWvWzC644AJbsmRJrm1YgAACCCCAAAIIIIDA3gQYg7Q3HdbFjcDbb79t1157re3evdtC\noZA9+uij9vrrr1udOnVcGb/77jvTJAxaV758eZs2bZpdeOGFbkIGzWBHQgABBBBAAAEEEECg\nIAK0IBVEiW1KVCA7O9v+/ve/W6lSpaxmzZrusXDhQhs1alS4XBMnTjRtV6NGDRcg1a5d2374\n4QdT4ERCAAEEEEAAAQQQQKCgAgRIBZViuxITyMzMtHXr1rkbvaoQaWlpVrp0aZs7d264TGol\n0nI/qSVJidYjX4T/EUAAAQQQQAABBAoiQIBUECW2CURAgY5u1LqnlJWVZdomZ1Jr0H777Wcb\nNmxwq9TNbteuXdauXbvwppq6u0KFCrZmzRrbuHGjrV692o444gg77LDDwtvwBAEEEEAAAQQQ\nQACB/AQIkPITYv0+C6g73GmnneYCmjZt2thjjz2WK8+RI0da+/btrVevXnbSSSe57nH+Rmot\nuv/++10AtHbtWtOjbdu2dtFFF/mbWJMmTeyll16yY445xurWrWv9+/e3Z555xnXLC2/EEwQQ\nQAABBBBAAAEE8hFg9Ho+QKwumIBahrZu3WrVq1ePeoO6ul188cU2b948Nz5I2913333WqFEj\nO/nkk922mn3unnvusYyMDKtcubIbN6T3aLmCI6XOnTvbjBkz7PPPP7cqVapYx44dc3WfU4vS\nyy+/7LbnHwQQQAABBBBAAAEEiiJAC1JR1HhPlIBmlFPLkFp1zjjjjKjptdV6NH/+fFM3ufT0\ndKtUqZKbaW7q1KnhPDTjnLrNaZ220bY//fRTVCuSNq5Vq5YLqo499thcwVE4M54ggAACCCCA\nAAIIILAPAgRI+4DHW83Gjx9vw4cPdxRq2fniiy/s8ssvD9OoRUizz2mGOT+pValq1ar+S6tW\nrVr4uZ7s3LnTTbiggImEAAIIIIAAAggggEBxChAgFad2Eu5r8uTJrvWnYsWKrvVH03B/8803\ntnTpUne0ag0666yz3MQJmoBh1apVrhvdueeeG9bo16+f6zandZqtbv369danTx/bf//9w9vw\nBAEEEEAAAQQQQACB4hBgDFJxKCfxPtT640+prcNU64/GDUW2/tx7773WvHlzU7e6evXq2cCB\nA+2QQw4Jq2g80rhx4+zpp5+2xYsX2wknnGAXXHBBeD1PEEAAAQQQQAABBBAoLgECpOKSTtL9\nnHfeeS64UeuPAiNNv33++edHdZvTvYg041zkrHM5OZo2bWpDhw51LU1qhSIhgAACCCCAAAII\nIFASAgRIJaGeRPts0aKFjR071k2prXsPHX/88TZgwIAkOkIOBQEEEEAAAQQQQCCVBAiQUuls\nx+hYW7VqZY888kiMcidbBBBAAAEEEEAAAQSKT4BJGorPmj0hgAACCCCAAAIIIIBAnAsQIMX5\nCaJ4CCCAAAIIIIAAAgggUHwCBEjFZ52we8rKyrJ58+bZmjVrEvYYKDgCCCCAAAIIIIAAAgUR\nIEAqiFIKb/P555/bsccea8cdd5x16NDBhg0blsIaHDoCCCCAAAIIIIBAsgsQICX7Gd6H49u+\nfbtdfvnltnLlSqtTp46VL1/eHn/8cXv//ff3IVfeigACCCCAAAIIIIBA/AoQIMXvuSnxkqlb\nnaburlWrlqWlpbkASTeF/fTTT0u8bBQAAQQQQAABBBBAAIFYCBAgxUI1SfL0A6Ps7OyoI1Jr\nEgkBBBBAAAEEEEAAgWQUIEBKxrMa0DHtv//+du6559qGDRts1apV7lG/fn0766yzAtoD2SCA\nAAIIIIAAAgggEF8C3Cg2vs5H3JXmnnvusfbt29tnn31mCo4UMNWuXTvuykmBEEAAAQQQQAAB\nBBAIQoAAKQjFBM4jMzPT3n33Xdu8ebN16dLFWrZsGXU0pUqVsjPPPNM9olbwAgEEEEAAAQQQ\nQACBJBQgQErCk1rQQ1qyZIkLfNR9TumBBx6w4cOHW9++fQuaBdshgAACCCCAAAIIIJBUAoxB\nSqrTWbiDefLJJ2358uVuljpNyFC6dGm78847befOnYXLiK0RQAABBBBAAAEEEEgSAQKkJDmR\nRTmMH3/80dLT08NvrVChgq1bt85NyhBeyBMEEEAAAQQQQAABBFJIgAAphU52zkPt0KGDay3a\ntWuXW7V+/Xpr1KiR1axZM+emvEYAAQQQQAABBBBAICUECJBS4jTnfZAXX3yxKUhSq5Ema6hR\no4Y9+OCDeW/MUgQQQAABBBBAAAEEUkCASRpS4CTv6RArV65sY8eOtdmzZ9uWLVusbdu2pmUk\nBBBAAAEEEEAAAQRSVYAAKVXP/O/HrWm8dZ8jEgIIIIAAAggggAACCJjRxY5agAACCCCAAAII\nIIAAAgj8LkCARFVAAAEEEEAAAQQQQAABBH4XIEBK8Kqgm71quu5QKJTnkWjyhQ8//NB+/vnn\nPNezEAEEEEAAAQQQQAABBP4QYAzSHxZx9Wz37t02YcIEmzt3rh100EHWu3dvK1euXLiMW7du\ntWuuucYmTZrkgqMWLVrYyJEjbf/99w9v8/bbb9uNN95oO3bscMvOP/98+8c//hFezxMEEEAA\nAQQQQAABBBCIFiBAivaIm1d/+ctfTAGOAqW0tDR77bXX7JVXXrGyZcu6Mo4YMcLef/99NzV3\n6dKl7bvvvrObbrrJxowZ49avWbPGBUd6f61atVyQ9MILL1iXLl2sV69ecXOcFAQBBBBAAAEE\nEEAAgXgSoItdPJ2N38syZ84cFxxVrVrV6tSp427c+uWXX9rEiRPDpZ0+fbplZGRYenq6aSY6\nbfvZZ59ZVlaW20YtT2o50nIlbatg6euvv3av+QcBBBBAAAEEEEAAAQRyCxAg5TYp8SXLly93\nQY+CHyUFQGpFWrZsWbhs++23n2VnZ4dfKxiqXr16uIWpfv36bp0fMGmMkvJp0KBB+D08QQAB\nBBBAAAEEEEAAgWgBAqRoj7h4dfjhh7tgZtOmTa4827dvd+OMWrduHS7fFVdcYeXLl7eVK1fa\nqlWrbOfOnXbDDTeE1zdp0sQGDBhgGzdudOvV5U7jlDSWiYQAAggggAACCCCAAAJ5CzAGKW+X\nEl2qiRbuvvtuu+2220yBjVqPFBB17NgxXK62bdva+PHjbdy4cbZt2zY74YQTrHPnzuH1eqI8\nunbt6rrVqeWoT58+VqFChahteIEAAggggAACCCCAAAJ/CBAg/WERV8/OOecc69mzpy1YsMAO\nPPBAO+CAA3KVr1mzZjZkyJBcyyMXHHfccaYHCQEEEEAAAQQQQAABBPIXIEDK36jEtqhdu7bp\nQUIAAQQQQAABBBBAAIHiEWAMUvE457mXFStW2K+//prnOhYigAACCCCAAAIIIIBA8QsQIBW/\nuW3evNkGDx7sxhR16tTJ1J0uMzOzBErCLhFAAAEEEEAAAQQQQCBSgAApUqOYnj/wwAPuJq+6\nR5Gm5v7oo4/sjjvuKKa9sxsEEEAAAQQQQAABBBDYk0DSjUHSzVBXr17tZn7b00EHsVz7UdK+\nCpvee+89d4NXvU/3J9J03ZMmTSpSXoXddzxvLwv//MVzOROpbH493bJli5vtMJHKHs9l3bVr\nl2n6fX8q/ngua6KUTaZK/sydiVLueC8nn6nBniF9Tynp3oNF+f4PtjTJk5v/XaXPVVIwAv5n\n6vr162N+TRxMiWOfi26J49e1/PaWdAGSboZas2bNmFcG/4NR+yps0k1edf+iMmV+41cl1o1d\ni5JXYfcdz9ur4uqCU61qpGAE9GWzYcMGN717xYoVg8mUXNz9xcqWLWvlypVDIyCBdevWmW5s\nXaNGDXcfuICyTfls9F2V6t8tQVYCfV8riM/IyDD1AiEFI6Dblci2UqVKwWRILu56auvWrVal\nShXT9xXJLDs7u8DfL0kXIKkCKEjSvYNinbQP7auw6ZprrrFBgwa5G7jqvenp6XbdddcVKa/C\n7juet/fPWVFM4/m4SrJsvmVR62pJlj2e9+17+r7xXNZEK5tMcQ32rOEZnGfkr8+4Bueqz1T/\nczW4XFM7J3kq8Zn6Rz0ozN9sUgZIf1DE57Pu3bvba6+9Zm+88Yap1aR3797WpUuX+CwspUIA\nAQQQQAABBBBAIIUECJBK6GS3b9/e9CAhgAACCCCAAAIIIIBA/AgUvn9Y/JSdkiCAAAIIIIAA\nAggggAACgQoQIAXKSWYIIIAAAggggAACCCCQyAIESIl89ig7AggggAACCCCAAAIIBCpAgBQo\nJ5khgAACCCCAAAIIIIBAIgsQICXy2aPsCCCAAAIIIIAAAgggEKgAAVKgnGSGAAIIIIAAAggg\ngAACiSxAgJTIZ4+yI4AAAggggAACCCCAQKACBEiBcpIZAggggAACCCCAAAIIJLIAAVIinz3K\njgACCCCAAAIIIIAAAoEKECAFyklmCCCAAAIIIIAAAgggkMgCBEiJfPYoOwIIIIAAAggggAAC\nCAQqQIAUKCeZIYAAAggggAACCCCAQCILECAl8tmj7AgggAACCCCAAAIIIBCoAAFSoJxkhgAC\nCCCAAAIIIIAAAoksQICUyGePsiOAAAIIIIAAAggggECgAgRIgXKSGQIIIIAAAggggAACCCSy\nAAFSIp89yo4AAggggAACCCCAAAKBChAgBcpJZggggAACCCCAAAIIIJDIAgRIiXz2KDsCCCCA\nAAIIIIAAAggEKkCAFCgnmSGAAAIIIIAAAggggEAiCxAgJfLZo+wIIIAAAggggAACCCAQqAAB\nUqCcZIYAAggggAACCCCAAAKJLECAlMhnj7IjgAACCCCAAAIIIIBAoAIESIFykhkCCCCAAAII\nIIAAAggksgABUiKfPcqOAAIIIIAAAggggAACgQoQIAXKSWYIIIAAAggggAACCCCQyAIESIl8\n9ig7AggggAACCCCAAAIIBCpAgBQoJ5khgAACCCCAAAIIIIBAIgsQICXy2aPsCCCAAAIIIIAA\nAgggEKgAAVKgnGSGAAIIIIAAAggggAACiSxAgJTIZ4+yI4AAAggggAACCCCAQKACBEiBcpIZ\nAggggAACCCCAAAIIJLIAAVIinz3KjgACCCCAAAIIIIAAAoEKECAFyklmCCCAAAIIIIAAAggg\nkMgCBEiJfPYoOwIIIIAAAggggAACCAQqQIAUKCeZIYAAAggggAACCCCAQCILECAl8tmj7Agg\ngAACCCCAAAIIIBCoAAFSoJxkhgACCCCAAAIIIIAAAoksQICUyGePsiOAAAIIIIAAAggggECg\nAgRIgXKSGQIIIIAAAggggAACCCSyAAFSIp89yo4AAggggAACCCCAAAKBChAgBcpJZggggAAC\nCCCAAAIIIJDIAgRIiXz2KDsCCCCAAAIIIIAAAggEKkCAFCgnmSGAAAIIIIAAAggggEAiCxAg\nJfLZo+wIIIAAAggggAACCCAQqAABUqCcZIYAAggggAACCCCAAAKJLECAlMhnj7IjgAACCCCA\nAAIIIIBAoAIESIFykhkCCCCAAAIIIIAAAggksgABUiKfPcqOAAIIIIAAAggggAACgQoQIAXK\nSWYIIIAAAggggAACCCCQyAIESIl89ig7AggggAACCCCAAAIIBCpAgBQoJ5khgAACCCCAAAII\nIIBAIguUSeTCx2vZV65caf/617/sm2++sZYtW9pVV11l++23X7wWl3IhgAACCCCAAAIIIIDA\n7wIESAFXhW3bttk555xjP/30k6Wnp9ucOXNs5syZ9t5771mlSpUC3hvZIYAAAggggAACCCCA\nQJACdLELUtPLa8aMGbZw4UKrXbu2Va9e3f2/ZMkSmzZtWsB7IjsEEEAAAQQQQAABBBAIWoAA\nKWBRtSCVKlXK0tLSXM76X6+1nIQAAggggAACCCCAAALxLUCAFPD56dSpk5UvX97WrFljWVlZ\nlpmZaRkZGXbssccGvCeyQwABBBBAAAEEEEAAgaAFCJACFq1bt66NGjXKGjZsaJs3b7YDDjjA\nRo4cafXr1w94T2SHAAIIIIAAAggggAACQQswSUPQol5+Rx99tE2fPt0FSEzMEANgskQAAQQQ\nQAABBBBAIEYCtCDFCFbZEhzFEJesEUAAAQQQQAABBBCIgQABUgxQyRIBBBBAAAEEEEAAAQQS\nU4AAKTHPG6VGAAEEEEAAAQQQQACBGAjEdYC0a9cuGz16tG3cuDEGh06WCCCAAAIIIIAAAggg\ngEC0QFwHSI8//ribEU6zwZEQQAABBBBAAAEEEEAAgVgLxOUsditXrrQHHnjAZs+eHevjJ38E\nEEAAAQQQQAABBBBAICwQly1IQ4cOtVAoZMOGDQsXlCcIIIAAAggggAACCCCAQKwF4rIF6eab\nbzbdcHXx4sV7Pf4lS5bYzz//HLXN7t27bfv27ZaWlha1POgXCuCUtC9SMAIac+afv2ByJJes\nrCyHsHPnTupqgNVBdTU7OzvAHMlKf/tKO3bsiPnndypp67uK76ngzrj+9pX0P67BuerzlO//\n4DyVk773lfSZ6n++ugUp/E9hvrfjMkBScFSQNHny5FytTLVq1bJ169YVyxesvni0L1KwApgG\n66nctm3b5h7B55y6OepLhxS8wPr164PPNMVz5DM1+AqgCy1cg3cl6AzelHH8f5gqaCxosFii\nAdI777xjkSeub9++Vq5cuT+OJJ9n7dq1s2uuuSZqq1deecXdoDXWLUhbtmxx+61YsWLU/nlR\ndAFVWl10li9fvuiZ8M4oAb/lqGzZsqYHKRgB1dPSpUtbmTIl+hEazMHESS4K4vWrvD5TY/35\nHSeHXCzF0HcV31PBUet7auvWre5vvzDXK8GVIDlzUsCpH535ngru/Op7Sq66ptL3Fem3VrWC\nfr+U6Lf7lClTbNWqVeFzduKJJxYqQGrdurXpEZnGjh1rlStXjvkXrD4ghax9kYIR0B+yLugx\nDcZTuejXOD0yMjLcDwfB5ZzaOekiSaZcIAVXD9QdVAFSpUqVrFSpuBweG9zBFmNO+q7iMzU4\ncH1H+QESrsG5ylR//5gGZ6qAU9dVFSpUIPD8nVUeCREgPfzww8HVBHJCAAEEEEAAAQQQQAAB\nBPZRgJ/p9hGQtyOAAAIIIIAAAggggEDyCBAgJc+55EgQQAABBBBAAAEEEEBgHwVKdAxSfmU/\n6KCDbObMmfltxnoEEEAAAQQQQAABBBBAIBABWpACYSQTBBBAAAEEEEAAAQQQSAYBAqRkOIsc\nAwIIIIAAAggggAACCAQiQIAUCCOZIIAAAggggAACCCCAQDIIECAlw1nkGBBAAAEEEEAAAQQQ\nQCAQAQKkQBjJBAEEEEAAAQQQQAABBJJBgAApGc4ix4AAAggggAACCCCAAAKBCBAgBcJIJggg\ngAACCCCAAAIIIJAMAgRIyXAWOQYEEEAAAQQQQAABBBAIRIAAKRBGMkEAAQQQQAABBBBAAIFk\nECBASoazyDEggAACCCCAAAIIIIBAIAIESIEwkgkCCCCAAAIIIIAAAggkgwABUjKcRY4BAQQQ\nQAABBBBAAAEEAhEgQAqEkUwQQAABBBBAAAEEEEAgGQQIkJLhLHIMCCCAAAIIIIAAAgggEIhA\nmUByibNM+vfvH/MSZWVluX2ULVs25vtKlR3s3r3bdu3aZenp6alyyDE/Tpnu3LnTSpcu7R4x\n32GK7ECmpUqVco8UOeSYH2Z2draFQiH395+Wlhbz/aXKDvRdxfdUcGdbdVR1VX//Zcok5SVU\ncFiFyL38IMgAABPCSURBVEnf/Ur6riIFI6DvKV0DqJ6qvpLMfccU1CHN+2MPFXTjRNhu5cqV\n7oIw1mW97LLL3Bf5o48+GutdpVT+qo5cHAV3ymfNmmUPPvignXvuuda7d+/gMk7xnKinwVeA\noUOH2tdff23PPvuslS9fPvgdpGiO1NVgT/zq1avt6quvtk6dOtlVV10VbOYpnJt/Kcr3f3CV\nYMyYMfaf//zHbr/9dmvWrFlwGSd4TgoY69atm+9RJN3PHwU56HxVCrBBZmamZWRkWIMGDQqw\nNZsgUDIC33//velHA30gUFdL5hyw14IJbN261dXVevXqWaVKlQr2JrZCoJgF9Iu8PlPVMsdn\najHjs7tCC6iuVq1albpaaDkz2tyKgMZbEEAAAQQQQAABBBBAIDkFCJCS87xyVAgggAACCCCA\nAAIIIFAEgaTrYlcEgyK95dhjj2UygSLJ8abiFKhTp4716NHDGjZsWJy7ZV8IFFqgTZs2bjIB\nBr4Xmo43FKNAhQoV3Gdqy5Yti3Gv7AqBwgscfPDBrq5Wq1at8G/mHZZ0kzRwThFAAAEEEEAA\nAQQQQACBogrQxa6ocrwPAQQQQAABBBBAAAEEkk6AACnpTikHhAACCCCAAAIIIIAAAkUVKH2H\nl4r65lR93y+//GLvvfeeLVu2zM2lzk34UrUmxNdxa/rZb775xiZOnGgrVqywAw44IOpGhroR\n31dffWVTp05litr4OnUpW5o1a9bYq6++aq1atYq6keGmTZts2rRprr5WrlzZTVObskgceIkJ\n6DP1448/tunTp7sbbuo2Ijnv08P1QImdHnYcIbB9+3b3mblgwQI3pXd6enrEWjO+/6M4CvSC\nMUgFYvpjoxdffNFGjRplXbt2dQHSjh07TDeLrV69+h8b8QyBYhbQheagQYPcvblat25tn3zy\nibuXzFNPPWVVqlRxH46XXnqpLV++3DTByEcffWTdu3e36667rphLyu4Q+E1AN4a84YYb7PPP\nP3dBu/9D08KFC23gwIHWuHFj90Wvunr33Xdbx44doUOg2ASys7PtlltuMdXHI4880j799FPT\nYHfdyLhUqd8633A9UGyngx3tRUA/Jt17773WokULN9GNfgjVzYxPPfVU9y4FR3z/7wVwT6u8\nLylSAQUWL14c8i4qQ17lc+/wPkBD3hd56IknnihgDmyGQGwEVAcvu+yycObeTTdDJ554Yujp\np592y7w7aofOOeec0ObNm93rRYsWhTp37hyaN29e+D08QaA4BcaOHevqqBewh7wfmsK7Hjx4\ncOihhx4Keb/eu2XPP/986Oyzzw6/Dm/IEwRiKPDOO++EvAvM0OrVq91evF/o3etJkya511wP\nxBCfrAslcO6554aGDRsWfo+uB/T973+G8v0fpinUE8Yg7SlyzGO5fumsX7++aTpaJU1H61VC\nmzx5ch5bswiB4hPQ1LPnn39+eIfly5e3Zs2auVZOLfzwww/tuOOOs4oVK7ptDjroINM0tdTd\nMBlPilFAv8qPHj3avKA+aq+ZmZk2d+5c69OnT7grk34FVXfm77//PmpbXiAQS4F///vf1q9f\nP6tVq5bbTUZGhj333HN29NFHu9dcD8RSn7wLI5CVlWW6pYef1L1eLaA7d+50i/j+92UK9z/3\nQSqEl7onNWjQIOodCpjUvUl9lf+/vTOPsaLY4vBBUFFQBERARwX3fQPF4BKCCIqAiiIax0Aw\nSkRRcEOFP8QoiUaIGo0LYgyKEgyoIeIG7sZEAVGDexQVTQRFcCNx61e/8153au7cSeh5M/f2\njN9JZm51dXVX9Ved2/f0WSo1u9dpwAYEKkAgVo7U3YYNGzx+49JLL/Xede/qXo1F2+vWrYur\nKEOg2QnowT19+nS7+OKL632fKnZOEt+rXbt2dbcR3atyIUEgUAkCii3SfShFfuXKle5GX1tb\na1pbRsLvgUrMAn1sCYHRo0fbvHnzrEuXLta+fXubO3eujRw5Mlurk+f/llCs3wYLUn0mDdbo\n4a14jlgUQCzlaNOmTXE1ZQhUjYDeJin3iqxEZ5xxhr9FkhJfeu9qW4oUAoFKEpg9e7a/7Rwx\nYkS9bvUg15t6/cWi79mffvoprqIMgWYjEFyUbfPmza4crVq1yuM29fwfP368rVmzxvvl90Cz\n4efEOQnIO6Rnz542a9Ysj0Vq27atBbdkP4usSDz/cwL9X3MUpBzclBUkNVmmh6XbcnFCIFBt\nAj///LNNnjzZf0zOnDnT3yDpy1LWzfReTceo7dTlLq3jEwLNSUBv4pUB9LrrrivbTbnvWDVU\nkDHfsWWRUdkMBHS/STp27GghHs5GjRplIa7DsymGeA7fV+5eTb9juVcdEf8qQED33NixY93a\nuXDhQpNrqJKKyKtEL+55/jd+EnCxy8FOvsjp26P0MP0gVQa70jee6X4+IVApAnpLNGnSJFd6\n7r777iw1stLSyvSu1Mmx6N7t0aNHXEUZAs1KQFkV9eMxBBR7P6nlfdq0aTZ8+HCP99CPU73B\nj39k6l7VG1IEApUgIIulnukDBgzIutP3qOKPFD8n4fdAhoZCFQnIwilPELksyx1ZovKSJUs8\n8+KQIUN4/jdyfrAg5QDXu3dvC1m/6ryJX716dT0/+hynpCkEmoTA999/bxMmTPC1j5R2vlOn\nTnXOq5TJuldjUdB7aUxdvJ8yBJqawGmnnWZDhw61gw46yP/kBipRQhEp8TU1NZ78Jr5XlbRB\nbsxxXFJTj4vzQaCUgJ73aUxcuu+LL75w12Vt83sgpcJnNQloqRlJ7A0iZV5/v/32m+/j+e8Y\ncv9DQcqBbNCgQd5awXB6YOvLUlr6BRdckOMsNIVA0xOQO53evMsVREr8e++953/p205lY1q6\ndKlnAgt5Lk2meMUq6ccqAoFKEVDc0ZgxY7K/wYMHe9cKflcCBin2qlO2sJCS3rT4odadU7bQ\nbt26VWqY9AMBC6mTbfHixbZ8+XL/bg3pvf37c+DAgU6H3wPcJEUgoEW29XJJL0YVNyervOI8\nJf379/dPnv+OIfc/ForNiUwLcCkDk1xAlEpZ6WjHjRuX8yw0h0DTEVAKZGWxKSf9+vWz22+/\n3XdpgUMtbCjfeVmOlOGub9++5Q6jDgIVIbBixQp3C122bJlnqlOnSsag71gp+XJz0sLHU6dO\nrZdkpCIDpJN/NYH58+fbnDlzTC+V9LzXYtx65qfC74GUBJ/VJKCXojNmzDBlXpTlSO6fU6ZM\nqfN85/mff4ZQkPIz8yPk0qQ3mqT2biRADqsKAVmNFM+Rru1RlUHQKQS2gIDuUwUYx64jW3AY\nTSDQpARkmVd8p9aZ0Y/PcsLvgXJUqKs0Ab1c0v3a0POd53++GUFByseL1hCAAAQgAAEIQAAC\nEIBAKyZADFIrnlwuDQIQgAAEIAABCEAAAhDIRwAFKR8vWkMAAhCAAAQgAAEIQAACrZgAClIr\nnlwuDQIQgAAEIAABCEAAAhDIRwAFKR8vWkMAAhCAAAQgAAEIQAACrZgAClIrnlwuDQIQgAAE\nIAABCEAAAhDIR6Bdvua0hgAEIACBfzMBpZJVCu5SadeunS/02rFjx9JdTb6tBWTT1Mrbb799\nk5+/MSfUeP766y9fY6wxx3MMBCAAAQgUhwAWpOLMBSOBAAQgUHgCN910k/Xq1aveX01Nje2w\nww6211572ZIlS5r1Ol555RXvv7n7yXMRtbW1dvzxx+c5hLYQgAAEIFBQAliQCjoxDAsCEIBA\nkQnMnDnT9txzz2yIGzdutKVLl9rzzz9vI0aMsCeffNKGDx+e7acAAQhAAAIQaCkEUJBaykwx\nTghAAAIFInDyySfboYceWmdEF154oStIp5xyis2dOxcFqQ4dNiAAAQhAoKUQQEFqKTPFOCEA\nAQi0AAKDBw+2HXfc0d555506o12/fr0rTR9//LEpjmnvvfe2YcOG2QknnJC1++yzz+zRRx+1\niRMn2ooVK9xVb926dXb00UfbJZdcYtttt13WtrQgC9add97pbSZNmmTbbLNNaZNse/Pmzfbg\ngw/a8uXL7e+//7bDDz/cLrroIttpp52yNvfee6/tsssutvvuu9s999xjBx54oEkB7Natm2lM\nixcvtpdeesl69+5tY8eOzY4rLXzwwQe2YMEC++ijj2yPPfbwax44cGDWTDwef/xxu+yyy+y+\n++6zNWvW2DnnnGNDhgzJ2lCAAAQgAIEKE0gQCEAAAhCAwBYSCMpHEh5Tyfvvv1/2iDfeeMP3\nn3TSSdl+1QXFIgkJHJJgeUpCrE7Stm3bpE2bNsmcOXOyds8884wfO27cOP884ogjkv3339/L\nRx11VBKUGW/77LPPet0TTzzh20E5SoISlYQYqOS1117Lzleu8O233yYhhioJCpSPJShpSYcO\nHZLddtstCQpTdoj6HjBggI97q622SvS3du3aJCh6SXAt9L5OP/305Nhjj026dOni59R5YwkK\nj/ejvoK7YaJrELurr746a/bUU0953ahRo/xT/YwfPz7bTwECEIAABCpPAAtShRVSuoMABCDQ\nGgi88MIL9sknn/ilhEeX/fjjj/bpp5/a/fffb1tvvbVNnz49u8wbbrjBlHnu888/t+7du3t9\nUDbc+iLrTFCIsrYqBEXJZHk55JBDvD4oDPbAAw94jJMsVLEoo56sLepbYwoKS7y7XllWIGWc\ne/311+2YY47x/Tr2xBNPdEvQu+++a8rIJ1EyiCuvvNJuvvlm+/rrrz1DnfqXtUrWp/3228/b\n3XbbbTZlyhRPHOEV4Z+u9fLLL7fjjjvOLUg777yz75o2bZrdcsstJjfEoESmze3ll1+21atX\nW8+ePe2PP/7I6ilAAAIQgEAVCFReJ6NHCEAAAhBoqQRSC1J4XLnFI/4MrnVJSNCQrFy5Mru8\nf/75J3nuueeSoGxkdWkhKA9JcGFLN5PUgnTjjTdmdSq8+OKL3ldQkrw+tSA99NBDSf/+/d2C\nE1t/6hwcbXzzzTd+nqC4RLX/Lc6aNcv3hUQTXiELkqw5wR0va7tp0ya3esUWIO0M6b2T4D7n\nVqS08eTJk/18IWlFWuWfGzZsSIICmZx55pm+nVqQrr/++jrt2IAABCAAgeoRwIJUBaWULiEA\nAQi0dAKLFi2yfffd1/78808LCpFdc801Hld066232gEHHJBdXnCjcwuP4o4WLlxoH374oVue\nZKlRObUoZQeEwj777BNveiyQKhQ7FIusO7LmKA4o7jNuE5cVByRJLUfxvn79+vmmYoJSy47i\nj9q3b581C26FFh7Xdthhh2V1KgR3QY9jktUrFVnXdO2yfCneKRat3SSrVSypNSquowwBCEAA\nAtUhwDpI1eFOrxCAAARaNAEpMXKBO/LIIz15gdzbpCCceuqppoQMsTzyyCOe7ODss8+2YPWx\n33//3caMGWN9+vSJm2Xl0sVfpWhIpJzEokQMUsi+/PJLu/baa+NdZctyA5QoiUSppAvcSuFL\npWvXrmnRP9Pj07bxzhCHFG/aDz/8YNtuu6276wVLlMV/cq8rdQUs7avOydiAAAQgAIGKEsCC\nVFHcdAYBCECgdRLo27evzZgxw6666irPwqYMb1JsFO+jGKPgsuYWJGVyS0WpwEuVnnTflnwG\ntzg7//zzTdYoZZ0766yzLM4QV3oOZc6TKFNcqaR1GmdDku5TPFKpfPfdd3WqtGDu22+/bVpY\nt9Q6FFzysjinOgexAQEIQAAChSCABakQ08AgIAABCLR8AiHuxkKGOk9uoGQNEikvUghCzI2n\nuU6vUim95WamfY0VWWgkd911l8mCowQMv/76a4OnU6ruzp0728MPP1xPMZNlS5IqQeVOooVx\n5RIoxS4WKUdK+hCLkjNIStvKTU8WqCuuuCJuThkCEIAABApEAAWpQJPBUCAAAQi0ZAKyGM2e\nPdvXIFJWt5BS2w4++GDPahfSeVtIrmBfffWVzZ8/3wYNGuRuZ7/88ks9ZSUvA61NdMcdd7hl\nSLFQDYkUE2WkU8zUyJEj7a233vL1lpQl7+mnn3YLWLwWUrnz6PpWrVpl5513nh+r7HNaz0nr\nKcWic0oh07i0PpMy1D322GN27rnnuoKkbHYIBCAAAQgUkwAKUjHnhVFBAAIQaJEElCxh6tSp\npvTbWvxUiQ7mzZvn6aulSPTq1cutJ1IQpDwo5iesk/R/X2ttba2nzpblatmyZQ2eb8KECR4H\n9eabb1rIgGdyDQwZ9kzueiGTXIPHpTvCekauBErJ0rFhXSePpRo9enTaxD+V6vzVV1+1oUOH\nWsh65/FacgeUxUwLw0qpQyAAAQhAoJgE2iiBXjGHxqggAAEIQKA1EVA8khSimpqaQlxWSPvt\nGeh23XXXRo1HcUta36hc0ob4hFrXSOsiderUydRXmnQibkMZAhCAAASKQwAFqThzwUggAAEI\nQAACEIAABCAAgSoTwMWuyhNA9xCAAAQgAAEIQAACEIBAcQigIBVnLhgJBCAAAQhAAAIQgAAE\nIFBlAihIVZ4AuocABCAAAQhAAAIQgAAEikMABak4c8FIIAABCEAAAhCAAAQgAIEqE0BBqvIE\n0D0EIAABCEAAAhCAAAQgUBwCKEjFmQtGAgEIQAACEIAABCAAAQhUmQAKUpUngO4hAAEIQAAC\nEIAABCAAgeIQQEEqzlwwEghAAAIQgAAEIAABCECgygRQkKo8AXQPAQhAAAIQgAAEIAABCBSH\nAApSceaCkUAAAhCAAAQgAAEIQAACVSaAglTlCaB7CEAAAhCAAAQgAAEIQKA4BP4DnTDJV1zD\nftwAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#pdf(\"RBP_enrichment_100nt_UPvsDN.pdf\",width=6,height=4.5)\n",
    "ggplot(df.ret,aes(x=order,y=y.log,label=rbp.name))+\n",
    "  #geom_point(aes(colour=fe.ratio))+\n",
    "  #scale_colour_gradientn(colours = yk.col(10)[1:10])\n",
    "  #scale_colour_gradient( low = \"#132B43\", high = \"#56B1F7\",space = \"Lab\", na.value = \"grey50\", guide = \"colourbar\")\n",
    "  geom_point(color = dplyr::case_when(abs(df.ret$adj.y.log) >0.7~ \"#EE7447\", \n",
    "                                      abs(df.ret$adj.y.log)<=0.7 ~ \"black\"), \n",
    "             size = 1, alpha = 0.8) +theme_bw()+\n",
    "  geom_text_repel(data        = subset(df.ret, abs(y.log)>1),\n",
    "                  #nudge_y       = -(5 - subset(df.ret, abs(y.log)>2)$y.log),\n",
    "                  #nudge_x       = 150 - subset(df.ret, abs(adj.y.log)> 4)$order,\n",
    "                  #nudge_x = 4.5 - subset(df.ret, abs(adj.y.log)> 4)$y.log,\n",
    "                  size          = 3,\n",
    "                  seed=40,\n",
    "                  box.padding   = 1,\n",
    "                  point.padding = 0.2,\n",
    "                  #force.pull   = 1,\n",
    "                  segment.size  = 0.2,\n",
    "                  segment.color = \"grey50\",\n",
    "                  direction     = \"both\")+xlab(\"Rank order\")+labs(title = expression(\"0h\"%->%\"1h\"))+ylab(\"-log10(p-value)\")+theme(legend.position = \"right\")#+xlim(c(0,200))#+ylim(c(-3,16))\n",
    "dev.off() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "4a505df4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/Users/yeonuikwak/Desktop/2021_Grimsonlab/RNA_Revision/reviewer2/Request5/Rcode/step10_Fig7G_RBP'"
      ],
      "text/latex": [
       "'/Users/yeonuikwak/Desktop/2021\\_Grimsonlab/RNA\\_Revision/reviewer2/Request5/Rcode/step10\\_Fig7G\\_RBP'"
      ],
      "text/markdown": [
       "'/Users/yeonuikwak/Desktop/2021_Grimsonlab/RNA_Revision/reviewer2/Request5/Rcode/step10_Fig7G_RBP'"
      ],
      "text/plain": [
       "[1] \"/Users/yeonuikwak/Desktop/2021_Grimsonlab/RNA_Revision/reviewer2/Request5/Rcode/step10_Fig7G_RBP\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "getwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "867a74a4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
